{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TUTORIAL\n",
    "**_Training and validating CLASTER_**\n",
    "\n",
    "*Authors:* \n",
    "\n",
    "Marc Pielies Avelli ([marc.pielies-avelli@cpr.ku.dk](mailto:marc.pielies-avelli@cpr.ku.dk), [mpielies@broadinstitute.org](mailto:mpielies@broadinstitute.org) )\n",
    "\n",
    "Arnor Sigurdsson ([arnor.sigurdsson@sund.ku.dk](mailto:arnor.sigurdsson@sund.ku.dk), [asigurds@broadinstitute.org](mailto:asigurds@broadinstitute.org))\n",
    "\n",
    "Henry Webel ([henry.webel@sund.ku.dk](mailto:henry.webel@sund.ku.dk))\n",
    " \n",
    "Simon Rasmussen ([srasmuss@sund.ku.dk](mailto:srasmuss@sund.ku.dk), [srasmuss@broadinstitute.org](mailto:srasmuss@broadinstitute.org))\n",
    "\n",
    "## Introduction ##\n",
    "\n",
    "Hi! \n",
    "\n",
    "Welcome to this small tutorial on how to build and run CLASTER using the EIR framework. Please clone this repository (CLASTER) on your computer to start.\n",
    "\n",
    "### About CLASTER\n",
    "\n",
    "The Chromatin Landscape and Structure Expression Regressor (CLASTER) is a deep convolutional neural network aimed to translate a given chromatin landscape and its matching 3D structure to the corresponding nascent RNA landscape. \n",
    "\n",
    "<p align=\"center\">\n",
    "<img src=\"../images/Claster_image.png\" alt=\"Claster image\" width=\"600\"/>\n",
    "</p>\n",
    "\n",
    "**Figure 1. Network overview.** *Chromatin landscape information is provided as ATAC-seq, H3K4me3 (promoter), H3K27ac (enhancer) and H3K27me3 (silencer) enrichment levels for a region of 1Mbp at a resolution of 100bp. Chromatin structure is provided as a Micro-C map matching the genomic region covered by the other branch of the model at a resolution of 1600kbp. A set of convolutions extract features from both branches separately, then these features are combined in a set of fully connected layers (MLP) and finally used to predict the nascent transcription profiles (measured by EU-seq) at a kbp resolution.*  \n",
    "\n",
    "**Inputs:** \n",
    "\n",
    "The network takes as inputs two matrices describing the chromatin landscape in a region of 1 Mbp and the matching Micro-C map describing the contact frequency between different loci in this region.\n",
    "1) The chromatin landscape is described by a matrix of shape (4,10001) where we have the enrichment of ATAC-seq, H3K4me3, H3K27ac and H3K27me3 in bins of 100bp exploring 1Mbp of genomic context.\n",
    "2) Micro-C contact maps explore the same region at a resolution of 1600 bp.\n",
    "\n",
    "**Network:**\n",
    "The network consists of:\n",
    "- **Feature extractors**: two separate feature extractors with a number of convolutional layers implementing dilated convolutions and some residual connections. \n",
    "    - The first one extracts patterns from the chromatin landscape, described by combinations of ATAC-seq, H3K4me3, H3K27ac and H3K27me3 enrichments. \n",
    "    - The second one extracts features from the Micro-C maps (e.g. point-like contacts, compartments or domains).\n",
    "- **Fusion module**: The extracted high-level, abstract features are then combined in a number of dense layers.\n",
    "- **Output module**: A final set of dense layers maps the feature vectors to the targets in a regression task per node, which will represent the EU-seq signal at a distance from the TSS of the gene defining the sample (located in the middle).\n",
    "\n",
    "CLASTER was built using the EIR framework, a python package developed by Arnor Sigurdsson ([arnor.sigurdsson@sund.ku.dk](mailto:arnor.sigurdsson@sund.ku.dk)) which makes it easy to replicate and adapt it to new tasks. Documentation on EIR can be found at [eir.readthedocs.io](https://eir.readthedocs.io/en/latest/). Have a look at the tutorials to get a feel for the config files required and all possibilities that EIR offers. The framework uses hydra (https://hydra.cc/docs/intro/) to manage a set of configuration files, which allow you to streamline the process."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Before we start ... \n",
    "\n",
    ">ðŸ’» **Create an environment for the project:**\n",
    ">The following steps will be performed from the terminal, and once the environment is set up we will run everything else from this notebook.\n",
    ">We will first create an environment for this project, where we will install all the required dependencies. The python package EIR, which is the core framework used to build, train and test CLASTER, will need `python >= 3.11`. \n",
    ">If you have anaconda, the environment can be created from the terminal by typing:\n",
    ">\n",
    ">```bash\n",
    ">conda create -n claster_env_tutorial python=3.11 pip -y #  Create environment\n",
    ">conda activate claster_env_tutorial #  Activate it\n",
    ">```\n",
    "> We can now install all the required packages by typing:\n",
    ">\n",
    ">``` bash\n",
    ">pip install numpy pandas matplotlib jupyter ipykernel eir-dl>=0.1.43 #  Install these packages to run the notebook\n",
    ">```\n",
    "> We ran this tutorial on eir version 0.1.43."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**From now on, you can run the pipeline from this notebook!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Import the required packages\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import os\n",
    "from pathlib import Path\n",
    "from scipy.stats import pearsonr, spearmanr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Inputs and outputs\n",
    "The obtention of inputs and outputs from publicly available sources is detailed in notebook ```1_Data_obtention.ipynb```. In this tutorial we will however provide you with the already created inputs and targets for all samples in the test set.\n",
    "\n",
    "Input samples and their matching targets are named after the ENSEMBL ID code for the protein coding gene located at the center of the region of interest. We kept the orientation of the genes, and hence the EU-seq signal can go both towards the right or towards the left. \n",
    "\n",
    "**Landscape branch:**\n",
    "\n",
    "1Mbp regions centered at the TSS of protein coding genes in chromosome 4. \n",
    "\n",
    "- Input arrays can be found at the folders ```inputs/landscape_arrays/test/``` and ```inputs/microC_rotated/test/```. \n",
    "- The matching target profiles are given in a tabular format and can be found in ```targets/test_targets.csv```.\n",
    "\n",
    "> *Note: As a standard data augmentation procedure, samples were provided in their natural orientation (SampleID_forward.npy) and flipped. (SampleID_forward.npy)*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "def visualize_input_array(a,\n",
    "                     cropped_bins,\n",
    "                     scaling_factor,\n",
    "                     track_dict):\n",
    "    \"\"\"\n",
    "    Function to visualize an input numpy array (chromatin landscape).\n",
    "\n",
    "    Args:\n",
    "        a: numpy array containing the data. (Shape (4,10.001) in our case)\n",
    "        cropped_bins: number of bins to crop from one side to visualize a zoomed in\n",
    "                      and centered chunk of the input sample.\n",
    "        scaling_factor: factor to multiply the profiles by. Default should be 1.\n",
    "        track_dict: dictionary with metadata about each track.\n",
    "    \n",
    "    Returns:\n",
    "        fig: the figure object to be saved or shown.\n",
    "    \"\"\"\n",
    "    fig, axs = plt.subplots(len(track_dict), 2, figsize=(20,8))\n",
    "\n",
    "    # Plot the arrays, track by track\n",
    "    for j,line in enumerate(a):\n",
    "        axs[j][0].plot(np.arange(-len(line)//2,len(line)//2),line*scaling_factor, color=track_dict[j][\"color\"], label=track_dict[j][\"name\"], lw=.2)\n",
    "        axs[j][0].fill_between(np.arange(-len(line)//2,len(line)//2),line*scaling_factor, color=track_dict[j][\"color\"], alpha=1)\n",
    "        axs[j][0].set_ylabel(track_dict[j][\"name\"])\n",
    "        axs[-1][0].set_xlabel(\"Distance to TSS (kbp)\", fontsize= 16)\n",
    "\n",
    "        axs[j][1].plot(np.arange(-len(line)//2,len(line)//2)[cropped_bins:-cropped_bins],line[cropped_bins:-cropped_bins], color=track_dict[j][\"color\"], label=track_dict[j][\"name\"], lw=.2)\n",
    "        axs[j][1].fill_between(np.arange(-len(line)//2,len(line)//2)[cropped_bins:-cropped_bins],line[cropped_bins:-cropped_bins], color=track_dict[j][\"color\"], alpha=.6)\n",
    "\n",
    "    return fig\n",
    "        \n",
    "def visualize_target_array(a,\n",
    "                           scaling_factor: float = 1.):\n",
    "    \"\"\"\n",
    "    Function to visualize target profiles.\n",
    "    \n",
    "    Args:\n",
    "        a: 1D numpy array containing the target EU-seq profile. (Shape = (401) in our case). \n",
    "        scaling_factor: factor to multiply the profiles by. Default should be 1.\n",
    "    \n",
    "    Returns:\n",
    "        fig: the figure object to be saved or shown.\n",
    "    \"\"\"\n",
    "    fig = plt.figure(figsize=(10,2))\n",
    "    plt.plot(np.arange(-len(a)//2,len(a)//2),a*scaling_factor,label=\"ctrl\",color=\"grey\",lw=.2)\n",
    "    plt.fill_between(np.arange(-len(a)//2,len(a)//2),a*scaling_factor, color = \"silver\" , alpha=1)\n",
    "    plt.xlabel(\"Distance to TSS (kbp)\")\n",
    "    plt.ylabel(\"EU-seq enrichment (reads)\")\n",
    "\n",
    "    return fig\n",
    "\n",
    "def plot_microc_original_and_rotated(pathlist: list, num_bins: int, window_of_observation: float, name: str):\n",
    "    \"\"\" \n",
    "    This function plots the rotated and cropped Micro-C contact maps.\n",
    "    Args:\n",
    "        pathlist: list of directory paths to scan.\n",
    "        num_bins: number of bins in the width (626).\n",
    "        window_of_observation: genomic distance that our samples scan in kbp.  Set to ca. 0.707 Mbp given the\n",
    "        sqrt(2) factor in distances when rotating 45 degrees the contact matrices.\n",
    "        name: name of the sample.\n",
    "\n",
    "    Returns:\n",
    "        fig: figure object to be saved or shown.\n",
    "    \"\"\"\n",
    "    bin_per_kbp = num_bins/window_of_observation #Unit (bin/kbp)\n",
    "    max_shift = int(200*bin_per_kbp) # Distance to crop because it was rolled from the other side\n",
    "\n",
    "    for p,path in enumerate(pathlist):\n",
    "        for sample in os.scandir(path):\n",
    "            if name in sample.name: # Nice example: \"ENSMUST00000105369.7.npy\" common \"ENSMUST00000001565.14.npy\" \n",
    "                microc_map = np.flipud(np.load(sample))\n",
    "                fig = plt.figure(figsize=(8,4))\n",
    "                #plt.imshow(microc_map, cmap=\"YlOrRd\")\n",
    "                # plt.plot(200*bin_per_kbp,50,marker=\"x\")\n",
    "                # plt.plot(-200*bin_per_kbp+313,50,marker=\"x\")\n",
    "                # plt.plot(313,50,marker=\"x\")\n",
    "                plt.imshow(microc_map, cmap=\"YlOrRd\", vmin = -5, vmax = 0)\n",
    "                # Real scale\n",
    "                center_bin = np.shape(microc_map)[1]//2\n",
    "                bin_pos = np.arange(13,np.shape(microc_map)[1],50)\n",
    "                kbp_per_bin = 1/bin_per_kbp\n",
    "                bin_val = (bin_pos-np.shape(microc_map)[1]//2)*kbp_per_bin #Unit (kbp/bin)\n",
    "                bin_val = [\"%.2f\"%item for item in bin_val]\n",
    "                plt.xticks(bin_pos,bin_val, rotation=90, fontsize=8)\n",
    "                plt.xlabel(\"Distance in 1D sequence (kbp)\", fontsize=10)\n",
    "                # Careful! Scale is not the distance between points but the distance between interacting points in 1D! \n",
    "                bin_pos = np.arange(0,np.shape(microc_map)[0],25)\n",
    "                bin_val = 2*bin_pos*kbp_per_bin #*(kbp_per_bin) + np.shape(avg_att_mat)[0]\n",
    "                bin_val = [\"%.2f\"%item for item in bin_val]\n",
    "                plt.yticks(bin_pos,bin_val, fontsize=8)\n",
    "                plt.ylabel(\"Separation between\\n interacting elements\\n in 1D sequence (kbp)\", fontsize=10)\n",
    "                plt.colorbar()\n",
    "    return fig\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Inputs:**\n",
    "\n",
    "Now we will visualize an input sample centered at the gene LcK (ENSMUSG00000000409.14). \n",
    "\n",
    "**Chromatin landscape input:**\n",
    "\n",
    "In black we visualize the ATAC-seq track, showing the open chromatin regions as peaks.\n",
    "In red we show the promoter track H3K4me3. Each peak marks an active promoter, i.e. the TSS of an expressed gene.\n",
    "In blue we show the enhancer track H3K27ac, marking cis and trans regulatory elements.\n",
    "In green we show the silencer track.\n",
    "\n",
    ">Note: \n",
    ">- The wide input region per sample (1Mbp) allows us to capture plenty of different genes in a single snapshot.\n",
    ">- The correlations between marks. Active promoters (red peaks) overlap with open chromatin regions (black peaks)\n",
    ">- The enhancer mark can also appear in active promoters.\n",
    ">- The Lck gene seems to be silenced in mESCs: The gene at 0 has no active promoter nor enhancer peak, low accessibility and >is marked with H3K27me3, the silencer mark.\n",
    "\n",
    "**Chromatin structure input:**\n",
    "\n",
    "The input to the chromatin structure branch of the model is a log 10 transformed Micro-C count matrix. The matrix was rotated 45 degrees and cropped to provide contact information of the region of interest in a CLASTER friendly way. On the top we would have the original \"diagonal\" of the Micro-C map. Now, a contact of any genomic locus in the top row of the matrix is shown as a darker region (orange-red) in the middle of the matrix. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 2,
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "cropped_bins = 4400\n",
    "scaling_factor = 1\n",
    "track_dict = {0:{\"name\":\"ATAC-seq\",\"function\":\"Chromatin accessibility\",\"color\":\"k\"},\n",
    "            1:{\"name\":\"H3K4me3\",\"function\":\"Promoter\",\"color\":\"r\"},\n",
    "            2:{\"name\":\"H3K27ac\",\"function\":\"Enhancer\",\"color\":\"b\"},\n",
    "            3:{\"name\":\"H3K27me3\",\"function\":\"Chromatin silencing\",\"color\":\"g\"}}\n",
    "\n",
    "sample_name = \"ENSMUSG00000000409.14\" # \"ENSMUSG00000000085.16\"\n",
    "pathlist = [Path('../inputs/microC_rotated/test/')]\n",
    "num_bins = 626\n",
    "window_of_observation = 707.10678118  # kbp (diagonal is not observing the same length in the genome.)\n",
    "\n",
    "for orientation in [\"forward\",\"rev\"]:\n",
    "    landscape_array = np.load(f\"../inputs/landscape_arrays/test/{sample_name}_{orientation}.npy\")\n",
    "    fig = visualize_input_array(landscape_array,\n",
    "                     cropped_bins,\n",
    "                     scaling_factor,\n",
    "                     track_dict)\n",
    "    fig.show()\n",
    "    plot_microc_original_and_rotated(pathlist, num_bins, window_of_observation, name=f\"{sample_name}_{orientation}.npy\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Outputs:**\n",
    "\n",
    "The goal of the network is to predict the nascent transcription levels matching the previous inputs. Nascent transcription was measured by EU-seq, reads were mapped to the reference mouse genome. The Reads were then binned in bins of 20bp and the obtained enrichment profiles were further smoothed and downsampled to obtain output profiles at a kbp resolution for the central 401 kbp.\n",
    "\n",
    "In EIR, the outputs are provided as a csv file containing the sample ID and as many columns as targets we want to predict. \n",
    "You can now have a look at the targets for this tutorial, i.e EU-seq read counts per bin (1kbp) at a certain distance from the TSS of the gene of interest, the one naming each sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Target csv file:\n",
    "targets = pd.read_csv(\"../targets/test_targets.csv\").set_index('ID')\n",
    "targets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final profiles look like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Target profiles\n",
    "sample_name = \"ENSMUSG00000000409.14\" \n",
    "for orientation in [\"forward\",\"rev\"]:\n",
    "    target_profile = targets.loc[f\"{sample_name}_{orientation}\"]\n",
    "    print(f\"Orientation: {orientation}\")\n",
    "    fig = visualize_target_array(target_profile,scaling_factor=1)\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Training and validating CLASTER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "Parameters",
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "epochs: int = 120"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Create config files:\n",
    "First we will create the config files required to train and validate CLASTER:\n",
    "\n",
    "We will need:\n",
    "- A global configuration file: this file describes a number of training hyperparameters, e.g. the device we want to use, number of epochs, learning rate...\n",
    "- An input configuration file: contains the description of the input feature extractor, including number of layers and type. We'll have one for each branch of the model.\n",
    "- A fusion configuration file: This file describes the type and number of layers we want to fuse our features.\n",
    "- An output configuration file: It describes how do we want to link our hidden representations of each sample and our targets.\n",
    "\n",
    "The following cell will create a folder with the configs. Have a look at the fields in each of them!\n",
    "We will also create a folder for the results called runs_tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "# Create folder to store configuration files\n",
    "config_paths = [Path(\"../configurations/conf_microc_rotated_pure_conv_tutorial/\")]\n",
    "for config_path in config_paths:\n",
    "  config_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Create folder to store results:\n",
    "output_path = Path(\"../runs_tutorial/\")\n",
    "output_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Write config files explicitly:\n",
    "training_microc_rotated_yaml_contents = {\"globals.yaml\": f\"\"\"\n",
    "output_folder: ../runs_tutorial/gene_expression_microc_rotated_pure_conv_tutorial/ \n",
    "checkpoint_interval: {int(860 /120 * epochs)}\n",
    "sample_interval: {int(860 /120 * epochs)} \n",
    "n_epochs: {epochs}\n",
    "batch_size: 64\n",
    "optimizer: \"adamw\"\n",
    "lr: 0.0001 \n",
    "device: \"cpu\"\n",
    "compute_attributions: false\n",
    "# latent_sampling: \n",
    "#   layers_to_sample:\n",
    "#     - \"input_modules.contact_maps.feature_extractor.conv.0.conv_1\"\n",
    "#     - \"input_modules.contact_maps.feature_extractor.conv.1.conv_2\"\n",
    "#     - \"input_modules.contact_maps.feature_extractor.conv.1.conv_2\"\n",
    "\"\"\",\n",
    "\"input_cnn.yaml\": \"\"\" \n",
    "input_info:\n",
    "  input_source: ../inputs/landscape_arrays/test/ \n",
    "  input_name: gene_expression\n",
    "  input_type: array\n",
    "\n",
    "model_config:\n",
    "  model_type: cnn\n",
    "  #pre_normalization: \"instancenorm\"\n",
    "  model_init_config:\n",
    "    num_output_features: 512 # before fc_repr_dim\n",
    "    layers: [4,4]\n",
    "    kernel_height: 1\n",
    "    down_stride_width: 2\n",
    "    first_stride_expansion_width: 1\n",
    "    first_kernel_expansion_height: 4 #5\n",
    "    kernel_width: 10  #10\n",
    "    dilation_factor_width: 2\n",
    "    dilation_factor_height: 1\n",
    "    channel_exp_base: 5  #3 #-1\n",
    "    first_channel_expansion: 2\n",
    "    rb_do: .3\n",
    "    stochastic_depth_p: .1\n",
    "    attention_inclusion_cutoff: 1 #50 #256\n",
    "\n",
    "\"\"\",\n",
    "\"\"\"input_cnn_microc_rotated.yaml\"\"\": \"\"\" \n",
    "input_info:\n",
    "  input_source: ../inputs/microC_rotated/test/ #arrays_train_100bp_no_H3K27ac_uncoupled/\n",
    "  input_name: contact_maps\n",
    "  input_type: array\n",
    "\n",
    "model_config:\n",
    "  model_type: cnn\n",
    "  #pre_normalization: \"instancenorm\"\n",
    "  model_init_config:\n",
    "    num_output_features: 512 #before fc_repr_dim\n",
    "    layers: [1,2]\n",
    "    kernel_height: 1\n",
    "    down_stride_width: 2\n",
    "    first_stride_expansion_width: 1\n",
    "    first_kernel_expansion_height: 129 #Height of the array\n",
    "    kernel_width: 10  #10\n",
    "    dilation_factor_width: 2\n",
    "    channel_exp_base: 4  #3 #-1\n",
    "    first_channel_expansion: 2\n",
    "    rb_do: .3\n",
    "    stochastic_depth_p: .1\n",
    "    attention_inclusion_cutoff: 1 # 50 #128 #256\n",
    "\"\"\",\n",
    "\"fusion.yaml\": \"\"\"\n",
    "model_config:\n",
    "  fc_do: 0.4\n",
    "  fc_task_dim: 256\n",
    "  layers:\n",
    "  - 2\n",
    "  rb_do: 0.4\n",
    "  stochastic_depth_p: 0.5\n",
    "model_type: \"mlp-residual\"\n",
    "\"\"\",\n",
    "\"outputs.yaml\":\"\"\"\n",
    "output_info:\n",
    "  output_name: expression_output\n",
    "  output_source: ../targets/test_targets.csv \n",
    "  output_type: tabular\n",
    "\n",
    "model_config: # <- new\n",
    "  model_type: linear # <- new\n",
    "\n",
    "output_type_info:\n",
    "  con_loss_name: \"SmoothL1Loss\"\n",
    "  target_con_columns:\n",
    "  - \"-200_ctrl\"\n",
    "  - \"-199_ctrl\"\n",
    "  - \"-198_ctrl\"\n",
    "  - \"-197_ctrl\"\n",
    "  - \"-196_ctrl\"\n",
    "  - \"-195_ctrl\"\n",
    "  - \"-194_ctrl\"\n",
    "  - \"-193_ctrl\"\n",
    "  - \"-192_ctrl\"\n",
    "  - \"-191_ctrl\"\n",
    "  - \"-190_ctrl\"\n",
    "  - \"-189_ctrl\"\n",
    "  - \"-188_ctrl\"\n",
    "  - \"-187_ctrl\"\n",
    "  - \"-186_ctrl\"\n",
    "  - \"-185_ctrl\"\n",
    "  - \"-184_ctrl\"\n",
    "  - \"-183_ctrl\"\n",
    "  - \"-182_ctrl\"\n",
    "  - \"-181_ctrl\"\n",
    "  - \"-180_ctrl\"\n",
    "  - \"-179_ctrl\"\n",
    "  - \"-178_ctrl\"\n",
    "  - \"-177_ctrl\"\n",
    "  - \"-176_ctrl\"\n",
    "  - \"-175_ctrl\"\n",
    "  - \"-174_ctrl\"\n",
    "  - \"-173_ctrl\"\n",
    "  - \"-172_ctrl\"\n",
    "  - \"-171_ctrl\"\n",
    "  - \"-170_ctrl\"\n",
    "  - \"-169_ctrl\"\n",
    "  - \"-168_ctrl\"\n",
    "  - \"-167_ctrl\"\n",
    "  - \"-166_ctrl\"\n",
    "  - \"-165_ctrl\"\n",
    "  - \"-164_ctrl\"\n",
    "  - \"-163_ctrl\"\n",
    "  - \"-162_ctrl\"\n",
    "  - \"-161_ctrl\"\n",
    "  - \"-160_ctrl\"\n",
    "  - \"-159_ctrl\"\n",
    "  - \"-158_ctrl\"\n",
    "  - \"-157_ctrl\"\n",
    "  - \"-156_ctrl\"\n",
    "  - \"-155_ctrl\"\n",
    "  - \"-154_ctrl\"\n",
    "  - \"-153_ctrl\"\n",
    "  - \"-152_ctrl\"\n",
    "  - \"-151_ctrl\"\n",
    "  - \"-150_ctrl\"\n",
    "  - \"-149_ctrl\"\n",
    "  - \"-148_ctrl\"\n",
    "  - \"-147_ctrl\"\n",
    "  - \"-146_ctrl\"\n",
    "  - \"-145_ctrl\"\n",
    "  - \"-144_ctrl\"\n",
    "  - \"-143_ctrl\"\n",
    "  - \"-142_ctrl\"\n",
    "  - \"-141_ctrl\"\n",
    "  - \"-140_ctrl\"\n",
    "  - \"-139_ctrl\"\n",
    "  - \"-138_ctrl\"\n",
    "  - \"-137_ctrl\"\n",
    "  - \"-136_ctrl\"\n",
    "  - \"-135_ctrl\"\n",
    "  - \"-134_ctrl\"\n",
    "  - \"-133_ctrl\"\n",
    "  - \"-132_ctrl\"\n",
    "  - \"-131_ctrl\"\n",
    "  - \"-130_ctrl\"\n",
    "  - \"-129_ctrl\"\n",
    "  - \"-128_ctrl\"\n",
    "  - \"-127_ctrl\"\n",
    "  - \"-126_ctrl\"\n",
    "  - \"-125_ctrl\"\n",
    "  - \"-124_ctrl\"\n",
    "  - \"-123_ctrl\"\n",
    "  - \"-122_ctrl\"\n",
    "  - \"-121_ctrl\"\n",
    "  - \"-120_ctrl\"\n",
    "  - \"-119_ctrl\"\n",
    "  - \"-118_ctrl\"\n",
    "  - \"-117_ctrl\"\n",
    "  - \"-116_ctrl\"\n",
    "  - \"-115_ctrl\"\n",
    "  - \"-114_ctrl\"\n",
    "  - \"-113_ctrl\"\n",
    "  - \"-112_ctrl\"\n",
    "  - \"-111_ctrl\"\n",
    "  - \"-110_ctrl\"\n",
    "  - \"-109_ctrl\"\n",
    "  - \"-108_ctrl\"\n",
    "  - \"-107_ctrl\"\n",
    "  - \"-106_ctrl\"\n",
    "  - \"-105_ctrl\"\n",
    "  - \"-104_ctrl\"\n",
    "  - \"-103_ctrl\"\n",
    "  - \"-102_ctrl\"\n",
    "  - \"-101_ctrl\"\n",
    "  - \"-100_ctrl\"\n",
    "  - \"-99_ctrl\"\n",
    "  - \"-98_ctrl\"\n",
    "  - \"-97_ctrl\"\n",
    "  - \"-96_ctrl\"\n",
    "  - \"-95_ctrl\"\n",
    "  - \"-94_ctrl\"\n",
    "  - \"-93_ctrl\"\n",
    "  - \"-92_ctrl\"\n",
    "  - \"-91_ctrl\"\n",
    "  - \"-90_ctrl\"\n",
    "  - \"-89_ctrl\"\n",
    "  - \"-88_ctrl\"\n",
    "  - \"-87_ctrl\"\n",
    "  - \"-86_ctrl\"\n",
    "  - \"-85_ctrl\"\n",
    "  - \"-84_ctrl\"\n",
    "  - \"-83_ctrl\"\n",
    "  - \"-82_ctrl\"\n",
    "  - \"-81_ctrl\"\n",
    "  - \"-80_ctrl\"\n",
    "  - \"-79_ctrl\"\n",
    "  - \"-78_ctrl\"\n",
    "  - \"-77_ctrl\"\n",
    "  - \"-76_ctrl\"\n",
    "  - \"-75_ctrl\"\n",
    "  - \"-74_ctrl\"\n",
    "  - \"-73_ctrl\"\n",
    "  - \"-72_ctrl\"\n",
    "  - \"-71_ctrl\"\n",
    "  - \"-70_ctrl\"\n",
    "  - \"-69_ctrl\"\n",
    "  - \"-68_ctrl\"\n",
    "  - \"-67_ctrl\"\n",
    "  - \"-66_ctrl\"\n",
    "  - \"-65_ctrl\"\n",
    "  - \"-64_ctrl\"\n",
    "  - \"-63_ctrl\"\n",
    "  - \"-62_ctrl\"\n",
    "  - \"-61_ctrl\"\n",
    "  - \"-60_ctrl\"\n",
    "  - \"-59_ctrl\"\n",
    "  - \"-58_ctrl\"\n",
    "  - \"-57_ctrl\"\n",
    "  - \"-56_ctrl\"\n",
    "  - \"-55_ctrl\"\n",
    "  - \"-54_ctrl\"\n",
    "  - \"-53_ctrl\"\n",
    "  - \"-52_ctrl\"\n",
    "  - \"-51_ctrl\"\n",
    "  - \"-50_ctrl\"\n",
    "  - \"-49_ctrl\"\n",
    "  - \"-48_ctrl\"\n",
    "  - \"-47_ctrl\"\n",
    "  - \"-46_ctrl\"\n",
    "  - \"-45_ctrl\"\n",
    "  - \"-44_ctrl\"\n",
    "  - \"-43_ctrl\"\n",
    "  - \"-42_ctrl\"\n",
    "  - \"-41_ctrl\"\n",
    "  - \"-40_ctrl\"\n",
    "  - \"-39_ctrl\"\n",
    "  - \"-38_ctrl\"\n",
    "  - \"-37_ctrl\"\n",
    "  - \"-36_ctrl\"\n",
    "  - \"-35_ctrl\"\n",
    "  - \"-34_ctrl\"\n",
    "  - \"-33_ctrl\"\n",
    "  - \"-32_ctrl\"\n",
    "  - \"-31_ctrl\"\n",
    "  - \"-30_ctrl\"\n",
    "  - \"-29_ctrl\"\n",
    "  - \"-28_ctrl\"\n",
    "  - \"-27_ctrl\"\n",
    "  - \"-26_ctrl\"\n",
    "  - \"-25_ctrl\"\n",
    "  - \"-24_ctrl\"\n",
    "  - \"-23_ctrl\"\n",
    "  - \"-22_ctrl\"\n",
    "  - \"-21_ctrl\"\n",
    "  - \"-20_ctrl\"\n",
    "  - \"-19_ctrl\"\n",
    "  - \"-18_ctrl\"\n",
    "  - \"-17_ctrl\"\n",
    "  - \"-16_ctrl\"\n",
    "  - \"-15_ctrl\"\n",
    "  - \"-14_ctrl\"\n",
    "  - \"-13_ctrl\"\n",
    "  - \"-12_ctrl\"\n",
    "  - \"-11_ctrl\"\n",
    "  - \"-10_ctrl\"\n",
    "  - \"-9_ctrl\"\n",
    "  - \"-8_ctrl\"\n",
    "  - \"-7_ctrl\"\n",
    "  - \"-6_ctrl\"\n",
    "  - \"-5_ctrl\"\n",
    "  - \"-4_ctrl\"\n",
    "  - \"-3_ctrl\"\n",
    "  - \"-2_ctrl\"\n",
    "  - \"-1_ctrl\"\n",
    "  - \"0_ctrl\"\n",
    "  - \"1_ctrl\"\n",
    "  - \"2_ctrl\"\n",
    "  - \"3_ctrl\"\n",
    "  - \"4_ctrl\"\n",
    "  - \"5_ctrl\"\n",
    "  - \"6_ctrl\"\n",
    "  - \"7_ctrl\"\n",
    "  - \"8_ctrl\"\n",
    "  - \"9_ctrl\"\n",
    "  - \"10_ctrl\"\n",
    "  - \"11_ctrl\"\n",
    "  - \"12_ctrl\"\n",
    "  - \"13_ctrl\"\n",
    "  - \"14_ctrl\"\n",
    "  - \"15_ctrl\"\n",
    "  - \"16_ctrl\"\n",
    "  - \"17_ctrl\"\n",
    "  - \"18_ctrl\"\n",
    "  - \"19_ctrl\"\n",
    "  - \"20_ctrl\"\n",
    "  - \"21_ctrl\"\n",
    "  - \"22_ctrl\"\n",
    "  - \"23_ctrl\"\n",
    "  - \"24_ctrl\"\n",
    "  - \"25_ctrl\"\n",
    "  - \"26_ctrl\"\n",
    "  - \"27_ctrl\"\n",
    "  - \"28_ctrl\"\n",
    "  - \"29_ctrl\"\n",
    "  - \"30_ctrl\"\n",
    "  - \"31_ctrl\"\n",
    "  - \"32_ctrl\"\n",
    "  - \"33_ctrl\"\n",
    "  - \"34_ctrl\"\n",
    "  - \"35_ctrl\"\n",
    "  - \"36_ctrl\"\n",
    "  - \"37_ctrl\"\n",
    "  - \"38_ctrl\"\n",
    "  - \"39_ctrl\"\n",
    "  - \"40_ctrl\"\n",
    "  - \"41_ctrl\"\n",
    "  - \"42_ctrl\"\n",
    "  - \"43_ctrl\"\n",
    "  - \"44_ctrl\"\n",
    "  - \"45_ctrl\"\n",
    "  - \"46_ctrl\"\n",
    "  - \"47_ctrl\"\n",
    "  - \"48_ctrl\"\n",
    "  - \"49_ctrl\"\n",
    "  - \"50_ctrl\"\n",
    "  - \"51_ctrl\"\n",
    "  - \"52_ctrl\"\n",
    "  - \"53_ctrl\"\n",
    "  - \"54_ctrl\"\n",
    "  - \"55_ctrl\"\n",
    "  - \"56_ctrl\"\n",
    "  - \"57_ctrl\"\n",
    "  - \"58_ctrl\"\n",
    "  - \"59_ctrl\"\n",
    "  - \"60_ctrl\"\n",
    "  - \"61_ctrl\"\n",
    "  - \"62_ctrl\"\n",
    "  - \"63_ctrl\"\n",
    "  - \"64_ctrl\"\n",
    "  - \"65_ctrl\"\n",
    "  - \"66_ctrl\"\n",
    "  - \"67_ctrl\"\n",
    "  - \"68_ctrl\"\n",
    "  - \"69_ctrl\"\n",
    "  - \"70_ctrl\"\n",
    "  - \"71_ctrl\"\n",
    "  - \"72_ctrl\"\n",
    "  - \"73_ctrl\"\n",
    "  - \"74_ctrl\"\n",
    "  - \"75_ctrl\"\n",
    "  - \"76_ctrl\"\n",
    "  - \"77_ctrl\"\n",
    "  - \"78_ctrl\"\n",
    "  - \"79_ctrl\"\n",
    "  - \"80_ctrl\"\n",
    "  - \"81_ctrl\"\n",
    "  - \"82_ctrl\"\n",
    "  - \"83_ctrl\"\n",
    "  - \"84_ctrl\"\n",
    "  - \"85_ctrl\"\n",
    "  - \"86_ctrl\"\n",
    "  - \"87_ctrl\"\n",
    "  - \"88_ctrl\"\n",
    "  - \"89_ctrl\"\n",
    "  - \"90_ctrl\"\n",
    "  - \"91_ctrl\"\n",
    "  - \"92_ctrl\"\n",
    "  - \"93_ctrl\"\n",
    "  - \"94_ctrl\"\n",
    "  - \"95_ctrl\"\n",
    "  - \"96_ctrl\"\n",
    "  - \"97_ctrl\"\n",
    "  - \"98_ctrl\"\n",
    "  - \"99_ctrl\"\n",
    "  - \"100_ctrl\"\n",
    "  - \"101_ctrl\"\n",
    "  - \"102_ctrl\"\n",
    "  - \"103_ctrl\"\n",
    "  - \"104_ctrl\"\n",
    "  - \"105_ctrl\"\n",
    "  - \"106_ctrl\"\n",
    "  - \"107_ctrl\"\n",
    "  - \"108_ctrl\"\n",
    "  - \"109_ctrl\"\n",
    "  - \"110_ctrl\"\n",
    "  - \"111_ctrl\"\n",
    "  - \"112_ctrl\"\n",
    "  - \"113_ctrl\"\n",
    "  - \"114_ctrl\"\n",
    "  - \"115_ctrl\"\n",
    "  - \"116_ctrl\"\n",
    "  - \"117_ctrl\"\n",
    "  - \"118_ctrl\"\n",
    "  - \"119_ctrl\"\n",
    "  - \"120_ctrl\"\n",
    "  - \"121_ctrl\"\n",
    "  - \"122_ctrl\"\n",
    "  - \"123_ctrl\"\n",
    "  - \"124_ctrl\"\n",
    "  - \"125_ctrl\"\n",
    "  - \"126_ctrl\"\n",
    "  - \"127_ctrl\"\n",
    "  - \"128_ctrl\"\n",
    "  - \"129_ctrl\"\n",
    "  - \"130_ctrl\"\n",
    "  - \"131_ctrl\"\n",
    "  - \"132_ctrl\"\n",
    "  - \"133_ctrl\"\n",
    "  - \"134_ctrl\"\n",
    "  - \"135_ctrl\"\n",
    "  - \"136_ctrl\"\n",
    "  - \"137_ctrl\"\n",
    "  - \"138_ctrl\"\n",
    "  - \"139_ctrl\"\n",
    "  - \"140_ctrl\"\n",
    "  - \"141_ctrl\"\n",
    "  - \"142_ctrl\"\n",
    "  - \"143_ctrl\"\n",
    "  - \"144_ctrl\"\n",
    "  - \"145_ctrl\"\n",
    "  - \"146_ctrl\"\n",
    "  - \"147_ctrl\"\n",
    "  - \"148_ctrl\"\n",
    "  - \"149_ctrl\"\n",
    "  - \"150_ctrl\"\n",
    "  - \"151_ctrl\"\n",
    "  - \"152_ctrl\"\n",
    "  - \"153_ctrl\"\n",
    "  - \"154_ctrl\"\n",
    "  - \"155_ctrl\"\n",
    "  - \"156_ctrl\"\n",
    "  - \"157_ctrl\"\n",
    "  - \"158_ctrl\"\n",
    "  - \"159_ctrl\"\n",
    "  - \"160_ctrl\"\n",
    "  - \"161_ctrl\"\n",
    "  - \"162_ctrl\"\n",
    "  - \"163_ctrl\"\n",
    "  - \"164_ctrl\"\n",
    "  - \"165_ctrl\"\n",
    "  - \"166_ctrl\"\n",
    "  - \"167_ctrl\"\n",
    "  - \"168_ctrl\"\n",
    "  - \"169_ctrl\"\n",
    "  - \"170_ctrl\"\n",
    "  - \"171_ctrl\"\n",
    "  - \"172_ctrl\"\n",
    "  - \"173_ctrl\"\n",
    "  - \"174_ctrl\"\n",
    "  - \"175_ctrl\"\n",
    "  - \"176_ctrl\"\n",
    "  - \"177_ctrl\"\n",
    "  - \"178_ctrl\"\n",
    "  - \"179_ctrl\"\n",
    "  - \"180_ctrl\"\n",
    "  - \"181_ctrl\"\n",
    "  - \"182_ctrl\"\n",
    "  - \"183_ctrl\"\n",
    "  - \"184_ctrl\"\n",
    "  - \"185_ctrl\"\n",
    "  - \"186_ctrl\"\n",
    "  - \"187_ctrl\"\n",
    "  - \"188_ctrl\"\n",
    "  - \"189_ctrl\"\n",
    "  - \"190_ctrl\"\n",
    "  - \"191_ctrl\"\n",
    "  - \"192_ctrl\"\n",
    "  - \"193_ctrl\"\n",
    "  - \"194_ctrl\"\n",
    "  - \"195_ctrl\"\n",
    "  - \"196_ctrl\"\n",
    "  - \"197_ctrl\"\n",
    "  - \"198_ctrl\"\n",
    "  - \"199_ctrl\"\n",
    "  - \"200_ctrl\"\n",
    "  \"\"\"\n",
    "}\n",
    "\n",
    "for file,content in training_microc_rotated_yaml_contents.items():\n",
    "  with open(config_paths[0] / file, 'w') as f:\n",
    "      f.write(content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Run CLASTER:\n",
    "\n",
    "> â° *This step can take a while if we train for the recommended 120 epochs. To accelerate computations set the device to \"cuda\" or \"mps\" (for Mac) in the global config file, if you have access to a GPU.* \n",
    ">```python\n",
    ">device: \"cuda\" # Alternatives are \"mps\" or \"cpu\"\n",
    ">```\n",
    ">*Otherwise, you can reduce the number of epochs by changing the following parameters in the global configs:*\n",
    ">\n",
    ">```python\n",
    ">checkpoint_interval: 860\n",
    ">sample_interval: 860 \n",
    ">n_epochs: 120\n",
    ">```\n",
    ">\n",
    ">*To* \n",
    ">\n",
    ">```python\n",
    ">checkpoint_interval: 72\n",
    ">sample_interval: 72 \n",
    ">n_epochs: 5\n",
    ">```\n",
    ">\n",
    ">*This will highly compromise the performance of the model but will run faster. `checkpoint_interval` and `sample_interval` define at which batch number should we keep the results (model weights and predictions). We have 2590 samples and train in batches of 64 samples, which yields 36 batches per epoch. We can hence save the model every 72 batches (i.e. 2 epochs).* \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the following cell to train in a command-line fashion from this jupyter notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! eirtrain \\\n",
    "--global_configs ../configurations/conf_microc_rotated_pure_conv_tutorial/globals.yaml \\\n",
    "--input_configs ../configurations/conf_microc_rotated_pure_conv_tutorial/input_cnn.yaml ../configurations/conf_microc_rotated_pure_conv_tutorial/input_cnn_microc_rotated.yaml \\\n",
    "--fusion_configs ../configurations/conf_microc_rotated_pure_conv_tutorial/fusion.yaml \\\n",
    "--output_configs ../configurations/conf_microc_rotated_pure_conv_tutorial/outputs.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Analyze results:\n",
    "\n",
    "The previous code will generate a number of files and folders.\n",
    "- Configs: a summary of the config files and complete description of model hyperparameters.\n",
    "- Meta: eir version used (tested with eir 0.1.43).\n",
    "- Results: predictions and true targets are given here for every batch we decided to store (e.g. 860). Performance plots per target node (e.g. -143_ctrl) are also provided here as csv and pdf files. Individual files per target node will be concatenated to yield predicted and target dataframes.\n",
    "- Saved models: saved model weights.\n",
    "- Serializations\n",
    "- Tensorboard_logs\n",
    "- A number of files including:\n",
    "    - logging_history.log\n",
    "    - model_info.txt: meta description of the model's architecture\n",
    "    - training and validation history logs\n",
    "    - *.pdf: performance curves\n",
    "\n",
    "Now we will visualize the results by:\n",
    "1) Plotting predicted vs True target values at a bin level (1kbp resolution).\n",
    "2) Visualizing the predicted profiles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "#### Functions:\n",
    "def _get_predictions(results_path: Path, N_BINS: int, condition_list: list, SPLIT: int):\n",
    "    \"\"\"\n",
    "    This function reads the prediction files that EIR yields for each target (bin_condition, e.g. -150_ctrl) in the \n",
    "    training setting. Predictions are then converted into dataframes. \n",
    "    \n",
    "    Args:\n",
    "        results_path: path where the predictions are stored\n",
    "        N_BINS: number of the bins to one side of the central bin (200 normally, 57 for the Enformer predictions).\n",
    "        condition_list: list of conditions. As of now [\"_ctrl\"], but can be extended to other target conditions.\n",
    "        SPLIT: The batch number we want to use for our visualization.\n",
    "    Returns:\n",
    "        ids: names of the samples\n",
    "        predicted: predicted untransformed values \n",
    "        actual: actual values \n",
    "        \n",
    "    \"\"\"\n",
    "\n",
    "    # Lists to store DataFrames for concatenation\n",
    "    predicted_dfs = []\n",
    "    actual_dfs = []\n",
    "\n",
    "    for condition in condition_list:\n",
    "        for i in range(-N_BINS, N_BINS + 1):\n",
    "            file_path = results_path / f\"expression_output/{i}{condition}/samples/{SPLIT}/regression_predictions.csv\"\n",
    "            predictions = pd.read_csv(file_path).set_index('ID')\n",
    "\n",
    "            # Apply ReLU to model predictions\n",
    "            predictions = predictions.clip(lower=0)\n",
    "\n",
    "            # Prepare data for concatenation.\n",
    "            predicted_column = predictions[\"Predicted\"].rename(f\"{i}{condition}\")\n",
    "            actual_column = predictions[\"Actual\"].rename(f\"{i}{condition}\")\n",
    "\n",
    "            predicted_dfs.append(predicted_column)\n",
    "            actual_dfs.append(actual_column)\n",
    "\n",
    "    # Concatenate all DataFrames horizontally\n",
    "    predicted = pd.concat(predicted_dfs, axis=1)\n",
    "    actual = pd.concat(actual_dfs, axis=1)\n",
    "    ids = list(predicted.index)\n",
    "\n",
    "    return ids, predicted, actual\n",
    "\n",
    "def plot_target_predictions(line_p, line_a):\n",
    "    \"\"\" \n",
    "    Plot the predicted and target EU-seq profiles. \n",
    "    Args:\n",
    "        line_p: 1D numpy array containing the predicted values for the 401 output nodes\n",
    "        line_a: \"                           \" ground truth \"                         \"\n",
    "\n",
    "    Returns:\n",
    "        fig: figure object to save or show.\n",
    "\n",
    "    \"\"\"\n",
    "    fig, axs =plt.subplots(2, figsize=(8,4), sharex = True)\n",
    "    axs[0].fill_between(np.arange(-200,201),line_p, lw=1, color='royalblue',alpha=.3, label=\"Predicted expression control\")\n",
    "    axs[1].fill_between(np.arange(-200,201),line_a, lw=1, color='silver', alpha=.6, label=\"Actual expression control\")\n",
    "    axs[1].set_xlabel(\"Distance from TSS (kbp)\", fontsize=12)\n",
    "    axs[1].set_ylabel('Reads')\n",
    "    fig.legend(loc='upper center', ncols=2)   \n",
    "    \n",
    "    return fig\n",
    "\n",
    "def plot_correlations(pred_list_values: list, actual_list_values: list, title: str, cmap: colors.Colormap = \"afmhot\", binlims: tuple = (0,50)):\n",
    "    \"\"\" \n",
    "    This function plots predicted vs. actual values as a density map and provides the Spearman and Pearson correlations of the regressions. \n",
    "\n",
    "    Args:\n",
    "        pred_list_values: list of predicted values flattened for all samples.\n",
    "        actual_list_values: list of target values flattened for all samples.\n",
    "        title: title of the plot\n",
    "        cmap: colormap to visualize the density plot.\n",
    "        binlims: limits of the bins\n",
    "\n",
    "    Returns:\n",
    "        fig: figure object to store or show.\n",
    "    \"\"\"\n",
    "    # Get the counts histogram to plot correlation with density colormap\n",
    "    counts, xedges, yedges = np.histogram2d(actual_list_values, pred_list_values, bins=[np.linspace(binlims[0],binlims[1],200),np.linspace(binlims[0],binlims[1],200)], density=False)\n",
    "    \n",
    "    # Plot regression of single predicted values vs. real\n",
    "    pearson_r = pearsonr(actual_list_values,pred_list_values)[0]\n",
    "    spearman_r = spearmanr(actual_list_values,pred_list_values)[0]\n",
    "    m, b = np.polyfit(actual_list_values, pred_list_values, deg=1)\n",
    "    # Actual fit:\n",
    "    a1, a2, a3 = np.polyfit(actual_list_values, pred_list_values, deg=2)\n",
    "    # Regression from origin:\n",
    "    #m_0 , _, _, _ = np.linalg.lstsq(np.array(actual_list_values)[:,np.newaxis], pred_list_values) #x needs to be a column vector for this function\n",
    "\n",
    "    # Create the figure\n",
    "    fig, ax = plt.subplots(figsize=(4,4))    \n",
    "    min_val, max_val, step = (0,10,.01)\n",
    "    eps= 1\n",
    "    # Counts matrix is transposed: the origin convention for plt.imshow is (top, left), for plt.hist2D is (bottom,left)\n",
    "    counts[counts < eps] =1\n",
    "    hist = ax.imshow(counts.T, origin='lower', extent=[xedges[0], xedges[-1], yedges[0], yedges[-1]], cmap=cmap,  norm=colors.LogNorm())\n",
    "    _ = fig.colorbar(hist)\n",
    "    _ = ax.axline(xy1=(0, b), slope=m, label=f'Linear fit:\\n$ y = {m:.3f}x {b:+.3f}$\\n$r_p:{pearson_r:.4f}$\\n$r_s:{spearman_r:.4f}$', color=\"royalblue\", ls=\"-\", lw=1)\n",
    "    #_ = ax.plot(np.arange(min_val,max_val,step), a1*(np.arange(min_val,max_val,step)**2)+a2*(np.arange(min_val,max_val,step))+a3, color='royalblue', ls = \"--\", lw=1, label= f\"Polynomial fit:\\n$y={a1:.3f}x^2+{a2:.3f}x+{a3:.3f}$\")\n",
    "    _ = ax.axis(\"scaled\")\n",
    "    _ = ax.set_xlim(min_val,max_val)\n",
    "    _ = ax.set_ylim(min_val,max_val)\n",
    "    _ = ax.set_xlabel(\"log$_2$(True +1)\")\n",
    "    _ = ax.set_ylabel(\"log$_2$(Predicted +1)\")\n",
    "    fig.legend(bbox_to_anchor=(.48,.8), fancybox=True, shadow=False, fontsize=7)\n",
    "\n",
    "    return fig\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now plot the overall validation performance (a fraction of samples are held out by default for validation). \n",
    "We will also plot the profiles for the sample centered around the gene Gm49337 (chr4, + strand)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "lines_to_next_cell": 2,
    "tags": [
     "hide-input"
    ]
   },
   "outputs": [],
   "source": [
    "results_path = Path(\"../runs_tutorial/gene_expression_microc_rotated_pure_conv_tutorial/results/\")\n",
    "N_BINS= 200\n",
    "condition_list = [\"_ctrl\"]\n",
    "SPLIT = int(860 /120 * epochs)\n",
    "# SPLIT = 4300 # for 860 samples, 120 epochs\n",
    "\n",
    "ids, predicted, actual = _get_predictions(results_path, N_BINS, condition_list, SPLIT)\n",
    "\n",
    "\n",
    "fig = plot_correlations(np.log2(predicted.values.flatten() + 1), np.log2(actual.values.flatten() +1 ), title= \"Binwise Predicted vs True\", cmap = \"binary\", binlims = (0,10))\n",
    "\n",
    "for name in ['ENSMUSG00000111410.1_forward']:\n",
    "    sample_name = name.split('_')[0]\n",
    "\n",
    "    line_p = predicted.loc[f\"{sample_name}_forward\"].values\n",
    "    line_a = actual.loc[f\"{sample_name}_forward\"].values\n",
    "\n",
    "    fig = plot_target_predictions(line_p, line_a)\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Summary\n",
    "\n",
    "Thanks for making it that far! We would like to end this tutorial with the following remark:\n",
    "\n",
    "CLASTER was designed to predict nascent transcription, but the EIR framework can be reused and expanded for any other purpose. It can handle plenty of other data modalities as well! You just need a well defined input containing some information or signal that can be used to infer the values of another, as an output. You can then customize your network using EIR, provide the inputs and targets in one of the supported data modalities, and name samples with the same ID for input and output.\n",
    "\n",
    "If in doubt, feel free to reach out to us!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py:percent"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
