
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>1. DATA OBTENTION &amp; PREPROCESSING</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=3ee479438cf8b5e0d341" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=3ee479438cf8b5e0d341" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=3ee479438cf8b5e0d341"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'scripts/1_Data_obtention';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title"></p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/RasmussenLab/CLASTER" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Claster
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="0_Tutorial.html">TUTORIAL</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/RasmussenLab/CLASTER/blob/master/scripts/1_Data_obtention.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER/edit/master/scripts/1_Data_obtention.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/scripts/1_Data_obtention.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1><center> 1. DATA OBTENTION & PREPROCESSING <center></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><center> 1. DATA OBTENTION &amp; PREPROCESSING <center></center></center></a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#before-we-start">Before we start …</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-obtention">1. Data obtention</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#micro-c-data-acquisition">Micro-C Data acquisition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#epigenomic-data-acquisition">Epigenomic Data acquisition:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-perturbations">2. Enhancer perturbations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmark-data-obtention">3. Benchmark data obtention</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-bed-file-with-dna-chunk-boundaries">Create bed file with DNA chunk boundaries:</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-obtention-preprocessing">
<h1><center> 1. DATA OBTENTION &amp; PREPROCESSING <center><a class="headerlink" href="#data-obtention-preprocessing" title="Link to this heading">#</a></h1>
<p>Hi!
This notebook is a guide to obtain and preprocess the data used to train and test CLASTER. It is meant to be run sequentially, i.e. one cell after the other, and it mainly contains:</p>
<ul class="simple">
<li><p>Markdown cells with descriptions of the data or the code</p></li>
<li><p>Code cells:</p>
<ul>
<li><p>that can be run from this notebook, i.e. fast.</p></li>
<li><p>that can be run as a separate python file. These cells, staring with the magic function %%writefile, create self-sufficient pieces of code were designed to speed up computations that can be parallelized, e.g. repeating a process per each gene / genomic region.</p></li>
</ul>
</li>
</ul>
<section id="before-we-start">
<h2>Before we start …<a class="headerlink" href="#before-we-start" title="Link to this heading">#</a></h2>
<p>A) It would be highly recommendable to create an environment for this project. The python package EIR, which is the core framework used to build, train and test CLASTER, will need python &gt;= 3.11. If you have anaconda, it can be done as follows from the terminal:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>claster_env<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.11<span class="w"> </span>-y<span class="w"> </span><span class="c1">#Create environment conda create -p ./claster_env python=3.11 -y</span>
conda<span class="w"> </span>activate<span class="w"> </span>claster_env<span class="w"> </span><span class="c1">#Activate it</span>
pip<span class="w"> </span>install<span class="w"> </span>ipykernel<span class="w"> </span><span class="c1">#Install ipykernel to run notebook</span>
</pre></div>
</div>
<blockquote>
<div><p>⚠️ Warning ⚠️</p>
<p>If you created a specific environment, make sure to have it activated.
In VS code we can simply choose the environment we want in the top right icon.</p>
</div></blockquote>
<p>B) We will first install a number of dependencies required to follow the pipeline:</p>
<ul class="simple">
<li><p>pyBigWig: python package to handle BigWig files, i.e. process genomic tracks as numpy arrays.</p></li>
<li><p>wget: python version of wget</p></li>
<li><p>pandas: python package to handle data tables as DataFrames</p></li>
<li><p>cooler: package to handle Micro-C data in mcool format.</p></li>
<li><p>matplotlib: python plotting package</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Install pyBigWig</span>
<span class="o">!</span><span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>pyBigWig<span class="w"> </span>wget<span class="w"> </span>pandas<span class="w"> </span>cooler<span class="w"> </span>matplotlib<span class="w"> </span>imageio
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="data-obtention">
<h1>1. Data obtention<a class="headerlink" href="#data-obtention" title="Link to this heading">#</a></h1>
<p><strong>Import libraries</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span> 
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">wget</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">gaussian_filter1d</span><span class="p">,</span> <span class="n">rotate</span>
<span class="kn">import</span> <span class="nn">cooler</span>
<span class="kn">import</span> <span class="nn">logging</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Functions</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">destination_folder</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function downloads the files in links and stores them in destination_folder.</span>
<span class="sd">    Args:</span>
<span class="sd">        links: dictionary where keys are file names and values are links.</span>
<span class="sd">        destination_folder: path where we&#39;ll store the files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span><span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Ensure the destination folder exists</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">)</span>

            <span class="c1"># Specify the output path for the downloaded file</span>
            <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

            <span class="c1"># Download the file to the specified destination folder</span>
            <span class="n">wget</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">link</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Successfully downloaded </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">destination_folder</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Failed to download </span><span class="si">{</span><span class="n">link</span><span class="si">}</span><span class="s2"> because : </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Gene annotations</strong></p>
<p>We obtained both the mouse reference genome and protein coding gene annotations from the release M25 (GRCm38.p6) from GENCODE.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Gene_annotations.gtf.gz&quot;</span><span class="p">:</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz&quot;</span><span class="p">}</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="s2">&quot;../annotations/&quot;</span>
<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>

<span class="o">!</span><span class="w"> </span>gunzip<span class="w"> </span>../annotations/Gene_annotations.gtf.gz

<span class="c1"># Create a file with annotations that correspond only to genes</span>
<span class="o">!</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;$3 == &quot;gene&quot;&#39;</span><span class="w"> </span>../annotations/Gene_annotations.gtf<span class="w"> </span>&gt;<span class="w"> </span>../annotations/Filtered_gene_annotations.gtf

<span class="c1"># Further filter the number of fields/columns required</span>

<span class="n">file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/Filtered_gene_annotations.gtf&quot;</span><span class="p">)</span>
<span class="n">outfile</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_gene_annotations.tsv&quot;</span><span class="p">)</span>
<span class="n">gene_annotations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

<span class="n">ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">final_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span><span class="s2">&quot;End&quot;</span><span class="p">,</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span><span class="s2">&quot;type&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">features</span><span class="p">)</span> <span class="ow">in</span> <span class="n">gene_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="n">splitline</span> <span class="o">=</span>  <span class="n">features</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
    <span class="nb">id</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">gene_type</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">gene_type</span> <span class="o">==</span> <span class="s1">&#39;&quot;protein_coding&quot;&#39;</span> <span class="ow">and</span> <span class="n">chrom</span> <span class="o">!=</span><span class="s1">&#39;chrM&#39;</span> <span class="p">:</span> <span class="c1"># or gene_type == &#39;&quot;lincRNA&quot;&#39;:</span>
        <span class="n">final_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">id</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="n">chrom</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">strand</span><span class="p">,</span><span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">gene_type</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>

<span class="n">final_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Enhancer annotations</strong></p>
<p>Coordinates for proximal Enhancer Like Signatures (pELS) and distal Enhancer Like Signatures (dELS) were obtained from Supplementary Table 11 in the paper:</p>
<ul class="simple">
<li><p>https://doi.org/10.1038/s41586-020-2493-4</p></li>
</ul>
<p>Download link:</p>
<ul class="simple">
<li><p>https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-020-2493-4/MediaObjects/41586_2020_2493_MOESM13_ESM.txt</p></li>
</ul>
<p>Once downloaded, you should move the file to the annotations folder to proceed. E.g.:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>!<span class="w"> </span>mv<span class="w"> </span>~/Downloads/41586_2020_2493_MOESM13_ESM.txt<span class="w"> </span>../annotations/<span class="w"> </span><span class="c1"># Edit if necessary</span>
</pre></div>
</div>
<blockquote>
<div><p><em>Note: mm10 and GRCm38 are synonyms to refer to a reference genome</em></p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">CRE_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/41586_2020_2493_MOESM13_ESM.txt&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;cCRE_accession&quot;</span><span class="p">)</span>
<span class="n">CRE_df</span> <span class="o">=</span> <span class="n">CRE_df</span><span class="p">[(</span><span class="n">CRE_df</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;ELS&quot;</span><span class="p">))]</span>
<span class="n">CRE_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_Enhancer_annotation.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="micro-c-data-acquisition">
<h2>Micro-C Data acquisition<a class="headerlink" href="#micro-c-data-acquisition" title="Link to this heading">#</a></h2>
<p>Micro-C matrices for mESCs were obtained from:</p>
<p>Tsung-Han S. Hsieh et al., Resolving the 3D Landscape of Transcription-Linked Mammalian Chromatin Folding, Molecular Cell,
2020 https://doi.org/10.1016/j.molcel.2020.03.002.</p>
<blockquote>
<div><p>⏰: The mcool file occupies quite some memory and it took us 33 minutes to download:</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;GSE130275_mESC_WT_combined_2.6B.mcool&quot;</span><span class="p">:</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE130275&amp;format=file&amp;file=GSE130275</span><span class="si">%5F</span><span class="s2">mESC</span><span class="si">%5F</span><span class="s2">WT</span><span class="si">%5F</span><span class="s2">combined</span><span class="si">%5F</span><span class="s2">2</span><span class="si">%2E</span><span class="s2">6B</span><span class="si">%2E</span><span class="s2">mcool&quot;</span><span class="p">}</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="s2">&quot;../GEO_files/&quot;</span>
<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The following cells will create two python files that generate the microC arrays in different ways. The first one does not use parallelization but the second one does, but they do the same.
You can run the file in the terminal as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">create_microC_arrays</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_microC_arrays.py

<span class="kn">import</span> <span class="nn">os</span> 
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">wget</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">rotate</span>
<span class="kn">import</span> <span class="nn">cooler</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="c1">########### Functions #####################################</span>
<span class="k">def</span> <span class="nf">find_dataset_min_max</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">))</span>

<span class="k">def</span> <span class="nf">plot_histogram_dataset</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span> <span class="nf">mcool_summary</span><span class="p">(</span><span class="n">microc_path</span><span class="p">):</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>

    <span class="c1"># Inspecting what is inside the cooler object:</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">info</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">chromnames</span>
    <span class="n">extent</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">extent</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Metadata:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Extent:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">create_Micro_C_arrays</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">500000</span><span class="p">,</span> <span class="n">imputation_value</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1e-14</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is aimed to retreive the contact maps for each sample.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    id_path: path to file with ids (including the file&#39;s name).</span>
<span class="sd">    input_path: path to the input mcool (contact matrix) file.</span>
<span class="sd">    output_path: path to the output folder where all the new cropped matrices </span>
<span class="sd">                 will be stored as npy arrays named after the central gene.</span>
<span class="sd">    shift: number of basepairs before and after the TSS of the central gene that we</span>
<span class="sd">           want to keep in the sample</span>
<span class="sd">    imputation_value: value that will be used to replace missing values. </span>
<span class="sd">                      Default to 1e-5, minimum in Hsieh et al. is ca. 1,98e-5.</span>

<span class="sd">    Returns:</span>
<span class="sd">        The function creates .npy arrays for the contact matrices matching the dimensions</span>
<span class="sd">        of the chromatin mark profiles.</span>

<span class="sd">    Notes:</span>
<span class="sd">        Signal ranges from 1e-5 to 1e0. Nans after balancing and 0 count bins are set to be 1e-10 </span>
<span class="sd">        (same number of orders of magnitude below the first actual contact bin detected). </span>
<span class="sd">        The signal is then upscaled 5 orders of magnitude and log10 transformed, leading to matrices with values </span>
<span class="sd">        from 0.5</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>

    <span class="n">split_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="n">split_list</span><span class="p">:</span>
        <span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="n">savepath</span><span class="o">/</span> <span class="s2">&quot;microC_rotated&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Read the file with gene coordinates</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">ID</span><span class="p">,</span><span class="n">chrom</span><span class="p">,</span><span class="n">Start</span><span class="p">,</span><span class="n">End</span><span class="p">,</span><span class="n">Strand</span><span class="p">,</span><span class="n">Name</span><span class="p">,</span><span class="n">CRE_type</span><span class="p">)</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>

        <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
        <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

        <span class="n">window_start</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">-</span> <span class="n">shift</span>
        <span class="n">window_end</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">+</span> <span class="n">shift</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">window_start</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">window_end</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">imputation_value</span><span class="p">)</span> <span class="c1"># Nans to imputation value</span>
            <span class="n">c_matrix</span><span class="p">[</span><span class="n">c_matrix</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputation_value</span> <span class="c1"># Zeros to imputation value</span>
            <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">)</span> <span class="c1"># Matrix is log10 transformed</span>
            
            <span class="k">if</span> <span class="n">c_matrix</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">626</span><span class="p">,</span><span class="mi">626</span><span class="p">):</span>
                
                <span class="n">anti_c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fliplr</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flipud</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">))</span>

                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">c_matrix</span> <span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">anti_c_matrix</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ID: </span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">, Original Shape: </span><span class="si">{</span><span class="n">c_matrix</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># Rotate and crop</span>
            <span class="n">rotated_array</span> <span class="o">=</span> <span class="n">rotate</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="n">reshape</span><span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">crop_pixel</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">rotated_array</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">final_array</span> <span class="o">=</span> <span class="n">rotated_array</span><span class="p">[</span><span class="n">crop_pixel</span><span class="p">:</span><span class="o">-</span><span class="n">crop_pixel</span><span class="p">,:]</span>
            <span class="n">cut_pixel</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">final_array</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">final_array</span> <span class="o">=</span> <span class="n">final_array</span><span class="p">[:</span><span class="n">cut_pixel</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">final_array</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">129</span><span class="p">,</span><span class="mi">626</span><span class="p">):</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC_rotated/&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">/&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">final_array</span><span class="p">,</span> <span class="n">allow_pickle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC_rotated/&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">/&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">fliplr</span><span class="p">(</span><span class="n">final_array</span><span class="p">),</span> <span class="n">allow_pickle</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1">#flip over sequence axis</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ID: </span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">, Rotated Shape: </span><span class="si">{</span><span class="n">final_array</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="k">except</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> raised an error&quot;</span><span class="p">)</span>


<span class="c1">############################### Script #################################################</span>

<span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;../&quot;</span> <span class="c1"># &quot;../data/raw_data/&quot;</span>
<span class="n">microc_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s2">&quot;/GEO_files/GSE130275_mESC_WT_combined_2.6B.mcool::/13&quot;</span> <span class="c1"># Until 16</span>
<span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations.tsv&quot;</span> <span class="c1"># Gene coordinates</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> 
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Start logger</span>
<span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;MicroC_data_obtention.log&quot;</span>
<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">path</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>  


<span class="n">mcool_summary</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>           
<span class="n">min_dataset</span><span class="p">,</span> <span class="n">max_dataset</span> <span class="o">=</span> <span class="n">find_dataset_min_max</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Log 10 (min)=&quot;</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">min_dataset</span><span class="p">),</span> <span class="s2">&quot;Max:&quot;</span><span class="p">,</span> <span class="n">max_dataset</span><span class="p">)</span>

<span class="n">create_Micro_C_arrays</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span> <span class="p">,</span> <span class="n">imputation_value</span> <span class="o">=</span> <span class="n">min_dataset</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_microC_arrays.py

<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">cooler</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">rotate</span>

<span class="c1">########### Functions #####################################</span>
<span class="k">def</span> <span class="nf">find_dataset_min_max</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">))</span>

<span class="k">def</span> <span class="nf">plot_histogram_dataset</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span> <span class="nf">mcool_summary</span><span class="p">(</span><span class="n">microc_path</span><span class="p">):</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>

    <span class="c1"># Inspecting what is inside the cooler object:</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">info</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">chromnames</span>
    <span class="n">extent</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">extent</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Metadata:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Extent:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">process_microC_data</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">imputation_value</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">ID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">Start</span><span class="p">,</span> <span class="n">End</span><span class="p">,</span> <span class="n">Strand</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">CRE_type</span> <span class="o">=</span> <span class="n">row</span>
    <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
    <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>
    <span class="n">window_start</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">-</span> <span class="n">shift</span>
    <span class="n">window_end</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">+</span> <span class="n">shift</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Fetch and process the matrix</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">window_start</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">window_end</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">imputation_value</span><span class="p">)</span>
        <span class="n">c_matrix</span><span class="p">[</span><span class="n">c_matrix</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputation_value</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">c_matrix</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">626</span><span class="p">,</span> <span class="mi">626</span><span class="p">):</span>
            <span class="c1"># Save direct contact matrices</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">c_matrix</span><span class="p">)</span>
            <span class="n">anti_c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fliplr</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flipud</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">))</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">anti_c_matrix</span><span class="p">)</span>

            <span class="c1"># Rotate, crop, and save rotated arrays</span>
            <span class="n">rotated_array</span> <span class="o">=</span> <span class="n">rotate</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="mi">45</span><span class="p">,</span> <span class="n">reshape</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">crop_pixel</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">rotated_array</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">final_array</span> <span class="o">=</span> <span class="n">rotated_array</span><span class="p">[</span><span class="n">crop_pixel</span><span class="p">:</span><span class="o">-</span><span class="n">crop_pixel</span><span class="p">,</span> <span class="p">:]</span>
            <span class="n">cut_pixel</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">final_array</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">final_rotated_array</span> <span class="o">=</span> <span class="n">final_array</span><span class="p">[:</span><span class="n">cut_pixel</span><span class="p">]</span>

            <span class="c1"># Ensuring rotated array has the expected shape before saving</span>
            <span class="k">if</span> <span class="n">final_rotated_array</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">129</span><span class="p">,</span> <span class="mi">626</span><span class="p">):</span>  <span class="c1"># Adjust the expected shape based on your specific requirements</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC_rotated&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">final_rotated_array</span><span class="p">)</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC_rotated&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">fliplr</span><span class="p">(</span><span class="n">final_rotated_array</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ID: </span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">, Rotated Shape: </span><span class="si">{</span><span class="n">final_rotated_array</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2"> does not meet expected criteria.&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> raised an error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">create_Micro_C_arrays_parallel</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="mi">500000</span><span class="p">,</span> <span class="n">imputation_value</span><span class="o">=</span><span class="mf">1e-14</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parallel version for creating Micro C arrays.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="n">args</span> <span class="o">=</span> <span class="p">[(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">imputation_value</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>

    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_microC_data</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="c1"># Define paths</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">microc_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files/GSE130275_mESC_WT_combined_2.6B.mcool::/13&quot;</span> 
    <span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations/Final_gene_annotations.tsv&quot;</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span>
    <span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Setup logging</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;MicroC_data_obtention.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">savepath</span> <span class="o">/</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># Create directories for output</span>
    <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]:</span>
        <span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC&quot;</span> <span class="o">/</span> <span class="n">split</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;microC_rotated&quot;</span> <span class="o">/</span> <span class="n">split</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Run the processing in parallel</span>
    <span class="n">create_Micro_C_arrays_parallel</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">microc_path</span><span class="p">),</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">imputation_value</span><span class="o">=</span><span class="mf">1e-14</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>python<span class="w"> </span>create_microC_arrays.py
</pre></div>
</div>
</div>
</div>
</section>
<section id="epigenomic-data-acquisition">
<h2>Epigenomic Data acquisition:<a class="headerlink" href="#epigenomic-data-acquisition" title="Link to this heading">#</a></h2>
<p><strong>Genomic tracks</strong></p>
<p>The used genomic tracks can be obtained from NCBI GEO under accession numbers GSE146328 and GSE186349, which refer to papers:</p>
<ul class="simple">
<li><p>Narita, T., Higashijima, Y., Kilic, S. et al. Acetylation of histone H2B marks active enhancers and predicts CBP/p300 target genes. Nat Genet 55, 679–692 (2023). https://doi.org/10.1038/s41588-023-01348-4</p></li>
<li><p>Takeo Narita, Shinsuke Ito, Yoshiki Higashijima, Wai Kit Chu, Katrin Neumann, Jonas Walter, Shankha Satpathy, Tim Liebner, William B. Hamilton, Elina Maskey, Gabriela Prus, Marika Shibata, Vytautas Iesmantavicius, Joshua M. Brickman, Konstantinos Anastassiadis, Haruhiko Koseki, Chunaram Choudhary,
Enhancers are activated by p300/CBP activity-dependent PIC assembly, RNAPII recruitment, and pause release,
Molecular Cell,
Volume 81, Issue 10,
2021,
Pages 2166-2182.e6,
ISSN 1097-2765,
https://doi.org/10.1016/j.molcel.2021.03.008.
(https://www.sciencedirect.com/science/article/pii/S1097276521001763)</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;EU_Seq_Ctrl.bw&quot;</span><span class="p">:</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE146326&amp;format=file&amp;file=GSE146326</span><span class="si">%5F</span><span class="s2">EUSeq</span><span class="si">%5F</span><span class="s2">mESC</span><span class="si">%5F</span><span class="s2">A485</span><span class="si">%5F</span><span class="s2">Timecourse</span><span class="si">%5F</span><span class="s2">0min</span><span class="si">%2E</span><span class="s2">n2</span><span class="si">%2E</span><span class="s2">smooth</span><span class="si">%2E</span><span class="s2">bw&quot;</span><span class="p">,</span> <span class="c1">#p300 paper</span>
<span class="c1">#&quot;EU_Seq_Treated.bw&quot;:&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE146326&amp;format=file&amp;file=GSE146326%5FEUSeq%5FmESC%5FA485%5FTimecourse%5F60min%2En2%2Esmooth%2Ebw&quot;, #p300</span>
<span class="s2">&quot;ATAC_Seq.bw&quot;</span><span class="p">:</span> <span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE146325&amp;format=file&amp;file=GSE146325</span><span class="si">%5F</span><span class="s2">ATACSeq</span><span class="si">%2E</span><span class="s2">ESC</span><span class="si">%5F</span><span class="s2">Ctrl</span><span class="si">%2E</span><span class="s2">bw&quot;</span><span class="p">,</span> <span class="c1">#p300</span>
<span class="s2">&quot;H3K4me3.bw&quot;</span><span class="p">:</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE186349&amp;format=file&amp;file=GSE186349</span><span class="si">%5F</span><span class="s2">ESC</span><span class="si">%5F</span><span class="s2">Ctrl</span><span class="si">%5F</span><span class="s2">TC0</span><span class="si">%5F</span><span class="s2">H3K4me3</span><span class="si">%5F</span><span class="s2">CC10</span><span class="si">%2E</span><span class="s2">n2</span><span class="si">%2E</span><span class="s2">smooth</span><span class="si">%2E</span><span class="s2">bw&quot;</span><span class="p">,</span><span class="c1">#H2B paper</span>
<span class="c1">#&quot;H2BK20ac.bw&quot;: &quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE186349&amp;format=file&amp;file=GSE186349%5FESC%5FCtrl%5FTC0%5FH2BK20ac%2Eab177430%5FCC8%2ECC15%2En2%2Esmooth%2Ebw&quot;, #H2B paper</span>
<span class="s2">&quot;H3K27ac.bw&quot;</span><span class="p">:</span><span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE186349&amp;format=file&amp;file=GSE186349</span><span class="si">%5F</span><span class="s2">ESC</span><span class="si">%5F</span><span class="s2">Ctrl</span><span class="si">%5F</span><span class="s2">TC0</span><span class="si">%5F</span><span class="s2">H3K27ac</span><span class="si">%2E</span><span class="s2">ab4729</span><span class="si">%5F</span><span class="s2">CC1</span><span class="si">%2E</span><span class="s2">CC11</span><span class="si">%2E</span><span class="s2">n2</span><span class="si">%2E</span><span class="s2">smooth</span><span class="si">%2E</span><span class="s2">bw&quot;</span><span class="p">,</span>
<span class="s2">&quot;H3K27me3.bw&quot;</span><span class="p">:</span> <span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE146324&amp;format=file&amp;file=GSE146324</span><span class="si">%5F</span><span class="s2">ChIP</span><span class="si">%2E</span><span class="s2">ESC</span><span class="si">%5F</span><span class="s2">Ctrl</span><span class="si">%5F</span><span class="s2">H3K27me3</span><span class="si">%5F</span><span class="s2">TC0</span><span class="si">%2E</span><span class="s2">n2</span><span class="si">%2E</span><span class="s2">smooth</span><span class="si">%2E</span><span class="s2">bw&quot;</span><span class="p">}</span> <span class="c1">#p300 paper</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../GEO_files/&quot;</span><span class="p">)</span>

<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Link to this heading">#</a></h2>
<p>From the bigwig files from NCBI’s GEO, we will obtain:</p>
<ul class="simple">
<li><p>Input numpy arrays of shape (<span class="math notranslate nohighlight">\(n_{tracks}\)</span>,<span class="math notranslate nohighlight">\(n_{bins}\)</span>) = (4,10001) with enrichments of ATAC, H3K4me3, H2BK20ac and H3K27me3 (in rows) at 100bp resolution (where bins are in columns).</p></li>
<li><p>Target csv files containing target EU-seq profiles matching each ID.npy input array cropped to the desired distance. Targets are given in reads unless otherwise stated.</p></li>
</ul>
<p>We will create samples for forward strand and reverse strand by flipping the profiles and targets. They will ve called ID_forward.npy and ID_rev.npy.</p>
<blockquote>
<div><p>⏰ Note: This step takes quite some time. To speed up computations, we allowed multiprocessing to parallelize the creation of samples. We run it using a SLURM-based cluster, which allowed us to specify the number of CPUs:</p>
<p><code class="docutils literal notranslate"><span class="pre">srun</span> <span class="pre">--cpus-per-task=128</span> <span class="pre">--</span> <span class="pre">python</span> <span class="pre">create_landscape_arrays_and_targets.py</span> </code></p>
<p>This took more or less a couple of hours.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_landscape_arrays_and_targets.py

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">gaussian_filter1d</span><span class="p">,</span> <span class="n">rotate</span>


<span class="c1"># Define your existing functions for creating input arrays, target arrays, and visualization here</span>
<span class="c1"># create_input_array, create_target_array, visualize_input_array, visualize_target_array</span>

<span class="n">track_dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
            <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
            <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Enhancer&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;b&quot;</span><span class="p">},</span>
            <span class="mi">3</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>

<span class="k">def</span> <span class="nf">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                       <span class="n">input_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">n_input_tracks</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
                       <span class="n">n_input_bins</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                       <span class="n">bw_input_track_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
                       <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                       <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Inputs will be centered at the TSS of the gene:</span>
<span class="sd">      o _pos sample will be oriented left to right and neg will be oriented right to left.</span>
<span class="sd">      o _neg sample &quot;&quot; right to left</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">n_input_tracks</span><span class="p">,</span><span class="n">n_input_bins</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">bw_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bw_input_track_list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span><span class="n">TSS</span><span class="o">-</span><span class="n">input_shift</span><span class="p">,</span><span class="n">TSS</span><span class="o">+</span><span class="n">input_shift</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span><span class="n">nBins</span><span class="o">=</span><span class="n">n_input_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span><span class="kc">None</span><span class="p">)</span> <span class="c1"># ReLU</span>
            <span class="n">input_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">stats</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> input landscape coordinates are out of bounds.&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">input_sample</span>


<span class="k">def</span> <span class="nf">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                        <span class="n">output_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                        <span class="n">n_output_bins</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                        <span class="n">bw_target_track_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
                        <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                        <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                        <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                        <span class="n">sigma</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                        <span class="n">binsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Targets are centered at the TSS, and obtained at a 20 bp resolution, smoothed and downsized to 1kbp resolution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">bw_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bw_target_track_list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span><span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span><span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span><span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span><span class="kc">None</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">gaussian_filter1d</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
            <span class="n">target_cond</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">stats</span><span class="p">)</span>

            <span class="c1"># Averaging over a number of bins</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">binsize</span><span class="p">):</span>
                <span class="n">target_cond</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">binsize</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="o">-</span><span class="n">i</span><span class="p">)</span>
            <span class="n">target_cond</span> <span class="o">=</span> <span class="n">target_cond</span><span class="p">[::</span><span class="n">binsize</span><span class="p">]</span>
            <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_pos</span><span class="p">,</span><span class="n">target_cond</span><span class="p">])</span>
            <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_neg</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">target_cond</span><span class="p">)])</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> target coordinates are out of bounds.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">target_sample_pos</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">target_sample_neg</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">visualize_input_array</span><span class="p">(</span><span class="n">a</span><span class="p">,</span>
                     <span class="n">cropped_bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4400</span><span class="p">,</span>
                     <span class="n">scaling_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.</span><span class="p">,</span>
                     <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to visualize an input numpy array.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">track_dict</span><span class="p">),</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">8</span><span class="p">),</span> <span class="n">sharex</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Plot the scaled arrays</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="c1">#axs[0][0].set_title(file.name[:-4])</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#axs[j][0].set_yticks([]) #([0,int(np.max(line))], [0,int(np.max(line))], fontsize=16)</span>
        <span class="c1">#axs[j][0].set_ylim(0,350)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">16</span><span class="p">)</span>
        <span class="c1">#axs[-1][0].set_xticks([])#([-5000,0,5000], [-500, 0, 500], fontsize= 16)</span>
        <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5000</span><span class="p">))</span>

        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">],</span><span class="n">line</span><span class="p">[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">],</span><span class="n">line</span><span class="p">[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.6</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>
        
<span class="k">def</span> <span class="nf">visualize_target_array</span><span class="p">(</span><span class="n">a</span><span class="p">,</span>
                           <span class="n">scaling_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to visualize target profiles.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ctrl</span> <span class="o">=</span> <span class="n">a</span><span class="p">[:</span><span class="nb">len</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">]</span>
    <span class="c1">#treated = a[len(a)//2:]</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">ctrl</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;ctrl&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">ctrl</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;silver&quot;</span> <span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1">#plt.plot(np.arange(-len(treated)//2,len(treated)//2),treated*scaling_factor,label=&quot;treated&quot;, color=&quot;royalblue&quot;, lw=.2)</span>
    <span class="c1">#plt.fill_between(np.arange(-len(treated)//2,len(treated)//2),treated*scaling_factor, label=&quot;treated&quot;, color=&quot;royalblue&quot;,alpha=1)</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span> <span class="nf">process_gene_data</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">,</span> <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">split_list</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">PLOT_IDS</span><span class="p">,</span> <span class="n">target_files</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
    <span class="n">ID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">Start</span><span class="p">,</span> <span class="n">End</span><span class="p">,</span> <span class="n">Strand</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">CRE_type</span> <span class="o">=</span> <span class="n">row</span>
    <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
    <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">)</span>
    <span class="n">target_sample_pos</span><span class="p">,</span> <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">input_sample</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">INPUT_SHAPE</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">)</span> <span class="o">==</span> <span class="n">OUTPUT_LENGTH</span><span class="p">):</span>
        <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">input_sample</span><span class="p">)</span>
        <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
        
        <span class="n">target_file_path</span> <span class="o">=</span> <span class="n">target_files</span><span class="p">[</span><span class="n">split</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">target_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_pos</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_neg</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> : Input or target did not match the expected shape.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">PLOT_IDS</span><span class="p">:</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">input_sample</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward_target.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev_target.png&quot;</span><span class="p">)</span>

<span class="c1">############################ Script ###################</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>

    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files&quot;</span>
    <span class="n">figure_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;figures&quot;</span>
    <span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span>
    <span class="n">target_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">split_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>
    <span class="n">output_dirs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="n">split_list</span><span class="p">]</span>

    <span class="n">PLOT_IDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000051951.5&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSMUSG00000025902&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSMUSG00000033845.13&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSMUSG00000025903.14&quot;</span><span class="p">]</span>

    <span class="c1"># Start logger</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;/Landscape_data_obtention.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>  

    <span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations.tsv&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">input_shift</span> <span class="o">=</span> <span class="mi">500050</span> <span class="c1"># In basepairs, from TSS to one side + central bin (100bp res)</span>
    <span class="n">output_shift</span> <span class="o">=</span> <span class="mi">200010</span> <span class="c1"># Output in 20 bp resolution (to be smoothed and downsampled)</span>
    <span class="n">n_input_bins</span> <span class="o">=</span> <span class="mi">10001</span>
    <span class="n">n_input_tracks</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">n_output_bins</span> <span class="o">=</span> <span class="mi">20001</span>
    <span class="n">sigma</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">binsize</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">INPUT_SHAPE</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">10001</span><span class="p">)</span>
    <span class="n">OUTPUT_LENGTH</span> <span class="o">=</span> <span class="mi">401</span> <span class="c1">#802</span>
    <span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
    <span class="n">bw_input_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC_Seq.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3.bw&quot;</span><span class="p">]</span>
    <span class="n">bw_target_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;EU_Seq_Ctrl.bw&quot;</span><span class="p">]</span> <span class="c1"># &quot;EU_Seq_Treated.bw&quot;</span>

    <span class="n">target_files</span> <span class="o">=</span> <span class="p">{</span><span class="n">split</span><span class="p">:</span> <span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_targets.csv&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]}</span>
    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">target_files</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ID,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">process_args</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">,</span> <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">],</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">PLOT_IDS</span><span class="p">,</span> <span class="n">target_files</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">process_gene_data</span><span class="p">,</span> <span class="p">[(</span><span class="n">process_args</span> <span class="o">+</span> <span class="p">(</span><span class="n">row</span><span class="p">,))</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overwriting create_landscape_arrays_and_targets.py
</pre></div>
</div>
</div>
</div>
<p>The created python file can be run from a jupyter notebook cell as:</p>
<blockquote>
<div><p><code class="docutils literal notranslate"><span class="pre">!</span> <span class="pre">python</span> <span class="pre">create_landscape_arrays_and_targets.py</span> </code></p>
</div></blockquote>
<p><strong>Create report of missing samples</strong></p>
<p>Not all samples could be created in all data modalities given boundary issues (genes close to the end of the chromosome). Hence, we will create a report of missing samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> keep_only_sample_intersection.py

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># Define base paths for inputs and targets</span>
<span class="n">base_input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs&quot;</span><span class="p">)</span>
<span class="n">base_target_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets&quot;</span><span class="p">)</span>

<span class="c1"># Modalities and splits</span>
<span class="n">data_modalities</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;landscape_arrays&#39;</span><span class="p">,</span> <span class="s1">&#39;microC&#39;</span><span class="p">,</span> <span class="s1">&#39;microC_rotated&#39;</span><span class="p">]</span>
<span class="n">splits</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;training&#39;</span><span class="p">,</span> <span class="s1">&#39;test&#39;</span><span class="p">]</span>

<span class="c1"># Initialize a list to collect rows for the new DataFrame</span>
<span class="n">rows_list</span> <span class="o">=</span> <span class="p">[]</span>

<span class="c1"># Read target samples and store available samples for each modality</span>
<span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="n">splits</span><span class="p">:</span>
    <span class="n">targets_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">base_target_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_targets.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">target_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">targets_df</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">modality</span> <span class="ow">in</span> <span class="n">data_modalities</span><span class="p">:</span>
        <span class="n">modality_path</span> <span class="o">=</span> <span class="n">base_input_path</span> <span class="o">/</span> <span class="n">modality</span> <span class="o">/</span> <span class="n">split</span>
        <span class="n">modality_samples</span> <span class="o">=</span> <span class="p">{</span><span class="n">file</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">modality_path</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.npy&#39;</span><span class="p">)}</span>

        <span class="c1"># Find missing samples</span>
        <span class="n">missing_samples</span> <span class="o">=</span> <span class="n">target_samples</span> <span class="o">-</span> <span class="n">modality_samples</span>
        <span class="n">extra_samples</span> <span class="o">=</span> <span class="n">modality_samples</span> <span class="o">-</span> <span class="n">target_samples</span>

        <span class="c1"># Collect missing samples</span>
        <span class="n">rows_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([{</span><span class="s1">&#39;Split&#39;</span><span class="p">:</span> <span class="n">split</span><span class="p">,</span> <span class="s1">&#39;Modality&#39;</span><span class="p">:</span> <span class="n">modality</span><span class="p">,</span> <span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="n">sample</span><span class="p">}</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">missing_samples</span><span class="p">])</span>
        <span class="c1"># Collect extra samples (in modalities but not in targets)</span>
        <span class="n">rows_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([{</span><span class="s1">&#39;Split&#39;</span><span class="p">:</span> <span class="n">split</span><span class="p">,</span> <span class="s1">&#39;Modality&#39;</span><span class="p">:</span> <span class="s1">&#39;targets&#39;</span><span class="p">,</span> <span class="s1">&#39;Sample&#39;</span><span class="p">:</span> <span class="n">sample</span><span class="p">}</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">extra_samples</span><span class="p">])</span>

<span class="c1"># Create a DataFrame from the collected rows</span>
<span class="n">missing_samples_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows_list</span><span class="p">)</span>

<span class="c1"># Save the DataFrame as a CSV</span>
<span class="n">missing_samples_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/missing_samples_report.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We will now remove all missing samples from the rest of the modalities, including targets.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> remove_missing_samples.py

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># Load the report of missing samples</span>
<span class="n">missing_samples_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/missing_samples_report.csv&quot;</span><span class="p">)</span>

<span class="c1"># Base paths for inputs and targets</span>
<span class="n">base_input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs&quot;</span><span class="p">)</span>
<span class="n">base_target_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets&quot;</span><span class="p">)</span>

<span class="c1"># Modalities and splits</span>
<span class="n">data_modalities</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;landscape_arrays&#39;</span><span class="p">,</span> <span class="s1">&#39;microC&#39;</span><span class="p">,</span> <span class="s1">&#39;microC_rotated&#39;</span><span class="p">,</span> <span class="s1">&#39;targets&#39;</span><span class="p">]</span>  <span class="c1"># Include &#39;targets&#39; as a modality for uniform processing</span>
<span class="n">splits</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;training&#39;</span><span class="p">,</span> <span class="s1">&#39;test&#39;</span><span class="p">]</span>

<span class="c1"># Function to remove a sample from a specific modality and split</span>
<span class="k">def</span> <span class="nf">remove_sample_from_modality</span><span class="p">(</span><span class="n">split</span><span class="p">,</span> <span class="n">modality</span><span class="p">,</span> <span class="n">sample</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">modality</span> <span class="o">==</span> <span class="s1">&#39;targets&#39;</span><span class="p">:</span>
        <span class="n">target_file_path</span> <span class="o">=</span> <span class="n">base_target_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_targets.csv&quot;</span>
        <span class="n">targets_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">target_file_path</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">targets_df</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
            <span class="n">targets_df</span> <span class="o">=</span> <span class="n">targets_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
            <span class="n">targets_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">target_file_path</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">modality_path</span> <span class="o">=</span> <span class="n">base_input_path</span> <span class="o">/</span> <span class="n">modality</span> <span class="o">/</span> <span class="n">split</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">modality_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.npy&quot;</span>  <span class="c1"># Assuming .npy extension</span>
        <span class="k">if</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">file_path</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

<span class="c1"># Iterate over the missing samples DataFrame</span>
<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">missing_samples_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">split</span><span class="p">,</span> <span class="n">missing_modality</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Split&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Modality&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Sample&#39;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">modality</span> <span class="ow">in</span> <span class="n">data_modalities</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">modality</span> <span class="o">!=</span> <span class="n">missing_modality</span><span class="p">:</span>  <span class="c1"># Remove the sample from all modalities except where it&#39;s missing</span>
            <span class="n">remove_sample_from_modality</span><span class="p">(</span><span class="n">split</span><span class="p">,</span> <span class="n">modality</span><span class="p">,</span> <span class="n">sample</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="enhancer-perturbations">
<h2>2. Enhancer perturbations<a class="headerlink" href="#enhancer-perturbations" title="Link to this heading">#</a></h2>
<p>The goal of this section is to create slightly modified inputs, emulating the effect of an epigenetic perturbation at a given locus. We tried two perturbational scenarios:</p>
<ul class="simple">
<li><p>P1: Only setting to zero the enhancer mark at enhancer loci (ELS) that fall within the boundaries of the predicted region.</p></li>
<li><p>P2: Setting all marks to zero except the silencing mark H3K27me3.</p></li>
</ul>
<p><strong>Create gene and enhancer coordinate table</strong></p>
<p>We will create a table with all enhancers and genes that appear in a sample given the target boundaries from the TSS.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>

<span class="k">def</span> <span class="nf">calculate_tss</span><span class="p">(</span><span class="n">genes_df</span><span class="p">):</span>
    <span class="c1"># Calculate TSS for genes based on strand information</span>
    <span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;TSS&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">genes_df</span>

<span class="k">def</span> <span class="nf">process_genes_chunk</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="n">genes_chunk</span><span class="p">,</span> <span class="n">genes_df</span><span class="p">,</span> <span class="n">enhancers_df</span><span class="p">,</span> <span class="n">boundary</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">results_list</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># Initialize an empty list to store intermediate results</span>

    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_chunk</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">tss</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;TSS&#39;</span><span class="p">]</span>
        <span class="n">chromosome</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span>

        <span class="c1"># Adjust selection criteria for gene entities: overlap at some point with the prediction window</span>
        <span class="n">gene_entities</span> <span class="o">=</span> <span class="n">genes_df</span><span class="p">[(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span> <span class="o">&amp;</span> 
                         <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">tss</span> <span class="o">+</span> <span class="n">boundary</span><span class="p">)</span> <span class="o">&amp;</span> 
                         <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">tss</span> <span class="o">-</span> <span class="n">boundary</span><span class="p">)]</span>

        <span class="c1"># Process gene entities</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">entity</span> <span class="ow">in</span> <span class="n">gene_entities</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">entity_start</span> <span class="o">=</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tss</span>
            <span class="n">entity_end</span> <span class="o">=</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tss</span>
            <span class="n">results_list</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;Ref_gene&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> 
                <span class="s1">&#39;Ref_gene_chromosome&#39;</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span> 
                <span class="s1">&#39;Ref_gene_strand&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">],</span>
                <span class="s1">&#39;Entity_ID&#39;</span><span class="p">:</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> 
                <span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">:</span> <span class="n">entity_start</span><span class="p">,</span> 
                <span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">:</span> <span class="n">entity_end</span>
            <span class="p">})</span>
        
        <span class="c1"># Adjust selection criteria for enhancer entities: fully contained inside prediction window</span>
        <span class="n">enhancer_entities</span> <span class="o">=</span> <span class="n">enhancers_df</span><span class="p">[(</span><span class="n">enhancers_df</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chromosome</span><span class="p">)</span> <span class="o">&amp;</span> 
                                         <span class="p">(</span><span class="n">enhancers_df</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">tss</span> <span class="o">+</span> <span class="n">boundary</span><span class="p">)</span> <span class="o">&amp;</span> 
                                         <span class="p">(</span><span class="n">enhancers_df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">tss</span> <span class="o">-</span> <span class="n">boundary</span><span class="p">)]</span>
        <span class="c1"># Process enhancer entities</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">entity</span> <span class="ow">in</span> <span class="n">enhancer_entities</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">entity_start</span> <span class="o">=</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tss</span>
            <span class="n">entity_end</span> <span class="o">=</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">tss</span>
            <span class="n">results_list</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;Ref_gene&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> 
                <span class="s1">&#39;Ref_gene_chromosome&#39;</span><span class="p">:</span> <span class="n">chromosome</span><span class="p">,</span> 
                <span class="s1">&#39;Ref_gene_strand&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">],</span>
                <span class="s1">&#39;Entity_ID&#39;</span><span class="p">:</span> <span class="n">entity</span><span class="p">[</span><span class="s1">&#39;cCRE_accession&#39;</span><span class="p">],</span> 
                <span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">:</span> <span class="n">entity_start</span><span class="p">,</span> 
                <span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">:</span> <span class="n">entity_end</span>
            <span class="p">})</span>

    <span class="c1"># Convert the results list to a DataFrame</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results_list</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_gene_annotations.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">enhancers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_Enhancer_annotation.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">boundary</span> <span class="o">=</span> <span class="mi">200010</span> 

    <span class="n">genes_df</span> <span class="o">=</span> <span class="n">calculate_tss</span><span class="p">(</span><span class="n">genes_df</span><span class="p">)</span>


    <span class="n">n_processes</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span>
    <span class="n">genes_chunks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array_split</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">n_processes</span><span class="p">)</span>
    <span class="n">args</span> <span class="o">=</span> <span class="p">[(</span><span class="n">chunk</span><span class="p">,</span> <span class="n">genes_df</span><span class="p">,</span> <span class="n">enhancers_df</span><span class="p">,</span> <span class="n">boundary</span><span class="p">)</span> <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">genes_chunks</span><span class="p">]</span>

    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_processes</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_genes_chunk</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

    <span class="n">result_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">result_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;../annotations/gene_enhancer_relationships_corrected.tsv&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/numpy/core/fromnumeric.py:59: FutureWarning: &#39;DataFrame.swapaxes&#39; is deprecated and will be removed in a future version. Please use &#39;DataFrame.transpose&#39; instead.
  return bound(*args, **kwds)
</pre></div>
</div>
</div>
</div>
<p><strong>Create perturbed inputs and corresponding targets.</strong></p>
<p>The goal of this part of the pipeline will be to assess what genes are affected when silencing single enhancers.</p>
<ul class="simple">
<li><p>We will only perturb/silence enhancers that were originally active, i.e. for which the average read count in the enhancer region is above a certain threshold for the enhancer activity mark H3K27ac. The threshold was set to an average signal of 10 reads across the range.</p></li>
<li><p>Previous H3K27me3 levels will be kept.</p></li>
</ul>
<blockquote>
<div><p><em>Note: EIR requires to have matching targets even when predicting new samples and therefore we will create dummy target tables filled with zeros.</em></p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_perturbed_arrays_and_targets.py 

<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">scipy.integrate</span> <span class="kn">import</span> <span class="n">simpson</span>
<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>

<span class="k">def</span> <span class="nf">get_track_index</span><span class="p">(</span><span class="n">track_name</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps track names to their respective index in the numpy array.&quot;&quot;&quot;</span>
    <span class="n">track_order</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;ATAC&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;H3K4me3&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;H3K27ac&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;H3K27me3&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}</span>
    <span class="k">return</span> <span class="n">track_order</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">track_name</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Returns -1 if track name not found</span>

<span class="k">def</span> <span class="nf">modify_array</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Modifies numpy arrays for a single reference gene by perturbing specified enhancers.</span>
<span class="sd">    Returns a list of kept enhancers based on a threshold in the H3K27ac channel.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ref_gene</span><span class="p">,</span> <span class="n">enhancers</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">modification_dict</span><span class="p">,</span> <span class="n">input_arrays_path</span><span class="p">,</span> <span class="n">perturbed_arrays_path</span><span class="p">,</span> <span class="n">threshold</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">kept_enhancers</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># To keep track of enhancers that meet the threshold criteria</span>

    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">enhancer</span> <span class="ow">in</span> <span class="n">enhancers</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">array_path</span> <span class="o">=</span> <span class="n">input_arrays_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_gene</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span>
            <span class="n">gene_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">array_path</span><span class="p">)</span>

            <span class="n">central_idx</span> <span class="o">=</span> <span class="n">gene_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">//</span> <span class="mi">2</span>
            <span class="n">start_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">enhancer</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">//</span> <span class="n">resolution</span><span class="p">)</span>
            <span class="n">end_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">enhancer</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">]</span> <span class="o">//</span> <span class="n">resolution</span><span class="p">)</span>

            <span class="n">h3k27ac_idx</span> <span class="o">=</span> <span class="n">get_track_index</span><span class="p">(</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">)</span>
            <span class="n">enhancer_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">gene_array</span><span class="p">[</span><span class="n">h3k27ac_idx</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">):</span><span class="nb">min</span><span class="p">(</span><span class="n">end_idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>

            <span class="k">if</span> <span class="n">enhancer_mean</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">track_name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">modification_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">track_idx</span> <span class="o">=</span> <span class="n">get_track_index</span><span class="p">(</span><span class="n">track_name</span><span class="p">)</span>
                    <span class="n">gene_array</span><span class="p">[</span><span class="n">track_idx</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">):</span><span class="nb">min</span><span class="p">(</span><span class="n">end_idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span> <span class="o">=</span> <span class="n">value</span>

                <span class="n">new_array_name</span> <span class="o">=</span> <span class="n">perturbed_arrays_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_gene</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">enhancer</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span>
                <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">new_array_name</span><span class="p">,</span> <span class="n">gene_array</span><span class="p">)</span>
                <span class="n">kept_enhancers</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">ref_gene</span><span class="p">,</span> <span class="n">enhancer</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]))</span>

        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Array file for </span><span class="si">{</span><span class="n">ref_gene</span><span class="si">}</span><span class="s2"> not found.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">kept_enhancers</span>

<span class="k">def</span> <span class="nf">parallel_modify_arrays</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">input_arrays_path</span><span class="p">,</span> <span class="n">perturbed_arrays_path</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">,</span> <span class="n">modification_dict</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Processes enhancer perturbations in parallel and generates a target file for kept enhancers.&quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene_chromosome&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">)]</span>
    <span class="n">ref_genes</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="p">[(</span><span class="n">ref_gene</span><span class="p">,</span> 
             <span class="n">filtered_df</span><span class="p">[(</span><span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref_gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="o">~</span><span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ENSMUSG&#39;</span><span class="p">))],</span>
             <span class="n">resolution</span><span class="p">,</span> <span class="n">modification_dict</span><span class="p">,</span> <span class="n">input_arrays_path</span><span class="p">,</span> <span class="n">perturbed_arrays_path</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span> <span class="k">for</span> <span class="n">ref_gene</span> <span class="ow">in</span> <span class="n">ref_genes</span><span class="p">]</span>

    <span class="n">kept_enhancers</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">cpu_count</span><span class="p">())</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">pool</span><span class="o">.</span><span class="n">imap</span><span class="p">(</span><span class="n">modify_array</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
            <span class="n">kept_enhancers</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

    <span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
    <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>
    <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_gene</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">enhancer</span><span class="si">}</span><span class="s2">_forward&quot;</span> <span class="k">for</span> <span class="n">ref_gene</span><span class="p">,</span> <span class="n">enhancer</span> <span class="ow">in</span> <span class="n">kept_enhancers</span><span class="p">]</span>
    <span class="n">perturbed_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">perturbed_targets_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../targets/&#39;</span><span class="p">)</span> <span class="o">/</span> <span class="s1">&#39;perturbed_targets.csv&#39;</span>
    <span class="n">perturbed_targets</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">perturbed_targets_path</span><span class="p">,</span> <span class="n">index_label</span><span class="o">=</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span>

<span class="c1"># Example usage, adjust paths and parameters as needed</span>
<span class="n">input_table_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">input_arrays_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)</span>
<span class="n">perturbed_arrays_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/perturbed_landscape_arrays/test_only_H3K27ac_silenced/&quot;</span><span class="p">)</span>
<span class="n">perturbed_arrays_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># Ensure the directory exists</span>

<span class="n">chromosomes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chr4&#39;</span><span class="p">]</span>
<span class="n">modification_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">}</span> <span class="c1"># If all active marks silenced: {&quot;ATAC&quot;: 0, &quot;H3K4me3&quot;: 0, &quot;H3K27ac&quot;: 0}</span>
<span class="n">resolution</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">threshold</span> <span class="o">=</span> <span class="mi">10</span>  <span class="c1"># Adjust the threshold as needed</span>

<span class="n">parallel_modify_arrays</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">input_arrays_path</span><span class="p">,</span> <span class="n">perturbed_arrays_path</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">,</span> <span class="n">modification_dict</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">threshold</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overwriting create_perturbed_arrays_and_targets.py
</pre></div>
</div>
</div>
</div>
<p>Now we have all the data required to train and test CLASTER!</p>
<p>(We’ll do that in the next python notebook.)</p>
</section>
<section id="benchmark-data-obtention">
<h2>3. Benchmark data obtention<a class="headerlink" href="#benchmark-data-obtention" title="Link to this heading">#</a></h2>
<p><strong>Comparing CLASTER with DNA-sequence based models: HyenaDNA and the Enformer</strong></p>
<p>We will now obtain the genomic sequences corresponding to the regions constituting our samples. These will be used to:</p>
<ul class="simple">
<li><p>Get HyenaDNA-160kb / Enformer pretrained embeddings or</p></li>
<li><p>Fine-tune HyenaDNA backbone with an addded head.</p></li>
</ul>
<p>First we need to install bedtools and samtools, two packages to handle genomic sequences:</p>
<blockquote>
<div><p>Note: it is easier to do it directly on the terminal. Make sure to have the right environment activated.</p>
</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>install<span class="w"> </span>bioconda:bedtools
conda<span class="w"> </span>install<span class="w"> </span>bioconda:samtools
</pre></div>
</div>
<p><strong>Get the genome sequence</strong></p>
<p>We will now obtain the mouse reference genome sequence. The genome sequence M25(GRCm38.p6) was obtained from GENCODE.</p>
<p>https://www.gencodegenes.org/mouse/release_M25.html</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;GRCm38.p6.genome.fa.gz&quot;</span><span class="p">:</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/GRCm38.p6.genome.fa.gz&quot;</span><span class="p">}</span> <span class="c1">#p300 paper</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../GEO_files/&quot;</span><span class="p">)</span>

<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>

<span class="o">!</span><span class="w"> </span>gunzip<span class="w"> </span>../GEO_files/GRCm38.p6.genome.fa.gz
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Successfully downloaded GRCm38.p6.genome.fa.gz to ../GEO_files
</pre></div>
</div>
</div>
</div>
<section id="create-bed-file-with-dna-chunk-boundaries">
<h3>Create bed file with DNA chunk boundaries:<a class="headerlink" href="#create-bed-file-with-dna-chunk-boundaries" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_sequence_boundaries.sh

<span class="c1">#!/bin/bash</span>

<span class="c1"># Script to create BED files based on transcript IDs found in specific CSV files</span>

<span class="c1"># Input file name</span>
<span class="n">input_file</span><span class="o">=</span><span class="s2">&quot;../annotations/Final_gene_annotations.tsv&quot;</span>

<span class="c1"># Additional CSV files</span>
<span class="n">test_targets_csv</span><span class="o">=</span><span class="s2">&quot;../targets/test_targets.csv&quot;</span>
<span class="n">train_targets_csv</span><span class="o">=</span><span class="s2">&quot;../targets/training_targets.csv&quot;</span>

<span class="c1"># Output file names</span>
<span class="n">output_file1</span><span class="o">=</span><span class="s2">&quot;../annotations/training_boundaries.bed&quot;</span>
<span class="n">output_file2</span><span class="o">=</span><span class="s2">&quot;../annotations/test_boundaries.bed&quot;</span>
<span class="n">output_file3</span><span class="o">=</span><span class="s2">&quot;../annotations/validation_boundaries.bed&quot;</span>

<span class="c1"># Variable &#39;shift&#39;</span>
<span class="n">shift</span><span class="o">=</span><span class="mi">80000</span>

<span class="c1"># Read IDs from CSV files into arrays</span>
<span class="n">IFS</span><span class="o">=</span><span class="err">$</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="n">read</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;&#39;</span> <span class="o">-</span><span class="n">r</span> <span class="o">-</span><span class="n">a</span> <span class="n">test_ids</span> <span class="o">&lt;</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">tail</span> <span class="o">-</span><span class="n">n</span> <span class="o">+</span><span class="mi">2</span> <span class="s2">&quot;$test_targets_csv&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;,&#39;</span> <span class="o">-</span><span class="n">f1</span><span class="p">)</span>
<span class="n">IFS</span><span class="o">=</span><span class="err">$</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="n">read</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;&#39;</span> <span class="o">-</span><span class="n">r</span> <span class="o">-</span><span class="n">a</span> <span class="n">train_ids</span> <span class="o">&lt;</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">tail</span> <span class="o">-</span><span class="n">n</span> <span class="o">+</span><span class="mi">2</span> <span class="s2">&quot;$train_targets_csv&quot;</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">d</span> <span class="s1">&#39;,&#39;</span> <span class="o">-</span><span class="n">f1</span><span class="p">)</span>

<span class="c1"># Function to check if an ID is in the arrays with _forward or _rev suffix</span>
<span class="n">id_in_targets</span><span class="p">()</span> <span class="p">{</span>
    <span class="n">local</span> <span class="n">id_forward</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{1}</span><span class="s2">_forward&quot;</span>
    <span class="n">local</span> <span class="n">id_rev</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{1}</span><span class="s2">_rev&quot;</span>
    <span class="k">for</span> <span class="n">target_id</span> <span class="ow">in</span> <span class="s2">&quot;$</span><span class="si">{test_ids[@]}</span><span class="s2">&quot;</span> <span class="s2">&quot;$</span><span class="si">{train_ids[@]}</span><span class="s2">&quot;</span><span class="p">;</span> <span class="n">do</span>
        <span class="k">if</span> <span class="p">[[</span> <span class="s2">&quot;$target_id&quot;</span> <span class="o">==</span> <span class="s2">&quot;$id_forward&quot;</span> <span class="p">]]</span> <span class="o">||</span> <span class="p">[[</span> <span class="s2">&quot;$target_id&quot;</span> <span class="o">==</span> <span class="s2">&quot;$id_rev&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="n">fi</span>
    <span class="n">done</span>
    <span class="k">return</span> <span class="mi">1</span>
<span class="p">}</span>

<span class="c1"># Check if input file exists</span>
<span class="k">if</span> <span class="p">[</span> <span class="err">!</span> <span class="o">-</span><span class="n">f</span> <span class="s2">&quot;$input_file&quot;</span> <span class="p">];</span> <span class="n">then</span>
    <span class="n">echo</span> <span class="s2">&quot;Error: Input file not found.&quot;</span>
    <span class="n">exit</span> <span class="mi">1</span>
<span class="n">fi</span>

<span class="c1"># Process the input file and create output files</span>
<span class="k">while</span> <span class="n">IFS</span><span class="o">=</span><span class="err">$</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="n">read</span> <span class="o">-</span><span class="n">r</span> <span class="n">ID</span> <span class="n">Chr</span> <span class="n">Start</span> <span class="n">End</span> <span class="n">Strand</span> <span class="n">Name</span> <span class="nb">type</span><span class="p">;</span> <span class="n">do</span>
    <span class="k">if</span> <span class="p">[</span> <span class="s2">&quot;$ID&quot;</span> <span class="o">!=</span> <span class="s2">&quot;ID&quot;</span> <span class="p">];</span> <span class="n">then</span>
        <span class="c1"># Define TSS based on the strand</span>
        <span class="k">if</span> <span class="p">[[</span> <span class="s2">&quot;$Strand&quot;</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="n">TSS</span><span class="o">=</span><span class="err">$</span><span class="n">Start</span>
        <span class="k">elif</span> <span class="p">[[</span> <span class="s2">&quot;$Strand&quot;</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span> <span class="p">]];</span> <span class="n">then</span>
            <span class="n">TSS</span><span class="o">=</span><span class="err">$</span><span class="n">End</span>
        <span class="n">fi</span>

        <span class="n">new_start</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="n">TSS</span> <span class="o">-</span> <span class="n">shift</span><span class="p">))</span>
        <span class="n">new_end</span><span class="o">=</span><span class="err">$</span><span class="p">((</span><span class="n">TSS</span> <span class="o">+</span> <span class="n">shift</span><span class="p">))</span>
        <span class="n">output_id</span><span class="o">=</span><span class="s2">&quot;$</span><span class="si">{ID}</span><span class="s2">_forward&quot;</span> <span class="c1"># Use ID with _forward suffix for output</span>

        <span class="k">if</span> <span class="n">id_in_targets</span> <span class="s2">&quot;$ID&quot;</span><span class="p">;</span> <span class="n">then</span>
            <span class="k">if</span> <span class="p">[</span> <span class="s2">&quot;$Chr&quot;</span> <span class="o">!=</span> <span class="s2">&quot;chr4&quot;</span> <span class="p">]</span> <span class="o">&amp;&amp;</span> <span class="p">[</span> <span class="s2">&quot;$Chr&quot;</span> <span class="o">!=</span> <span class="s2">&quot;chr17&quot;</span> <span class="p">];</span> <span class="n">then</span>
                <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;$</span><span class="si">{Chr}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_start}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_end}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{output_id}</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="s2">&quot;$output_file1&quot;</span>
            <span class="n">fi</span>

            <span class="k">if</span> <span class="p">[</span> <span class="s2">&quot;$Chr&quot;</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="p">];</span> <span class="n">then</span>
                <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;$</span><span class="si">{Chr}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_start}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_end}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{output_id}</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="s2">&quot;$output_file2&quot;</span>
            <span class="n">fi</span>

            <span class="k">if</span> <span class="p">[</span> <span class="s2">&quot;$Chr&quot;</span> <span class="o">==</span> <span class="s2">&quot;chr17&quot;</span> <span class="p">];</span> <span class="n">then</span>
                <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;$</span><span class="si">{Chr}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_start}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{new_end}</span><span class="se">\t</span><span class="s2">$</span><span class="si">{output_id}</span><span class="s2">&quot;</span> <span class="o">&gt;&gt;</span> <span class="s2">&quot;$output_file3&quot;</span>
            <span class="n">fi</span>
        <span class="n">fi</span>
    <span class="n">fi</span>
<span class="n">done</span> <span class="o">&lt;</span> <span class="s2">&quot;$input_file&quot;</span>

<span class="n">echo</span> <span class="s2">&quot;Processing complete. Files generated: $output_file1, $output_file2, $output_file3&quot;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overwriting create_sequence_boundaries.sh
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span><span class="w"> </span>bash<span class="w"> </span>create_sequence_boundaries.sh<span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Processing complete. Files generated: ../annotations/training_boundaries.bed, ../annotations/test_boundaries.bed, ../annotations/validation_boundaries.bed
</pre></div>
</div>
</div>
</div>
<p><strong>Create fasta files with the sequences:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="c1"># Define paths</span>
<span class="n">ref_genome_path</span> <span class="o">=</span> <span class="s2">&quot;../GEO_files/&quot;</span>
<span class="n">reference_genome</span> <span class="o">=</span> <span class="s2">&quot;GRCm38.p6.genome.fa&quot;</span>
<span class="n">bed_path</span> <span class="o">=</span> <span class="s2">&quot;../annotations/&quot;</span>
<span class="n">sequence_fasta_path</span> <span class="o">=</span> <span class="s2">&quot;../inputs/DNA_sequences/&quot;</span>

<span class="c1"># Ensure the output directory exists</span>
<span class="n">Path</span><span class="p">(</span><span class="n">sequence_fasta_path</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Loop through the splits and run bedtools getfasta</span>
<span class="n">splits</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="n">splits</span><span class="p">:</span>
    <span class="c1"># Construct file paths</span>
    <span class="n">input_fasta</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_genome_path</span><span class="si">}{</span><span class="n">reference_genome</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">input_bed</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bed_path</span><span class="si">}{</span><span class="n">split</span><span class="si">}</span><span class="s2">_boundaries.bed&quot;</span>
    <span class="n">output_fasta</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sequence_fasta_path</span><span class="si">}{</span><span class="n">split</span><span class="si">}</span><span class="s2">_boundaries.fasta&quot;</span>
    
    <span class="c1"># Construct the command</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;bedtools&quot;</span><span class="p">,</span> <span class="s2">&quot;getfasta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-fi&quot;</span><span class="p">,</span> <span class="n">input_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-bed&quot;</span><span class="p">,</span> <span class="n">input_bed</span><span class="p">,</span>
        <span class="s2">&quot;-fo&quot;</span><span class="p">,</span> <span class="n">output_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-name&quot;</span>
    <span class="p">]</span>
    
    <span class="c1"># Run the command</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Create Enformer benchmark targets</strong></p>
<p>Enformer targets are the same targets, just cropped to fit the new target window (-57 to 57 bins). Test targets are de-duplicated, i.e. we only want forward strand oriented predictions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="k">def</span> <span class="nf">filter_columns_by_bins</span><span class="p">(</span><span class="n">input_csv</span><span class="p">,</span> <span class="n">output_csv</span><span class="p">,</span> <span class="n">n_bins</span><span class="p">,</span> <span class="n">TEST</span><span class="p">):</span>
    <span class="c1"># Read the CSV file</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_csv</span><span class="p">)</span>

    <span class="c1"># Generate column names of interest based on the range from -N_BINS to N_BINS</span>
    <span class="n">cols_of_interest</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_ctrl&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">n_bins</span><span class="p">,</span> <span class="n">n_bins</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)]</span>

    <span class="c1"># Select the columns that exist in the DataFrame</span>
    <span class="n">selected_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cols_of_interest</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

    <span class="c1"># Filter the DataFrame to keep only the desired columns</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">selected_columns</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">TEST</span><span class="p">:</span>
        <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="p">[</span><span class="o">~</span><span class="n">filtered_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;_rev&#39;</span><span class="p">)]</span>

    <span class="c1"># Write the filtered DataFrame to a new CSV file</span>
    <span class="n">filtered_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_csv</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File &#39;</span><span class="si">{</span><span class="n">output_csv</span><span class="si">}</span><span class="s2">&#39; has been created with columns from -</span><span class="si">{</span><span class="n">n_bins</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">n_bins</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>


<span class="k">for</span> <span class="n">input_name</span><span class="p">,</span><span class="n">output_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">([</span><span class="s1">&#39;../targets/training_targets.csv&#39;</span><span class="p">,</span><span class="s1">&#39;../targets/test_targets.csv&#39;</span><span class="p">],[</span><span class="s1">&#39;../targets/training_targets_Enformer.csv&#39;</span><span class="p">,</span><span class="s1">&#39;../targets/test_targets_Enformer.csv&#39;</span><span class="p">]):</span>
    <span class="n">TEST</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="s1">&#39;test&#39;</span> <span class="ow">in</span> <span class="n">input_name</span> <span class="k">else</span> <span class="kc">False</span>
    <span class="n">filter_columns_by_bins</span><span class="p">(</span><span class="n">input_name</span><span class="p">,</span> <span class="n">output_name</span><span class="p">,</span> <span class="mi">57</span><span class="p">,</span> <span class="n">TEST</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File &#39;../targets/training_targets_Enformer.csv&#39; has been created with columns from -57 to 57.
File &#39;../targets/test_targets_Enformer.csv&#39; has been created with columns from -57 to 57.
</pre></div>
</div>
</div>
</div>
<p>That’s it!
Now you have all files required to proceed to create, train and test CLASTER. This is done in the next notebook.</p>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><center> 1. DATA OBTENTION &amp; PREPROCESSING <center></center></center></a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#before-we-start">Before we start …</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#data-obtention">1. Data obtention</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#micro-c-data-acquisition">Micro-C Data acquisition</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#epigenomic-data-acquisition">Epigenomic Data acquisition:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-perturbations">2. Enhancer perturbations</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmark-data-obtention">3. Benchmark data obtention</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#create-bed-file-with-dna-chunk-boundaries">Create bed file with DNA chunk boundaries:</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Marc Pielies Avelli
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022, Marc Pielies Avelli.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=3ee479438cf8b5e0d341"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=3ee479438cf8b5e0d341"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>