{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <center> BENCHMARKING CLASTER WITH HYENA-DNA AND ENFORMER <center>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Hyena-DNA:\n",
    "\n",
    "The following is an adaptation of the open source code provided by the authors of Hyena-DNA, who are to be credited for it. \n",
    "\n",
    "The paper can be found as:\n",
    "Eric Nguyen and Michael Poli and Marjan Faizi and Armin Thomas and Callum Birch-Sykes and Michael Wornow and Aman Patel and Clayton Rabideau and Stefano Massaroli and Yoshua Bengio and Stefano Ermon and Stephen A. Baccus and Chris Ré, _HyenaDNA: Long-Range Genomic Sequence Modeling at Single Nucleotide Resolution_ , 2023.\n",
    "\n",
    "https://doi.org/10.48550/arXiv.2306.15794\n",
    "\n",
    "The original public colab notebook can be found here:\n",
    "\n",
    "https://colab.research.google.com/drive/1wyVEQd4R3HYLTUOXEEQmp_I8aNC_aLhL\n",
    "\n",
    "**Edits:**\n",
    "The goal was to add a head to the Hyena-DNA backbone. We tried obtaining only the embeddings and then adding the head on top of those, but we could not predict EU-seq profiles from the pretrained embeddings.\n",
    "\n",
    "The structure of the backbone remains untouched. We added custom regression heads aimed to\n",
    "- Perform dimensionality reduction from the high dimensional embeddings.\n",
    "- Link the resulting sequence representations to our outputs, i.e. EU-seq levels at a kbp resolution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: einops in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (0.7.0)\n",
      "Requirement already satisfied: torchvision in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (0.17.1)\n",
      "Collecting transformers==4.26.1\n",
      "  Using cached transformers-4.26.1-py3-none-any.whl.metadata (100 kB)\n",
      "Requirement already satisfied: nvidia-ml-py3 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (7.352.0)\n",
      "Requirement already satisfied: genomic-benchmarks in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (0.0.9)\n",
      "Requirement already satisfied: OmegaConf in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (2.3.0)\n",
      "Requirement already satisfied: filelock in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (3.13.3)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.11.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (0.22.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (1.26.4)\n",
      "Requirement already satisfied: packaging>=20.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (24.0)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (6.0.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (2023.12.25)\n",
      "Requirement already satisfied: requests in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (2.31.0)\n",
      "Collecting tokenizers!=0.11.3,<0.14,>=0.11.1 (from transformers==4.26.1)\n",
      "  Downloading tokenizers-0.13.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)\n",
      "Requirement already satisfied: tqdm>=4.27 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from transformers==4.26.1) (4.66.2)\n",
      "Requirement already satisfied: torch==2.2.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torchvision) (2.2.1)\n",
      "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torchvision) (10.2.0)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (4.10.0)\n",
      "Requirement already satisfied: sympy in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (1.12)\n",
      "Requirement already satisfied: networkx in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (3.2.1)\n",
      "Requirement already satisfied: jinja2 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (3.1.3)\n",
      "Requirement already satisfied: fsspec in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (2024.3.1)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.19.3 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (2.19.3)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (12.1.105)\n",
      "Requirement already satisfied: triton==2.2.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from torch==2.2.1->torchvision) (2.2.0)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch==2.2.1->torchvision) (12.4.99)\n",
      "Requirement already satisfied: biopython>=1.79 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from genomic-benchmarks) (1.83)\n",
      "Requirement already satisfied: pip>=20.0.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from genomic-benchmarks) (23.3.1)\n",
      "Requirement already satisfied: pandas>=1.1.4 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from genomic-benchmarks) (2.2.1)\n",
      "Requirement already satisfied: gdown>=4.2.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from genomic-benchmarks) (5.1.0)\n",
      "Requirement already satisfied: yarl in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from genomic-benchmarks) (1.9.4)\n",
      "Requirement already satisfied: antlr4-python3-runtime==4.9.* in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from OmegaConf) (4.9.3)\n",
      "Requirement already satisfied: beautifulsoup4 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from gdown>=4.2.0->genomic-benchmarks) (4.12.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from pandas>=1.1.4->genomic-benchmarks) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from pandas>=1.1.4->genomic-benchmarks) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from pandas>=1.1.4->genomic-benchmarks) (2024.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from requests->transformers==4.26.1) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from requests->transformers==4.26.1) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from requests->transformers==4.26.1) (2.0.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from requests->transformers==4.26.1) (2024.2.2)\n",
      "Requirement already satisfied: multidict>=4.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from yarl->genomic-benchmarks) (6.0.5)\n",
      "Requirement already satisfied: six>=1.5 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from python-dateutil>=2.8.2->pandas>=1.1.4->genomic-benchmarks) (1.16.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from beautifulsoup4->gdown>=4.2.0->genomic-benchmarks) (2.5)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from jinja2->torch==2.2.1->torchvision) (2.1.5)\n",
      "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from requests[socks]->gdown>=4.2.0->genomic-benchmarks) (1.7.1)\n",
      "Requirement already satisfied: mpmath>=0.19 in /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages (from sympy->torch==2.2.1->torchvision) (1.3.0)\n",
      "Using cached transformers-4.26.1-py3-none-any.whl (6.3 MB)\n",
      "Using cached tokenizers-0.13.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.8 MB)\n",
      "Installing collected packages: tokenizers, transformers\n",
      "  Attempting uninstall: tokenizers\n",
      "    Found existing installation: tokenizers 0.15.2\n",
      "    Uninstalling tokenizers-0.15.2:\n",
      "      Successfully uninstalled tokenizers-0.15.2\n",
      "  Attempting uninstall: transformers\n",
      "    Found existing installation: transformers 4.39.1\n",
      "    Uninstalling transformers-4.39.1:\n",
      "      Successfully uninstalled transformers-4.39.1\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "eir-dl 0.1.42 requires tokenizers<0.16.0,>=0.15.0, but you have tokenizers 0.13.3 which is incompatible.\n",
      "eir-dl 0.1.42 requires transformers<5.0.0,>=4.34.0, but you have transformers 4.26.1 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed tokenizers-0.13.3 transformers-4.26.1\n"
     ]
    }
   ],
   "source": [
    "! pip install  einops torchvision transformers==4.26.1 nvidia-ml-py3 genomic-benchmarks OmegaConf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Train and test HyenaDNA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Hyena_DNA_Esrum.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile Hyena_DNA_Esrum.py \n",
    "\n",
    "\"\"\"\n",
    "Python file containing the HyenaDNA ipynb cells stacked:\n",
    "https://colab.research.google.com/drive/1wyVEQd4R3HYLTUOXEEQmp_I8aNC_aLhL\n",
    "All credit is for the original developers.\n",
    "\n",
    "We added custom heads and modified and added training functions accordingly.\n",
    "\"\"\"\n",
    "\n",
    "#@title Installs\n",
    "# ! pip install einops\n",
    "# ! pip install torchvision\n",
    "# ! pip install transformers==4.26.1\n",
    "# ! pip install genomic-benchmarks\n",
    "# ! pip install OmegaConf\n",
    "\n",
    "\n",
    "#@title Imports\n",
    "# for HyenaDNA specifically\n",
    "from pathlib import Path\n",
    "import torch\n",
    "import math\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from functools import partial\n",
    "from einops import rearrange\n",
    "from typing import Optional\n",
    "from functools import partial\n",
    "from torch import Tensor\n",
    "from torchvision.ops import StochasticDepth\n",
    "from collections import namedtuple\n",
    "\n",
    "import os\n",
    "import requests\n",
    "import shutil\n",
    "import subprocess\n",
    "\n",
    "\n",
    "checkpoints_path = Path(\"../checkpoints/\")\n",
    "checkpoints_path.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "#@title Hyena layer\n",
    "\n",
    "def fftconv(u, k, D):\n",
    "    \"\"\"\n",
    "    We apply a convolution through the fourier domain (from the Convolution Theorem)\n",
    "\n",
    "    \"\"\"\n",
    "    seqlen = u.shape[-1]\n",
    "    fft_size = 2 * seqlen\n",
    "\n",
    "    k_f = torch.fft.rfft(k, n=fft_size) / fft_size\n",
    "    u_f = torch.fft.rfft(u.to(dtype=k.dtype), n=fft_size)\n",
    "\n",
    "    if len(u.shape) > 3: k_f = k_f.unsqueeze(1)\n",
    "    y = torch.fft.irfft(u_f * k_f, n=fft_size, norm='forward')[..., :seqlen]\n",
    "\n",
    "    out = y + u * D.unsqueeze(-1)\n",
    "    return out.to(dtype=u.dtype)\n",
    "\n",
    "\n",
    "@torch.jit.script\n",
    "def mul_sum(q, y):\n",
    "    return (q * y).sum(dim=1)\n",
    "\n",
    "class OptimModule(nn.Module):\n",
    "    \"\"\" Interface for Module that allows registering buffers/parameters with configurable optimizer hyperparameters \"\"\"\n",
    "\n",
    "    def register(self, name, tensor, lr=None, wd=0.0):\n",
    "        \"\"\"Register a tensor with a configurable learning rate and 0 weight decay\"\"\"\n",
    "\n",
    "        if lr == 0.0:\n",
    "            self.register_buffer(name, tensor)\n",
    "        else:\n",
    "            self.register_parameter(name, nn.Parameter(tensor))\n",
    "\n",
    "            optim = {}\n",
    "            if lr is not None: optim[\"lr\"] = lr\n",
    "            if wd is not None: optim[\"weight_decay\"] = wd\n",
    "            setattr(getattr(self, name), \"_optim\", optim)\n",
    "\n",
    "\n",
    "class Sin(nn.Module):\n",
    "    \"\"\"The Sin activation function for the Hyena Filter function.\"\"\"\n",
    "    def __init__(self, dim, w=10, train_freq=True):\n",
    "        super().__init__()\n",
    "        self.freq = nn.Parameter(w * torch.ones(1, dim)) if train_freq else w * torch.ones(1, dim)\n",
    "\n",
    "    def forward(self, x):\n",
    "        return torch.sin(self.freq * x)\n",
    "\n",
    "\n",
    "class PositionalEmbedding(OptimModule):\n",
    "    def __init__(self, emb_dim: int, seq_len: int, lr_pos_emb: float=1e-5, **kwargs):\n",
    "        \"\"\"Complex exponential positional embeddings for Hyena filters.\"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.seq_len = seq_len\n",
    "        # The time embedding fed to the filteres is normalized so that t_f = 1\n",
    "        t = torch.linspace(0, 1, self.seq_len)[None, :, None] # 1, L, 1\n",
    "\n",
    "        if emb_dim > 1:\n",
    "            bands = (emb_dim - 1) // 2\n",
    "        # To compute the right embeddings we use the \"proper\" linspace\n",
    "        t_rescaled = torch.linspace(0, seq_len - 1, seq_len)[None, :, None]\n",
    "        w = 2 * math.pi * t_rescaled / seq_len # 1, L, 1\n",
    "\n",
    "        f = torch.linspace(1e-4, bands - 1, bands)[None, None]\n",
    "        z = torch.exp(-1j * f * w)\n",
    "        z = torch.cat([t, z.real, z.imag], dim=-1)\n",
    "        self.register(\"z\", z, lr=lr_pos_emb)\n",
    "        self.register(\"t\", t, lr=0.0)\n",
    "\n",
    "    def forward(self, L):\n",
    "        return self.z[:, :L], self.t[:, :L]\n",
    "\n",
    "\n",
    "class ExponentialModulation(OptimModule):\n",
    "    \"\"\"The window function applied to the output of the (MLP) filter function.\"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        d_model,\n",
    "        fast_decay_pct=0.3,\n",
    "        slow_decay_pct=1.5,\n",
    "        target=1e-2,\n",
    "        modulation_lr=0.0,\n",
    "        modulate: bool=True,\n",
    "        shift: float = 0.05,\n",
    "        **kwargs\n",
    "    ):\n",
    "        super().__init__()\n",
    "        self.modulate = modulate\n",
    "        self.shift = shift\n",
    "        max_decay = math.log(target) / fast_decay_pct\n",
    "        min_decay = math.log(target) / slow_decay_pct\n",
    "        deltas = torch.linspace(min_decay, max_decay, d_model)[None, None]\n",
    "        self.register(\"deltas\", deltas, lr=modulation_lr)\n",
    "\n",
    "    def forward(self, t, x):\n",
    "        if self.modulate:\n",
    "            decay = torch.exp(-t * self.deltas.abs())\n",
    "            x = x * (decay + self.shift)\n",
    "        return x\n",
    "\n",
    "\n",
    "class HyenaFilter(OptimModule):\n",
    "    def __init__(\n",
    "            self,\n",
    "            d_model,\n",
    "            emb_dim=3, # dim of input to MLP, augments with positional encoding\n",
    "            order=16, # width of the implicit MLP\n",
    "            fused_fft_conv=False,\n",
    "            seq_len=1024,\n",
    "            lr=1e-3,\n",
    "            lr_pos_emb=1e-5,\n",
    "            dropout=0.0,\n",
    "            w=1, # frequency of periodic activations\n",
    "            wd=0, # weight decay of kernel parameters\n",
    "            bias=True,\n",
    "            num_inner_mlps=2,\n",
    "            normalized=False,\n",
    "            **kwargs\n",
    "        ):\n",
    "        \"\"\"\n",
    "        Implicit long filter with modulation.\n",
    "\n",
    "        Args:\n",
    "            d_model: number of channels in the input\n",
    "            emb_dim: dimension of the positional encoding (`emb_dim` - 1) // 2 is the number of bands\n",
    "            order: width of the FFN\n",
    "            num_inner_mlps: number of inner linear layers inside filter MLP\n",
    "\n",
    "        Note:\n",
    "            filter_dropout is not implemented\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.d_model = d_model\n",
    "        self.use_bias = bias\n",
    "        self.fused_fft_conv = fused_fft_conv\n",
    "        self.bias = nn.Parameter(torch.randn(self.d_model))\n",
    "        self.dropout = nn.Dropout(dropout)\n",
    "\n",
    "        act = Sin(dim=order, w=w)\n",
    "        self.emb_dim = emb_dim\n",
    "        assert emb_dim % 2 != 0 and emb_dim >= 3, \"emb_dim must be odd and greater or equal to 3 (time, sine and cosine)\"\n",
    "        self.seq_len = seq_len\n",
    "\n",
    "        self.pos_emb = PositionalEmbedding(emb_dim, seq_len, lr_pos_emb)\n",
    "\n",
    "        self.implicit_filter = nn.Sequential(\n",
    "            nn.Linear(emb_dim, order),\n",
    "            act,\n",
    "        )\n",
    "        for i in range(num_inner_mlps):\n",
    "            self.implicit_filter.append(nn.Linear(order, order))\n",
    "            self.implicit_filter.append(act)\n",
    "\n",
    "        self.implicit_filter.append(nn.Linear(order, d_model, bias=False))\n",
    "\n",
    "        self.modulation = ExponentialModulation(d_model, **kwargs)\n",
    "\n",
    "        self.normalized = normalized\n",
    "        for c in self.implicit_filter.children():\n",
    "            for name, v in c.state_dict().items():\n",
    "                optim = {\"weight_decay\": wd, \"lr\": lr}\n",
    "                setattr(getattr(c, name), \"_optim\", optim)\n",
    "\n",
    "    def filter(self, L, *args, **kwargs):\n",
    "        z, t = self.pos_emb(L)\n",
    "        h = self.implicit_filter(z)\n",
    "        h = self.modulation(t, h)\n",
    "        return h\n",
    "\n",
    "    def forward(self, x, L, k=None, bias=None, *args, **kwargs):\n",
    "        if k is None: k = self.filter(L)\n",
    "\n",
    "        # Ensure compatibility with filters that return a tuple\n",
    "        k = k[0] if type(k) is tuple else k\n",
    "\n",
    "        y = fftconv(x, k, bias)\n",
    "        return y\n",
    "\n",
    "\n",
    "class HyenaOperator(nn.Module):\n",
    "    def __init__(\n",
    "            self,\n",
    "            d_model,\n",
    "            l_max,\n",
    "            order=2,\n",
    "            filter_order=64,\n",
    "            dropout=0.0,\n",
    "            filter_dropout=0.0,\n",
    "            **filter_args,\n",
    "        ):\n",
    "        r\"\"\"\n",
    "        Hyena operator described in the paper https://arxiv.org/pdf/2302.10866.pdf\n",
    "\n",
    "        Args:\n",
    "            d_model (int): Dimension of the input and output embeddings (width of the layer)\n",
    "            l_max: (int): Maximum input sequence length. Defaults to None\n",
    "            order: (int): Depth of the Hyena recurrence. Defaults to 2\n",
    "            dropout: (float): Dropout probability. Defaults to 0.0\n",
    "            filter_dropout: (float): Dropout probability for the filter. Defaults to 0.0\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "\n",
    "        self.d_model = d_model\n",
    "        self.l_max = l_max\n",
    "        self.order = order\n",
    "        inner_width = d_model * (order + 1)\n",
    "        self.dropout = nn.Dropout(dropout)\n",
    "        self.in_proj = nn.Linear(d_model, inner_width)\n",
    "        self.out_proj = nn.Linear(d_model, d_model)\n",
    "\n",
    "        self.short_filter = nn.Conv1d(\n",
    "            inner_width,\n",
    "            inner_width,\n",
    "            3,\n",
    "            padding=2,\n",
    "            groups=inner_width\n",
    "        )\n",
    "        self.filter_fn = HyenaFilter(\n",
    "            d_model * (order - 1),\n",
    "            order=filter_order,\n",
    "            seq_len=l_max,\n",
    "            channels=1,\n",
    "            dropout=filter_dropout,\n",
    "            **filter_args\n",
    "        )\n",
    "\n",
    "    def forward(self, u, *args, **kwargs):\n",
    "        l = u.size(-2)\n",
    "        l_filter = min(l, self.l_max)\n",
    "        u = self.in_proj(u)\n",
    "        u = rearrange(u, 'b l d -> b d l')\n",
    "\n",
    "        uc = self.short_filter(u)[...,:l_filter]\n",
    "        *x, v = uc.split(self.d_model, dim=1)\n",
    "\n",
    "        k = self.filter_fn.filter(l_filter)[0]\n",
    "        k = rearrange(k, 'l (o d) -> o d l', o=self.order - 1)\n",
    "        bias = rearrange(self.filter_fn.bias, '(o d) -> o d', o=self.order - 1)\n",
    "\n",
    "        for o, x_i in enumerate(reversed(x[1:])):\n",
    "            v = self.dropout(v * x_i)\n",
    "            v = self.filter_fn(v, l_filter, k=k[o], bias=bias[o])\n",
    "\n",
    "        y = rearrange(v * x[0], 'b d l -> b l d')\n",
    "\n",
    "        y = self.out_proj(y)\n",
    "        return y\n",
    "\n",
    "#@title Self-Attention (alternative)\n",
    "\n",
    "\"\"\"\n",
    "If you'd like to try the HyenaDNA model using attention instead, you can. ie,\n",
    "use a regular decoder only Transformer.\n",
    "\n",
    "Borrowed from the FlashAttention library by Tri Dao.\n",
    "\"\"\"\n",
    "\n",
    "class SelfAttention(nn.Module):\n",
    "    \"\"\"Implement the scaled dot product attention with softmax.\n",
    "    Arguments\n",
    "    ---------\n",
    "        softmax_scale: The temperature to use for the softmax attention.\n",
    "                      (default: 1/sqrt(d_keys) where d_keys is computed at\n",
    "                      runtime)\n",
    "        attention_dropout: The dropout rate to apply to the attention\n",
    "                           (default: 0.0)\n",
    "    \"\"\"\n",
    "    def __init__(self, causal=False, softmax_scale=None, attention_dropout=0.0):\n",
    "        super().__init__()\n",
    "        self.causal = causal\n",
    "        self.softmax_scale = softmax_scale\n",
    "        self.dropout_p = attention_dropout\n",
    "\n",
    "    def forward(self, qkv, causal=None, key_padding_mask=None):\n",
    "        \"\"\"Implements the multihead softmax attention.\n",
    "        Arguments\n",
    "        ---------\n",
    "            qkv: The tensor containing the query, key, and value. (B, S, 3, H, D)\n",
    "            causal: if passed, will override self.causal\n",
    "            key_padding_mask: boolean mask to apply to the attention weights. True means to keep,\n",
    "                False means to mask out. (B, S)\n",
    "        \"\"\"\n",
    "        batch_size, seqlen = qkv.shape[0], qkv.shape[1]\n",
    "        causal = self.causal if causal is None else causal\n",
    "        q, k, v = qkv.unbind(dim=2)\n",
    "        softmax_scale = self.softmax_scale or 1.0 / math.sqrt(q.shape[-1])\n",
    "        scores = torch.einsum('bthd,bshd->bhts', q, k * softmax_scale)\n",
    "        if key_padding_mask is not None:\n",
    "            padding_mask = torch.full((batch_size, seqlen), -10000.0, dtype=scores.dtype,\n",
    "                                      device=scores.device)\n",
    "            padding_mask.masked_fill_(key_padding_mask, 0.0)\n",
    "            scores = scores + rearrange(padding_mask, 'b s -> b 1 1 s')\n",
    "        if causal:\n",
    "            # \"triu_tril_cuda_template\" not implemented for 'BFloat16'\n",
    "            # So we have to construct the mask in float\n",
    "            causal_mask = torch.triu(torch.full((seqlen, seqlen), -10000.0, device=scores.device), 1)\n",
    "            scores = scores + causal_mask.to(dtype=scores.dtype)\n",
    "        attention = torch.softmax(scores, dim=-1, dtype=v.dtype)\n",
    "        attention_drop = F.dropout(attention, self.dropout_p if self.training else 0.0)\n",
    "        output = torch.einsum('bhts,bshd->bthd', attention_drop, v)\n",
    "        return output\n",
    "\n",
    "class MHA(nn.Module):\n",
    "    \"\"\"Multi-head self-attention and cross-attention\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, embed_dim, num_heads, bias=True, dropout=0.0,\n",
    "                 softmax_scale=None, causal=False, layer_idx=None, dwconv=False,return_residual=False,device=None, dtype=None) -> None:\n",
    "        \"\"\"\n",
    "            return_residual: whether to return the input x along with the output. This is for\n",
    "                performance reason: for post-norm architecture, returning the input allows us\n",
    "                to fuse the backward of nn.Linear with the residual connection.\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        self.embed_dim = embed_dim\n",
    "        self.causal = causal\n",
    "        self.layer_idx = layer_idx\n",
    "        self.dwconv = dwconv\n",
    "        self.return_residual = return_residual\n",
    "\n",
    "        self.num_heads = num_heads\n",
    "        assert self.embed_dim % num_heads == 0, \"self.kdim must be divisible by num_heads\"\n",
    "        self.head_dim = self.embed_dim // num_heads\n",
    "\n",
    "        linear_cls = nn.Linear\n",
    "        linear_resid_cls = LinearResidual\n",
    "        inner_attn_cls =  SelfAttention\n",
    "\n",
    "        if not self.return_residual:\n",
    "            self.Wqkv = linear_cls(embed_dim, 3 * embed_dim, bias=bias, **factory_kwargs)\n",
    "        else:\n",
    "            self.Wqkv = linear_resid_cls(embed_dim, 3 * embed_dim, bias=bias, **factory_kwargs)\n",
    "        if self.dwconv:\n",
    "            self.dwconv_qkv = nn.Conv1d(3 * embed_dim, 3 * embed_dim, kernel_size=3, padding=2,\n",
    "                                        groups=3 * embed_dim)\n",
    "\n",
    "        self.inner_attn = inner_attn_cls(causal=causal, softmax_scale=softmax_scale,\n",
    "                                         attention_dropout=dropout)\n",
    "\n",
    "        # output projection always have the bias (for now)\n",
    "        self.out_proj = linear_cls(embed_dim, embed_dim, **factory_kwargs)\n",
    "\n",
    "    def forward(self, x, key_padding_mask=None, **kwargs):\n",
    "        \"\"\"\n",
    "        Arguments:\n",
    "            x: (batch, seqlen, hidden_dim) (where hidden_dim = num heads * head dim) if\n",
    "                cu_seqlens is None and max_seqlen is None, else (total, hidden_dim) where total\n",
    "                is the is the sum of the sequence lengths in the batch.\n",
    "            cu_seqlens: (batch_size + 1,), dtype torch.int32. The cumulative sequence lengths\n",
    "                of the sequences in the batch, used to index into x. Only applicable when using\n",
    "                FlashAttention.\n",
    "            max_seqlen: int. Maximum sequence length in the batch.\n",
    "            key_padding_mask: boolean mask, True means to keep, False means to mask out.\n",
    "                (batch, seqlen). Only applicable when not using FlashAttention.\n",
    "            mixer_subset: for cross-attention only. If not None, will take a subset of x\n",
    "                before applying the query projection. Useful for e.g., ViT where we only care\n",
    "                about the CLS token in the last layer.\n",
    "            inference_params: for generation. Adapted from Megatron-LM (and Apex)\n",
    "            https://github.com/NVIDIA/apex/blob/3ff1a10f72ec07067c4e44759442329804ac5162/apex/transformer/testing/standalone_transformer_lm.py#L470\n",
    "        \"\"\"\n",
    "\n",
    "        kwargs = ({'key_padding_mask': key_padding_mask, **kwargs})\n",
    "\n",
    "        if not self.return_residual:\n",
    "            qkv = self.Wqkv(x)\n",
    "        else:\n",
    "            qkv, x = self.Wqkv(x)\n",
    "        if self.dwconv:\n",
    "            qkv = rearrange(self.dwconv_qkv(rearrange(qkv, 'b s d -> b d s'))[..., :-2],\n",
    "                            'b d s -> b s d').contiguous()\n",
    "        qkv = rearrange(qkv, '... (three h d) -> ... three h d', three=3, d=self.head_dim)\n",
    "\n",
    "        context = self.inner_attn(qkv, **kwargs)\n",
    "\n",
    "        out = self.out_proj(rearrange(context, '... h d -> ... (h d)'))\n",
    "        return out if not self.return_residual else (out, x)\n",
    "\n",
    "#@title MLP layer\n",
    "\n",
    "\"\"\"\n",
    "The MLP layer after the mixer layer (HyenaOperator).\n",
    "\"\"\"\n",
    "\n",
    "class Mlp(nn.Module):\n",
    "\n",
    "    def __init__(self, in_features, hidden_features=None, out_features=None, activation=F.gelu,\n",
    "                 return_residual=False, device=None, dtype=None):\n",
    "        \"\"\"\n",
    "        From https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/modules/mlp.py\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        out_features = out_features or in_features\n",
    "        hidden_features = hidden_features or in_features\n",
    "        self.return_residual = return_residual\n",
    "        self.fc1 = nn.Linear(in_features, hidden_features, **factory_kwargs)\n",
    "        self.activation = activation\n",
    "        self.fc2 = nn.Linear(hidden_features, out_features, **factory_kwargs)\n",
    "\n",
    "    def forward(self, x):\n",
    "        y = self.fc1(x)\n",
    "        y = self.activation(y)\n",
    "        y = self.fc2(y)\n",
    "        return y if not self.return_residual else (y, x)\n",
    "\n",
    "#@title Block layer (Hyena + MLP layers)\n",
    "\n",
    "\"\"\"\n",
    "A block consists of a Mixer layer (Hyena or attention), and a MLP layer.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "class LinearResidual(nn.Linear):\n",
    "    \"\"\"Wrap nn.Linear to return the residual as well. For compatibility with FusedDense.\n",
    "    \"\"\"\n",
    "\n",
    "    def forward(self, input: torch.Tensor) -> torch.Tensor:\n",
    "        return super().forward(input), input\n",
    "\n",
    "class Block(nn.Module):\n",
    "\n",
    "    def __init__(self, dim, mixer_cls=None, mlp_cls=None, norm_cls=nn.LayerNorm,\n",
    "                 dropout_cls=nn.Dropout, prenorm=True, resid_dropout1=0., resid_dropout2=0.,\n",
    "                 drop_path1=0., drop_path2=0.,\n",
    "                 return_residual=False,\n",
    "                 residual_in_fp32=False):\n",
    "        \"\"\"\n",
    "        From https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/modules/block.py\n",
    "        For prenorm=True, this Block has a slightly different structure compared to a regular\n",
    "        prenorm Transformer block.\n",
    "        The standard block is: LN -> MHA -> Dropout -> Add -> LN -> MLP -> Dropout -> Add.\n",
    "        [Ref: https://arxiv.org/abs/2002.04745]\n",
    "        Here we have: Dropout -> Add -> LN -> MHA -> Dropout -> Add -> LN -> MLP, returning both\n",
    "        the hidden_states (output of the MLP) and the residual.\n",
    "        This is for performance reasons, as we can fuse the dropout, add and LayerNorm.\n",
    "        The residual needs to be provided (except for the very first block).\n",
    "        For prenorm=False, this Block has the same structure as a regular postnorm Transformer\n",
    "        block: MHA -> Dropout -> Add -> LN -> MLP -> Dropout -> Add -> LN.\n",
    "        return_residual: whether each of the sub-layers (mixer and mlp) will return the residual.\n",
    "        This is for performance reason: for post-norm architecture, returning the input allows us\n",
    "        to fuse the backward of nn.Linear with the residual connectio\n",
    "        \"\"\"\n",
    "        super().__init__()\n",
    "        self.prenorm = prenorm\n",
    "        self.return_residual = return_residual\n",
    "        self.residual_in_fp32 = residual_in_fp32\n",
    "        if self.residual_in_fp32:\n",
    "            assert self.prenorm, 'residual_in_fp32 is only compatible with prenorm=True'\n",
    "        if mixer_cls is None:\n",
    "            mixer_cls = partial(MHA, num_heads=dim // 64)\n",
    "        if mlp_cls is None:\n",
    "            mlp_cls = partial(Mlp, hidden_features=4 * dim)\n",
    "        self.mixer = mixer_cls()\n",
    "        self.dropout1 = dropout_cls(resid_dropout1)\n",
    "        self.drop_path1 = StochasticDepth(drop_path1, mode='row')\n",
    "        self.norm1 = norm_cls(dim)\n",
    "        self.mlp = mlp_cls(dim)\n",
    "        if not isinstance(self.mlp, nn.Identity):\n",
    "            self.dropout2 = dropout_cls(resid_dropout2)\n",
    "            self.drop_path2 = StochasticDepth(drop_path2, mode='row')\n",
    "            self.norm2 = norm_cls(dim)\n",
    "\n",
    "    def forward(self, hidden_states, residual = None,\n",
    "                mixer_subset=None, mixer_kwargs=None):\n",
    "        r\"\"\"Pass the input through the encoder layer.\n",
    "        Args:\n",
    "            hidden_states: the sequence to the encoder layer (required).\n",
    "            residual: if postnorm, residual=None, If prenorm, hidden_states = Attn/MLP(LN(residual))\n",
    "            mixer_subset: for cross-attention only. If not None, will take a subset of x\n",
    "                before applying the query projection. Useful for e.g., ViT where we only care\n",
    "                about the CLS token in the last layer.\n",
    "        \"\"\"\n",
    "        if self.prenorm:\n",
    "            dropped = self.drop_path1(self.dropout1(hidden_states))\n",
    "            residual = (dropped + residual) if residual is not None else dropped\n",
    "            hidden_states = self.norm1(residual.to(dtype=self.norm1.weight.dtype))\n",
    "            if self.residual_in_fp32:\n",
    "                residual = residual.to(torch.float32)\n",
    "            if mixer_kwargs is None:\n",
    "                mixer_kwargs = {}\n",
    "            if mixer_subset is not None:\n",
    "                mixer_kwargs['mixer_subset'] = mixer_subset\n",
    "            hidden_states = self.mixer(hidden_states, **mixer_kwargs)\n",
    "            if mixer_subset is not None:\n",
    "                residual = residual[:, mixer_subset]\n",
    "            if not isinstance(self.mlp, nn.Identity):\n",
    "                dropped = self.drop_path2(self.dropout2(hidden_states))\n",
    "                residual = (dropped + residual) if residual is not None else dropped\n",
    "                hidden_states = self.norm2(residual.to(dtype=self.norm2.weight.dtype))\n",
    "                if self.residual_in_fp32:\n",
    "                    residual = residual.to(torch.float32)\n",
    "\n",
    "                hidden_states = self.mlp(hidden_states)\n",
    "            return hidden_states, residual\n",
    "        else:\n",
    "            assert residual is None\n",
    "            mixer_out = self.mixer(\n",
    "                hidden_states, **(mixer_kwargs if mixer_kwargs is not None else {})\n",
    "            )\n",
    "            if self.return_residual:  # mixer out is actually a pair here\n",
    "                mixer_out, hidden_states = mixer_out\n",
    "\n",
    "            hidden_states = self.norm1((self.drop_path1(self.dropout1(mixer_out))\n",
    "                                        + hidden_states).to(dtype=self.norm1.weight.dtype))\n",
    "\n",
    "            if not isinstance(self.mlp, nn.Identity):\n",
    "                mlp_out = self.mlp(hidden_states)\n",
    "                if self.return_residual:  # mlp out is actually a pair here\n",
    "                    mlp_out, hidden_states = mlp_out\n",
    "\n",
    "                hidden_states = self.norm2((self.drop_path2(self.dropout2(mlp_out))\n",
    "                                            + hidden_states).to(dtype=self.norm2.weight.dtype))\n",
    "\n",
    "            return hidden_states\n",
    "\n",
    "def create_mixer_cls(layer=None,\n",
    "                     attn_layer_idx=None, attn_cfg=None, layer_idx=None,\n",
    "                     device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    if attn_layer_idx is not None and layer_idx in attn_layer_idx:\n",
    "        causal = True if attn_cfg is None else attn_cfg.pop('causal', True)\n",
    "\n",
    "        mha_cls = MHA\n",
    "\n",
    "        mixer_cls = partial(mha_cls, causal=causal, layer_idx=layer_idx,\n",
    "                            **(attn_cfg if attn_cfg is not None else {}),**factory_kwargs)\n",
    "    else:\n",
    "        # mixer_cls = instantiate(registry.layer, layer, partial=True, layer_idx=layer_idx, **factory_kwargs)\n",
    "\n",
    "        mixer_cls = partial(HyenaOperator, **layer)\n",
    "\n",
    "    return mixer_cls\n",
    "\n",
    "def create_mlp_cls(d_model, d_inner=None, device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    inner_dim = d_inner if d_inner is not None else 4 * d_model\n",
    "\n",
    "    mlp_cls = partial(Mlp, hidden_features=inner_dim,\n",
    "                          activation=partial(F.gelu, approximate='tanh'), **factory_kwargs)\n",
    "\n",
    "    return mlp_cls\n",
    "\n",
    "\n",
    "def create_block(d_model, d_inner=None,\n",
    "                 layer=None, attn_layer_idx=None,\n",
    "                 attn_cfg=None, layer_norm_epsilon=1e-5,\n",
    "                 resid_dropout1=0.0, resid_dropout2=0.0, residual_in_fp32=False,\n",
    "                 layer_idx=None,\n",
    "                 device=None, dtype=None):\n",
    "    factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "    mixer_cls = create_mixer_cls(layer=layer,\n",
    "                                 attn_layer_idx=attn_layer_idx,\n",
    "                                 attn_cfg=attn_cfg, layer_idx=layer_idx,\n",
    "                                 **factory_kwargs)\n",
    "    mlp_cls = create_mlp_cls(d_model, d_inner=d_inner,\n",
    "                             **factory_kwargs)\n",
    "    norm_cls = partial(nn.LayerNorm, eps=layer_norm_epsilon, **factory_kwargs)\n",
    "    block = Block(d_model, mixer_cls, mlp_cls, norm_cls=norm_cls,\n",
    "                  prenorm=True, resid_dropout1=resid_dropout1, resid_dropout2=resid_dropout2,residual_in_fp32=residual_in_fp32)\n",
    "    block.layer_idx = layer_idx\n",
    "    return block\n",
    "\n",
    "\n",
    "# https://github.com/huggingface/transformers/blob/c28d04e9e252a1a099944e325685f14d242ecdcd/src/transformers/models/gpt2/modeling_gpt2.py#L454\n",
    "def _init_weights(module, n_layer, initializer_range=0.02, rescale_prenorm_residual=True,\n",
    "                  glu_act=False):\n",
    "    if isinstance(module, nn.Linear):\n",
    "        nn.init.normal_(module.weight, std=initializer_range)\n",
    "        if module.bias is not None:\n",
    "            nn.init.zeros_(module.bias)\n",
    "    elif isinstance(module, nn.Embedding):\n",
    "        nn.init.normal_(module.weight, std=initializer_range)\n",
    "\n",
    "    if rescale_prenorm_residual:\n",
    "        # Reinitialize selected weights subject to the OpenAI GPT-2 Paper Scheme:\n",
    "        #   > A modified initialization which accounts for the accumulation on the residual path with model depth. Scale\n",
    "        #   > the weights of residual layers at initialization by a factor of 1/√N where N is the # of residual layers.\n",
    "        #   >   -- GPT-2 :: https://openai.com/blog/better-language-models/\n",
    "        #\n",
    "        # Reference (Megatron-LM): https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py\n",
    "        for name, p in module.named_parameters():\n",
    "            if name in [\"out_proj.weight\", \"fc2.weight\"]:\n",
    "                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block\n",
    "                nn.init.normal_(p, mean=0.0, std=initializer_range / math.sqrt(2 * n_layer))\n",
    "            # If using GLU activation for now, we scale the std by 2\n",
    "            elif name in [\"output_linear.0.weight\"]:\n",
    "                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block\n",
    "                if not glu_act:\n",
    "                    nn.init.normal_(p, mean=0.0, std=initializer_range / math.sqrt(2 * n_layer))\n",
    "                else:\n",
    "                    out_features = p.shape[0]\n",
    "                    # Multiplying the first half of the matrix by 2 since sigmoid scales it down by 0.5\n",
    "                    # on average.\n",
    "                    nn.init.normal_(p[:out_features // 2], mean=0.0, std=initializer_range / math.sqrt(2 * n_layer) * 2)\n",
    "\n",
    "#@title Backbone model (stack of blocks)\n",
    "\n",
    "\"\"\"\n",
    "A backbone model consists of a stack of blocks. If you use attention, then\n",
    "positional embeddings are included. When using Hyena, then the pos emb\n",
    "revert to doing nothing.\n",
    "\"\"\"\n",
    "\n",
    "class GPT2Embeddings(nn.Module):\n",
    "\n",
    "    def __init__(self, embed_dim, vocab_size, max_position_embeddings, padding_idx=None,\n",
    "                 word_embed_proj_dim=None, device=None, dtype=None):\n",
    "        \"\"\"\n",
    "            If max_position_embeddings <= 0, there's no position embeddings\n",
    "            If word_embe_proj_dim is not None (e.g., OPT-350m), we embed to that dimension\n",
    "                the project up to embed_dim\n",
    "        \"\"\"\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        if word_embed_proj_dim is None:\n",
    "            self.word_embeddings = nn.Embedding(vocab_size, embed_dim, padding_idx=padding_idx,\n",
    "                                                **factory_kwargs)\n",
    "            self.project_in = None\n",
    "        else:\n",
    "            self.word_embeddings = nn.Embedding(vocab_size, word_embed_proj_dim,\n",
    "                                                padding_idx=padding_idx, **factory_kwargs)\n",
    "            self.project_in = nn.Linear(word_embed_proj_dim, embed_dim, bias=False,\n",
    "                                        **factory_kwargs)\n",
    "        self.max_position_embeddings = max_position_embeddings\n",
    "        if self.max_position_embeddings > 0:\n",
    "            self.position_embeddings = nn.Embedding(max_position_embeddings, embed_dim,\n",
    "                                                    **factory_kwargs)\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None):\n",
    "        \"\"\"\n",
    "            input_ids: (batch, seqlen)\n",
    "            position_ids: (batch, seqlen)\n",
    "        \"\"\"\n",
    "        batch_size, seqlen = input_ids.shape\n",
    "        embeddings = self.word_embeddings(input_ids)\n",
    "        if self.project_in is not None:\n",
    "            embeddings = self.project_in(embeddings)\n",
    "        if self.max_position_embeddings > 0:\n",
    "            if position_ids is None:\n",
    "                position_ids = torch.arange(seqlen, dtype=torch.long, device=input_ids.device)\n",
    "            position_embeddings = self.position_embeddings(position_ids)\n",
    "            embeddings = embeddings + position_embeddings\n",
    "        return embeddings\n",
    "\n",
    "class LMBackbone(nn.Module):\n",
    "\n",
    "    def __init__(self, d_model: int, n_layer: int, d_inner: int, vocab_size: int,\n",
    "                 process_group=None, layer=None,\n",
    "                 attn_layer_idx=None, attn_cfg=None, max_position_embeddings=0,\n",
    "                 resid_dropout: float = 0.0, embed_dropout: float = 0.1,\n",
    "                 layer_norm_epsilon: float = 1e-5, initializer_cfg=None,residual_in_fp32=False,\n",
    "                 device=None, dtype=None, **kwargs) -> None:\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        self.process_group = process_group\n",
    "        self.residual_in_fp32 = residual_in_fp32\n",
    "        # note max_position_embeddings is 0 for Hyena, and therefore isn't used\n",
    "        self.embeddings = GPT2Embeddings(d_model, vocab_size, max_position_embeddings,\n",
    "                                             **factory_kwargs)\n",
    "\n",
    "        self.layers = nn.ModuleList([create_block(\n",
    "            d_model, d_inner=d_inner,\n",
    "            layer=layer, attn_layer_idx=attn_layer_idx,\n",
    "            attn_cfg=attn_cfg, layer_norm_epsilon=layer_norm_epsilon,\n",
    "            resid_dropout1=embed_dropout if i == 0 else resid_dropout,\n",
    "            resid_dropout2=resid_dropout, residual_in_fp32=residual_in_fp32,layer_idx=i,\n",
    "            **factory_kwargs,\n",
    "        ) for i in range(n_layer)])\n",
    "\n",
    "        self.drop_f = nn.Dropout(resid_dropout)\n",
    "        self.ln_f = nn.LayerNorm(d_model, eps=layer_norm_epsilon, **factory_kwargs)\n",
    "\n",
    "        self.apply(partial(_init_weights, n_layer=n_layer,\n",
    "                           **(initializer_cfg if initializer_cfg is not None else {})))\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None):\n",
    "        #print(\"Before embeddings:\", torch.cuda.memory_allocated(device='cuda'))\n",
    "        hidden_states = self.embeddings(input_ids, position_ids=position_ids,)\n",
    "        residual = None\n",
    "\n",
    "        #print(hidden_states.shape, hidden_states.dtype)\n",
    "        #print(\"After embeddings:\", torch.cuda.memory_allocated(device='cuda'))\n",
    "\n",
    "        for layer in self.layers:\n",
    "            hidden_states, residual = layer(hidden_states, residual)\n",
    "            #print(hidden_states.shape)\n",
    "            #print(torch.cuda.memory_allocated(device='cuda'))\n",
    "\n",
    "        dropped = self.drop_f(hidden_states)\n",
    "        residual = (dropped + residual) if residual is not None else dropped\n",
    "        hidden_states = self.ln_f(residual.to(dtype=self.ln_f.weight.dtype))\n",
    "\n",
    "        return hidden_states\n",
    "\n",
    "\n",
    "#@title Decoder head layer\n",
    "\n",
    "\"\"\"\n",
    "A simple decoder head (using MLP) to predict a sequence level classification.\n",
    "You have the option to average across all the tokens in a sequence or using the\n",
    "\"last\" token to classify.  At least, those 2 worked best for us, but we provide\n",
    "other \"modes\" as well.\n",
    "\n",
    "We only need this for classification.  Otherwise we'll use the hidden\n",
    "states of the backbone as embeddings.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "class SequenceDecoder(nn.Module):\n",
    "    def __init__(\n",
    "        self, d_model, d_output=None, l_output=None, use_lengths=False, mode=\"last\"\n",
    "    ):\n",
    "        super().__init__()\n",
    "\n",
    "        self.output_transform = nn.Identity() if d_output is None else nn.Linear(d_model, d_output)\n",
    "\n",
    "        if l_output is None:\n",
    "            self.l_output = None\n",
    "            self.squeeze = False\n",
    "        elif l_output == 0:\n",
    "            # Equivalent to getting an output of length 1 and then squeezing\n",
    "            self.l_output = 1\n",
    "            self.squeeze = True\n",
    "        else:\n",
    "            assert l_output > 0\n",
    "            self.l_output = l_output\n",
    "            self.squeeze = False\n",
    "\n",
    "        self.use_lengths = use_lengths\n",
    "        self.mode = mode\n",
    "\n",
    "        if mode == 'ragged':\n",
    "            assert not use_lengths\n",
    "\n",
    "    def forward(self, x, state=None, lengths=None, l_output=None):\n",
    "        \"\"\"\n",
    "        x: (n_batch, l_seq, d_model)\n",
    "        Returns: (n_batch, l_output, d_output)\n",
    "        \"\"\"\n",
    "\n",
    "        if self.l_output is None:\n",
    "            if l_output is not None:\n",
    "                assert isinstance(l_output, int)  # Override by pass in\n",
    "            else:\n",
    "                # Grab entire output\n",
    "                l_output = x.size(-2)\n",
    "            squeeze = False\n",
    "        else:\n",
    "            l_output = self.l_output\n",
    "            squeeze = self.squeeze\n",
    "\n",
    "        if self.mode == \"last\":\n",
    "            restrict = lambda x: x[..., -l_output:, :]\n",
    "        elif self.mode == \"first\":\n",
    "            restrict = lambda x: x[..., :l_output, :]\n",
    "        elif self.mode == \"pool\":\n",
    "            restrict = lambda x: (\n",
    "                torch.cumsum(x, dim=-2)\n",
    "                / torch.arange(\n",
    "                    1, 1 + x.size(-2), device=x.device, dtype=x.dtype\n",
    "                ).unsqueeze(-1)\n",
    "            )[..., -l_output:, :]\n",
    "\n",
    "            def restrict(x):\n",
    "                L = x.size(-2)\n",
    "                s = x.sum(dim=-2, keepdim=True)\n",
    "                if l_output > 1:\n",
    "                    c = torch.cumsum(x[..., -(l_output - 1) :, :].flip(-2), dim=-2)\n",
    "                    c = F.pad(c, (0, 0, 1, 0))\n",
    "                    s = s - c  # (B, l_output, D)\n",
    "                    s = s.flip(-2)\n",
    "                denom = torch.arange(\n",
    "                    L - l_output + 1, L + 1, dtype=x.dtype, device=x.device\n",
    "                )\n",
    "                s = s / denom\n",
    "                return s\n",
    "\n",
    "        elif self.mode == \"sum\":\n",
    "            restrict = lambda x: torch.cumsum(x, dim=-2)[..., -l_output:, :]\n",
    "            # TODO use same restrict function as pool case\n",
    "        elif self.mode == 'ragged':\n",
    "            assert lengths is not None, \"lengths must be provided for ragged mode\"\n",
    "            # remove any additional padding (beyond max length of any sequence in the batch)\n",
    "            restrict = lambda x: x[..., : max(lengths), :]\n",
    "        else:\n",
    "            raise NotImplementedError(\n",
    "                \"Mode must be ['last' | 'first' | 'pool' | 'sum']\"\n",
    "            )\n",
    "\n",
    "        # Restrict to actual length of sequence\n",
    "        if self.use_lengths:\n",
    "            assert lengths is not None\n",
    "            x = torch.stack(\n",
    "                [\n",
    "                    restrict(out[..., :length, :])\n",
    "                    for out, length in zip(torch.unbind(x, dim=0), lengths)\n",
    "                ],\n",
    "                dim=0,\n",
    "            )\n",
    "        else:\n",
    "            x = restrict(x)\n",
    "\n",
    "        if squeeze:\n",
    "            assert x.size(-2) == 1\n",
    "            x = x.squeeze(-2)\n",
    "\n",
    "        x = self.output_transform(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "    def step(self, x, state=None):\n",
    "        # Ignore all length logic\n",
    "        return self.output_transform(x)\n",
    "\n",
    "\n",
    "class ConvLinearModel(nn.Module):\n",
    "    def __init__(self, input_depth, input_length, output_length,avg_pool_size, kernel_size, stride, padding, dilation, dropout_rate, hidden_depth):\n",
    "        super(ConvLinearModel, self).__init__()\n",
    "        # 2D Average Pooling\n",
    "        self.avgpool2d = nn.AvgPool2d(kernel_size=avg_pool_size)\n",
    "\n",
    "        # Dropout\n",
    "        self.dropout = nn.Dropout(dropout_rate)\n",
    "\n",
    "        # Adjust the input depth and sequence length based on the 2D pooling\n",
    "        new_depth = input_depth // avg_pool_size[0]\n",
    "        new_seq_length = input_length // avg_pool_size[1]\n",
    "\n",
    "        # 1D Convolution\n",
    "        self.conv1 = nn.Conv1d(in_channels=new_depth, out_channels=hidden_depth, kernel_size=kernel_size, stride=stride, padding=padding, dilation=dilation)\n",
    "\n",
    "        # Calculate the output length after convolution\n",
    "        self.seq_length_after_conv = self.calculate_output_length(new_seq_length, kernel_size, stride, padding, dilation)\n",
    "\n",
    "        # Fully connected layer with calculated input size\n",
    "        self.fc = nn.Linear(hidden_depth * self.seq_length_after_conv, output_length)\n",
    "        self.softplus = nn.Softplus()\n",
    "\n",
    "    def calculate_output_length(self, input_length, kernel_size, stride, padding, dilation):\n",
    "        return ((input_length + 2 * padding - dilation * (kernel_size - 1) - 1) // stride) + 1\n",
    "\n",
    "    def forward(self, x):\n",
    "        # Reshape x to (batch_size, 1, input_depth, seq_length) for 2D avg pooling\n",
    "        x = x.view(x.size(0),1, x.size(2), x.size(1))\n",
    "        #print(x.shape)\n",
    "        x = self.avgpool2d(x)\n",
    "       # print(x.shape)\n",
    "\n",
    "        # Reshape back to (batch_size, new_depth, new_seq_length) for 1D convolution\n",
    "        x = x.view(x.size(0), -1, x.size(3))\n",
    "\n",
    "        # Apply 1D convolution\n",
    "        x = self.conv1(x)\n",
    "\n",
    "        # Apply dropout to the inputs of the final layer\n",
    "        x = self.dropout(x)\n",
    "\n",
    "        # Reshape x for the linear layer\n",
    "        x = x.view(x.size(0), -1)\n",
    "\n",
    "        # Pass through the fully connected layer\n",
    "        x = self.fc(x)\n",
    "\n",
    "        # Apply Softplus activation\n",
    "        x = self.softplus(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "\n",
    "#@title Model (backbone + head)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Putting it all together, the model consists of a backbone model\n",
    "and a decoder head (you can turn off head for embeddings only too).\n",
    "\n",
    "Here we use a simple head to do multi-classification, but\n",
    "can also swap the head to do next token prediction too.  We defer to the main\n",
    "HyenaDNA for that code, since pretraining with next token prediction isn't quite\n",
    "feasible on colab.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "class HyenaDNAModel(nn.Module):\n",
    "    def __init__(self, d_model: int, n_layer: int, d_inner: int, vocab_size: int,\n",
    "                 layer=None, attn_layer_idx=None, attn_cfg=None, max_position_embeddings=0,\n",
    "                 resid_dropout: float = 0.0, embed_dropout: float = 0.1,\n",
    "                 layer_norm_epsilon: float = 1e-5, initializer_cfg=None,residual_in_fp32=False,\n",
    "                 pad_vocab_size_multiple: int = 1, use_head=False, n_classes: int = 2,\n",
    "                 device=None, dtype=None, **kwargs) -> None:\n",
    "        factory_kwargs = {'device': device, 'dtype': dtype}\n",
    "        super().__init__()\n",
    "        if vocab_size % pad_vocab_size_multiple != 0:\n",
    "            vocab_size += pad_vocab_size_multiple - (vocab_size % pad_vocab_size_multiple)\n",
    "\n",
    "        self.use_head = use_head\n",
    "\n",
    "        # check if layer (config) has d_model (HF code differs from main Safari code)\n",
    "        if 'd_model' not in layer:\n",
    "            layer['d_model'] = d_model\n",
    "\n",
    "        self.backbone = LMBackbone(\n",
    "            d_model=d_model, n_layer=n_layer, d_inner=d_inner, vocab_size=vocab_size,\n",
    "            layer=layer, attn_layer_idx=attn_layer_idx, attn_cfg=attn_cfg,\n",
    "            max_position_embeddings=max_position_embeddings,\n",
    "            resid_dropout=resid_dropout, embed_dropout=embed_dropout,\n",
    "            layer_norm_epsilon=layer_norm_epsilon,\n",
    "            initializer_cfg=initializer_cfg, residual_in_fp32=residual_in_fp32,\n",
    "            **factory_kwargs, **kwargs\n",
    "        )\n",
    "\n",
    "        # we only need a head if doing classification, otherwise we'll use the\n",
    "        # hidden states as embeddings\n",
    "        if self.use_head:\n",
    "\n",
    "            #self.head = SequenceDecoder(d_model=d_model, d_output=1, l_output=230, mode='pool')\n",
    "            input_depth = 256\n",
    "            input_length = 32768 #160000\n",
    "            output_length = (15*2+1)# *2 only one condition now #(57*2+1)*2\n",
    "            avg_pool_size= (1,128) #(depth_axis, seqlen_axis)\n",
    "            kernel_size = 9\n",
    "            dilation=2\n",
    "            stride = 2\n",
    "            padding= (dilation * (kernel_size - 1)) // 2\n",
    "            dropout_rate=.3\n",
    "            hidden_depth=10\n",
    "    \n",
    "\n",
    "            self.head = ConvLinearModel(input_depth, input_length, output_length,avg_pool_size, kernel_size, stride, padding, dilation, dropout_rate, hidden_depth)\n",
    "        \n",
    "        # Initialize weights and apply final processing\n",
    "        self.apply(partial(_init_weights, n_layer=n_layer,\n",
    "                           **(initializer_cfg if initializer_cfg is not None else {})))\n",
    "\n",
    "        # if self.use_head:\n",
    "        #     self.tie_weights()\n",
    "\n",
    "    # def tie_weights(self):\n",
    "    #     self.head.weight = self.backbone.embeddings.word_embeddings.weight\n",
    "\n",
    "    def forward(self, input_ids, position_ids=None, state=None): # state for the repo interface\n",
    "        hidden_states = self.backbone(input_ids, position_ids=position_ids)\n",
    "        if self.use_head:\n",
    "            return self.head(hidden_states)\n",
    "\n",
    "        else:\n",
    "            return hidden_states\n",
    "\n",
    "\n",
    "#@title Huggingface Pretrained Wrapper\n",
    "# for Huggingface integration, we use a wrapper class around the model\n",
    "# to load weights\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "import re\n",
    "\n",
    "def inject_substring(orig_str):\n",
    "    \"\"\"Hack to handle matching keys between models trained with and without\n",
    "    gradient checkpointing.\"\"\"\n",
    "\n",
    "    # modify for mixer keys\n",
    "    pattern = r\"\\.mixer\"\n",
    "    injection = \".mixer.layer\"\n",
    "\n",
    "    modified_string = re.sub(pattern, injection, orig_str)\n",
    "\n",
    "    # modify for mlp keys\n",
    "    pattern = r\"\\.mlp\"\n",
    "    injection = \".mlp.layer\"\n",
    "\n",
    "    modified_string = re.sub(pattern, injection, modified_string)\n",
    "\n",
    "    return modified_string\n",
    "\n",
    "def load_weights(scratch_dict, pretrained_dict, checkpointing=False):\n",
    "    \"\"\"Loads pretrained (backbone only) weights into the scratch state dict.\n",
    "\n",
    "    scratch_dict: dict, a state dict from a newly initialized HyenaDNA model\n",
    "    pretrained_dict: dict, a state dict from the pretrained ckpt\n",
    "    checkpointing: bool, whether the gradient checkpoint flag was used in the\n",
    "    pretrained model ckpt. This slightly changes state dict keys, so we patch\n",
    "    that if used.\n",
    "\n",
    "    return:\n",
    "    dict, a state dict with the pretrained weights loaded (head is scratch)\n",
    "\n",
    "    # loop thru state dict of scratch\n",
    "    # find the corresponding weights in the loaded model, and set it\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    # need to do some state dict \"surgery\"\n",
    "    for key, value in scratch_dict.items():\n",
    "        if 'backbone' in key:\n",
    "            # the state dicts differ by one prefix, '.model', so we add that\n",
    "            key_loaded = 'model.' + key\n",
    "            # breakpoint()\n",
    "            # need to add an extra \".layer\" in key\n",
    "            if checkpointing:\n",
    "                key_loaded = inject_substring(key_loaded)\n",
    "            try:\n",
    "                scratch_dict[key] = pretrained_dict[key_loaded]\n",
    "            except:\n",
    "                raise Exception('key mismatch in the state dicts!')\n",
    "\n",
    "    # scratch_dict has been updated\n",
    "    return scratch_dict\n",
    "\n",
    "def download_files(url, output_dir):\n",
    "    # Create the output directory if it doesn't exist\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "    \n",
    "    # Download the files\n",
    "    response = requests.get(url, stream=True)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        # Get the filename from the URL\n",
    "        filename = os.path.join(output_dir, url.split('/')[-1])\n",
    "        \n",
    "        # Write the downloaded content to the file\n",
    "        with open(filename, 'wb') as f:\n",
    "            shutil.copyfileobj(response.raw, f)\n",
    "        \n",
    "        return filename\n",
    "    else:\n",
    "        raise ValueError(f\"Failed to download files from {url}\")\n",
    "\n",
    "def download_pretrained_model(model_name, output_dir):\n",
    "    hf_url = f'https://huggingface.co/LongSafari/{model_name}/resolve/main/config.json'\n",
    "    config_file = download_files(hf_url, output_dir)\n",
    "    config = json.load(open(config_file))\n",
    "    \n",
    "    weights_url = f'https://huggingface.co/LongSafari/{model_name}/resolve/main/weights.ckpt'\n",
    "    weights_file = download_files(weights_url, output_dir)\n",
    "    \n",
    "    return config, weights_file\n",
    "\n",
    "\n",
    "class HyenaDNAPreTrainedModel(PreTrainedModel):\n",
    "    \"\"\"\n",
    "    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained\n",
    "    models.\n",
    "    \"\"\"\n",
    "    base_model_prefix = \"hyenadna\"\n",
    "\n",
    "    def __init__(self, config):\n",
    "        pass\n",
    "\n",
    "    def forward(self, input_ids, **kwargs):\n",
    "        return self.model(input_ids, **kwargs)\n",
    "\n",
    "    @classmethod\n",
    "    def from_pretrained(cls,\n",
    "                        path,\n",
    "                        model_name,\n",
    "                        download=False,\n",
    "                        config=None,\n",
    "                        device='cpu',\n",
    "                        use_head=False,\n",
    "                        n_classes=2,\n",
    "                      ):\n",
    "        # first check if it is a local path\n",
    "        pretrained_model_name_or_path = os.path.join(path, model_name)\n",
    "        if os.path.isdir(pretrained_model_name_or_path) and download == False:\n",
    "            if config is None:\n",
    "                config = json.load(open(os.path.join(pretrained_model_name_or_path, 'config.json')))\n",
    "        else:\n",
    "            # hf_url = f'https://huggingface.co/LongSafari/{model_name}'\n",
    "            # print(\"here again\")\n",
    "            # subprocess.run(f'rm -rf {pretrained_model_name_or_path}', shell=True)\n",
    "            # command = f'mkdir -p {path} && cd {path} && git lfs install && git clone {hf_url}'\n",
    "            # subprocess.run(command, shell=True)\n",
    "            config, weights_file = download_pretrained_model(model_name, pretrained_model_name_or_path)\n",
    "\n",
    "            if config is None:\n",
    "                config = json.load(open(os.path.join(pretrained_model_name_or_path, 'config.json')))\n",
    "\n",
    "        scratch_model = HyenaDNAModel(**config, use_head=use_head, n_classes=n_classes)  # the new model format\n",
    "        loaded_ckpt = torch.load(\n",
    "            os.path.join(pretrained_model_name_or_path, 'weights.ckpt'),\n",
    "            map_location=torch.device(device)\n",
    "        )\n",
    "\n",
    "        # need to load weights slightly different if using gradient checkpointing\n",
    "        if config.get(\"checkpoint_mixer\", False):\n",
    "            checkpointing = config[\"checkpoint_mixer\"] == True or config[\"checkpoint_mixer\"] == True\n",
    "        else:\n",
    "            checkpointing = False\n",
    "\n",
    "        # grab state dict from both and load weights\n",
    "        state_dict = load_weights(scratch_model.state_dict(), loaded_ckpt['state_dict'], checkpointing=checkpointing)\n",
    "\n",
    "        # scratch model has now been updated\n",
    "        scratch_model.load_state_dict(state_dict)\n",
    "        print(\"Loaded pretrained weights ok!\")\n",
    "        return scratch_model\n",
    "\n",
    "#@title Tokenizer\n",
    "\n",
    "\"\"\"\n",
    "Just a simple character level tokenizer.\n",
    "\n",
    "From: https://github.com/dariush-bahrami/character-tokenizer/blob/master/charactertokenizer/core.py\n",
    "\n",
    "CharacterTokenzier for Hugging Face Transformers.\n",
    "This is heavily inspired from CanineTokenizer in transformers package.\n",
    "\"\"\"\n",
    "import json\n",
    "import os\n",
    "from pathlib import Path\n",
    "from typing import Dict, List, Optional, Sequence, Union\n",
    "\n",
    "from transformers.tokenization_utils import AddedToken, PreTrainedTokenizer\n",
    "\n",
    "class CharacterTokenizer(PreTrainedTokenizer):\n",
    "    def __init__(self, characters: Sequence[str], model_max_length: int, padding_side: str='left', **kwargs):\n",
    "        \"\"\"Character tokenizer for Hugging Face transformers.\n",
    "        Args:\n",
    "            characters (Sequence[str]): List of desired characters. Any character which\n",
    "                is not included in this list will be replaced by a special token called\n",
    "                [UNK] with id=6. Following are list of all of the special tokens with\n",
    "                their corresponding ids:\n",
    "                    \"[CLS]\": 0\n",
    "                    \"[SEP]\": 1\n",
    "                    \"[BOS]\": 2\n",
    "                    \"[MASK]\": 3\n",
    "                    \"[PAD]\": 4\n",
    "                    \"[RESERVED]\": 5\n",
    "                    \"[UNK]\": 6\n",
    "                an id (starting at 7) will be assigned to each character.\n",
    "            model_max_length (int): Model maximum sequence length.\n",
    "        \"\"\"\n",
    "        self.characters = characters\n",
    "        self.model_max_length = model_max_length\n",
    "        bos_token = AddedToken(\"[BOS]\", lstrip=False, rstrip=False)\n",
    "        eos_token = AddedToken(\"[SEP]\", lstrip=False, rstrip=False)\n",
    "        sep_token = AddedToken(\"[SEP]\", lstrip=False, rstrip=False)\n",
    "        cls_token = AddedToken(\"[CLS]\", lstrip=False, rstrip=False)\n",
    "        pad_token = AddedToken(\"[PAD]\", lstrip=False, rstrip=False)\n",
    "        unk_token = AddedToken(\"[UNK]\", lstrip=False, rstrip=False)\n",
    "\n",
    "        mask_token = AddedToken(\"[MASK]\", lstrip=True, rstrip=False)\n",
    "\n",
    "        super().__init__(\n",
    "            bos_token=bos_token,\n",
    "            eos_token=sep_token,\n",
    "            sep_token=sep_token,\n",
    "            cls_token=cls_token,\n",
    "            pad_token=pad_token,\n",
    "            mask_token=mask_token,\n",
    "            unk_token=unk_token,\n",
    "            add_prefix_space=False,\n",
    "            model_max_length=model_max_length,\n",
    "            padding_side=padding_side,\n",
    "            **kwargs,\n",
    "        )\n",
    "\n",
    "        self._vocab_str_to_int = {\n",
    "            \"[CLS]\": 0,\n",
    "            \"[SEP]\": 1,\n",
    "            \"[BOS]\": 2,\n",
    "            \"[MASK]\": 3,\n",
    "            \"[PAD]\": 4,\n",
    "            \"[RESERVED]\": 5,\n",
    "            \"[UNK]\": 6,\n",
    "            **{ch: i + 7 for i, ch in enumerate(characters)},\n",
    "        }\n",
    "        self._vocab_int_to_str = {v: k for k, v in self._vocab_str_to_int.items()}\n",
    "\n",
    "    @property\n",
    "    def vocab_size(self) -> int:\n",
    "        return len(self._vocab_str_to_int)\n",
    "\n",
    "    def _tokenize(self, text: str) -> List[str]:\n",
    "        return list(text)\n",
    "\n",
    "    def _convert_token_to_id(self, token: str) -> int:\n",
    "        return self._vocab_str_to_int.get(token, self._vocab_str_to_int[\"[UNK]\"])\n",
    "\n",
    "    def _convert_id_to_token(self, index: int) -> str:\n",
    "        return self._vocab_int_to_str[index]\n",
    "\n",
    "    def convert_tokens_to_string(self, tokens):\n",
    "        return \"\".join(tokens)\n",
    "\n",
    "    def build_inputs_with_special_tokens(\n",
    "        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None\n",
    "    ) -> List[int]:\n",
    "        sep = [self.sep_token_id]\n",
    "        cls = [self.cls_token_id]\n",
    "        result = cls + token_ids_0 + sep\n",
    "        if token_ids_1 is not None:\n",
    "            result += token_ids_1 + sep\n",
    "        return result\n",
    "\n",
    "    def get_special_tokens_mask(\n",
    "        self,\n",
    "        token_ids_0: List[int],\n",
    "        token_ids_1: Optional[List[int]] = None,\n",
    "        already_has_special_tokens: bool = False,\n",
    "    ) -> List[int]:\n",
    "        if already_has_special_tokens:\n",
    "            return super().get_special_tokens_mask(\n",
    "                token_ids_0=token_ids_0,\n",
    "                token_ids_1=token_ids_1,\n",
    "                already_has_special_tokens=True,\n",
    "            )\n",
    "\n",
    "        result = [1] + ([0] * len(token_ids_0)) + [1]\n",
    "        if token_ids_1 is not None:\n",
    "            result += ([0] * len(token_ids_1)) + [1]\n",
    "        return result\n",
    "\n",
    "    def create_token_type_ids_from_sequences(\n",
    "        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None\n",
    "    ) -> List[int]:\n",
    "        sep = [self.sep_token_id]\n",
    "        cls = [self.cls_token_id]\n",
    "\n",
    "        result = len(cls + token_ids_0 + sep) * [0]\n",
    "        if token_ids_1 is not None:\n",
    "            result += len(token_ids_1 + sep) * [1]\n",
    "        return result\n",
    "\n",
    "    def get_config(self) -> Dict:\n",
    "        return {\n",
    "            \"char_ords\": [ord(ch) for ch in self.characters],\n",
    "            \"model_max_length\": self.model_max_length,\n",
    "        }\n",
    "\n",
    "    @classmethod\n",
    "    def from_config(cls, config: Dict) -> \"CharacterTokenizer\":\n",
    "        cfg = {}\n",
    "        cfg[\"characters\"] = [chr(i) for i in config[\"char_ords\"]]\n",
    "        cfg[\"model_max_length\"] = config[\"model_max_length\"]\n",
    "        return cls(**cfg)\n",
    "\n",
    "    def save_pretrained(self, save_directory: Union[str, os.PathLike], **kwargs):\n",
    "        cfg_file = Path(save_directory) / \"tokenizer_config.json\"\n",
    "        cfg = self.get_config()\n",
    "        with open(cfg_file, \"w\") as f:\n",
    "            json.dump(cfg, f, indent=4)\n",
    "\n",
    "    @classmethod\n",
    "    def from_pretrained(cls, save_directory: Union[str, os.PathLike], **kwargs):\n",
    "        cfg_file = Path(save_directory) / \"tokenizer_config.json\"\n",
    "        with open(cfg_file) as f:\n",
    "            cfg = json.load(f)\n",
    "        return cls.from_config(cfg)\n",
    "\n",
    "\n",
    "#@title GenomicBenchmark dataset\n",
    "\n",
    "\"\"\"\n",
    "The GenomicBenchmarks dataset will automatically download to /contents on colab.\n",
    "There are 8 datasets to choose from.\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "from random import random\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "#from genomic_benchmarks.loc2seq import download_dataset\n",
    "#from genomic_benchmarks.data_check import is_downloaded\n",
    "\n",
    "\n",
    "# helper functions\n",
    "def exists(val):\n",
    "    return val is not None\n",
    "\n",
    "def coin_flip():\n",
    "    return random() > 0.5\n",
    "\n",
    "\n",
    "string_complement_map = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'a': 't', 'c': 'g', 'g': 'c', 't': 'a'}\n",
    "# augmentation\n",
    "def string_reverse_complement(seq):\n",
    "    rev_comp = ''\n",
    "    for base in seq[::-1]:\n",
    "        if base in string_complement_map:\n",
    "            rev_comp += string_complement_map[base]\n",
    "        # if bp not complement map, use the same bp\n",
    "        else:\n",
    "            rev_comp += base\n",
    "    return rev_comp\n",
    "\n",
    "\n",
    "from random import random\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "from genomic_benchmarks.loc2seq import download_dataset\n",
    "from genomic_benchmarks.data_check import is_downloaded\n",
    "\n",
    "\n",
    "class GenomicBenchmarkDataset(torch.utils.data.Dataset):\n",
    "\n",
    "    '''\n",
    "    Loop thru bed file, retrieve (chr, start, end), query fasta file for sequence.\n",
    "    Returns a generator that retrieves the sequence.\n",
    "\n",
    "    Genomic Benchmarks Dataset, from:\n",
    "    https://github.com/ML-Bioinfo-CEITEC/genomic_benchmarks\n",
    "\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        split,\n",
    "        max_length,\n",
    "        dataset_name='human_enhancers_cohn',\n",
    "        d_output=2, # default binary classification\n",
    "        dest_path=\"./data/\", # default for colab\n",
    "        tokenizer=None,\n",
    "        tokenizer_name=None,\n",
    "        use_padding=None,\n",
    "        add_eos=False,\n",
    "        rc_aug=False,\n",
    "        return_augs=False,\n",
    "    ):\n",
    "\n",
    "        self.max_length = max_length\n",
    "        self.use_padding = use_padding\n",
    "        self.tokenizer_name = tokenizer_name\n",
    "        self.tokenizer = tokenizer\n",
    "        self.return_augs = return_augs\n",
    "        self.add_eos = add_eos\n",
    "        self.d_output = d_output  # needed for decoder to grab\n",
    "        self.rc_aug = rc_aug\n",
    "\n",
    "        if not is_downloaded(dataset_name, cache_path=dest_path):\n",
    "            print(\"downloading {} to {}\".format(dataset_name, dest_path))\n",
    "            download_dataset(dataset_name, version=0, dest_path=dest_path)\n",
    "        else:\n",
    "            print(\"already downloaded {}-{}\".format(split, dataset_name))\n",
    "\n",
    "        # use Path object\n",
    "        base_path = Path(dest_path) / dataset_name / split\n",
    "\n",
    "        self.all_paths = []\n",
    "        self.all_labels = []\n",
    "        label_mapper = {}\n",
    "\n",
    "        for i, x in enumerate(base_path.iterdir()):\n",
    "            label_mapper[x.stem] = i\n",
    "\n",
    "        for label_type in label_mapper.keys():\n",
    "            for x in (base_path / label_type).iterdir():\n",
    "                self.all_paths.append(x)\n",
    "                self.all_labels.append(label_mapper[label_type])\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.all_paths)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        txt_path = self.all_paths[idx]\n",
    "        with open(txt_path, \"r\") as f:\n",
    "            content = f.read()\n",
    "        x = content\n",
    "        y = self.all_labels[idx]\n",
    "\n",
    "        # apply rc_aug here if using\n",
    "        if self.rc_aug and coin_flip():\n",
    "            x = string_reverse_complement(x)\n",
    "\n",
    "        seq = self.tokenizer(x,\n",
    "            add_special_tokens=False,\n",
    "            padding=\"max_length\" if self.use_padding else None,\n",
    "            max_length=self.max_length,\n",
    "            truncation=True,\n",
    "        )  # add cls and eos token (+2)\n",
    "        seq = seq[\"input_ids\"]  # get input_ids\n",
    "\n",
    "        # need to handle eos here\n",
    "        if self.add_eos:\n",
    "            # append list seems to be faster than append tensor\n",
    "            seq.append(self.tokenizer.sep_token_id)\n",
    "\n",
    "        # convert to tensor\n",
    "        seq = torch.LongTensor(seq)\n",
    "\n",
    "        # need to wrap in list\n",
    "        target = torch.LongTensor([y])\n",
    "\n",
    "        return seq, target\n",
    "\n",
    "############################## Our Datasets ###############\n",
    "\n",
    "import torch\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import zlib\n",
    "import sys\n",
    "\n",
    "\n",
    "class CLASTER_HYENA_Dataset(torch.utils.data.Dataset):\n",
    "\n",
    "    '''\n",
    "    Returns a generator that retrieves the sequence.\n",
    "    '''\n",
    "\n",
    "    def __init__(\n",
    "        self,\n",
    "        fasta_file,\n",
    "        csv_file,\n",
    "        max_length,\n",
    "        target_columns,\n",
    "        d_output=None,\n",
    "        tokenizer=None,\n",
    "        tokenizer_name=None,\n",
    "        use_padding=None,\n",
    "        add_eos=False,\n",
    "        rc_aug=False,\n",
    "        return_augs=False,\n",
    "    ):\n",
    "\n",
    "        self.max_length = max_length\n",
    "        self.use_padding = use_padding\n",
    "        self.tokenizer_name = tokenizer_name\n",
    "        self.tokenizer = tokenizer\n",
    "        self.return_augs = return_augs\n",
    "        self.add_eos = add_eos\n",
    "        self.rc_aug = rc_aug\n",
    "\n",
    "        # Read CSV file\n",
    "        self.targets_df = pd.read_csv(csv_file).set_index('ID')\n",
    "\n",
    "        # Extract continuous target vectors\n",
    "        self.continuous_targets = self.targets_df[target_columns].values\n",
    "\n",
    "        # Read FASTA file and store sequences and corresponding IDs in a dictionary\n",
    "        self.sequences = {}\n",
    "\n",
    "        for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            header_parts = record.description.split(\"::\")\n",
    "            seq_id = header_parts[0]  # Remove '>'\n",
    "            sequence = str(record.seq)\n",
    "\n",
    "\n",
    "            # Create reverse complement and store it with the \"_flipped\" identifier\n",
    "            reverse_complement = string_reverse_complement(sequence)\n",
    "            flipped_seq_id = seq_id + \"_flipped\"\n",
    "            # Crop distances:\n",
    "            crop_dist = (160000-32768)//2\n",
    "            # Store the sequence with its ID\n",
    "            self.sequences[seq_id] = sequence[crop_dist:-crop_dist]\n",
    "            self.sequences[flipped_seq_id] = reverse_complement[crop_dist:-crop_dist]\n",
    "\n",
    "        # Extract IDs from the target CSV file (assuming the ID column is named 'ID')\n",
    "        self.target_ids = self.targets_df.index\n",
    "\n",
    "        # Match sequence IDs with target IDs and store their indices\n",
    "        self.matched_indices = [i for i, seq_id in enumerate(self.target_ids) if seq_id in self.sequences]\n",
    "\n",
    "        # Determine the number of output dimensions if not provided\n",
    "        if d_output is None:\n",
    "            self.d_output = len(target_columns)\n",
    "        else:\n",
    "            self.d_output = d_output\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.matched_indices)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        # Get sequence ID and retrieve the sequence\n",
    "        seq_id = self.target_ids[self.matched_indices[idx]]\n",
    "        sequence = self.sequences[seq_id]\n",
    "\n",
    "        # Get continuous target vector\n",
    "        target = self.continuous_targets[self.matched_indices[idx]]\n",
    "\n",
    "        # Tokenize sequence\n",
    "        encoded_sequence = self.tokenizer(\n",
    "            sequence,\n",
    "            add_special_tokens=False,\n",
    "            padding=\"max_length\" if self.use_padding else None,\n",
    "            max_length=self.max_length,\n",
    "            truncation=True,\n",
    "        )[\"input_ids\"]    # Get input_ids\n",
    "\n",
    "        # Need to handle eos here\n",
    "        if self.add_eos:\n",
    "            # Append list seems to be faster than append tensor\n",
    "            encoded_sequence.append(self.tokenizer.sep_token_id)\n",
    "\n",
    "        # Convert to tensor\n",
    "        encoded_sequence = torch.LongTensor(encoded_sequence)\n",
    "\n",
    "        # Convert target to tensor\n",
    "        target = torch.FloatTensor(target)\n",
    "\n",
    "        return encoded_sequence, target\n",
    "\n",
    "\n",
    "import torch.optim as optim\n",
    "import logging\n",
    "\n",
    "LOG_FILENAME = \"../checkpoints/hyenadna_32k.log\"\n",
    "logging.basicConfig(filename=LOG_FILENAME, level=logging.INFO)  \n",
    "\n",
    "\"\"\"\n",
    "We provide simple training code for the GenomicBenchmark datasets.\n",
    "\"\"\"\n",
    "\n",
    "def train(model, device, train_loader, optimizer, epoch, loss_fn):\n",
    "    \"\"\"Training loop.\"\"\"\n",
    "    model.train()\n",
    "    train_loss = 0\n",
    "    for batch_idx, (data, target) in enumerate(train_loader):\n",
    "        data, target = data.to(device), target.to(device)\n",
    "        optimizer.zero_grad()\n",
    "        output = model(data)\n",
    "        loss = loss_fn(output, target.squeeze())\n",
    "        train_loss += loss_fn(output, target.squeeze()).item()  # sum up batch loss\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "\n",
    "    train_loss /= len(train_loader.dataset)\n",
    "    logging.info('Train Epoch: {}\\tAverage Loss: {:.6f}'.format(\n",
    "        epoch, train_loss))\n",
    "    torch.save(model,'../checkpoints/hyenadna-small-32k-seqlen/model_32k.pt')\n",
    "\n",
    "def test(model, device, test_loader, loss_fn):\n",
    "    \"\"\"Test loop.\"\"\"\n",
    "    model.eval()\n",
    "    test_loss = 0\n",
    "    correct = 0\n",
    "\n",
    "    targets = np.array([])\n",
    "    predictions = np.array([])\n",
    "\n",
    "    with torch.no_grad():\n",
    "        for data, target in test_loader:\n",
    "            data, target = data.to(device), target.to(device)\n",
    "            output = model(data)\n",
    "            test_loss += loss_fn(output, target.squeeze()).item()  # sum up batch loss\n",
    "            targets = np.concatenate([targets,torch.flatten(torch.clone(target)).cpu().numpy()])\n",
    "            predictions = np.concatenate([predictions,torch.flatten(torch.clone(output)).cpu().numpy()]) \n",
    "            \n",
    "    test_loss /= len(test_loader.dataset)\n",
    "\n",
    "    np.save(\"Hyena_finetunned_targets.npy\",targets)\n",
    "    np.save(\"Hyena_finetunned_predictions.npy\",predictions)\n",
    "\n",
    "    logging.info('\\nTest set: Average loss: {:.6f}\\t'.format(\n",
    "        test_loss))\n",
    "\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "import nvidia_smi\n",
    "\n",
    "def run_train():\n",
    "    '''\n",
    "    Main entry point for training.  Select the dataset name and metadata, as\n",
    "    well as model and training args, and you're off to the genomic races!\n",
    "\n",
    "    ### GenomicBenchmarks Metadata\n",
    "    # there are 8 datasets in this suite, choose 1 at a time, with their corresponding settings\n",
    "    # name                                num_seqs        num_classes     median len    std\n",
    "    # dummy_mouse_enhancers_ensembl       1210            2               2381          984.4\n",
    "    # demo_coding_vs_intergenomic_seqs    100_000         2               200           0\n",
    "    # demo_human_or_worm                  100_000         2               200           0\n",
    "    # human_enhancers_cohn                27791           2               500           0\n",
    "    # human_enhancers_ensembl             154842          2               269           122.6\n",
    "    # human_ensembl_regulatory            289061          3               401           184.3\n",
    "    # human_nontata_promoters             36131           2               251           0\n",
    "    # human_ocr_ensembl                   174756          2               315           108.1\n",
    "\n",
    "    '''\n",
    "    # experiment settings:\n",
    "    num_epochs = 100  # ~100 seems fine\n",
    "    max_length = 500  # max len of sequence of dataset (of what you want)\n",
    "    use_padding = True\n",
    "    dataset_name = 'human_enhancers_cohn'\n",
    "    batch_size = 256\n",
    "    learning_rate = 6e-4  # good default for Hyena\n",
    "    rc_aug = True  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "    weight_decay = 0.1\n",
    "\n",
    "    # for fine-tuning, only the 'tiny' model can fit on colab\n",
    "    pretrained_model_name = 'hyenadna-tiny-1k-seqlen'  # use None if training from scratch\n",
    "\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = True\n",
    "    n_classes = 2\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one by default\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    print(\"Using device:\", device)\n",
    "    logging.info(f\"Using device: {device}\")\n",
    "\n",
    "    # instantiate the model (pretrained here)\n",
    "    if pretrained_model_name in ['hyenadna-tiny-1k-seqlen']:\n",
    "        # use the pretrained Huggingface wrapper instead\n",
    "        model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "            './checkpoints',\n",
    "            pretrained_model_name,\n",
    "            download=False,\n",
    "            config=backbone_cfg,\n",
    "            device=device,\n",
    "            use_head=use_head,\n",
    "            n_classes=n_classes,\n",
    "        )\n",
    "\n",
    "    # from scratch\n",
    "    else:\n",
    "        model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "\n",
    "    # create tokenizer\n",
    "    tokenizer = CharacterTokenizer(\n",
    "        characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "        model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "        add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "        padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    )\n",
    "\n",
    "    # create datasets\n",
    "    ds_train = GenomicBenchmarkDataset(\n",
    "        max_length = max_length,\n",
    "        use_padding = use_padding,\n",
    "        split = 'train',\n",
    "        tokenizer=tokenizer,\n",
    "        dataset_name=dataset_name,\n",
    "        rc_aug=rc_aug,\n",
    "        add_eos=add_eos,\n",
    "    )\n",
    "\n",
    "    ds_test = GenomicBenchmarkDataset(\n",
    "        max_length = max_length,\n",
    "        use_padding = use_padding,\n",
    "        split = 'test',\n",
    "        tokenizer=tokenizer,\n",
    "        dataset_name=dataset_name,\n",
    "        rc_aug=rc_aug,\n",
    "        add_eos=add_eos,\n",
    "    )\n",
    "\n",
    "\n",
    "    train_loader = DataLoader(ds_train, batch_size=batch_size, shuffle=True)\n",
    "    test_loader = DataLoader(ds_test, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "    # loss function\n",
    "    loss_fn = nn.CrossEntropyLoss()\n",
    "\n",
    "    # create optimizer\n",
    "    optimizer = optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)\n",
    "\n",
    "    model.to(device)\n",
    "\n",
    "    for epoch in range(num_epochs):\n",
    "        train(model, device, train_loader, optimizer, epoch, loss_fn)\n",
    "        test(model, device, test_loader, loss_fn)\n",
    "        optimizer.step()\n",
    "\n",
    "\n",
    "def run_train_CLASTER_HYENA():\n",
    "\n",
    "    '''\n",
    "    Main entry point for training.  Select the dataset name and metadata, as\n",
    "    well as model and training args, and you're off to the genomic races!\n",
    "\n",
    "    '''\n",
    "    # experiment settings:\n",
    "    ONLY_TEST = True\n",
    "    num_epochs = 200 #50  # ~100 seems fine\n",
    "    max_length = 32768 #160000  # max len of sequence of dataset (of what you want)\n",
    "    use_padding = True\n",
    "    batch_size = 16\n",
    "    learning_rate = 5e-7  # Super small batch size -> lower learning rate\n",
    "    rc_aug = True  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "    weight_decay = 0.1\n",
    "    n_classes=2\n",
    "    # create datasets\n",
    "\n",
    "    train_fasta_file = \"../inputs/DNA_sequences/training_boundaries.fasta\"\n",
    "    test_fasta_file = \"../inputs/DNA_sequences/test_boundaries.fasta\"\n",
    "    train_csv_file = \"../targets/training_targets.csv\"\n",
    "    test_csv_file = \"../targets/test_targets.csv\"\n",
    "\n",
    "    #max_length = 160000\n",
    "    N_BINS = 15 #57\n",
    "    target_columns = [f\"{i}{cond}\" for cond in [\"_ctrl\"] for i in range(-N_BINS, N_BINS+1)]\n",
    "\n",
    "    # for fine-tuning, only the 'tiny' model can fit on colab\n",
    "    pretrained_model_name = 'hyenadna-small-32k-seqlen' #'hyenadna-medium-160k-seqlen'  # use None if training from scratch\n",
    "    #model_name_or_path = './checkpoints/hyenadna-medium-160k-seqlen-custom/'\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = True\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one by default\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    print(\"Using device:\", device)\n",
    "\n",
    "\n",
    "    #instantiate the model (pretrained here)\n",
    "    if pretrained_model_name in ['hyenadna-small-32k-seqlen']: #['hyenadna-medium-160k-seqlen']:\n",
    "        # use the pretrained Huggingface wrapper instead\n",
    "        print('Here')\n",
    "        model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "            '../checkpoints/',\n",
    "            pretrained_model_name,\n",
    "            download=True,\n",
    "            config=backbone_cfg,\n",
    "            device=device,\n",
    "            use_head=use_head,\n",
    "            n_classes=n_classes,\n",
    "        )\n",
    "\n",
    "    # from scratch\n",
    "    else:\n",
    "        model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "\n",
    "    # backbone_cfg = json.load(open(os.path.join(model_name_or_path, 'config.json')))\n",
    "    # model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "    #print(torch.cuda.memory_allocated(device=device))\n",
    "\n",
    "    # create tokenizer\n",
    "    tokenizer = CharacterTokenizer(\n",
    "        characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "        model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "        add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "        padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    )\n",
    "\n",
    "\n",
    "    ds_train = CLASTER_HYENA_Dataset(train_fasta_file,\n",
    "        train_csv_file,\n",
    "        max_length,\n",
    "        target_columns,\n",
    "        d_output=None, # No need for default here, it will be inferred from target_columns\n",
    "        tokenizer=tokenizer,\n",
    "        tokenizer_name=None,\n",
    "        use_padding=use_padding,\n",
    "        add_eos=add_eos,\n",
    "        rc_aug=rc_aug,\n",
    "        return_augs=False)\n",
    "\n",
    "    ds_test = CLASTER_HYENA_Dataset(test_fasta_file,\n",
    "        test_csv_file,\n",
    "        max_length,\n",
    "        target_columns,\n",
    "        d_output=None, # No need for default here, it will be inferred from target_columns\n",
    "        tokenizer=tokenizer,\n",
    "        tokenizer_name=None,\n",
    "        use_padding=use_padding,\n",
    "        add_eos=add_eos,\n",
    "        rc_aug=rc_aug,\n",
    "        return_augs=False)\n",
    "\n",
    "\n",
    "    train_loader = DataLoader(ds_train, batch_size=batch_size, shuffle=True)\n",
    "    test_loader = DataLoader(ds_test, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "    # loss function\n",
    "    loss_fn = nn.SmoothL1Loss()\n",
    "\n",
    "    # create optimizer\n",
    "    optimizer = optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)\n",
    "\n",
    "    model.to(device)\n",
    "\n",
    "    # nvidia_smi.nvmlInit()\n",
    "    # deviceCount = nvidia_smi.nvmlDeviceGetCount()\n",
    "    # for i in range(deviceCount):\n",
    "    #     handle = nvidia_smi.nvmlDeviceGetHandleByIndex(i)\n",
    "    #     util = nvidia_smi.nvmlDeviceGetUtilizationRates(handle)\n",
    "    #     mem = nvidia_smi.nvmlDeviceGetMemoryInfo(handle)\n",
    "    #     print(f\"|Device {i}| Mem Free: {mem.free/1024**2:5.2f}MB / {mem.total/1024**2:5.2f}MB | gpu-util: {util.gpu/100.0:3.1%} | gpu-mem: {util.memory/100.0:3.1%} |\")\n",
    "\n",
    "    # print(torch.cuda.memory_allocated(device=device))\n",
    "\n",
    "    if not ONLY_TEST:\n",
    "        for epoch in range(num_epochs):\n",
    "            train(model, device, train_loader, optimizer, epoch, loss_fn)\n",
    "            test(model, device, test_loader, loss_fn)\n",
    "            optimizer.step()\n",
    "\n",
    "    else:\n",
    "        model = torch.load('../checkpoints/hyenadna-small-32k-seqlen/model_32k.pt')\n",
    "        model.eval()\n",
    "        test(model, device, test_loader, loss_fn)\n",
    "\n",
    "\n",
    "######################################################################################################################################\n",
    "#@title Single example\n",
    "import json\n",
    "import os\n",
    "import subprocess\n",
    "import transformers\n",
    "from transformers import PreTrainedModel, AutoModelForCausalLM, PretrainedConfig\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "\n",
    "def inference_single_sequential(path: Path, filename: str, savepath: Path, checkpoint_folder: str ):\n",
    "\n",
    "    '''\n",
    "    this selects which backbone to use, and grabs weights/ config from HF\n",
    "    4 options:\n",
    "      'hyenadna-tiny-1k-seqlen'   # fine-tune on colab ok\n",
    "      'hyenadna-small-32k-seqlen'\n",
    "      'hyenadna-medium-160k-seqlen'  # inference only on colab\n",
    "      'hyenadna-medium-450k-seqlen'  # inference only on colab\n",
    "      'hyenadna-large-1m-seqlen'  # inference only on colab\n",
    "    '''\n",
    "\n",
    "    # you only need to select which model to use here, we'll do the rest!\n",
    "    pretrained_model_name = 'hyenadna-medium-160k-seqlen' #'hyenadna-tiny-1k-seqlen'# 'hyenadna-medium-450k-seqlen'\n",
    "\n",
    "    max_lengths = {\n",
    "        'hyenadna-tiny-1k-seqlen': 1024,\n",
    "        'hyenadna-small-32k-seqlen': 32768,\n",
    "        'hyenadna-medium-160k-seqlen': 160000,\n",
    "        'hyenadna-medium-450k-seqlen': 450000,  # T4 up to here\n",
    "        'hyenadna-large-1m-seqlen': 1_000_000,  # only A100 (paid tier)\n",
    "    }\n",
    "\n",
    "    max_length = max_lengths[pretrained_model_name]  # auto selects\n",
    "\n",
    "    # data settings:\n",
    "    use_padding = True\n",
    "    rc_aug = False  # reverse complement augmentation\n",
    "    add_eos = False  # add end of sentence token\n",
    "\n",
    "    # we need these for the decoder head, if using\n",
    "    use_head = False\n",
    "    n_classes = 2  # not used for embeddings only\n",
    "\n",
    "    # you can override with your own backbone config here if you want,\n",
    "    # otherwise we'll load the HF one in None\n",
    "    backbone_cfg = None\n",
    "\n",
    "    device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "    print(\"Using device:\", device)\n",
    "\n",
    "    # instantiate the model (pretrained here)\n",
    "    if pretrained_model_name in ['hyenadna-tiny-1k-seqlen',\n",
    "                                 'hyenadna-small-32k-seqlen',\n",
    "                                 'hyenadna-medium-160k-seqlen',\n",
    "                                 'hyenadna-medium-450k-seqlen',\n",
    "                                 'hyenadna-large-1m-seqlen']:\n",
    "        # use the pretrained Huggingface wrapper instead\n",
    "        model = HyenaDNAPreTrainedModel.from_pretrained(\n",
    "            checkpoint_folder,\n",
    "            pretrained_model_name,\n",
    "            download=True,\n",
    "            config=backbone_cfg,\n",
    "            device=device,\n",
    "            use_head=use_head,\n",
    "            n_classes=n_classes,\n",
    "        )\n",
    "\n",
    "    # from scratch\n",
    "    elif pretrained_model_name is None:\n",
    "        model = HyenaDNAModel(**backbone_cfg, use_head=use_head, n_classes=n_classes)\n",
    "\n",
    "    # create tokenizer\n",
    "    tokenizer = CharacterTokenizer(\n",
    "        characters=['A', 'C', 'G', 'T', 'N'],  # add DNA characters, N is uncertain\n",
    "        model_max_length=max_length + 2,  # to account for special tokens, like EOS\n",
    "        add_special_tokens=False,  # we handle special tokens elsewhere\n",
    "        padding_side='left', # since HyenaDNA is causal, we pad on the left\n",
    "    )\n",
    "\n",
    "    #### Get named embeddings ####\n",
    "\n",
    "    with open(path / filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if line[0] == '>':\n",
    "                seq_id = line.strip().split(':')[0][1:]\n",
    "            else:\n",
    "                seq = line.strip()\n",
    "                rev_seq = string_reverse_complement(seq)\n",
    "                for seq_name,sequence in zip([seq_id, seq_id + '_flipped'],[seq, rev_seq]):\n",
    "                    savefile = savepath / f\"{seq_name}.npy\"\n",
    "                    if not savefile.is_file():\n",
    "                        #sequence = 'ACTG' * int(max_length/4)\n",
    "                        tok_seq = tokenizer(sequence)\n",
    "                        tok_seq = tok_seq[\"input_ids\"]  # grab ids\n",
    "\n",
    "                        # place on device, convert to tensor\n",
    "                        tok_seq = torch.LongTensor(tok_seq).unsqueeze(0)  # unsqueeze for batch dim\n",
    "                        tok_seq = tok_seq.to(device)\n",
    "\n",
    "                        # prep model and forward\n",
    "                        model.to(device)\n",
    "                        model.eval()\n",
    "                        with torch.inference_mode():\n",
    "                            embeddings = model(tok_seq) \n",
    "                            embeddings = embeddings.to(dtype=torch.float16).cpu()\n",
    "                            print(torch.max(embeddings), torch.min(embeddings))\n",
    "                            np.save(savepath / f\"{seq_name}.npy\",embeddings)  # embeddings here!\n",
    "\n",
    "\n",
    "################################################################################\n",
    "######################### S C R I P T ##########################################\n",
    "\n",
    "path = Path(\"../inputs/DNA_sequences/\")\n",
    "filename : str = \"test_boundaries.fasta\"\n",
    "savepath: Path = Path(\"../inputs/DNA_sequences/test_embeddings_Hyena-DNA/\")\n",
    "savepath.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "checkpoint_folder : str = \"../checkpoints/\"\n",
    "#Path(checkpoint_folder).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "# Uncomment if we want to get pretrained embeddings\n",
    "inference_single_sequential(path, filename, savepath, checkpoint_folder)\n",
    "\n",
    "# Uncomment if we want to train and test HyenaDNA:\n",
    "#run_train_CLASTER_HYENA()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Enformer\n",
    "\n",
    "The original paper can be found as:\n",
    "\n",
    "Avsec, Ž., Agarwal, V., Visentin, D. et al. Effective gene expression prediction from sequence by integrating long-range interactions. Nat Methods 18, 1196–1203 (2021). https://doi.org/10.1038/s41592-021-01252-x . In the case of the Enformer, we simply obtained the embeddings from the pretrained model matching our sequences. \n",
    "\n",
    "\n",
    "Code to build the Enformer using pytorch and load pretrained weights was obtained from:\n",
    "https://github.com/lucidrains/enformer-pytorch.\n",
    "\n",
    ">KUDOS: Huge kudos to Phil Wang (lucidrains) for open sourcing the pytorch version of the Enformer\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install the Enformer pytorch package:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "! pip install enformer-pytorch>=0.5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Get Enformer embeddings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Enformer_GPU.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile Enformer_GPU.py\n",
    "\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import torch\n",
    "from enformer_pytorch import from_pretrained\n",
    "import torch.nn.functional as F\n",
    "import os\n",
    "\n",
    "# Set the device to GPU if available\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "\n",
    "# Load the model globally and move it to the specified device\n",
    "model = from_pretrained('EleutherAI/enformer-official-rough').to(device)\n",
    "\n",
    "###### Functions #######\n",
    "\n",
    "def pad_tensor_symmetrically(input_tensor, target_length, pad_value=4):\n",
    "    if input_tensor.shape != (1, 160000):\n",
    "        raise ValueError(\"Input tensor must have shape (1, 160000)\")\n",
    "\n",
    "    total_padding = target_length - input_tensor.shape[1]\n",
    "    padding_per_side = total_padding // 2\n",
    "    padded_tensor = F.pad(input_tensor, (padding_per_side, padding_per_side), \"constant\", pad_value)\n",
    "    return padded_tensor\n",
    "\n",
    "def map_dna_to_numeric_tensor(sequence):\n",
    "    mapping = {'A': 0, 'C': 1, 'G': 2, 'T': 3, 'N': 4}\n",
    "    numeric_sequence = [mapping[nucleotide] for nucleotide in sequence if nucleotide in mapping]\n",
    "    sequence_tensor = torch.tensor(numeric_sequence, dtype=torch.long).unsqueeze(0)\n",
    "    return sequence_tensor\n",
    "\n",
    "string_complement_map = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'a': 't', 'c': 'g', 'g': 'c', 't': 'a'}\n",
    "\n",
    "def string_reverse_complement(seq):\n",
    "    return ''.join(string_complement_map.get(base, base) for base in reversed(seq))\n",
    "\n",
    "def process_sequence(path, seq_id, sequence, savepath):\n",
    "    rev_seq = string_reverse_complement(sequence)\n",
    "    for seq_name, seq in zip([seq_id, seq_id[:-8] + '_rev'], [sequence, rev_seq]):\n",
    "        savefile = savepath / f\"{seq_name}.npy\"\n",
    "        if not savefile.is_file():\n",
    "            seq_tensor = map_dna_to_numeric_tensor(seq)\n",
    "            seq_tensor = pad_tensor_symmetrically(seq_tensor, target_length=196608)\n",
    "            seq_tensor = seq_tensor.to(device)  # Move data to GPU\n",
    "            _, embeddings = model(seq_tensor, return_embeddings=True)\n",
    "            embeddings = embeddings.detach().cpu().numpy()  # Move data back to CPU for saving\n",
    "            np.save(savefile, embeddings.astype(np.float32))\n",
    "\n",
    "def create_Enformer_embeddings(path, filename, savepath):\n",
    "    with open(path / filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                seq_id = line.strip().split(':')[0][1:]\n",
    "            else:\n",
    "                seq = line.strip()\n",
    "                process_sequence(path, seq_id, seq, savepath)\n",
    "\n",
    "#### Main script ####\n",
    "\n",
    "path = Path(\"../inputs/DNA_sequences/\")\n",
    "split_list = [\"training\", \"validation\", \"test\"]\n",
    "\n",
    "for split in split_list:\n",
    "    filename = f\"{split}_boundaries.fasta\"\n",
    "    savepath = Path(f\"../inputs/DNA_sequences/{split}_embeddings_Enformer/\")\n",
    "    savepath.mkdir(parents=True, exist_ok=True)\n",
    "    create_Enformer_embeddings(path, filename, savepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now run it as a python file in a slurm based system. We will greatly benefit from the multiprocessing functionality.\n",
    "\n",
    "```bash\n",
    "srun -- python Enformer.py\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Train and test head on top of embeddings**\n",
    "\n",
    ">_Note: The test set is quite large. I predicted it all at once using 100 CPUs, but the code can be splitted to predict in batches and join the predictions afterwards if needed. Additionally, I removed all reversed (duplicated) samples from inputs and targets for test predictions to match those of Hyena._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting Enformer_head.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile Enformer_head.py\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "#from sklearn.metrics import r2_score\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "import logging\n",
    "from tqdm import tqdm\n",
    "\n",
    "###### Data class and head Model ######################\n",
    "class EnformerHeadDataset(Dataset):\n",
    "    \"\"\"\n",
    "    This class reads input numpy arrays and target csv files and returns input-target torch tensors.\n",
    "    Args:\n",
    "        - data_folder: Directory containing our input files stored as separate numpy arrays (sampleID.npy).\n",
    "        - targets_file: CSV file containing the IDs of the samples and their corresponding targets.\n",
    "        - N_kbp: Number of kilobase pairs from TSS to include in the target arrays. For example, N_kbp = 32 means targets from -32 kbp to +32 kbp.\n",
    "        - stack_type: \"hstack\" or \"vstack\" for the concatenation method of control and treated target arrays.\n",
    "    Returns:\n",
    "        - Input or data tensor.\n",
    "        - Target tensor.\n",
    "    \"\"\"\n",
    "    def __init__(self, data_folder, targets_file, stack_type, bin_size):\n",
    "        assert os.path.exists(data_folder), f\"The specified data folder does not exist: {data_folder}\"\n",
    "        assert stack_type in [\"hstack\", \"vstack\"], f\"Invalid stack type: {stack_type}\"\n",
    "        self.data_folder = data_folder\n",
    "        self.stack_type = stack_type\n",
    "        self.bin_size = bin_size\n",
    "        targets_df = pd.read_csv(targets_file)\n",
    "\n",
    "        \n",
    "        ctrl_cols = [col for col in targets_df.columns if \"ctrl\" in col]\n",
    "        treated_cols = [col for col in targets_df.columns if \"treated\" in col]\n",
    "\n",
    "        # Filter out targets for which input files do not exist and separate control and treated targets\n",
    "        self.sample_info = []\n",
    "        for _, row in targets_df.iterrows():\n",
    "            sample_id = row['ID']\n",
    "            file_path = os.path.join(data_folder, f\"{sample_id}.npy\")\n",
    "            if os.path.exists(file_path):\n",
    "                ctrl_target = row[ctrl_cols].values\n",
    "                treated_target = row[treated_cols].values\n",
    "                self.sample_info.append((sample_id, ctrl_target, treated_target))\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sample_info)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        sample_name, ctrl_target, treated_target = self.sample_info[idx]\n",
    "        data_path = os.path.join(self.data_folder, f\"{sample_name}.npy\")\n",
    "        data = np.load(data_path)\n",
    "\n",
    "        # Stack the targets based on stack_type\n",
    "        if self.stack_type == \"hstack\":\n",
    "            target = np.hstack((ctrl_target, treated_target))\n",
    "        else:  # vstack\n",
    "            target = np.vstack((ctrl_target, treated_target))\n",
    "\n",
    "        target = torch.from_numpy(np.array(target, dtype=np.float32))\n",
    "        return torch.tensor(data.reshape(data.shape[1], data.shape[2]), dtype=torch.float32), torch.tensor(target, dtype=torch.float32)\n",
    "\n",
    "class DNAConvNet(nn.Module):\n",
    "    def __init__(self, input_depth, output_size, n_conditions, kernel_size, stride, padding, dilation, dropout_prob):\n",
    "        super(DNAConvNet, self).__init__()\n",
    "        self.dropout = nn.Dropout(p=dropout_prob)\n",
    "        self.conv1 = nn.Conv1d(in_channels=input_depth, out_channels=n_conditions, kernel_size=kernel_size, stride=stride, padding=padding, dilation=dilation)\n",
    "        self.softplus = nn.Softplus()  # Adding the Softplus activation layer as in Enformer\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = x.view(x.size(0), x.size(2), x.size(1)) # (64, 3072,896)\n",
    "        # Apply dropout to the input\n",
    "        x = self.dropout(x)\n",
    "\n",
    "        x = self.conv1(x)\n",
    "        x = self.softplus(x)  # Applying softplus activation after convolution\n",
    "        return x\n",
    "\n",
    "class DNAConvandDense(nn.Module):\n",
    "    def __init__(self, input_depth, input_length, output_length, kernel_size, stride, padding, dilation, dropout_prob, hidden_depth):\n",
    "        super(DNAConvandDense, self).__init__()\n",
    "        self.dropout = nn.Dropout(p=dropout_prob)\n",
    "        self.conv1d = nn.Conv1d(in_channels=input_depth, out_channels=hidden_depth, kernel_size=kernel_size,stride=stride, padding=padding, dilation=dilation)\n",
    "        \n",
    "        # Calculate the output length after convolution\n",
    "        self.seq_length_after_conv = self.calculate_output_length(input_length, kernel_size, stride, padding, dilation)\n",
    "        \n",
    "        # Fully connected layer with calculated input size\n",
    "        self.fc = nn.Linear(hidden_depth * self.seq_length_after_conv, output_length)\n",
    "        \n",
    "        # Softplus activation\n",
    "        self.softplus = nn.Softplus()\n",
    "\n",
    "\n",
    "    def calculate_output_length(self, input_length, kernel_size, stride, padding, dilation):\n",
    "        return ((input_length + 2 * padding - dilation * (kernel_size - 1) - 1) // stride) + 1\n",
    "\n",
    "    def forward(self, x):\n",
    "        x = x.view(x.size(0), x.size(2), x.size(1)) # (64, 3072,896)\n",
    "        # Apply 1D convolution\n",
    "        x = self.conv1d(x)\n",
    "        # Apply dropout to the input\n",
    "        x = self.dropout(x)\n",
    "\n",
    "        # Flatten the convolution output\n",
    "        x = x.view(x.size(0), -1)\n",
    "\n",
    "        # Pass through the fully connected layer\n",
    "        x = self.fc(x)\n",
    "        \n",
    "        # Apply Softplus activation\n",
    "        x = self.softplus(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "\n",
    "class ConvLinearModel(nn.Module):\n",
    "    def __init__(self, input_depth, input_length, output_length, kernel_size, stride, padding, dilation, dropout_prob, hidden_depth):\n",
    "        super(ConvLinearModel, self).__init__()\n",
    "        # 2D Average Pooling\n",
    "        self.avgpool2d = nn.AvgPool2d(kernel_size=avg_pool_size)\n",
    "\n",
    "        # Dropout\n",
    "        self.dropout = nn.Dropout(dropout_rate)\n",
    "\n",
    "        # Adjust the input depth and sequence length based on the 2D pooling\n",
    "        new_depth = input_depth // avg_pool_size[0]\n",
    "        new_seq_length = input_length // avg_pool_size[1]\n",
    "\n",
    "        # 1D Convolution\n",
    "        self.conv1 = nn.Conv1d(in_channels=new_depth, out_channels=hidden_depth, kernel_size=kernel_size, stride=stride, padding=padding, dilation=dilation)\n",
    "\n",
    "        # Calculate the output length after convolution\n",
    "        self.seq_length_after_conv = self.calculate_output_length(new_seq_length, kernel_size, stride, padding, dilation)\n",
    "        \n",
    "        # Fully connected layer with calculated input size\n",
    "        self.fc = nn.Linear(hidden_depth * self.seq_length_after_conv, output_length)\n",
    "        self.softplus = nn.Softplus()\n",
    "        \n",
    "    def calculate_output_length(self, input_length, kernel_size, stride, padding, dilation):\n",
    "        return ((input_length + 2 * padding - dilation * (kernel_size - 1) - 1) // stride) + 1\n",
    "\n",
    "    def forward(self, x):\n",
    "        # Reshape x to (batch_size, 1, input_depth, seq_length) for 2D avg pooling\n",
    "        x = x.view(x.size(0),1, x.size(2), x.size(1)) \n",
    "        x = self.avgpool2d(x)\n",
    "        print(x.shape)\n",
    "\n",
    "        # Reshape back to (batch_size, new_depth, new_seq_length) for 1D convolution\n",
    "        x = x.view(x.size(0), -1, x.size(3))\n",
    "\n",
    "        # Apply dropout to the inputs of the convolutions\n",
    "        x = self.dropout(x)\n",
    "\n",
    "        # Apply 1D convolution\n",
    "        x = self.conv1(x)\n",
    "\n",
    "        # Apply dropout to the inputs of the final layer\n",
    "        x = self.dropout(x)\n",
    "\n",
    "        # Reshape x for the linear layer\n",
    "        x = x.view(x.size(0), -1)\n",
    "        \n",
    "        # Pass through the fully connected layer\n",
    "        x = self.fc(x)\n",
    "\n",
    "        # Apply Softplus activation\n",
    "        x = self.softplus(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "\n",
    "#### Custom loss #############\n",
    "class ZIPLoss(nn.Module):\n",
    "    def __init__(self, alpha=.5, eps=.02):\n",
    "        super(ZIPLoss, self).__init__()\n",
    "        self.alpha = alpha\n",
    "        self.eps = eps\n",
    "\n",
    "    def forward(self, inputs, targets):\n",
    "        total_loss = 0\n",
    "        for i in range(inputs.shape[1]):  # Iterate over channels\n",
    "            lambda_ = inputs[:, i, :]\n",
    "            logit_zero = inputs[:, i, :]\n",
    "            target = targets[:, i, :]\n",
    "\n",
    "            # Adjust for non-true zeros\n",
    "            mask = (target < self.eps).float()\n",
    "            target = torch.where(mask == 1, torch.zeros_like(target), target)\n",
    "\n",
    "            # Poisson loss part\n",
    "            poisson_loss = lambda_ - target * torch.log(lambda_ + 1e-8)\n",
    "\n",
    "            # Zero-inflation part\n",
    "            zero_inflation_prob = torch.sigmoid(logit_zero)\n",
    "            zero_inflation_loss = -torch.where(mask == 1,\n",
    "                                               self.alpha * torch.log(zero_inflation_prob + 1e-8),\n",
    "                                               torch.log(1 - zero_inflation_prob + 1e-8))\n",
    "\n",
    "            # Combine the Poisson and zero-inflation losses for this channel\n",
    "            channel_loss = zero_inflation_loss + poisson_loss\n",
    "\n",
    "            total_loss += channel_loss.mean()\n",
    "\n",
    "        return total_loss / inputs.shape[1]  # Average loss over channels\n",
    "\n",
    "class ZINBLoss(nn.Module):\n",
    "    def __init__(self, dispersion=None, zero_inflation_prob=None, channel_weights=None):\n",
    "        \"\"\"\n",
    "        Zero-Inflated Negative Binomial loss module for multi-channel data with predefined extra parameters.\n",
    "\n",
    "        Parameters:\n",
    "        - dispersion: Optional, predefined dispersion parameter (alpha) for the negative binomial distribution.\n",
    "                      Can be a scalar or a tensor with shape (channel,).\n",
    "        - zero_inflation_prob: Optional, predefined tensor of zero-inflation probabilities with shape (batch_size, channel, seq_length).\n",
    "        - channel_weights: Optional tensor with shape (channel,) representing the weight of each channel in the loss calculation.\n",
    "        \"\"\"\n",
    "        super(ZINBLoss, self).__init__()\n",
    "        self.epsilon = 1e-10\n",
    "\n",
    "        if dispersion is not None:\n",
    "            self.register_buffer('dispersion', torch.tensor(dispersion))\n",
    "        else:\n",
    "            self.register_buffer('dispersion', torch.tensor(1.0))\n",
    "\n",
    "        self.zero_inflation_prob = zero_inflation_prob\n",
    "        self.channel_weights = channel_weights\n",
    "\n",
    "    def forward(self, true_counts, predicted_counts):\n",
    "        batch_size, num_channels, seq_length = true_counts.shape\n",
    "\n",
    "        if self.channel_weights is None:\n",
    "            channel_weights = torch.ones(num_channels, device=true_counts.device)\n",
    "        else:\n",
    "            channel_weights = self.channel_weights\n",
    "\n",
    "        if self.zero_inflation_prob is None:\n",
    "            zero_inflation_prob = torch.zeros_like(predicted_counts)\n",
    "        else:\n",
    "            zero_inflation_prob = self.zero_inflation_prob\n",
    "\n",
    "        total_loss = 0.0\n",
    "\n",
    "        for channel in range(num_channels):\n",
    "            # Extract data for the current channel\n",
    "            true_counts_channel = true_counts[:, channel, :]\n",
    "            predicted_counts_channel = predicted_counts[:, channel, :]\n",
    "            predicted_zero_inflation_channel = zero_inflation_prob[:, channel, :]\n",
    "            dispersion_channel = self.dispersion[channel] if self.dispersion.ndim > 0 else self.dispersion\n",
    "\n",
    "            # Negative Binomial term\n",
    "            theta_channel = 1.0 / (dispersion_channel + self.epsilon)\n",
    "            t1 = torch.lgamma(true_counts_channel + theta_channel) - torch.lgamma(theta_channel) - torch.lgamma(true_counts_channel + 1)\n",
    "            t2 = (theta_channel + true_counts_channel) * torch.log1p(predicted_counts_channel / theta_channel) + true_counts_channel * (torch.log(theta_channel) - torch.log(predicted_counts_channel + self.epsilon))\n",
    "\n",
    "            nb_term = t1 + t2\n",
    "\n",
    "            # Zero-Inflation term\n",
    "            zero_inflation_term = predicted_zero_inflation_channel + F.softplus(-predicted_zero_inflation_channel - nb_term) - F.softplus(-nb_term)\n",
    "\n",
    "            # Combine terms\n",
    "            zero_inflation_mask = (true_counts_channel == 0).type(torch.float32)\n",
    "            channel_loss = -torch.mean(zero_inflation_term * zero_inflation_mask + nb_term * (1 - zero_inflation_mask))\n",
    "\n",
    "            # Weighted sum of the channel losses\n",
    "            total_loss += channel_weights[channel] * channel_loss\n",
    "\n",
    "        # Average loss over channels\n",
    "        loss = total_loss / num_channels\n",
    "\n",
    "        return loss\n",
    "############## Training and auxilliary functions ##############\n",
    "\n",
    "\n",
    "def count_parameters(model):\n",
    "    total_params = sum(p.numel() for p in model.parameters())\n",
    "    trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)\n",
    "    return total_params, trainable_params\n",
    "\n",
    "def r2_score(outputs, targets):\n",
    "    \"\"\"\n",
    "    Function to compute R2 coefficient between predictions and targets.\n",
    "    This allows for all operations to happen in the GPU )if available)\n",
    "    without moving back and forth to the CPU.\n",
    "    \"\"\"\n",
    "    target_mean = torch.mean(targets)\n",
    "    ss_tot = torch.sum((targets - target_mean) ** 2)\n",
    "    ss_res = torch.sum((targets - outputs) ** 2)\n",
    "    r2 = 1 - ss_res / ss_tot\n",
    "    return r2.cpu().detach().numpy()\n",
    "\n",
    "# Train and validate the model\n",
    "def train_with_validation(model, train_loader, val_loader, criterion, optimizer, num_epochs, device, savepath):\n",
    "    train_losses, val_losses = [], []\n",
    "    train_accuracies, val_accuracies = [], []\n",
    "\n",
    "    logging.info(\"Training the model\")\n",
    "    for epoch in range(num_epochs):\n",
    "        model.train()\n",
    "        total_loss, total_r2, count = 0, 0, 0\n",
    "        # Initialize progress bar\n",
    "        progress_bar = tqdm(train_loader, desc=f'Epoch {epoch+1}/{num_epochs}', leave=False)\n",
    "        for inputs, targets in train_loader:\n",
    "            # Send data to GPU\n",
    "            inputs, targets = inputs.to(device), targets.to(device)\n",
    "\n",
    "            optimizer.zero_grad()\n",
    "            outputs = model(inputs)\n",
    "            loss = criterion(outputs, targets)\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "\n",
    "            total_loss += loss.item()\n",
    "            # Calculate R^2 score\n",
    "            total_r2 += r2_score(outputs, targets)\n",
    "            count += 1\n",
    "            # Update progress bar\n",
    "            progress_bar.set_postfix({'train_loss': loss.item()})\n",
    "\n",
    "        avg_train_loss = total_loss / count\n",
    "        avg_train_r2 = total_r2 / count\n",
    "        train_losses.append(avg_train_loss)\n",
    "        train_accuracies.append(avg_train_r2)\n",
    "\n",
    "        # Validation step\n",
    "        val_loss, val_r2 = validate(model, val_loader, criterion, device)\n",
    "        val_losses.append(val_loss)\n",
    "        val_accuracies.append(val_r2)\n",
    "\n",
    "        logging.info(f\"Epoch [{epoch + 1}/{num_epochs}], Training Loss: {avg_train_loss:.4f}, Training R^2: {avg_train_r2:.4f}, Validation Loss: {val_loss:.4f}, Validation R^2: {val_r2:.4f}\")\n",
    "\n",
    "        # Close the progress bar at the end of the epoch\n",
    "        progress_bar.close()\n",
    "\n",
    "    # Save the final model\n",
    "    torch.save(model, savepath / 'model_smoothl1.pt')\n",
    "\n",
    "    return train_losses, val_losses, train_accuracies, val_accuracies\n",
    "\n",
    "\n",
    "def validate(model, val_loader, criterion, device):\n",
    "    model.eval()\n",
    "    total_loss, total_r2, count = 0, 0, 0\n",
    "    with torch.no_grad():\n",
    "        for inputs, targets in val_loader:\n",
    "            inputs, targets = inputs.to(device), targets.to(device)\n",
    "            outputs = model(inputs)\n",
    "            loss = criterion(outputs, targets)\n",
    "            total_loss += loss.item()\n",
    "            total_r2 += r2_score(outputs, targets)\n",
    "            count += 1\n",
    "    avg_loss = total_loss / count\n",
    "    avg_r2 = total_r2 / count\n",
    "    return avg_loss, avg_r2\n",
    "\n",
    "def plot_losses_accuracies(train_losses, val_losses, train_accuracies, val_accuracies, savepath):\n",
    "    logging.info(\"Plotting results\")\n",
    "    fig = plt.figure(figsize=(12, 5))\n",
    "\n",
    "    # Plot losses\n",
    "    plt.subplot(1, 2, 1)\n",
    "    plt.plot(np.arange(1,len(train_losses)+1), train_losses, label='Training Loss')\n",
    "    plt.plot(np.arange(1,len(val_losses)+1), val_losses, label='Validation Loss')\n",
    "    plt.title('Training and Validation Loss')\n",
    "    plt.xlabel('Epoch')\n",
    "    plt.ylabel('Loss')\n",
    "    plt.legend()\n",
    "\n",
    "    # Plot point-wise R2 evolution\n",
    "    plt.subplot(1, 2, 2)\n",
    "    plt.plot(np.arange(1,len(train_accuracies)+1),train_accuracies, label='Training R²')\n",
    "    plt.plot(np.arange(1,len(val_accuracies)+1), val_accuracies, label='Validation R²')\n",
    "    plt.title('Training and Validation R² Score')\n",
    "    plt.xlabel('Epoch')\n",
    "    plt.ylabel('R² Score')\n",
    "    plt.legend()\n",
    "\n",
    "    fig.savefig(savepath / \"Reconstruction_metrics.png\")\n",
    "\n",
    "\n",
    "def test_model_with_predictions(model, test_loader, device):\n",
    "    model.eval()  # Set the model to evaluation mode\n",
    "    predictions = []\n",
    "    targets_list = []  # if you also want to store true targets\n",
    "\n",
    "    with torch.no_grad():  # No gradient calculation\n",
    "        for inputs, targets in test_loader:\n",
    "            inputs = inputs.to(device)\n",
    "            outputs = model(inputs)\n",
    "            predictions.append(outputs.cpu().numpy())\n",
    "            targets_list.append(targets.cpu().numpy())  # if storing targets\n",
    "\n",
    "    return predictions, targets_list\n",
    "########################## Script #########################\n",
    "# Train model and test on unseen data\n",
    "IS_HYENA = False # We will not train on HyenaDNA's embeddings\n",
    "TRAIN = False\n",
    "TEST = True\n",
    "# Paths:\n",
    "train_data_folder = \"../inputs/DNA_sequences/training_embeddings_Enformer/\"\n",
    "val_data_folder = \"../inputs/DNA_sequences/validation_embeddings_Enformer/\"\n",
    "test_data_folder = \"../inputs/DNA_sequences/test_embeddings_Enformer/\"\n",
    "\n",
    "targets_file = '../targets/training_targets_Enformer.csv'\n",
    "test_targets_file = '../targets/test_targets_Enformer.csv'\n",
    "#savepath = Path(\"./results/\")\n",
    "savepath = Path(\"../benchmarks/Enformer/\")\n",
    "savepath.mkdir(parents=True, exist_ok=True)\n",
    "# Hyperparameters:\n",
    "input_length = 896 # Embeddings add 2 characters to the 160kbp long sequences\n",
    "input_depth = 3072\n",
    "hidden_depth = 10\n",
    "output_size = 115 # 57kbp per side #input_length\n",
    "n_conditions = 1\n",
    "stack_type = 'hstack' #'vstack' # vstack when outputting directly from convolutions\n",
    "kernel_size = 9\n",
    "stride = 2\n",
    "dilation = 2\n",
    "padding = (dilation * (kernel_size - 1)) // 2 # proper padding\n",
    "num_epochs = 20\n",
    "\n",
    "batch_size = 64\n",
    "dropout_rate = 0.4\n",
    "lr = 1e-4\n",
    "#################\n",
    "\n",
    "if IS_HYENA:\n",
    "    train_data_folder = \"/projects/cbmr_shared/scratch/Hyena_DNA/training/\"\n",
    "    val_data_folder = \"/projects/cbmr_shared/scratch/Hyena_DNA/validation/\"\n",
    "    test_data_folder = \"/projects/cbmr_shared/scratch/Hyena_DNA/test/\"\n",
    "\n",
    "    targets_file = '/projects/cbmr_shared/scratch/Hyena_DNA/targets/benchmark_sequence_based_target_arrays_1kbp_57_bins_2_conditions_decareads_abs.csv'\n",
    "    test_targets_file = '/projects/cbmr_shared/scratch/Hyena_DNA/targets/benchmark_sequence_based_target_arrays_1kbp_57_bins_2_conditions_test_decareads.csv'\n",
    "\n",
    "    savepath = Path(\"/projects/cbmr_shared/scratch/Hyena_DNA/results/Hyena/\")\n",
    "    savepath.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "    # Hyperparameters: \n",
    "    input_length = 160002 # Embeddings add 2 characters to the 160kbp long sequences\n",
    "    input_depth = 256\n",
    "    hidden_depth = 20\n",
    "    output_size = 115 # 2 conditions from - N_kbt to + N_kbt (adding 0, central bin)\n",
    "    n_conditions = 1\n",
    "    stack_type = 'hstack' #'hstack'\n",
    "    kernel_size = 9\n",
    "    stride = 333\n",
    "    num_epochs = 200\n",
    "    dilation = 100\n",
    "    padding = (dilation * (kernel_size - 1)) // 2  \n",
    "\n",
    "    batch_size = 64\n",
    "    avg_pool_size = (1,3)\n",
    "    dropout_rate = .3\n",
    "    lr = 1e-4\n",
    "\n",
    "\n",
    "# Setup logging\n",
    "logging.basicConfig(filename=str(savepath) + '/training.log', level=logging.INFO, format='%(asctime)s %(levelname)s: %(message)s')\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "logging.info(f\"Device:{device}\")\n",
    "model = DNAConvandDense(input_depth=input_depth, input_length=input_length, output_length=n_conditions*output_size, kernel_size=kernel_size, stride=stride, padding=padding, dilation=dilation, dropout_prob=dropout_rate, hidden_depth=hidden_depth).to(device)\n",
    "if IS_HYENA:\n",
    "    model = ConvLinearModel(input_depth, input_length, n_conditions*output_size, kernel_size, stride, padding, dilation, dropout_rate, hidden_depth).to(device)\n",
    "total_params, trainable_params = count_parameters(model)\n",
    "logging.info(f\"Total parameters: {total_params}\")\n",
    "logging.info(f\"Trainable parameters: {trainable_params}\")\n",
    "\n",
    "if TRAIN:\n",
    "    logging.info(\"Beginning training\")\n",
    "\n",
    "    train_dataset = EnformerHeadDataset(train_data_folder, targets_file, stack_type, kernel_size)\n",
    "    val_dataset = EnformerHeadDataset(val_data_folder, targets_file, stack_type, kernel_size)\n",
    "\n",
    "    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)\n",
    "    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)\n",
    "\n",
    "    criterion = torch.nn.SmoothL1Loss() #torch.nn.PoissonNLLLoss(log_input=False) #ZINBLoss() #zinb_loss_channelwise()#ZeroInflatedPoissonLoss() #torch.nn.PoissonNLLLoss(log_input=False)\n",
    "    optimizer = torch.optim.AdamW(model.parameters(), lr=lr)\n",
    "\n",
    "    train_losses, val_losses, train_accuracies, val_accuracies= train_with_validation(model, train_loader, val_loader, criterion, optimizer, num_epochs=num_epochs, device=device, savepath=savepath)\n",
    "    plot_losses_accuracies(train_losses, val_losses, train_accuracies, val_accuracies, savepath)\n",
    "\n",
    "############# Test predictions #############\n",
    "if TEST:\n",
    "    #device = torch.device(\"cpu\")\n",
    "    logging.info(\"Beginning test predictions\")\n",
    "\n",
    "    model = torch.load(Path(\"../benchmarks/Enformer/model_smoothl1.pt\"), map_location=device)\n",
    "    model.to(device)\n",
    "\n",
    "    test_dataset = EnformerHeadDataset(test_data_folder, test_targets_file, stack_type, kernel_size)\n",
    "    test_loader = DataLoader(test_dataset, batch_size=len(test_dataset), shuffle=False)\n",
    "\n",
    "    predictions, targets = test_model_with_predictions(model, test_loader, device)\n",
    "\n",
    "    # Save Enformer predictions and targets in numpy arrays:\n",
    "    np.save(savepath / \"enformer_test_predictions.npy\",np.array(predictions), allow_pickle=\"False\")\n",
    "    np.save(savepath / \"enformer_test_targets.npy\",np.array(targets), allow_pickle=\"False\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "c305a8ed7bd3fb5b79b1e9049e998ba6b84af3f1b497d7bcc87b4717f669b9d3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
