
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>DATA ANALYSIS NOTEBOOK</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=3ee479438cf8b5e0d341" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=384b581d" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=3ee479438cf8b5e0d341" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=3ee479438cf8b5e0d341" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=3ee479438cf8b5e0d341"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'scripts/3_Data_analysis';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title"></p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/RasmussenLab/CLASTER" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Claster
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="0_Tutorial.html">TUTORIAL</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/RasmussenLab/CLASTER/blob/master/scripts/3_Data_analysis.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER/edit/master/scripts/3_Data_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/scripts/3_Data_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1><center> DATA ANALYSIS NOTEBOOK <center></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><center> DATA ANALYSIS NOTEBOOK <center></center></center></a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#packages">Packages</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functions">Functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model-performance-on-the-baseline-prediction-task"><center> Model performance on the baseline prediction task <center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-silico-perturbations"><center> <em>In silico</em> perturbations <center></center></center></a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#attributions"><center> Attributions <center></center></center></a></li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="data-analysis-notebook">
<h1><center> DATA ANALYSIS NOTEBOOK <center><a class="headerlink" href="#data-analysis-notebook" title="Link to this heading">#</a></h1>
<p>This notebook was used to run the analyses and produce most of the images in the paper.</p>
<blockquote>
<div><p><em>Overview:</em></p>
<ul class="simple">
<li><p>Get baseline predictions of the different models:</p>
<ul>
<li><p>Bin by bin (pointwise)</p></li>
<li><p>Integrating the area of the central gene.</p></li>
</ul>
</li>
<li><p>Compare model performances</p></li>
<li><p><strong>Input perturbations</strong>: <em>How does nascent transcription change when we perturb the inputs?</em></p></li>
<li><p><strong>Input attributions</strong>: <em>What is required in the input to predict a given output?</em></p></li>
</ul>
</div></blockquote>
<section id="packages">
<h2>Packages<a class="headerlink" href="#packages" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.colors</span> <span class="k">as</span> <span class="nn">colors</span>
<span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="k">as</span> <span class="nn">patches</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">scipy.integrate</span> <span class="kn">import</span> <span class="n">simpson</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span><span class="p">,</span> <span class="n">spearmanr</span>
<span class="kn">from</span> <span class="nn">sklearn.metrics</span> <span class="kn">import</span> <span class="n">r2_score</span>

<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">gridspec</span>
<span class="kn">from</span> <span class="nn">PIL</span> <span class="kn">import</span> <span class="n">Image</span>

<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">gaussian_filter</span>

<span class="kn">import</span> <span class="nn">matplotlib.cm</span> <span class="k">as</span> <span class="nn">cm</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">rotate</span>

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="functions">
<h2>Functions<a class="headerlink" href="#functions" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">read_gene_positions</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function reads a csv containing genes and enhancers (entities), and creates a dictionary with the relative coordinates of</span>
<span class="sd">    all entities to the TSS of the reference gene (central gene), which gives name to the sample.</span>

<span class="sd">    Args:</span>
<span class="sd">        csv_path: path to the csv with ref genes, entities and their relative coordinates.</span>
<span class="sd">        resolution: distance unit we want our relative distances to be given with (1-&gt; bp, 1000-&gt;kbp)</span>

<span class="sd">    Returns:</span>
<span class="sd">        pos_dict: dictionary given pos_dict[ref gene][entity_name] = (rel_start, rel_end)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the results from a CSV file</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Initialize the dictionary to store positions</span>
    <span class="n">pos_dict</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="c1"># Filter the DataFrame to get only gene entities </span>
    <span class="n">gene_entities</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ENSMUSG&#39;</span><span class="p">)]</span>  <span class="c1"># Adjust the condition according to your data</span>
    
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_entities</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">ref_gene_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span>
        <span class="n">entity_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span>
        <span class="n">rel_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span><span class="p">))</span>
        <span class="n">rel_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span><span class="p">))</span>
        
        <span class="c1"># Ensure the dictionary has the necessary structure</span>
        <span class="k">if</span> <span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pos_dict</span><span class="p">:</span>
            <span class="n">pos_dict</span><span class="p">[</span><span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        
        <span class="c1"># Store the scaled and rounded start and end positions</span>
        <span class="n">pos_dict</span><span class="p">[</span><span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">entity_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">pos_dict</span>


<span class="k">def</span> <span class="nf">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">condition_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function reads the prediction files that EIR yields for each target (bin_condition, e.g. -150_ctrl) in the </span>
<span class="sd">    test (eirpredict) setting. Predictions are then converted into dataframes. </span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: path where the predictions are stored</span>
<span class="sd">        N_BINS: number of the bins to one side of the central bin (200 normally, 57 for the Enformer predictions).</span>
<span class="sd">        condition_list: list of conditions. As of now [&quot;_ctrl&quot;], but can be extended to other target conditions.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        ids: names of the samples</span>
<span class="sd">        predicted: predicted untransformed values</span>
<span class="sd">        actual: actual values</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Lists to store DataFrames for concatenation</span>
    <span class="n">predicted_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_dfs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/predictions.csv&quot;</span>
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>

            <span class="c1"># Apply ReLU to model predictions</span>
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Prepare data for concatenation.</span>
            <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2"> Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;True Label Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">predicted_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_column</span><span class="p">)</span>
            <span class="n">actual_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_column</span><span class="p">)</span>

    <span class="c1"># Concatenate all DataFrames horizontally</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predicted_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">actual</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">actual_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>


    <span class="k">return</span> <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span>

<span class="k">def</span> <span class="nf">plot_target_predictions</span><span class="p">(</span><span class="n">savepath</span><span class="p">,</span> <span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="nb">id</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span><span class="p">,</span> <span class="n">pred_area</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Plot the predicted profiles. </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">sharex</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">201</span><span class="p">),</span><span class="n">line_p</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;royalblue&#39;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Predicted expression control&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">201</span><span class="p">),</span><span class="n">line_a</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;silver&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.6</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Actual expression control&quot;</span><span class="p">)</span>
    <span class="c1">#plt.plot(np.arange(len(line_p)),line_p, lw=1, color=&#39;darkred&#39;, label=f&quot;{simpson(line_p[TSS:END])/WIDTH}&quot;)</span>
    <span class="c1">#plt.plot(np.arange(len(line_a)),line_a, lw=1, color=&#39;darkgreen&#39;, label=f&quot;{simpson(line_a[TSS:END])/WIDTH} {END}&quot;)</span>
    <span class="c1">#plt.plot(TSS,1,&#39;go&#39;)</span>
    <span class="c1">#plt.plot(END,1,&#39;ro&#39;)</span>
    <span class="c1"># for i in range(len(axs)):</span>
    <span class="c1">#     axs[i].set_ylim(-10,50)</span>
    <span class="c1">#     axs[i].set_xlim(-200,200)</span>
    <span class="c1">#     axs[i].set_ylabel(&quot;Reads&quot;, fontsize=12)</span>
    <span class="c1">#     rect= patches.Rectangle((-75,-5),70,40, edgecolor=&#39;k&#39;, facecolor=&#39;none&#39;, linewidth=2)</span>
    <span class="c1">#     axs[i].add_patch(rect) </span>
    <span class="c1">#     if i &gt; 1:</span>
    <span class="c1">#         axs[i].annotate(&#39;&#39;, xy=(-50,10), xytext=(-50, 30), arrowprops=dict(arrowstyle=&#39;-|&gt;,head_width=0.2,head_length=0.5&#39;, color=&#39;k&#39;))             </span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">rel_start</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span> <span class="n">rel_end</span><span class="o">-</span><span class="mi">200</span><span class="p">],[</span><span class="mi">50</span><span class="p">,</span><span class="mi">50</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Rel_area=</span><span class="si">{</span><span class="n">pred_area</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>     
    <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">id</span><span class="si">}</span><span class="s2">_prediction.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">extend_df</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Extend the gene-enhancer relationships df to add meaningful columns, e.g.</span>
<span class="sd">    * Number of genes between the given enhancer and the promoter of the central gene.</span>
<span class="sd">    * Whether the enhancer is adjacent to the measured gene or not.</span>
<span class="sd">    * Area under the central gene (baseline and when the given enhancer is silenced in silico).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># EP_distance calculation</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;EP_distance&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">((</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
    
    <span class="c1"># Initialize new columns</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Genes_in_between&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Adjacent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1"># Filter for entities related to the current reference gene and are genes (ENSMUSG)</span>
        <span class="n">related_genes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">])</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ENSMUSG&#39;</span><span class="p">))]</span>
        
        <span class="c1"># Determine the side of TSS the current entity is on</span>
        <span class="n">entity_side</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        
        <span class="c1"># Count genes that are closer to the TSS than the current entity and on the same side</span>
        <span class="n">closer_genes</span> <span class="o">=</span> <span class="n">related_genes</span><span class="p">[(</span><span class="n">np</span><span class="o">.</span><span class="n">sign</span><span class="p">((</span><span class="n">related_genes</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">related_genes</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">==</span> <span class="n">entity_side</span><span class="p">)</span> <span class="o">&amp;</span>
                                     <span class="p">(</span><span class="nb">abs</span><span class="p">((</span><span class="n">related_genes</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">related_genes</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">abs</span><span class="p">((</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))]</span>
        
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;Genes_in_between&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">closer_genes</span><span class="p">)</span>
        
        <span class="c1"># Set &#39;Adjacent&#39; to 1 if there are no genes between this entity and the ref_gene</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">index</span><span class="p">,</span> <span class="s1">&#39;Adjacent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">closer_genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    
    <span class="k">return</span> <span class="n">df</span>

<span class="k">def</span> <span class="nf">filter_and_extend_gene_enhancer_table</span><span class="p">(</span><span class="n">samples_index_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">samples_data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">perturbed_table_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">chrom_to_keep</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;chr4&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function filters the genes and gene-enhancer combinations to keep and adds the extra columns required for downstream tasks.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the first table and extract Ref_gene and Entity_ID pairs</span>
    <span class="n">samples_index_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples_index_path</span><span class="p">)</span>
    <span class="n">ref_entity_pairs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">samples_index_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)[:</span><span class="mi">2</span><span class="p">])))</span>

    <span class="c1"># Additionally, include Ref_gene as both Ref_gene and Entity_ID</span>
    <span class="n">ref_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">pair</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">ref_entity_pairs</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">ref_gene</span> <span class="ow">in</span> <span class="n">ref_genes</span><span class="p">:</span>
        <span class="n">ref_entity_pairs</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">ref_gene</span><span class="p">,</span> <span class="n">ref_gene</span><span class="p">))</span>

    <span class="c1"># Load the second table</span>
    <span class="n">samples_data_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples_data_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">samples_data_df</span> <span class="o">=</span> <span class="n">samples_data_df</span><span class="p">[(</span><span class="n">samples_data_df</span><span class="p">[</span><span class="s1">&#39;Ref_gene_chromosome&#39;</span><span class="p">]</span><span class="o">==</span> <span class="n">chrom_to_keep</span><span class="p">)]</span> <span class="c1"># Filter only genes from test chromosome</span>

    <span class="n">samples_data_df</span> <span class="o">=</span> <span class="n">extend_df</span><span class="p">(</span><span class="n">samples_data_df</span><span class="p">)</span>

    <span class="c1"># Filter rows based on whether the (Ref_gene, Entity_ID) tuple matches the reference set</span>
    <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">samples_data_df</span><span class="p">[</span>
        <span class="n">samples_data_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">])</span> <span class="ow">in</span> <span class="n">ref_entity_pairs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Save the filtered DataFrame to a new CSV file if needed</span>
    <span class="n">filtered_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">perturbed_table_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">calculate_area_gene</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Args:</span>
<span class="sd">        line_p: vector of predicted values</span>
<span class="sd">        line_a: vector of actual values</span>
<span class="sd">        ID: reference gene, i.e. sample name</span>
<span class="sd">        pos_dict: dictionary with relative gene coordinates</span>
<span class="sd">        N_BINS: which length do we set as a threshold to integrate (in the new resolution).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_center</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">line_p</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">rel_start</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">rel_end</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span> <span class="k">else</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span>

    <span class="n">WIDTH</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">rel_end</span> <span class="o">-</span> <span class="n">rel_start</span><span class="p">))</span>

    <span class="c1"># Appending area</span>
    <span class="n">pred_area</span> <span class="o">=</span> <span class="n">simpson</span><span class="p">(</span><span class="n">line_p</span><span class="p">[</span><span class="n">rel_start</span><span class="p">:</span><span class="n">rel_end</span><span class="p">])</span><span class="o">/</span><span class="n">WIDTH</span>
    <span class="n">actual_area</span> <span class="o">=</span> <span class="n">simpson</span><span class="p">(</span><span class="n">line_a</span><span class="p">[</span><span class="n">rel_start</span><span class="p">:</span><span class="n">rel_end</span><span class="p">])</span><span class="o">/</span><span class="n">WIDTH</span>

    <span class="k">return</span> <span class="n">pred_area</span><span class="p">,</span> <span class="n">actual_area</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span>



<span class="k">def</span> <span class="nf">calculate_correlations</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function returns a set of dictionaries containing the actual EU-seq values and the predicted values. </span>
<span class="sd">    This is done both on a bin by bin basis and integrated over the target gene length (normalized by gene length).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Storing gene position&#39;s dict:</span>

    <span class="n">pos_dict</span> <span class="o">=</span> <span class="n">read_gene_positions</span><span class="p">(</span><span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">,</span> <span class="n">condition_list</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">)</span>

    <span class="c1"># Normalized Area per bin for the target gene (central gene in observation window)</span>
    <span class="n">pred_list_A_per_bin_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">actual_list_A_per_bin_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Predicted and actual EU-seq values in all positions:</span>
    <span class="n">pred_list_values_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">actual_list_values_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;_rev&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ID</span><span class="p">):</span> <span class="c1">#Only add the forward strand to the analyses</span>
            <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
                <span class="c1"># Extend binwise prediction and target lists</span>
                <span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">line_p</span><span class="p">))</span>
                <span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">line_a</span><span class="p">))</span>

                <span class="c1"># Extend area predictions</span>
                <span class="n">pred_area</span><span class="p">,</span> <span class="n">actual_area</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span> <span class="o">=</span> <span class="n">calculate_area_gene</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">)</span>

                <span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">pred_area</span><span class="p">])</span>
                <span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">actual_area</span><span class="p">])</span>

                <span class="c1"># Adding area to extended table:</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">ID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">),</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred_area</span> <span class="c1">#Edit existing table</span>

            <span class="n">PLOT</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">ID_LIST</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000078673.10_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000067261.4_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000028234.6_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000028280.9_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000035969.15_forward&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">PLOT</span> <span class="ow">and</span> <span class="p">(</span><span class="n">ID</span> <span class="ow">in</span> <span class="n">ID_LIST</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">pred_area</span><span class="p">)</span>
                <span class="n">plot_target_predictions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">line_p</span><span class="p">,</span> <span class="n">line_p</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span><span class="p">,</span> <span class="n">pred_area</span><span class="p">)</span>

    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pred_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">actual_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">pred_list_values_dict</span><span class="p">,</span> <span class="n">actual_list_values_dict</span>

<span class="k">def</span> <span class="nf">add_gene_area_after_enhancer_perturbation</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function reopens the csv file with reference genes and enhancers and edits the area under the reference gene when silencing a given enhancer.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Storing gene position&#39;s dict:</span>

    <span class="n">pos_dict</span> <span class="o">=</span> <span class="n">read_gene_positions</span><span class="p">(</span><span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">,</span> <span class="n">condition_list</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        <span class="n">seq_center</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">line_p</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">ID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">enhancer</span> <span class="o">=</span> <span class="n">ID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">rel_start</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
        <span class="n">rel_end</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">gene</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span> <span class="k">else</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span>

        <span class="n">WIDTH</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">rel_end</span> <span class="o">-</span> <span class="n">rel_start</span><span class="p">))</span>

        <span class="c1"># Appending area</span>
        <span class="n">pred_area</span> <span class="o">=</span> <span class="n">simpson</span><span class="p">(</span><span class="n">line_p</span><span class="p">[</span><span class="n">rel_start</span><span class="p">:</span><span class="n">rel_end</span><span class="p">])</span><span class="o">/</span><span class="n">WIDTH</span>

        <span class="c1"># Adding area to extended table:</span>
        <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">enhancer</span><span class="p">),</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred_area</span> <span class="c1">#Edit existing table</span>

        <span class="n">PLOT</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">ID_LIST</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000078673.10&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000067261.4&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">PLOT</span> <span class="ow">and</span> <span class="p">(</span><span class="n">gene</span> <span class="ow">in</span> <span class="n">ID_LIST</span><span class="p">):</span>
            <span class="n">plot_target_predictions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span><span class="p">,</span> <span class="n">pred_area</span><span class="p">)</span>

    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cmap</span><span class="p">:</span> <span class="n">colors</span><span class="o">.</span><span class="n">Colormap</span> <span class="o">=</span> <span class="s2">&quot;afmhot&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">50</span><span class="p">),</span> <span class="n">density</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">DELTA</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function plots predicted vs. actual values and provides the Spearman and Pearson correlations of the regressions. </span>
<span class="sd">    The style of the plot changes depending on whether the predictions are at a bin level or a gene level.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the counts histogram to plot correlation with density colormap</span>
    <span class="n">counts</span><span class="p">,</span> <span class="n">xedges</span><span class="p">,</span> <span class="n">yedges</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist2d</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">binlims</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">binlims</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">200</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">binlims</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">binlims</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">200</span><span class="p">)],</span> <span class="n">density</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Plot regression of single predicted values vs. real</span>
    <span class="n">pearson_r</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spearman_r</span> <span class="o">=</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">m</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">deg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># Actual fit:</span>
    <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">a3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">deg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># Regression from origin:</span>
    <span class="c1">#m_0 , _, _, _ = np.linalg.lstsq(np.array(actual_list_values)[:,np.newaxis], pred_list_values) #x needs to be a column vector for this function</span>

    <span class="c1"># Create the figure</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>    
    <span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">,</span> <span class="n">step</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mf">.1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">density</span><span class="p">:</span>
        <span class="n">eps</span><span class="o">=</span> <span class="mi">1</span>
        <span class="c1"># Counts matrix is transposed: the origin convention for plt.imshow is (top, left), for plt.hist2D is (bottom,left)</span>
        <span class="n">counts</span><span class="p">[</span><span class="n">counts</span> <span class="o">&lt;</span> <span class="n">eps</span><span class="p">]</span> <span class="o">=</span><span class="mi">1</span>
        <span class="n">hist</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">&#39;lower&#39;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="n">xedges</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">xedges</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">yedges</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">yedges</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]],</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>  <span class="n">norm</span><span class="o">=</span><span class="n">colors</span><span class="o">.</span><span class="n">LogNorm</span><span class="p">())</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">hist</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axline</span><span class="p">(</span><span class="n">xy1</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="n">slope</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Linear fit:</span><span class="se">\n</span><span class="s1">$ y = </span><span class="si">{</span><span class="n">m</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">x </span><span class="si">{</span><span class="n">b</span><span class="si">:</span><span class="s1">+.3f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_p:</span><span class="si">{</span><span class="n">pearson_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_s:</span><span class="si">{</span><span class="n">spearman_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;royalblue&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#_ = ax.plot(np.arange(min_val,max_val,step), a1*(np.arange(min_val,max_val,step)**2)+a2*(np.arange(min_val,max_val,step))+a3, color=&#39;royalblue&#39;, ls = &quot;--&quot;, lw=1, label= f&quot;Polynomial fit:\n$y={a1:.3f}x^2+{a2:.3f}x+{a3:.3f}$&quot;)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;scaled&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_val</span><span class="p">,</span><span class="n">max_val</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">min_val</span><span class="p">,</span><span class="n">max_val</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;log$_2$(True +1)&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;log$_2$(Predicted +1)&quot;</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.48</span><span class="p">,</span><span class="mf">.8</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>  <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">markeredgewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;scaled&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axline</span><span class="p">(</span><span class="n">xy1</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="n">slope</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Linear fit:</span><span class="se">\n</span><span class="s1">$ y = </span><span class="si">{</span><span class="n">m</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">x </span><span class="si">{</span><span class="n">b</span><span class="si">:</span><span class="s1">+.3f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_p:</span><span class="si">{</span><span class="n">pearson_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_s:</span><span class="si">{</span><span class="n">spearman_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;royalblue&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#ax.plot(np.arange(-20,200,1), a1*(np.arange(-20,200,1)**2)+a2*(np.arange(-20,200,1))+a3, color=&#39;royalblue&#39;, ls = &quot;--&quot;, lw=1, label= f&quot;Polynomial fit:\n$y={a1:.3f}x^2+{a2:.3f}x{a3:.3f}$&quot;)</span>
        <span class="k">if</span> <span class="n">DELTA</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Actual $\Delta_</span><span class="si">{exp}</span><span class="s2">$  (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted $\Delta_</span><span class="si">{exp}</span><span class="s2">$ (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.53</span><span class="p">,</span><span class="mf">.87</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Actual averaged expression (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted averaged expression (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.53</span><span class="p">,</span><span class="mf">.87</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

    <span class="c1">#ax.plot(np.arange(0,500,1), np.arange(0,500,1), color=&#39;k&#39;, ls = &quot;dashed&quot;, lw=1)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Regression_histogram_</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
    

<span class="k">def</span> <span class="nf">plot_chromatin_state_combinations</span><span class="p">(</span><span class="n">array_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">skip_n_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is aimed to visualize the different combinations of values for different chromatin marks, which define the chromatin states.   </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">H3K4me3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="n">H3K27ac</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="n">H3K27me3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>

    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">array_path</span><span class="p">):</span>
            <span class="k">if</span> <span class="s2">&quot;_rev&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="p">:</span>
                    <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

                    <span class="c1"># Adding mark enrichment values to a long array</span>
                    <span class="n">H3K4me3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">H3K4me3</span><span class="p">,</span><span class="n">array</span><span class="p">[</span><span class="mi">1</span><span class="p">][::</span><span class="n">skip_n_bins</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">H3K27ac</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">H3K27ac</span><span class="p">,</span><span class="n">array</span><span class="p">[</span><span class="mi">2</span><span class="p">][::</span><span class="n">skip_n_bins</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                    <span class="n">H3K27me3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">H3K27me3</span><span class="p">,</span><span class="n">array</span><span class="p">[</span><span class="mi">3</span><span class="p">][::</span><span class="n">skip_n_bins</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


    <span class="c1"># Plotting the 3d combinations at all sites</span>
    <span class="n">scaling_factor</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># Inputs were directly given in reads</span>
    <span class="n">altitude</span><span class="p">,</span> <span class="n">azimuth</span> <span class="o">=</span> <span class="mi">45</span><span class="p">,</span><span class="mi">45</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">))</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">projection</span><span class="o">=</span><span class="s2">&quot;3d&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">view_init</span><span class="p">(</span><span class="n">altitude</span><span class="p">,</span> <span class="n">azimuth</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1">#ax.plot_surface(x_val, y_val, avg_att_mat, rstride=1,cstride=1,linewidth=0, cmap=&quot;terrain&quot;)#,  vmin=-2e-4, vmax=2e-4)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">H3K4me3</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">H3K27ac</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">H3K27me3</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span><span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;All states combined&quot;</span><span class="p">)</span><span class="c1">#,  vmin=-2e-4, vmax=2e-4)</span>

    <span class="c1"># Directions</span>
    <span class="n">zdirs</span> <span class="o">=</span> <span class="p">((</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">.2</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span> <span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="s1">&#39;z&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">)</span>
    <span class="n">xs</span> <span class="o">=</span> <span class="p">(</span><span class="mi">250</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="mi">60</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="mi">30</span><span class="p">)</span>
    <span class="n">ys</span> <span class="o">=</span> <span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">120</span><span class="p">)</span>
    <span class="n">zs</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span><span class="mi">17</span><span class="p">,</span><span class="mi">5</span><span class="p">)</span>

    <span class="n">labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Active promoters&quot;</span><span class="p">,</span> <span class="s2">&quot;Active enhancers&quot;</span><span class="p">,</span> <span class="s2">&quot;Silenced</span><span class="se">\n</span><span class="s2">chromatin&quot;</span><span class="p">,</span><span class="s2">&quot;$P2$&quot;</span><span class="p">,</span><span class="s2">&quot;P1&quot;</span><span class="p">)</span>
    <span class="n">label_colors</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;#ff0000&quot;</span><span class="p">,</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span><span class="s2">&quot;darkgreen&quot;</span><span class="p">,</span><span class="s2">&quot;purple&quot;</span><span class="p">,</span> <span class="s2">&quot;goldenrod&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">zdir</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">zdirs</span><span class="p">,</span> <span class="n">xs</span><span class="p">,</span> <span class="n">ys</span><span class="p">,</span> <span class="n">zs</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">label_colors</span><span class="p">):</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">zdir</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">c</span><span class="p">)</span>


    <span class="c1"># Arrow towards silenced branch:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;goldenrod&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>

    <span class="c1"># Inwards arrow towards silenced branch for P2:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;purple&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">quiver</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="p">,</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;purple&#39;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.7</span><span class="p">)</span>
    <span class="c1">#ax.quiver(50, 10, 10, -20,0,0, color=&#39;purple&#39;,lw=.7)</span>




    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">,</span><span class="mi">400</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">,</span><span class="mi">400</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">40</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">200</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">200</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=-</span><span class="mi">40</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_zlabel</span><span class="p">(</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">180</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_zticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="mi">80</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="mi">80</span><span class="p">])</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">left</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">top</span><span class="o">=</span><span class="mf">.95</span><span class="p">,</span> <span class="n">bottom</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;chromstates.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">n_central_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
                             <span class="n">l_in</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10001</span><span class="p">,</span>
                             <span class="n">n_out</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">401</span><span class="p">,</span>
                             <span class="n">n_in</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                             <span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                             <span class="n">SPLIT</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is aimed to read attribution arrays from the chromatin landscape branch. Attribution arrays score the </span>
<span class="sd">    contribution of each position in the input arrays (4,10.001) towards each output node/position (401).</span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: path where the attribution arrays are stored.</span>
<span class="sd">        figure_path: path where we want to store the outputs</span>
<span class="sd">        n_central_bins: number of central bins (from -N_BINS to  N_BINS+1)</span>
<span class="sd">        track_dict: dictionary containing the names of the tracks and plot details</span>
<span class="sd">        l_in: input length </span>
<span class="sd">        n_out: number of output nodes</span>
<span class="sd">        n_in: number of input channels</span>
<span class="sd">        TRAIN_ATTR: whether this is a training or test run (folder structure changes).</span>
<span class="sd">        SPLIT: if Train, we need to know which of the stored batches it is</span>

<span class="sd">    Returns:</span>
<span class="sd">        Plots with attribution scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">n_central_bins</span> <span class="o">//</span> <span class="mi">2</span>  <span class="c1"># Number of bins from TSS that we want on one side</span>
    <span class="n">N_IN_BINS</span> <span class="o">=</span> <span class="n">l_in</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">x_val</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_IN_BINS</span><span class="p">,</span><span class="n">N_IN_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">500</span><span class="p">)</span>

    <span class="c1"># Getting attributions:</span>
    <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
    <span class="n">landscape_avg</span> <span class="o">=</span> <span class="p">{</span><span class="n">condition</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_in</span><span class="p">,</span><span class="n">l_in</span><span class="p">))</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">}</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
            <span class="c1"># CHROMATIN MARK ATTRIBUTIONS:</span>
            <span class="n">added_folders</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;/samples/</span><span class="si">{</span><span class="n">SPLIT</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">TRAIN_ATTR</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="c1"># Different folder structure</span>
            <span class="n">att_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}{</span><span class="n">added_folders</span><span class="si">}</span><span class="s2">/attributions/gene_expression/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">att_path</span><span class="p">):</span>
                <span class="n">attributions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">att_path</span><span class="p">)</span>
                <span class="c1">#if (i == -73) and (condition == &quot;_ctrl&quot;):</span>
                <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span> 
                <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="n">j</span><span class="p">]),</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Chromatin_landscape_attributions_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_ctrl.png&quot;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

                <span class="c1"># Add a small phase to remove high frequency fluctuations:</span>
                <span class="n">eps</span> <span class="o">=</span> <span class="mi">10</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                
                <span class="n">centered_attr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">attributions</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="nb">int</span><span class="p">(</span><span class="n">eps</span><span class="p">),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="n">n_out</span><span class="o">*</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">centered_attr</span><span class="p">))</span>

    <span class="n">SCALING_FACTOR</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">landscape_avg</span><span class="p">[</span><span class="n">cond</span><span class="p">]</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">])</span>

    
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span>   
        <span class="n">sub_axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">axes</span><span class="p">([</span><span class="mf">.615</span><span class="p">,</span> <span class="mf">.57</span><span class="p">,</span> <span class="mf">.27</span><span class="p">,</span> <span class="mf">.27</span><span class="p">])</span>  
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
            <span class="n">sub_axes</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">markersize</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span><span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
        
        <span class="c1">#sub_axes.set_xlim(xlim)</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">,</span><span class="mi">5000</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">501</span><span class="p">,</span><span class="mi">500</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5000</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">xlim</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#(0,.002))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Attribution score&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to the predicted locus (kbp)&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">400</span><span class="p">,</span><span class="mi">401</span><span class="p">,</span><span class="mi">100</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">40</span><span class="p">,</span><span class="mi">41</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Landscape_absolute_average_</span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">model_comparison</span><span class="p">(</span><span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">performance_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">color_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates bar plots comparin pearson and spearman correlations of the different models.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">shift</span> <span class="o">=</span> <span class="mi">8</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">model</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">performance_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="c1"># if i == 0:</span>
        <span class="c1">#     for j in range(2):</span>
        <span class="c1">#         plt.plot(pos,complementary_dict[i][&quot;Spearman&quot;], marker=&#39;d&#39;, color=&#39;k&#39;, markersize=10, markeredgewidth=0)</span>
        <span class="c1">#         plt.plot(pos+shift,complementary_dict[i][&quot;Pearson&quot;], marker=&#39;d&#39;, color=&#39;k&#39;, markersize=10, markeredgewidth=0)</span>
                
        <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">performance_dict</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="s2">&quot;Spearman&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">color_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">pos</span><span class="o">+</span><span class="n">shift</span><span class="p">,</span> <span class="n">performance_dict</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="s2">&quot;Pearson&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">color_list</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="c1">#plt.plot(pos,performance_dict[model][&quot;Spearman&quot;], marker=&#39;d&#39;, color=&#39;k&#39;, markersize=3, markeredgewidth=0)</span>
        <span class="c1">#plt.plot(pos+shift,performance_dict[model][&quot;Pearson&quot;], marker=&#39;d&#39;, color=&#39;k&#39;, markersize=3, markeredgewidth=0)</span>
        <span class="n">pos</span> <span class="o">+=</span> <span class="mi">1</span>


    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([</span><span class="mf">3.5</span><span class="p">,</span><span class="mf">11.5</span><span class="p">],[</span><span class="s2">&quot;Spearman&quot;</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;lower left&quot;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">0.04</span><span class="p">,</span><span class="mf">1.08</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot; Binwise correlation for log$_2$(reads +1)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;Model_comparison.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">model_comparison_II</span><span class="p">(</span><span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">performance_dicts</span><span class="p">:</span> <span class="nb">list</span> <span class="p">,</span> <span class="n">color_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates bar plots comparing pearson and spearman correlations of the different models.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">performance_dict</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">performance_dicts</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">performance_dict</span><span class="p">)</span>
        <span class="n">marker</span> <span class="o">=</span> <span class="s1">&#39;o&#39;</span> <span class="k">if</span> <span class="n">j</span><span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;^&#39;</span>
        <span class="n">task</span> <span class="o">=</span> <span class="s1">&#39;binwise&#39;</span> <span class="k">if</span> <span class="n">j</span><span class="o">==</span><span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;area&#39;</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">model</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">performance_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">performance_dict</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="s2">&quot;Pearson&quot;</span><span class="p">],</span><span class="n">performance_dict</span><span class="p">[</span><span class="n">model</span><span class="p">][</span><span class="s2">&quot;Spearman&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">color_list</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="n">marker</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">model</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">task</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;lower center&quot;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mf">.4</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">((</span><span class="mf">.4</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Spearman&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Pearson&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;Model_comparison_II.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="c1">####################### ATTRIBUTIONS ####################</span>

 
<span class="k">def</span> <span class="nf">plot_microc_original_and_rotated</span><span class="p">(</span><span class="n">pathlist</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function plots the Micro-C contact maps (both original and rotated-cropped)  given a sample name.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bin_per_kbp</span> <span class="o">=</span> <span class="n">num_bins</span><span class="o">/</span><span class="n">window_of_observation</span> <span class="c1">#Unit (bin/kbp)</span>
    <span class="n">max_shift</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mi">200</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">)</span> <span class="c1"># Distance to crop because it was rolled from the other side</span>

    <span class="k">for</span> <span class="n">p</span><span class="p">,</span><span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">pathlist</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">sample</span><span class="o">.</span><span class="n">name</span><span class="p">:</span> <span class="c1"># Nice example: &quot;ENSMUST00000105369.7.npy&quot; common &quot;ENSMUST00000001565.14.npy&quot; </span>
                <span class="n">microc_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
                <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
                <span class="c1">#plt.imshow(microc_map, cmap=&quot;YlOrRd&quot;)</span>
                <span class="c1"># plt.plot(200*bin_per_kbp,50,marker=&quot;x&quot;)</span>
                <span class="c1"># plt.plot(-200*bin_per_kbp+313,50,marker=&quot;x&quot;)</span>
                <span class="c1"># plt.plot(313,50,marker=&quot;x&quot;)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">microc_map</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;YlOrRd&quot;</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
                <span class="c1"># Real scale</span>
                <span class="n">center_bin</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">microc_map</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span>
                <span class="n">bin_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">13</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">microc_map</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span><span class="mi">50</span><span class="p">)</span>
                <span class="n">kbp_per_bin</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="n">bin_per_kbp</span>
                <span class="n">bin_val</span> <span class="o">=</span> <span class="p">(</span><span class="n">bin_pos</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">microc_map</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">kbp_per_bin</span> <span class="c1">#Unit (kbp/bin)</span>
                <span class="n">bin_val</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span><span class="o">%</span><span class="k">item</span> for item in bin_val]
                <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">bin_pos</span><span class="p">,</span><span class="n">bin_val</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Distance in 1D sequence (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
                <span class="c1"># Careful! Scale is not the distance between points but the distance between interacting points in 1D! </span>
                <span class="n">bin_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">microc_map</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span><span class="mi">25</span><span class="p">)</span>
                <span class="n">bin_val</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">bin_pos</span><span class="o">*</span><span class="n">kbp_per_bin</span> <span class="c1">#*(kbp_per_bin) + np.shape(avg_att_mat)[0]</span>
                <span class="n">bin_val</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span><span class="o">%</span><span class="k">item</span> for item in bin_val]
                <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">bin_pos</span><span class="p">,</span><span class="n">bin_val</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">p</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
                    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Separation between</span><span class="se">\n</span><span class="s2"> interacting elements</span><span class="se">\n</span><span class="s2"> in 1D sequence (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;MicroC_input_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">p</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>



<span class="k">def</span> <span class="nf">plot_microc_attributions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">,</span> <span class="n">ATTR_type_list</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">sigma</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Plot the attribution maps for micro-c matrices</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bin_per_kbp</span> <span class="o">=</span> <span class="n">num_bins</span><span class="o">/</span><span class="n">window_of_observation</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">bin_per_kbp</span><span class="p">)</span>
    <span class="n">max_shift</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mi">200</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">)</span> <span class="c1"># Distance to crop because it was rolled from the other side</span>

    <span class="n">avg_att_mat_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ATTR_type</span> <span class="ow">in</span> <span class="n">ATTR_type_list</span><span class="p">:</span>
            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">ATTR_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">129</span><span class="p">,</span><span class="mi">626</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sample_ids</span><span class="p">:</span>
            <span class="c1">#path = Path(f&quot;../runs/training_runs/gene_expression_exformer_unlimited_all_rotated/results/expression_output/{i}{condition}/samples/17880/attributions/contact_maps/{i}{condition}.npy&quot;) </span>
            <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_rotated_pure_conv/expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/attributions/contact_maps/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span> 
            <span class="n">att_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
            <span class="n">att_mat_rolled</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">att_mat</span><span class="p">,</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="s2">&quot;abs&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">))</span><span class="o">*</span><span class="nb">abs</span><span class="p">(</span><span class="n">att_mat_rolled</span><span class="p">)</span>
            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="s2">&quot;signed&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">))</span><span class="o">*</span><span class="n">att_mat_rolled</span>

            <span class="k">if</span> <span class="p">(</span><span class="n">i</span><span class="o">==</span><span class="mi">199</span><span class="p">):</span>

                <span class="n">fig_1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span><span class="c1">#(11.5, 5))</span>
                <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>
                <span class="n">ax</span> <span class="o">=</span> <span class="n">fig_1</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
                <span class="n">fig_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">att_mat</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>  <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.0002</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.0002</span><span class="p">)</span>
                <span class="n">ax_II</span> <span class="o">=</span> <span class="n">fig_1</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
                <span class="n">fig_display</span> <span class="o">=</span> <span class="n">ax_II</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">att_mat_rolled</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>  <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.0002</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.0002</span><span class="p">)</span>
                <span class="n">cax</span> <span class="o">=</span> <span class="n">fig_1</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
                <span class="c1"># Add the colorbar</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">fig_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
                <span class="n">fig_1</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Signed_attributions_</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig_1</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">CONDITION</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ATTR_type</span> <span class="ow">in</span> <span class="n">ATTR_type_list</span><span class="p">:</span>
            <span class="n">avg_att_mat</span> <span class="o">=</span> <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">ATTR_type</span><span class="p">]</span> 
            <span class="n">avg_att_mat</span> <span class="o">=</span> <span class="n">gaussian_filter</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
            <span class="n">avg_att_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">[:,</span><span class="n">max_shift</span><span class="p">:</span><span class="o">-</span><span class="n">max_shift</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Assuming &#39;img&#39; is your image data</span>
            <span class="n">y_projected_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># Compress along x-axis</span>
            <span class="n">x_projected_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Compress along y-axis</span>

            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span><span class="c1">#(11.5, 5))</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>

            <span class="c1"># Main image</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
            <span class="c1"># Create an axis for the colorbar</span>
            <span class="n">cax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">ATTR_type</span> <span class="o">==</span> <span class="s2">&quot;abs&quot;</span><span class="p">:</span>
                <span class="n">img_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;inferno&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.5e-4</span><span class="p">)</span>
                <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">img_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">5e-5</span><span class="p">])</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticklabels</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">img_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mf">0.5e-4</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.5e-4</span><span class="p">)</span>
                <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">img_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">([</span><span class="o">-</span><span class="mf">5e-5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">5e-5</span><span class="p">])</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticklabels</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>


            <span class="c1"># Plot vertical dashed line</span>
            <span class="n">ref_y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">))</span>
            <span class="n">ref_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">ones_like</span><span class="p">(</span><span class="n">ref_y</span><span class="p">)</span>
            <span class="n">im</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ref_x</span><span class="p">,</span><span class="n">ref_y</span><span class="p">,</span> <span class="s2">&quot;k--&quot;</span><span class="p">)</span>


            <span class="c1"># Plot diagonals from prediction point:</span>
            <span class="n">diag_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span>
            <span class="n">diag_x2</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span>
            <span class="n">diag_y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">diag_x</span><span class="p">,</span><span class="n">diag_y</span><span class="p">,</span> <span class="s2">&quot;b--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.5</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">diag_x2</span><span class="p">,</span><span class="n">diag_y</span><span class="p">,</span> <span class="s2">&quot;b--&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.5</span><span class="p">)</span>

            <span class="c1"># Real scale</span>
            <span class="n">bin_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span><span class="mi">25</span><span class="p">)</span>
            <span class="c1">#bin_val = (bin_pos-np.shape(avg_att_mat)[1]//2)*kbp_per_bin</span>
            <span class="n">kbp_per_bin</span> <span class="o">=</span> <span class="mi">1</span><span class="o">/</span><span class="n">bin_per_kbp</span>
            <span class="n">bin_val</span> <span class="o">=</span> <span class="p">(</span><span class="n">bin_pos</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="n">kbp_per_bin</span> <span class="c1">#Unit (kbp/bin)</span>
            <span class="n">bin_val</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span><span class="o">%</span><span class="k">item</span> for item in bin_val]
            <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">bin_pos</span><span class="p">,</span><span class="n">bin_val</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>

            <span class="c1"># Careful! Scale is not the distance between points but the distance between interacting points in 1D! </span>
            <span class="n">bin_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">shape</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span><span class="mi">25</span><span class="p">)</span>
            <span class="n">bin_val</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">bin_pos</span><span class="o">*</span><span class="n">kbp_per_bin</span> <span class="c1"># Unit (kbp/bin)</span>
            <span class="n">bin_val</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span><span class="o">%</span><span class="k">item</span> for item in bin_val]
            <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">bin_pos</span><span class="p">,</span><span class="n">bin_val</span><span class="p">)</span>


            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from the predicted point (kbp)&quot;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Separation between</span><span class="se">\n</span><span class="s2"> interacting elements</span><span class="se">\n</span><span class="s2"> in 1D sequence (kbp)&quot;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>

            <span class="c1"># # Top plot</span>
            <span class="c1"># axb = fig.add_subplot(gs[0, 0], sharex=ax)</span>
            <span class="c1"># axb.plot(np.arange(len(y_projected_mat)), y_projected_mat, color=&#39;k&#39;)</span>
            <span class="c1"># plt.setp(axb.get_xticklabels(), visible=False)</span>
            <span class="c1"># axb.set_ylabel(&quot;Averaged attribution score&quot;, fontsize= 8)</span>

            <span class="c1"># # Right plot</span>
            <span class="c1"># axl = fig.add_subplot(gs[1, 1], sharey=ax)</span>
            <span class="c1"># axl.plot(x_projected_mat, np.arange(len(x_projected_mat)), color=&#39;k&#39;)</span>
            <span class="c1"># plt.setp(axl.get_yticklabels(), visible=False)</span>
            <span class="c1"># axb.set_ylabel(&quot;Averaged attribution score&quot;, fontsize = 8)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mf">.20</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Structural_attributions_</span><span class="si">{</span><span class="n">ATTR_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">CONDITION</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sigma</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="c1">##### Model benchmarks ######</span>

<span class="k">def</span> <span class="nf">plot_benchmark_models_predictions</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">,</span> <span class="n">results_path</span><span class="p">,</span><span class="n">target_samples</span><span class="p">,</span> <span class="n">condition_list</span><span class="p">,</span> <span class="n">READ_SCALING_FACTOR</span><span class="p">,</span><span class="n">PLOT_LIST</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;binary&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function is aimed to compare the predicted profiles and performances between different models.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">CREATE_DFS</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">param_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;CLASTER&quot;</span><span class="p">:{</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">:</span><span class="mi">401</span><span class="p">,</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span><span class="mi">200</span><span class="p">,</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">:</span><span class="mi">401</span><span class="p">},</span>
                             <span class="s2">&quot;hyenadna&quot;</span><span class="p">:{</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">:</span><span class="mi">31</span><span class="p">,</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span><span class="mi">15</span><span class="p">,</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">:</span><span class="mi">31</span><span class="p">},</span>
                             <span class="s2">&quot;Enformer&quot;</span><span class="p">:{</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">:</span><span class="mi">115</span><span class="p">,</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span><span class="mi">57</span><span class="p">,</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">:</span><span class="mi">115</span><span class="p">}}</span>


    <span class="n">N_SAMPLES</span> <span class="o">=</span> <span class="mi">1295</span>
    <span class="n">pos_dict</span> <span class="o">=</span> <span class="n">read_gene_positions</span><span class="p">(</span><span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>  <span class="c1"># Storing gene position&#39;s dict:</span>

    <span class="c1">### Claster and targets </span>
    <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;CLASTER&quot;</span> <span class="c1">#&quot;Enformer&quot;</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span> <span class="n">condition_list</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="n">ids</span><span class="p">[::</span><span class="mi">2</span><span class="p">]</span> <span class="c1"># Keep only forward!</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">predicted</span><span class="p">[</span><span class="o">~</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;_rev&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span> 
    <span class="n">actual</span> <span class="o">=</span> <span class="n">actual</span><span class="p">[</span><span class="o">~</span><span class="n">actual</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;_rev&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>


    <span class="c1">### hyenadna and Enformer</span>
    <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;hyenadna&quot;</span>
    <span class="n">pred_list_df_hyena</span><span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;predictions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">N_SAMPLES</span><span class="p">,</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">]),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span> 
    <span class="n">actual_list_df_hyena</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">N_SAMPLES</span><span class="p">,</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">]),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span> 
    
    <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;Enformer&quot;</span>
    <span class="n">pred_list_df_enformer</span><span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;predictions&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">N_SAMPLES</span><span class="p">,</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">]),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span> 
    <span class="n">actual_list_df_enformer</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">N_SAMPLES</span><span class="p">,</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;OUT_LEN&quot;</span><span class="p">]),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">)],</span> <span class="n">index</span><span class="o">=</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span> 
    <span class="c1">###</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Normalized Area per bin for the target gene (central gene in observation window)</span>
    <span class="n">A_per_gene_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Enformer&quot;</span><span class="p">:{</span><span class="s2">&quot;Predicted&quot;</span><span class="p">:[],</span><span class="s2">&quot;Actual&quot;</span><span class="p">:[]},</span><span class="s2">&quot;hyenadna&quot;</span><span class="p">:{</span><span class="s2">&quot;Predicted&quot;</span><span class="p">:[],</span><span class="s2">&quot;Actual&quot;</span><span class="p">:[]}}</span>

    <span class="k">for</span> <span class="nb">id</span><span class="p">,</span> <span class="n">line_pc</span><span class="p">,</span> <span class="n">line_ac</span><span class="p">,</span> <span class="n">line_ph</span><span class="p">,</span> <span class="n">line_pe</span>  <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">pred_list_df_hyena</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">pred_list_df_enformer</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        <span class="c1">#if counter &lt; 50:   </span>
         
        <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;Enformer&quot;</span>
        <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span>
        <span class="n">MODEL_SHIFT</span> <span class="o">=</span> <span class="p">(</span><span class="n">param_dict</span><span class="p">[</span><span class="s2">&quot;CLASTER&quot;</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span>
        <span class="c1">#MODEL_SHIFT = (param_dict[&quot;CLASTER&quot;][&quot;n_central_bins&quot;] - param_dict[&quot;hyenadna&quot;][&quot;n_central_bins&quot;])//2</span>
        <span class="c1">#ENFORMER_SHIFT = (param_dict[&quot;Enformer&quot;][&quot;n_central_bins&quot;] - param_dict[&quot;hyenadna&quot;][&quot;n_central_bins&quot;])//2</span>
        <span class="n">pred_area_e</span><span class="p">,</span> <span class="n">actual_area_e</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calculate_area_gene</span><span class="p">(</span><span class="n">line_pe</span><span class="p">,</span> <span class="n">line_ac</span><span class="p">[</span><span class="n">MODEL_SHIFT</span><span class="p">:</span><span class="o">-</span><span class="n">MODEL_SHIFT</span><span class="p">],</span> <span class="nb">id</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">)</span>
        <span class="c1">#pred_area_e, actual_area_e, *_ = calculate_area_gene(line_pe[ENFORMER_SHIFT:-ENFORMER_SHIFT], line_ac[MODEL_SHIFT:-MODEL_SHIFT], id, pos_dict, N_BINS)</span>
        <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pred_area_e</span><span class="p">)</span>
        <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_area_e</span><span class="p">)</span>

        <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;hyenadna&quot;</span>
        <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span>
        <span class="n">MODEL_SHIFT</span> <span class="o">=</span> <span class="p">(</span><span class="n">param_dict</span><span class="p">[</span><span class="s2">&quot;CLASTER&quot;</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">])</span><span class="o">//</span><span class="mi">2</span>
        <span class="n">pred_area_h</span><span class="p">,</span> <span class="n">actual_area_h</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">calculate_area_gene</span><span class="p">(</span><span class="n">line_ph</span><span class="p">,</span> <span class="n">line_ac</span><span class="p">[</span><span class="n">MODEL_SHIFT</span><span class="p">:</span><span class="o">-</span><span class="n">MODEL_SHIFT</span><span class="p">],</span> <span class="nb">id</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span><span class="n">N_BINS</span><span class="p">)</span>
        <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pred_area_h</span><span class="p">)</span>
        <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_area_h</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">PLOT_LIST</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">nrows</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">sharex</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
            <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;CLASTER&quot;</span> 
            <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_pc</span><span class="p">[:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">*</span><span class="n">READ_SCALING_FACTOR</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;silver&#39;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.8</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;CLASTER&quot;</span><span class="p">)</span>
            <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_ac</span><span class="p">[:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">*</span><span class="n">READ_SCALING_FACTOR</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;EU-seq target&quot;</span><span class="p">)</span>
            <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="n">MODEL_SHIFT</span><span class="p">,</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span> <span class="n">MODEL_SHIFT</span><span class="o">+</span><span class="n">param_dict</span><span class="p">[</span><span class="s2">&quot;hyenadna&quot;</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]],[</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
            <span class="c1">#axs[2][1].fill_between(np.arange(-param_dict[MODEL][&quot;N_BINS&quot;],param_dict[MODEL][&quot;N_BINS&quot;]+1),line_pc[param_dict[MODEL][&quot;n_central_bins&quot;]:]*READ_SCALING_FACTOR, lw=1, color=&#39;royalblue&#39;, alpha=.9, label=&quot;Predicted expression treated&quot;)</span>
            <span class="c1">#axs[3][1].fill_between(np.arange(-param_dict[MODEL][&quot;N_BINS&quot;],param_dict[MODEL][&quot;N_BINS&quot;]+1),line_ac[param_dict[MODEL][&quot;n_central_bins&quot;]:]*READ_SCALING_FACTOR, lw=1, color=&#39;k&#39;, alpha=.6, label=&quot;Actual expression treated&quot;)</span>
            
            <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;hyenadna&quot;</span>
            <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_ph</span><span class="p">[:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">*</span><span class="n">READ_SCALING_FACTOR</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;silver&#39;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.8</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;HyenaDNA-32k </span><span class="si">{</span><span class="n">pred_area_h</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">actual_area_h</span><span class="si">}</span><span class="s2">&quot;</span> <span class="p">)</span>
            <span class="c1">#axs[0][1].fill_between(np.arange(-param_dict[MODEL][&quot;N_BINS&quot;],param_dict[MODEL][&quot;N_BINS&quot;]+1),line_ph[param_dict[MODEL][&quot;n_central_bins&quot;]:]*READ_SCALING_FACTOR, lw=1, color=&#39;royalblue&#39;, alpha=.9)</span>

            <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;Enformer&quot;</span>
            <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_pe</span><span class="p">[:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">*</span><span class="n">READ_SCALING_FACTOR</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;silver&#39;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.8</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Enformer </span><span class="si">{</span><span class="n">pred_area_e</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">actual_area_e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1">#axs[1][1].fill_between(np.arange(-param_dict[MODEL][&quot;N_BINS&quot;],param_dict[MODEL][&quot;N_BINS&quot;]+1),line_pe[param_dict[MODEL][&quot;n_central_bins&quot;]:]*READ_SCALING_FACTOR, lw=1, color=&#39;royalblue&#39;, alpha=.9)</span>
            
            <span class="n">model_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;HyenaDNA-32k&#39;</span><span class="p">,</span><span class="s1">&#39;Enformer&#39;</span><span class="p">,</span><span class="s1">&#39;CLASTER&#39;</span><span class="p">,</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">axs</span><span class="p">)):</span>
                <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">50</span><span class="p">))</span>
                <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
                <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
                    
            <span class="n">axs</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Reads&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>     
            <span class="c1">#fig.legend( bbox_to_anchor=(.84,1.003),ncol=2, fancybox=True, shadow=True, fontsize=8)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">id</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

        <span class="n">counter</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># Correlations Hyena:</span>
    <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;hyenadna&quot;</span>
    <span class="n">plot_correlations</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;figure_path&quot;</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">READ_SCALING_FACTOR</span><span class="o">*</span><span class="n">pred_list_df_hyena</span><span class="o">.</span><span class="n">values</span><span class="p">[:,:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>  <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">READ_SCALING_FACTOR</span><span class="o">*</span><span class="n">actual_list_df_hyena</span><span class="o">.</span><span class="n">values</span><span class="p">[:,:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">MODEL</span><span class="si">}</span><span class="s2">_performance_ctrl_smoothL1.png&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span> <span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">density</span> <span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">DELTA</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">plot_correlations</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;figure_path&quot;</span><span class="p">],</span> <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Predicted&quot;</span><span class="p">],</span><span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">MODEL</span><span class="si">}</span><span class="s2">_performance_ctrl_gene_avg.png&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span> <span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">density</span> <span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">DELTA</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="c1"># Correlations Enformer:</span>
    <span class="n">MODEL</span> <span class="o">=</span> <span class="s2">&quot;Enformer&quot;</span>
    <span class="n">plot_correlations</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;figure_path&quot;</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">READ_SCALING_FACTOR</span><span class="o">*</span><span class="n">pred_list_df_enformer</span><span class="o">.</span><span class="n">values</span><span class="p">[:,:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span>  <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">READ_SCALING_FACTOR</span><span class="o">*</span><span class="n">actual_list_df_enformer</span><span class="o">.</span><span class="n">values</span><span class="p">[:,:</span><span class="n">param_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;n_central_bins&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">MODEL</span><span class="si">}</span><span class="s2">_performance_ctrl_smoothL1.png&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span> <span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">density</span> <span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">DELTA</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">plot_correlations</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;figure_path&quot;</span><span class="p">],</span> <span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Predicted&quot;</span><span class="p">],</span><span class="n">A_per_gene_dict</span><span class="p">[</span><span class="n">MODEL</span><span class="p">][</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span> <span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">MODEL</span><span class="si">}</span><span class="s2">_performance_ctrl_gene_avg.png&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span> <span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">density</span> <span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">DELTA</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">plot_perturbed_profiles</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">SCALE</span> <span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">sample_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">PERT_TYPE</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>

    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">])</span>

    <span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">terrain</span>  <span class="c1"># define the colormap</span>
    <span class="c1"># extract all colors from the .jet map</span>
    <span class="n">cmaplist</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cmap</span><span class="o">.</span><span class="n">N</span><span class="p">)]</span>
    <span class="c1"># force the first color entry to be grey</span>

    <span class="c1"># create the new map</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
        <span class="s1">&#39;Custom cmap&#39;</span><span class="p">,</span> <span class="n">cmaplist</span><span class="p">,</span> <span class="n">cmap</span><span class="o">.</span><span class="n">N</span><span class="p">)</span>

    <span class="n">norm</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">18</span><span class="p">)</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
    <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="n">ref_gene</span><span class="p">,</span> <span class="n">entity_id</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">index</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">sample_name</span> <span class="ow">in</span> <span class="n">index</span><span class="p">:</span>
            <span class="c1"># Get the normalized color index for the current loop iteration</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">cmap</span><span class="p">(</span><span class="n">norm</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.8</span><span class="p">)</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">table</span><span class="p">[(</span><span class="n">table</span><span class="p">[</span><span class="s2">&quot;Ref_gene&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref_gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="s2">&quot;Entity_ID&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">entity_id</span><span class="p">)][</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">SCALE</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">table</span><span class="p">[(</span><span class="n">table</span><span class="p">[</span><span class="s2">&quot;Ref_gene&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ref_gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">table</span><span class="p">[</span><span class="s2">&quot;Entity_ID&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">entity_id</span><span class="p">)][</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">SCALE</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">],</span><span class="n">i</span><span class="o">+</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">25</span><span class="p">,</span><span class="mi">25</span><span class="p">]),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">1.5</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">200</span><span class="p">))</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">40</span><span class="p">))</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Reads&quot;</span><span class="p">)</span>
            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Perturbed_profiles_</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">PERT_TYPE</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>


<span class="c1"># Latent representation analysis</span>
<span class="k">def</span> <span class="nf">visualize_filters</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> 
                       <span class="n">savepath</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                       <span class="n">first_validation_sample_index</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> 
                       <span class="n">last_validation_sample_index</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> 
                       <span class="n">step</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                       <span class="n">layer_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span> 
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     This function plots the intermediate outputs of the network. This is useful to understand what filters  </span>
<span class="sd">     are being created and what are their roles. </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
  
    <span class="n">latents</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;latents.npy&quot;</span><span class="p">)</span> 
  
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">sample</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">latents</span><span class="p">[</span><span class="n">first_validation_sample_index</span><span class="p">:</span><span class="n">last_validation_sample_index</span><span class="p">:</span><span class="n">step</span><span class="p">]):</span> 
        <span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> 
        <span class="n">sample_values</span> <span class="o">=</span> <span class="n">sample</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">n_filters_in_layer</span> <span class="o">=</span> <span class="n">sample_values</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">sample_values</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">sample_values</span><span class="p">))</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">n_filters_in_layer</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="c1">#fig_stack, axs_stack = plt.subplots(n_filters_in_layer,1, figsize=(8,8))</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="nb">filter</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sample_values</span><span class="p">):</span> 
            <span class="n">x</span> <span class="o">=</span> <span class="n">j</span><span class="o">//</span><span class="nb">int</span><span class="p">(</span><span class="n">n_filters_in_layer</span><span class="o">/</span><span class="mi">4</span><span class="p">)</span>
            <span class="n">y</span> <span class="o">=</span> <span class="n">j</span><span class="o">%</span><span class="k">int</span>(n_filters_in_layer/4)
            <span class="c1">#filter = rotate(filter, angle=45, reshape=False)</span>
            <span class="n">im</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="nb">filter</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">)</span> 
            <span class="c1">#plt.colorbar(cm.ScalarMappable( cmap=&quot;inferno&quot;), ax=axs[j//4, j%4])</span>
            <span class="c1">#cb = plt.colorbar(im, ax=axs[x, y], fraction=0.046, pad=0.04) </span>
            <span class="c1">#cb.ax.tick_params(labelsize=12)  </span>
            <span class="c1">#axs[x,y].set_title(f&quot;filter {j}&quot;, fontsize=12) </span>
            <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
            <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

            <span class="c1"># Stacked plot</span>
            <span class="c1"># fx = np.zeros(len(filter))</span>
            <span class="c1"># shift = -int(len(filter)/2)</span>
            <span class="c1"># for i in range(len(filter)):</span>
            <span class="c1">#     roll_im = np.roll(filter[i],shift-i)</span>
            <span class="c1">#     fx += roll_im/len(filter)</span>
            <span class="c1"># axs_stack[j].plot(np.arange(len(fx)),fx, lw=.5, label=j)</span>
            <span class="c1"># gradient = np.convolve(fx,np.array([1,0,-1]))</span>
            <span class="c1"># axs_stack[j].plot(np.arange(len(gradient)), gradient, ls = &quot;dashed&quot;,lw=.5, label=f&quot;{j}_conv&quot;)  </span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">layer_name</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
        <span class="c1">#fig_stack.savefig(savepath / f&quot;{sample_name}_{layer_name}_stacked.png&quot;, dpi=250)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>


    <span class="k">return</span> <span class="n">fig</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="model-performance-on-the-baseline-prediction-task">
<h2><center> Model performance on the baseline prediction task <center><a class="headerlink" href="#model-performance-on-the-baseline-prediction-task" title="Link to this heading">#</a></h2>
<p><strong>Figure 1 and related supplementary figures</strong></p>
<blockquote>
<div><p>Note: ⏰ This took 42 min when analysing all genes and enhancers in chr4:</p>
</div></blockquote>
<p>We will first add meaningful columns to the gene-enhancer coordinate information table like EP-distance, number of genes between E and P… We will do that for the samples in the test set (chr4). This information will be used in the following sections.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">samples_index_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets/perturbed_targets.csv&quot;</span><span class="p">)</span>
<span class="n">samples_data_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">perturbed_table_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_perturbed.csv&quot;</span><span class="p">)</span>

<span class="n">filter_and_extend_gene_enhancer_table</span><span class="p">(</span><span class="n">samples_index_path</span><span class="p">,</span> <span class="n">samples_data_path</span><span class="p">,</span> <span class="n">perturbed_table_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can load the predictions and compare them with the ground truth values. We will do that for point-wise predictions and also integrating the area under each reference gene. The area under the curve within the central gene’s coordinates, a proxy for length normalized nascent expression, will be added to the table that we just created.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Compute the correlations and add area for the central gene to the table</span>
<span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span> <span class="c1">#Path(&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;) #chrom_and_microc_with_attention&quot;) #Path(&quot;/Users/wjq311/Desktop/PhD/Enhancer_logic_project/Data/results_all/results_no_H3K27ac_uncoupled/expression_output&quot;)</span>
<span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_perturbed.csv&quot;</span><span class="p">)</span>

<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_1/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="c1">#[&quot;&quot;]</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;binary&quot;</span>

<span class="n">pred_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">actual_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">pred_list_values_dict</span><span class="p">,</span> <span class="n">actual_list_values_dict</span> <span class="o">=</span> <span class="n">calculate_correlations</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">,</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">input_table_path</span><span class="p">)</span>

<span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
    <span class="c1"># Prediction of all bins (1kbp resolution)</span>
    <span class="n">pred_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># Multiplying by 10 to convert from reads to decareads</span>
    <span class="n">actual_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1">#pointwise = plot_correlations(figure_path, pred_list_values, actual_list_values, condition + &quot;_binpred_noH3K27ac&quot;, cmap=cmap, binlims=(-10,100))</span>
    <span class="n">pointwise</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">,</span> <span class="n">condition</span> <span class="o">+</span> <span class="s2">&quot;_pointwise_prediction&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>

    <span class="c1"># Central gene area (divided by gene length, equiv. to averaged EU-seq signal for the target gene)</span>
    <span class="n">pred_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span> 
    <span class="n">actual_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span>
    <span class="n">area</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">,</span> <span class="n">condition</span> <span class="o">+</span> <span class="s2">&quot;_gene_area_lengthnorm_prediction&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


<span class="n">pointwise_avg</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">]),</span> <span class="s2">&quot;_binpred_noH3K27ac_both_conditions&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;afmhot&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span><span class="mi">100</span><span class="p">))</span>
<span class="n">area</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">]),</span> <span class="n">condition</span> <span class="o">+</span> <span class="s2">&quot;_gene_area_norm_noH3K27ac_both_conditions&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;afmhot&quot;</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ENSMUSG00000028234.6 77.86259545833333
ENSMUSG00000028280.9 0.08941720612903226
ENSMUSG00000035969.15 2.8091807965185187
ENSMUSG00000067261.4 6.58254075
ENSMUSG00000078673.10 0.31774953625
</pre></div>
</div>
<img alt="../_images/33e46936f33b2540690d075e831c37e4d48cf1957634803a035c753713b056f2.png" src="../_images/33e46936f33b2540690d075e831c37e4d48cf1957634803a035c753713b056f2.png" />
</div>
</div>
<p><strong>Benchmark models predictions: Enformer and Hyenadna</strong></p>
<blockquote>
<div><p>Note: Create a folder ../benchmarks/Hyena/ and move the predictions (which were by default saved in the scripts directory).</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#####</span>
<span class="c1"># Paths:    </span>
<span class="n">data_and_path_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;hyenadna&quot;</span><span class="p">:{</span><span class="s2">&quot;predictions&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Hyena/Hyena_finetunned_predictions.npy&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
                                  <span class="s2">&quot;actual&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Hyena/Hyena_finetunned_targets.npy&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
                                  <span class="s2">&quot;figure_path&quot;</span><span class="p">:</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/Hyena_benchmarks/&quot;</span><span class="p">)},</span>
                      <span class="s2">&quot;Enformer&quot;</span><span class="p">:{</span><span class="s2">&quot;predictions&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Enformer/enformer_test_predictions.npy&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
                                  <span class="s2">&quot;actual&quot;</span><span class="p">:</span><span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Enformer/enformer_test_targets.npy&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span>
                                  <span class="s2">&quot;figure_path&quot;</span><span class="p">:</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/Enformer_benchmarks/&quot;</span><span class="p">)}}</span>

<span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span> <span class="c1">#Path(&quot;/Users/wjq311/Desktop/PhD/Enhancer_logic_project/Data/results_all/results_no_H3K27ac_uncoupled/expression_output&quot;)</span>
<span class="n">target_samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets/test_targets.csv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span>
<span class="n">target_samples</span> <span class="o">=</span> <span class="n">target_samples</span><span class="p">[</span><span class="o">~</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;_rev&#39;</span><span class="p">)]</span>

<span class="nb">print</span><span class="p">(</span><span class="n">target_samples</span><span class="o">.</span><span class="n">index</span><span class="p">[::</span><span class="o">-</span><span class="mi">23</span><span class="p">])</span>
<span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="c1">#[&quot;&quot;]</span>

<span class="n">READ_SCALING_FACTOR</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">PLOT_LIST</span> <span class="o">=</span>  <span class="p">[</span><span class="s1">&#39;ENSMUSG00000078673.10_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000054885.11_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000089773.7_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000046133.3_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000040372.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028784.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000050234.7_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000057375.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000111410.1_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000006442.10_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078672.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028602.12_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000024793.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000078626.2_forward&#39;</span><span class="p">,</span>       
       <span class="s1">&#39;ENSMUSG00000078490.10_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028902.4_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000042380.8_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000115115.1_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000033253.18_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028788.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000036896.5_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000017264.16_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028743.7_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028699.9_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000063172.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000111611.1_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000033326.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000057722.17_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000023153.9_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028635.7_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000036896.5_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000115115.1_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000078490.10_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028476.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000037692.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028950.3_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000045589.7_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078674.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028634.17_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000036887.5_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028948.16_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028214.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028410.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000061455.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028576.12_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000020220.16_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000057375.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028339.17_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000039774.12_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000066191.12_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000041120.6_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028917.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000010517.7_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000061887.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000048626.5_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028549.17_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000009640.11_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078719.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000037443.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028467.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028553.12_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028847.8_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028976.10_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028245.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028980.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000039911.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000042608.15_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028312.19_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000041153.9_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000054659.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000047613.10_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000039546.9_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000039577.17_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028688.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000005045.16_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000043383.5_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000063077.15_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000070985.3_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000062545.4_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000061894.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028974.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028857.16_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000055900.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028779.16_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000003810.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028614.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000040372.2_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028648.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000029005.4_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078716.10_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000043085.14_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000050395.9_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000094293.1_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028417.3_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000006215.12_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000025413.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000073988.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000029076.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000039410.16_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000060268.12_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000002384.2_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078772.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028560.11_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000029029.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028403.15_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028656.14_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000087166.9_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000042608.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000048706.3_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000037157.8_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000042616.8_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000043257.15_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028786.15_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000078513.2_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028393.10_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028277.13_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000040536.15_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000045699.4_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000025328.9_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028851.6_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028794.13_forward&#39;</span><span class="p">,</span>
       <span class="s1">&#39;ENSMUSG00000028772.19_forward&#39;</span><span class="p">,</span> <span class="s1">&#39;ENSMUSG00000028763.18_forward&#39;</span><span class="p">,]</span> 

<span class="n">plot_benchmark_models_predictions</span><span class="p">(</span><span class="n">data_and_path_dict</span><span class="p">,</span> <span class="n">results_path</span><span class="p">,</span><span class="n">target_samples</span><span class="p">,</span><span class="n">condition_list</span><span class="p">,</span> <span class="n">READ_SCALING_FACTOR</span><span class="p">,</span><span class="n">PLOT_LIST</span><span class="p">,</span>  <span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Index([&#39;ENSMUSG00000028635.7_forward&#39;, &#39;ENSMUSG00000028989.3_forward&#39;,
       &#39;ENSMUSG00000034785.14_forward&#39;, &#39;ENSMUSG00000028496.17_forward&#39;,
       &#39;ENSMUSG00000039298.16_forward&#39;, &#39;ENSMUSG00000094066.1_forward&#39;,
       &#39;ENSMUSG00000028820.13_forward&#39;, &#39;ENSMUSG00000040728.15_forward&#39;,
       &#39;ENSMUSG00000028619.15_forward&#39;, &#39;ENSMUSG00000029047.13_forward&#39;,
       &#39;ENSMUSG00000028648.13_forward&#39;, &#39;ENSMUSG00000029005.4_forward&#39;,
       &#39;ENSMUSG00000078716.10_forward&#39;, &#39;ENSMUSG00000043085.14_forward&#39;,
       &#39;ENSMUSG00000050395.9_forward&#39;, &#39;ENSMUSG00000094293.1_forward&#39;,
       &#39;ENSMUSG00000028417.3_forward&#39;, &#39;ENSMUSG00000006215.12_forward&#39;,
       &#39;ENSMUSG00000025413.13_forward&#39;, &#39;ENSMUSG00000073988.13_forward&#39;,
       &#39;ENSMUSG00000029076.14_forward&#39;, &#39;ENSMUSG00000039410.16_forward&#39;,
       &#39;ENSMUSG00000060268.12_forward&#39;, &#39;ENSMUSG00000002384.2_forward&#39;,
       &#39;ENSMUSG00000078772.2_forward&#39;, &#39;ENSMUSG00000028560.11_forward&#39;,
       &#39;ENSMUSG00000029029.14_forward&#39;, &#39;ENSMUSG00000028403.15_forward&#39;,
       &#39;ENSMUSG00000028656.14_forward&#39;, &#39;ENSMUSG00000087166.9_forward&#39;,
       &#39;ENSMUSG00000042608.15_forward&#39;, &#39;ENSMUSG00000048706.3_forward&#39;,
       &#39;ENSMUSG00000037157.8_forward&#39;, &#39;ENSMUSG00000042616.8_forward&#39;,
       &#39;ENSMUSG00000043257.15_forward&#39;, &#39;ENSMUSG00000028786.15_forward&#39;,
       &#39;ENSMUSG00000078513.2_forward&#39;, &#39;ENSMUSG00000028393.10_forward&#39;,
       &#39;ENSMUSG00000028277.13_forward&#39;, &#39;ENSMUSG00000040536.15_forward&#39;,
       &#39;ENSMUSG00000045699.4_forward&#39;, &#39;ENSMUSG00000025328.9_forward&#39;,
       &#39;ENSMUSG00000028851.6_forward&#39;, &#39;ENSMUSG00000028794.13_forward&#39;,
       &#39;ENSMUSG00000028772.19_forward&#39;, &#39;ENSMUSG00000028763.18_forward&#39;,
       &#39;ENSMUSG00000066090.5_forward&#39;, &#39;ENSMUSG00000028971.4_forward&#39;,
       &#39;ENSMUSG00000028367.5_forward&#39;, &#39;ENSMUSG00000028419.5_forward&#39;,
       &#39;ENSMUSG00000028389.12_forward&#39;, &#39;ENSMUSG00000040520.7_forward&#39;,
       &#39;ENSMUSG00000095935.1_forward&#39;, &#39;ENSMUSG00000049225.14_forward&#39;,
       &#39;ENSMUSG00000070999.2_forward&#39;, &#39;ENSMUSG00000058523.2_forward&#39;,
       &#39;ENSMUSG00000050234.7_forward&#39;],
      dtype=&#39;object&#39;, name=&#39;ID&#39;)
</pre></div>
</div>
<img alt="../_images/6357a80c397027c3a8b52fed0805a3230352f85e51a143deaf183b44bb7b09a1.png" src="../_images/6357a80c397027c3a8b52fed0805a3230352f85e51a143deaf183b44bb7b09a1.png" />
</div>
</div>
<p><strong>Model performance comparison</strong></p>
<p>Comparison shown in Figure 3.</p>
<blockquote>
<div><p><em>Note: Since we only want forward strand sequence predictions, we will remove the “_rev” samples from the enformer’s test (both in target csv and input npy arrays)</em></p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_2/&quot;</span><span class="p">)</span>
<span class="n">performance_dicts</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;Enformer embeddings&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.8067</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8739</span><span class="p">},</span>
                    <span class="s2">&quot;HyenaDNA-32k fine tuned&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7514</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.7403</span><span class="p">},</span>
                    <span class="s2">&quot;Convolutions </span><span class="se">\n</span><span class="s2">on chomatin marks&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7814</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8852</span><span class="p">},</span>
                    <span class="s2">&quot;Convolutions and attention </span><span class="se">\n</span><span class="s2">on chromatin marks&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7815</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8808</span><span class="p">},</span>
                    <span class="s2">&quot;Convolutions </span><span class="se">\n</span><span class="s2">on chromatin marks and structure rotated&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7717</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8690</span><span class="p">},</span>
                    <span class="s2">&quot;Convolutions </span><span class="se">\n</span><span class="s2">on chromatin marks and structure original&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7066</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8025</span><span class="p">}},</span>
                    <span class="p">{</span><span class="s2">&quot;Enformer embeddings&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.9093</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.7874</span><span class="p">},</span> <span class="c1">#Areas from here</span>
                    <span class="s2">&quot;HyenaDNA-32k fine tuned&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.8053</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.5247</span><span class="p">},</span>
                    <span class="s2">&quot;CLASTER&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.9275</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8859</span><span class="p">}}]</span>
                    
<span class="n">color_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;orange&quot;</span><span class="p">,</span><span class="s2">&quot;darkred&quot;</span><span class="p">,</span><span class="s2">&quot;silver&quot;</span><span class="p">,</span><span class="s2">&quot;black&quot;</span><span class="p">,</span><span class="s2">&quot;royalblue&quot;</span><span class="p">,</span><span class="s2">&quot;purple&quot;</span><span class="p">]</span>

<span class="n">cmap</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">inferno</span>  <span class="c1"># define the colormap</span>
<span class="c1"># extract all colors from the .jet map</span>
<span class="n">cmaplist</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cmap</span><span class="o">.</span><span class="n">N</span><span class="p">)]</span>
<span class="c1"># force the first color entry to be grey</span>

<span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span>
        <span class="s1">&#39;Custom cmap&#39;</span><span class="p">,</span> <span class="n">cmaplist</span><span class="p">,</span> <span class="n">cmap</span><span class="o">.</span><span class="n">N</span><span class="p">)</span>
<span class="n">norm</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">6</span><span class="p">)</span>
<span class="n">color_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmap</span><span class="p">(</span><span class="n">norm</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">)]</span>

<span class="n">model_comparison</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">performance_dicts</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">color_list</span><span class="p">)</span>

<span class="n">performance_dicts</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;Enformer embeddings&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.8067</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8739</span><span class="p">},</span>
                    <span class="s2">&quot;HyenaDNA-32k fine tuned&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7514</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.7403</span><span class="p">},</span>
                    <span class="s2">&quot;CLASTER&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.7717</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8690</span><span class="p">}},</span>
                    <span class="p">{</span><span class="s2">&quot;Enformer embeddings&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.9093</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.7874</span><span class="p">},</span> <span class="c1">#Areas from here</span>
                    <span class="s2">&quot;HyenaDNA-32k fine tuned&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.8053</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.5247</span><span class="p">},</span>
                    <span class="s2">&quot;CLASTER&quot;</span><span class="p">:{</span><span class="s2">&quot;Spearman&quot;</span><span class="p">:</span><span class="mf">0.9275</span><span class="p">,</span><span class="s2">&quot;Pearson&quot;</span><span class="p">:</span><span class="mf">0.8859</span><span class="p">}}]</span>

<span class="n">model_comparison_II</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">performance_dicts</span><span class="p">,</span> <span class="n">color_list</span><span class="p">[::</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;Enformer embeddings&#39;: {&#39;Spearman&#39;: 0.8067, &#39;Pearson&#39;: 0.8739}, &#39;HyenaDNA-32k fine tuned&#39;: {&#39;Spearman&#39;: 0.7514, &#39;Pearson&#39;: 0.7403}, &#39;CLASTER&#39;: {&#39;Spearman&#39;: 0.7717, &#39;Pearson&#39;: 0.869}}
{&#39;Enformer embeddings&#39;: {&#39;Spearman&#39;: 0.9093, &#39;Pearson&#39;: 0.7874}, &#39;HyenaDNA-32k fine tuned&#39;: {&#39;Spearman&#39;: 0.8053, &#39;Pearson&#39;: 0.5247}, &#39;CLASTER&#39;: {&#39;Spearman&#39;: 0.9275, &#39;Pearson&#39;: 0.8859}}
</pre></div>
</div>
</div>
</div>
</section>
<section id="in-silico-perturbations">
<h2><center> <em>In silico</em> perturbations <center><a class="headerlink" href="#in-silico-perturbations" title="Link to this heading">#</a></h2>
<p><strong>Figure 2 and related supplementary figures</strong></p>
<blockquote>
<div><p>⏰ The following step takes a few minutes</p>
</div></blockquote>
<p>We will first add the area under the reference gene when silencing a given enhancer.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_1/&quot;</span><span class="p">)</span>
<span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_perturbed.csv&quot;</span><span class="p">)</span>

<span class="n">add_gene_area_after_enhancer_perturbation</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">input_table_path</span><span class="p">,</span>  <span class="n">N_BINS</span> <span class="o">=</span>  <span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_1912029/3114433919.py:251: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value &#39;1.5751251853866663&#39; has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[(df[&#39;Ref_gene&#39;] == gene) &amp; (df[&#39;Entity_ID&#39;] == enhancer),&#39;Area&#39;] = pred_area #Edit existing table
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_perturbed.csv&quot;</span><span class="p">)</span>
<span class="n">table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">)</span>

<span class="n">table</span><span class="p">[</span><span class="n">table</span><span class="p">[</span><span class="s2">&quot;Ref_gene&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ENSMUSG00000039693.11&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ref_gene</th>
      <th>Ref_gene_chromosome</th>
      <th>Ref_gene_strand</th>
      <th>Entity_ID</th>
      <th>Entity_rel_Start</th>
      <th>Entity_rel_End</th>
      <th>EP_distance</th>
      <th>Genes_in_between</th>
      <th>Adjacent</th>
      <th>Area</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3341</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>ENSMUSG00000039693.11</td>
      <td>0</td>
      <td>21984</td>
      <td>10992.0</td>
      <td>0</td>
      <td>1</td>
      <td>9.970527</td>
    </tr>
    <tr>
      <th>3342</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755362</td>
      <td>-67572</td>
      <td>-67288</td>
      <td>67430.0</td>
      <td>0</td>
      <td>1</td>
      <td>10.147307</td>
    </tr>
    <tr>
      <th>3343</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755363</td>
      <td>-67276</td>
      <td>-67039</td>
      <td>67157.5</td>
      <td>0</td>
      <td>1</td>
      <td>10.256321</td>
    </tr>
    <tr>
      <th>3344</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755376</td>
      <td>847</td>
      <td>1162</td>
      <td>1004.5</td>
      <td>0</td>
      <td>1</td>
      <td>8.202687</td>
    </tr>
    <tr>
      <th>3345</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755407</td>
      <td>44761</td>
      <td>44952</td>
      <td>44856.5</td>
      <td>1</td>
      <td>0</td>
      <td>9.915334</td>
    </tr>
    <tr>
      <th>3346</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755409</td>
      <td>45533</td>
      <td>45856</td>
      <td>45694.5</td>
      <td>1</td>
      <td>0</td>
      <td>9.558456</td>
    </tr>
    <tr>
      <th>3347</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755410</td>
      <td>45866</td>
      <td>46172</td>
      <td>46019.0</td>
      <td>1</td>
      <td>0</td>
      <td>9.395018</td>
    </tr>
    <tr>
      <th>3348</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755411</td>
      <td>46222</td>
      <td>46424</td>
      <td>46323.0</td>
      <td>1</td>
      <td>0</td>
      <td>9.555614</td>
    </tr>
    <tr>
      <th>3349</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755412</td>
      <td>46508</td>
      <td>46849</td>
      <td>46678.5</td>
      <td>1</td>
      <td>0</td>
      <td>9.795846</td>
    </tr>
    <tr>
      <th>3350</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755364</td>
      <td>-66973</td>
      <td>-66770</td>
      <td>66871.5</td>
      <td>0</td>
      <td>1</td>
      <td>10.172708</td>
    </tr>
    <tr>
      <th>3351</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755368</td>
      <td>-48568</td>
      <td>-48226</td>
      <td>48397.0</td>
      <td>0</td>
      <td>1</td>
      <td>10.145588</td>
    </tr>
    <tr>
      <th>3352</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755377</td>
      <td>1682</td>
      <td>1888</td>
      <td>1785.0</td>
      <td>0</td>
      <td>1</td>
      <td>9.415998</td>
    </tr>
    <tr>
      <th>3353</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755378</td>
      <td>2013</td>
      <td>2207</td>
      <td>2110.0</td>
      <td>0</td>
      <td>1</td>
      <td>9.511173</td>
    </tr>
    <tr>
      <th>3354</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755394</td>
      <td>33811</td>
      <td>34156</td>
      <td>33983.5</td>
      <td>1</td>
      <td>0</td>
      <td>9.974123</td>
    </tr>
    <tr>
      <th>3355</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755402</td>
      <td>39582</td>
      <td>39930</td>
      <td>39756.0</td>
      <td>1</td>
      <td>0</td>
      <td>10.098386</td>
    </tr>
    <tr>
      <th>3356</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755423</td>
      <td>83048</td>
      <td>83393</td>
      <td>83220.5</td>
      <td>1</td>
      <td>0</td>
      <td>10.092491</td>
    </tr>
    <tr>
      <th>3357</th>
      <td>ENSMUSG00000039693.11</td>
      <td>chr4</td>
      <td>+</td>
      <td>EM10E0755403</td>
      <td>40099</td>
      <td>40271</td>
      <td>40185.0</td>
      <td>1</td>
      <td>0</td>
      <td>10.050406</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>

<span class="c1"># Plot perturbed profile predictions</span>
<span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_pure_conv_perturbed_only_H3K27ac&quot;</span><span class="p">)</span> <span class="c1"># Path(&quot;../runs/perturbation_runs/gene_expression_only_chrom_pure_conv/&quot;) # Change plot name to _original</span>
<span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">SCALE</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">PERT_TYPE</span> <span class="o">=</span> <span class="s2">&quot;only_H3K27ac&quot;</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_2/&quot;</span><span class="p">)</span>
<span class="n">sample_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000028345.15&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSMUSG00000028347.14&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000028344.12&quot;</span> <span class="p">]</span><span class="c1">#table[&#39;Ref_gene&#39;].unique()[::10] #[ &quot;ENSMUSG00000000409.14&quot;, &quot;ENSMUSG00000000085.16&quot;, ]</span>

<span class="k">for</span> <span class="n">sample_name</span> <span class="ow">in</span> <span class="n">sample_names</span><span class="p">:</span>
    <span class="n">plot_perturbed_profiles</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">,</span> <span class="n">SCALE</span><span class="p">,</span> <span class="n">sample_name</span><span class="p">,</span> <span class="n">PERT_TYPE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/449cee4954332e7119386ed3e1faa3af482e472d0e2507a8680de12c4850ff18.png" src="../_images/449cee4954332e7119386ed3e1faa3af482e472d0e2507a8680de12c4850ff18.png" />
<img alt="../_images/0f6abd2cad8eb345367f22051722d5f4918ec8c02d08b0bc8f42191ada31a3ae.png" src="../_images/0f6abd2cad8eb345367f22051722d5f4918ec8c02d08b0bc8f42191ada31a3ae.png" />
<img alt="../_images/a65d669906ff2291a680bff355dcb8f02d20abe121733ebfd4b9194684276358.png" src="../_images/a65d669906ff2291a680bff355dcb8f02d20abe121733ebfd4b9194684276358.png" />
</div>
</div>
<p>We will now compute:</p>
<ol class="arabic simple">
<li><p>The distance dependence of the in-silico enhancer silencing impact.</p></li>
<li><p>Which genes are the most affected and what’s their relative location to the affected gene.</p></li>
</ol>
<ul class="simple">
<li><p>We will first compute all the E-P pairs that are available.</p></li>
<li><p>We will store separately, for each enhancer, the gene that was the most affected and its relative properties to the enhancer (distance, adjacency and number of genes in between E and P).</p></li>
<li><p>We will histogram the properties of the most affected gene and normalize by the background frequencies: While most enhancers might have zero or 1 gene in between, it is rare that an enhancer will have 15 genes between it and another gene that appears in the same sample. Weapplied a reciprocal normalization, i.e. as if we had a dataset with the same number of cases for all categories, which allows us to compare the enrichment on certain categories relative to the “background”.</p></li>
</ul>
<p>⚠️ The different histograms follow parallel normalization schemes, yielding an apparent discrepancy between adjacent cases and “0 genes in between”.</p>
<blockquote>
<div><p>⚠️ IMPORTANT NOTE: Run first the cell where correlations are predicted for the model you used for the perturbed predictions. Otherwise the comparisons do not make sense!!!</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="nn">gridspec</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="k">def</span> <span class="nf">calculate_histogram</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">bins</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the histogram of given data.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        data (array-like): The data for which to calculate the histogram.</span>
<span class="sd">        bins (array-like): The bin edges.</span>

<span class="sd">    Returns:</span>
<span class="sd">        hist (np.ndarray): The counts of data in each bin.</span>
<span class="sd">        bins (np.ndarray): The edges of the bins.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hist</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">hist</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">),</span> <span class="n">bins</span>

<span class="k">def</span> <span class="nf">plot_normalized_histogram</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">reference_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">15.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">eps</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;royalblue&#39;</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span> <span class="n">xticks</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot a normalized histogram, optionally normalizing against a reference distribution, with explicit x-ticks.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        ax (matplotlib.axes.Axes): The axes object to plot on.</span>
<span class="sd">        data (array-like): The data to plot.</span>
<span class="sd">        reference_data (array-like, optional): The reference data for normalization.</span>
<span class="sd">        bins (array-like): The bin edges.</span>
<span class="sd">        eps (float): Epsilon value for adjusting bar width.</span>
<span class="sd">        color (str): Color of the histogram bars.</span>
<span class="sd">        xlabel (str): X-axis label.</span>
<span class="sd">        ylabel (str): Y-axis label.</span>
<span class="sd">        xticks (list, optional): Positions of the x-ticks. If None, automatic x-ticks will be used.</span>
<span class="sd">        xticklabels (list of str, optional): Labels of the x-ticks. If None, labels are generated from xticks.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">hist</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="n">calculate_histogram</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">bins</span><span class="p">)</span>
    <span class="n">frequency_vector</span> <span class="o">=</span> <span class="n">hist</span> <span class="o">/</span> <span class="n">hist</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># Normalize to sum to 1</span>
    
    <span class="k">if</span> <span class="n">reference_data</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">reference_hist</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">calculate_histogram</span><span class="p">(</span><span class="n">reference_data</span><span class="p">,</span> <span class="n">bins</span><span class="p">)</span>
        <span class="n">reference_frequency</span> <span class="o">=</span> <span class="n">reference_hist</span> <span class="o">/</span> <span class="n">reference_hist</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
        <span class="n">frequency_vector</span> <span class="o">=</span> <span class="n">frequency_vector</span> <span class="o">/</span> <span class="n">reference_frequency</span>
        <span class="n">frequency_vector</span> <span class="o">/=</span> <span class="n">frequency_vector</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># Re-normalize to sum to 1</span>
    
    <span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">bins</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">frequency_vector</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="p">(</span><span class="n">bins</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">-</span><span class="n">bins</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">eps</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">xlabel</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">ylabel</span><span class="p">)</span>
    
    <span class="c1"># Set x-ticks if specified</span>
    <span class="k">if</span> <span class="n">xticks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">xticklabels</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">xticks</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">xticklabels</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>  <span class="c1"># Adjust rotation as needed for readability</span>
    <span class="k">elif</span> <span class="n">xticks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">xticks</span><span class="p">)</span>
        <span class="c1"># Generate automatic xticklabels if only xticks are provided</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">tick</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">tick</span> <span class="ow">in</span> <span class="n">xticks</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">plot_perturbation_histograms</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">):</span>
    <span class="c1"># Filter out rows where Area is 0.0 for reference genes and their enhancers</span>
    <span class="n">non_zero_ref_genes</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.0</span><span class="p">][</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">df_filtered</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">non_zero_ref_genes</span><span class="p">)]</span>
    
    <span class="c1"># Calculate Absolute Relative Change</span>
    <span class="n">df_filtered</span><span class="p">[</span><span class="s1">&#39;Absolute_Relative_Change&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span>
        <span class="p">(</span><span class="n">df_filtered</span><span class="p">[</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">)[</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s1">&#39;first&#39;</span><span class="p">))</span> <span class="o">/</span>
        <span class="n">df_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">)[</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="s1">&#39;first&#39;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Exclude rows where Ref_gene equals Entity_ID (gene rows) to focus on enhancers</span>
    <span class="n">enhancers_filtered</span> <span class="o">=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="n">df_filtered</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">df_filtered</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]]</span>

    <span class="c1"># Bin EP_distance into 1000 bp bins</span>
    <span class="n">max_distance</span> <span class="o">=</span> <span class="mi">15000</span>
    <span class="n">min_distance</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">scale</span> <span class="o">=</span> <span class="mi">1000</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">min_distance</span> <span class="o">-</span> <span class="n">min_distance</span> <span class="o">%</span> <span class="n">scale</span><span class="p">,</span> <span class="n">max_distance</span> <span class="o">+</span> <span class="n">scale</span><span class="p">,</span> <span class="n">scale</span><span class="p">)</span>  <span class="c1"># Create bins edges</span>
    <span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;EP_distance_bin&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;EP_distance&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">include_lowest</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Plot Swarm Plots with Binned EP_distance</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">swarmplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;EP_distance_bin&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;Absolute_Relative_Change&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">enhancers_filtered</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Enhancer-Promoter distance (kbp)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Change&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">15</span><span class="p">),</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">a</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">15</span><span class="p">)])</span>  <span class="c1"># Improve label readability</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mf">.6</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;Distance_effect.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

    <span class="c1"># Filter DataFrame to keep only rows with the largest Absolute_Relative_Change for each enhancer</span>
    <span class="n">enhancers_largest_change</span> <span class="o">=</span> <span class="n">enhancers_filtered</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Absolute_Relative_Change&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>

    <span class="c1"># Plot frequency histograms for all Enhancer-Promoter interactions</span>
    <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span> 
    <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>

    <span class="n">plot_normalized_histogram</span><span class="p">(</span>
        <span class="n">ax1</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;Genes_in_between&#39;</span><span class="p">],</span>
        <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">15.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Genes in between&#39;</span><span class="p">,</span>
        <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span>
    <span class="p">)</span>
    <span class="n">plot_normalized_histogram</span><span class="p">(</span>
        <span class="n">ax2</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;Adjacent&#39;</span><span class="p">],</span>
        <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span>
        <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span>
        <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span>
        <span class="n">xticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span>
        <span class="n">xticklabels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Skip&#39;</span><span class="p">,</span><span class="s1">&#39;Adjacent&#39;</span><span class="p">],</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;Background_frequencies.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

    <span class="c1"># Plot histograms for most relevant EP interactions with normalization against the background distribution</span>

    <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

    <span class="c1"># Plot frequency histograms for all Enhancer-Promoter interactions</span>
    <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span> 
    <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>

    <span class="n">plot_normalized_histogram</span><span class="p">(</span>
    <span class="n">ax1</span><span class="p">,</span>
    <span class="n">data</span><span class="o">=</span><span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;Genes_in_between&#39;</span><span class="p">],</span>
    <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">15.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
    <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Genes in between&#39;</span><span class="p">,</span>
    <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span>
    <span class="p">)</span>

    <span class="n">plot_normalized_histogram</span><span class="p">(</span>
        <span class="n">ax2</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="n">enhancers_largest_change</span><span class="p">[</span><span class="s1">&#39;Genes_in_between&#39;</span><span class="p">],</span>
        <span class="n">reference_data</span><span class="o">=</span><span class="n">enhancers_filtered</span><span class="p">[</span><span class="s1">&#39;Genes_in_between&#39;</span><span class="p">],</span>
        <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mf">0.01</span><span class="p">,</span> <span class="mf">15.6</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
        <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Genes in between&#39;</span><span class="p">,</span>
        <span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Frequency&#39;</span><span class="p">,</span>
        <span class="n">xticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">15</span><span class="p">),</span>
        <span class="n">xticklabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">15</span><span class="p">),</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span>

    <span class="p">)</span>
    <span class="c1"># plot_normalized_histogram(</span>
    <span class="c1">#     ax2,</span>
    <span class="c1">#     data=enhancers_largest_change[&#39;Adjacent&#39;],</span>
    <span class="c1">#     #reference_data=enhancers_filtered[&#39;Adjacent&#39;],</span>
    <span class="c1">#     bins=[0,1,2],</span>
    <span class="c1">#     xlabel=&#39;&#39;,</span>
    <span class="c1">#     ylabel=&#39;Frequency&#39;,</span>
    <span class="c1">#     xticks = np.arange(2),</span>
    <span class="c1">#     xticklabels = [&#39;Skip&#39;,&#39;Adjacent&#39;],</span>
    <span class="c1">#     color=&#39;k&#39;</span>
    <span class="c1"># )</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="s2">&quot;Normalized_frequency_hist.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>


<span class="n">plot_perturbation_histograms</span><span class="p">(</span><span class="n">table</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_2/&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_2263363/2841167992.py:67: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered[&#39;Absolute_Relative_Change&#39;] = abs(
/tmp/ipykernel_2263363/2841167992.py:80: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  enhancers_filtered[&#39;EP_distance_bin&#39;] = pd.cut(enhancers_filtered[&#39;EP_distance&#39;], bins=bins, include_lowest=True)
/maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/seaborn/categorical.py:3399: UserWarning: 25.4% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.
  warnings.warn(msg, UserWarning)
/maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/seaborn/categorical.py:3399: UserWarning: 6.6% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.
  warnings.warn(msg, UserWarning)
/tmp/ipykernel_2263363/2841167992.py:93: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  enhancers_largest_change = enhancers_filtered.groupby(&#39;Entity_ID&#39;).apply(lambda x: x.nlargest(1, &#39;Absolute_Relative_Change&#39;)).reset_index(drop=True)
/maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/seaborn/categorical.py:3399: UserWarning: 6.6% of the points cannot be placed; you may want to decrease the size of the markers or use stripplot.
  warnings.warn(msg, UserWarning)
</pre></div>
</div>
<img alt="../_images/3394ae945e2caf4559fc014506e9d94463cfb7a09a4cc91f140fdd776509ed12.png" src="../_images/3394ae945e2caf4559fc014506e9d94463cfb7a09a4cc91f140fdd776509ed12.png" />
<img alt="../_images/1bb88624f2ddd64445df3d62554118a97f73ec55350fc16f8117fd08e7c625d9.png" src="../_images/1bb88624f2ddd64445df3d62554118a97f73ec55350fc16f8117fd08e7c625d9.png" />
<img alt="../_images/c4571533ade19708e1a870471efb33c567fb7a8d918da5311e1535c7da588450.png" src="../_images/c4571533ade19708e1a870471efb33c567fb7a8d918da5311e1535c7da588450.png" />
</div>
</div>
<p><strong>Figure 2b: Chromatin mark correlations</strong></p>
<p>Visualize possible chromatin states found in the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">array_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_2/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">skip_n_bins</span> <span class="o">=</span> <span class="mi">20</span> <span class="c1">#20</span>

<span class="n">plot_chromatin_state_combinations</span><span class="p">(</span><span class="n">array_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">skip_n_bins</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="attributions">
<h1><center> Attributions <center><a class="headerlink" href="#attributions" title="Link to this heading">#</a></h1>
<p>Now we will compute the attribution scores for every input position (4,10.001) towards every output (401). This mechanism tells how important is every input position (i.e. a given bin in each input track in the landscape or a given bin in a Micro-C matrix) for the prediction of the EU-seq levels at the position described by a given target node (401 targets). It also tells us the direction of the association.
The original paper can be found at:</p>
<p>https://arxiv.org/abs/1703.01365</p>
<p>EIR uses the integrated gradients method as it is implemented in the captum library:</p>
<p>https://captum.ai/docs/extension/integrated_gradients</p>
<p><strong>Figure 3 and related supplementary figures</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_rotated_pure_conv/expression_output/&quot;</span><span class="p">)</span> <span class="c1">#Path(&quot;../runs/gene_expression_only_chrom_pure_conv/results/expression_output/&quot;) #Path(&quot;../runs/training_runs/gene_expression_exformer_unlimited_all_rotated/results/expression_output/&quot;)</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_3/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">n_central_bins</span> <span class="o">=</span> <span class="mi">401</span>
<span class="n">l_in</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10001</span>
<span class="n">n_out</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="mi">401</span>
<span class="n">n_in</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">4</span>

<span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
        <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
        <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Enhancer&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="mi">3</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>
<span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">SPLIT</span> <span class="o">=</span> <span class="mi">60600</span> <span class="c1">#17880</span>
<span class="n">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">n_central_bins</span><span class="p">,</span> <span class="n">track_dict</span><span class="p">,</span><span class="n">l_in</span><span class="p">,</span><span class="n">n_out</span><span class="p">,</span><span class="n">n_in</span><span class="p">,</span> <span class="n">TRAIN_ATTR</span><span class="p">,</span><span class="n">SPLIT</span><span class="p">)</span>


<span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv/results/expression_output/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_3_only_chrom/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">SPLIT</span> <span class="o">=</span> <span class="mi">60600</span> <span class="c1">#17880</span>
<span class="n">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">n_central_bins</span><span class="p">,</span> <span class="n">track_dict</span><span class="p">,</span><span class="n">l_in</span><span class="p">,</span><span class="n">n_out</span><span class="p">,</span><span class="n">n_in</span><span class="p">,</span> <span class="n">TRAIN_ATTR</span><span class="p">,</span><span class="n">SPLIT</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_122633/2179293752.py:449: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  centered_attr = np.roll(attributions, -10*i+int(eps),axis=1)
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyboardInterrupt</span><span class="g g-Whitespace">                         </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">61</span><span class="p">],</span> <span class="n">line</span> <span class="mi">15</span>
<span class="g g-Whitespace">     </span><span class="mi">13</span> <span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="g g-Whitespace">     </span><span class="mi">14</span> <span class="n">SPLIT</span> <span class="o">=</span> <span class="mi">60600</span> <span class="c1">#17880</span>
<span class="ne">---&gt; </span><span class="mi">15</span> <span class="n">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">n_central_bins</span><span class="p">,</span> <span class="n">track_dict</span><span class="p">,</span><span class="n">l_in</span><span class="p">,</span><span class="n">n_out</span><span class="p">,</span><span class="n">n_in</span><span class="p">,</span> <span class="n">TRAIN_ATTR</span><span class="p">,</span><span class="n">SPLIT</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">18</span> <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv/results/expression_output/&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">19</span> <span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_3_only_chrom/&quot;</span><span class="p">)</span>

<span class="nn">Cell In[60], line 443,</span> in <span class="ni">read_target_attributions</span><span class="nt">(results_path, figure_path, n_central_bins, track_dict, l_in, n_out, n_in, TRAIN_ATTR, SPLIT)</span>
<span class="g g-Whitespace">    </span><span class="mi">441</span>     <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
<span class="g g-Whitespace">    </span><span class="mi">442</span>     <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="ne">--&gt; </span><span class="mi">443</span> <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Chromatin_landscape_attributions_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_ctrl.png&quot;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">444</span> <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">446</span> <span class="c1"># Add a small phase to remove high frequency fluctuations:</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/figure.py:3285,</span> in <span class="ni">Figure.savefig</span><span class="nt">(self, fname, transparent, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">3281</span>     <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">axes</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">3282</span>         <span class="n">stack</span><span class="o">.</span><span class="n">enter_context</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">3283</span>             <span class="n">ax</span><span class="o">.</span><span class="n">patch</span><span class="o">.</span><span class="n">_cm_set</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">))</span>
<span class="ne">-&gt; </span><span class="mi">3285</span> <span class="bp">self</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">print_figure</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/backend_bases.py:2338,</span> in <span class="ni">FigureCanvasBase.print_figure</span><span class="nt">(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">2334</span> <span class="k">try</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">2335</span>     <span class="c1"># _get_renderer may change the figure dpi (as vector formats</span>
<span class="g g-Whitespace">   </span><span class="mi">2336</span>     <span class="c1"># force the figure dpi to 72), so we need to set it again here.</span>
<span class="g g-Whitespace">   </span><span class="mi">2337</span>     <span class="k">with</span> <span class="n">cbook</span><span class="o">.</span><span class="n">_setattr_cm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">figure</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">):</span>
<span class="ne">-&gt; </span><span class="mi">2338</span>         <span class="n">result</span> <span class="o">=</span> <span class="n">print_method</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">2339</span>             <span class="n">filename</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2340</span>             <span class="n">facecolor</span><span class="o">=</span><span class="n">facecolor</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2341</span>             <span class="n">edgecolor</span><span class="o">=</span><span class="n">edgecolor</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2342</span>             <span class="n">orientation</span><span class="o">=</span><span class="n">orientation</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2343</span>             <span class="n">bbox_inches_restore</span><span class="o">=</span><span class="n">_bbox_inches_restore</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2344</span>             <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">2345</span> <span class="k">finally</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">2346</span>     <span class="k">if</span> <span class="n">bbox_inches</span> <span class="ow">and</span> <span class="n">restore_bbox</span><span class="p">:</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/backend_bases.py:2204,</span> in <span class="ni">FigureCanvasBase._switch_canvas_and_return_print_method.&lt;locals&gt;.&lt;lambda&gt;</span><span class="nt">(*args, **kwargs)</span>
<span class="g g-Whitespace">   </span><span class="mi">2200</span>     <span class="n">optional_kws</span> <span class="o">=</span> <span class="p">{</span>  <span class="c1"># Passed by print_figure for other renderers.</span>
<span class="g g-Whitespace">   </span><span class="mi">2201</span>         <span class="s2">&quot;dpi&quot;</span><span class="p">,</span> <span class="s2">&quot;facecolor&quot;</span><span class="p">,</span> <span class="s2">&quot;edgecolor&quot;</span><span class="p">,</span> <span class="s2">&quot;orientation&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">   </span><span class="mi">2202</span>         <span class="s2">&quot;bbox_inches_restore&quot;</span><span class="p">}</span>
<span class="g g-Whitespace">   </span><span class="mi">2203</span>     <span class="n">skip</span> <span class="o">=</span> <span class="n">optional_kws</span> <span class="o">-</span> <span class="p">{</span><span class="o">*</span><span class="n">inspect</span><span class="o">.</span><span class="n">signature</span><span class="p">(</span><span class="n">meth</span><span class="p">)</span><span class="o">.</span><span class="n">parameters</span><span class="p">}</span>
<span class="ne">-&gt; </span><span class="mi">2204</span>     <span class="n">print_method</span> <span class="o">=</span> <span class="n">functools</span><span class="o">.</span><span class="n">wraps</span><span class="p">(</span><span class="n">meth</span><span class="p">)(</span><span class="k">lambda</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="n">meth</span><span class="p">(</span>
<span class="nn">   2205         *args, **{k: v for k, v</span> in <span class="ni">kwargs.items</span><span class="nt">() if k not in skip}))</span>
<span class="g g-Whitespace">   </span><span class="mi">2206</span> <span class="k">else</span><span class="p">:</span>  <span class="c1"># Let third-parties do as they see fit.</span>
<span class="g g-Whitespace">   </span><span class="mi">2207</span>     <span class="n">print_method</span> <span class="o">=</span> <span class="n">meth</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/_api/deprecation.py:410,</span> in <span class="ni">delete_parameter.&lt;locals&gt;.wrapper</span><span class="nt">(*inner_args, **inner_kwargs)</span>
<span class="g g-Whitespace">    </span><span class="mi">400</span>     <span class="n">deprecation_addendum</span> <span class="o">=</span> <span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">401</span>         <span class="sa">f</span><span class="s2">&quot;If any parameter follows </span><span class="si">{</span><span class="n">name</span><span class="si">!r}</span><span class="s2">, they should be passed as &quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">402</span>         <span class="sa">f</span><span class="s2">&quot;keyword, not positionally.&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">403</span>     <span class="n">warn_deprecated</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">404</span>         <span class="n">since</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">405</span>         <span class="n">name</span><span class="o">=</span><span class="nb">repr</span><span class="p">(</span><span class="n">name</span><span class="p">),</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">408</span>                  <span class="k">else</span> <span class="n">deprecation_addendum</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">409</span>         <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">410</span> <span class="k">return</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">inner_args</span><span class="p">,</span> <span class="o">**</span><span class="n">inner_kwargs</span><span class="p">)</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/backends/backend_agg.py:517,</span> in <span class="ni">FigureCanvasAgg.print_png</span><span class="nt">(self, filename_or_obj, metadata, pil_kwargs, *args)</span>
<span class="g g-Whitespace">    </span><span class="mi">468</span> <span class="nd">@_api</span><span class="o">.</span><span class="n">delete_parameter</span><span class="p">(</span><span class="s2">&quot;3.5&quot;</span><span class="p">,</span> <span class="s2">&quot;args&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">469</span> <span class="k">def</span> <span class="nf">print_png</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_or_obj</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">470</span>               <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pil_kwargs</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">471</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">472</span><span class="sd">     Write the figure to a PNG file.</span>
<span class="g g-Whitespace">    </span><span class="mi">473</span><span class="sd"> </span>
<span class="sd">   (...)</span>
<span class="g g-Whitespace">    </span><span class="mi">515</span><span class="sd">         *metadata*, including the default &#39;Software&#39; key.</span>
<span class="g g-Whitespace">    </span><span class="mi">516</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">517</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_print_pil</span><span class="p">(</span><span class="n">filename_or_obj</span><span class="p">,</span> <span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">pil_kwargs</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/backends/backend_agg.py:463,</span> in <span class="ni">FigureCanvasAgg._print_pil</span><span class="nt">(self, filename_or_obj, fmt, pil_kwargs, metadata)</span>
<span class="g g-Whitespace">    </span><span class="mi">458</span> <span class="k">def</span> <span class="nf">_print_pil</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename_or_obj</span><span class="p">,</span> <span class="n">fmt</span><span class="p">,</span> <span class="n">pil_kwargs</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="g g-Whitespace">    </span><span class="mi">459</span><span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="g g-Whitespace">    </span><span class="mi">460</span><span class="sd">     Draw the canvas, then save it using `.image.imsave` (to which</span>
<span class="g g-Whitespace">    </span><span class="mi">461</span><span class="sd">     *pil_kwargs* and *metadata* are forwarded).</span>
<span class="g g-Whitespace">    </span><span class="mi">462</span><span class="sd">     &quot;&quot;&quot;</span>
<span class="ne">--&gt; </span><span class="mi">463</span>     <span class="n">FigureCanvasAgg</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">464</span>     <span class="n">mpl</span><span class="o">.</span><span class="n">image</span><span class="o">.</span><span class="n">imsave</span><span class="p">(</span>
<span class="g g-Whitespace">    </span><span class="mi">465</span>         <span class="n">filename_or_obj</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">buffer_rgba</span><span class="p">(),</span> <span class="nb">format</span><span class="o">=</span><span class="n">fmt</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s2">&quot;upper&quot;</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">466</span>         <span class="n">dpi</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">dpi</span><span class="p">,</span> <span class="n">metadata</span><span class="o">=</span><span class="n">metadata</span><span class="p">,</span> <span class="n">pil_kwargs</span><span class="o">=</span><span class="n">pil_kwargs</span><span class="p">)</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/backends/backend_agg.py:405,</span> in <span class="ni">FigureCanvasAgg.draw</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">401</span> <span class="c1"># Acquire a lock on the shared font cache.</span>
<span class="g g-Whitespace">    </span><span class="mi">402</span> <span class="k">with</span> <span class="n">RendererAgg</span><span class="o">.</span><span class="n">lock</span><span class="p">,</span> \
<span class="g g-Whitespace">    </span><span class="mi">403</span>      <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">toolbar</span><span class="o">.</span><span class="n">_wait_cursor_for_draw_cm</span><span class="p">()</span> <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">toolbar</span>
<span class="g g-Whitespace">    </span><span class="mi">404</span>       <span class="k">else</span> <span class="n">nullcontext</span><span class="p">()):</span>
<span class="ne">--&gt; </span><span class="mi">405</span>     <span class="bp">self</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">406</span>     <span class="c1"># A GUI class may be need to update a window using this draw, so</span>
<span class="g g-Whitespace">    </span><span class="mi">407</span>     <span class="c1"># don&#39;t forget to call the superclass.</span>
<span class="g g-Whitespace">    </span><span class="mi">408</span>     <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/artist.py:74,</span> in <span class="ni">_finalize_rasterization.&lt;locals&gt;.draw_wrapper</span><span class="nt">(artist, renderer, *args, **kwargs)</span>
<span class="g g-Whitespace">     </span><span class="mi">72</span> <span class="nd">@wraps</span><span class="p">(</span><span class="n">draw</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">73</span> <span class="k">def</span> <span class="nf">draw_wrapper</span><span class="p">(</span><span class="n">artist</span><span class="p">,</span> <span class="n">renderer</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="ne">---&gt; </span><span class="mi">74</span>     <span class="n">result</span> <span class="o">=</span> <span class="n">draw</span><span class="p">(</span><span class="n">artist</span><span class="p">,</span> <span class="n">renderer</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">75</span>     <span class="k">if</span> <span class="n">renderer</span><span class="o">.</span><span class="n">_rasterizing</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">76</span>         <span class="n">renderer</span><span class="o">.</span><span class="n">stop_rasterizing</span><span class="p">()</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/artist.py:51,</span> in <span class="ni">allow_rasterization.&lt;locals&gt;.draw_wrapper</span><span class="nt">(artist, renderer)</span>
<span class="g g-Whitespace">     </span><span class="mi">48</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">49</span>         <span class="n">renderer</span><span class="o">.</span><span class="n">start_filter</span><span class="p">()</span>
<span class="ne">---&gt; </span><span class="mi">51</span>     <span class="k">return</span> <span class="n">draw</span><span class="p">(</span><span class="n">artist</span><span class="p">,</span> <span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">52</span> <span class="k">finally</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">53</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/figure.py:3082,</span> in <span class="ni">Figure.draw</span><span class="nt">(self, renderer)</span>
<span class="g g-Whitespace">   </span><span class="mi">3079</span>         <span class="c1"># ValueError can occur when resizing a window.</span>
<span class="g g-Whitespace">   </span><span class="mi">3081</span> <span class="bp">self</span><span class="o">.</span><span class="n">patch</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">3082</span> <span class="n">mimage</span><span class="o">.</span><span class="n">_draw_list_compositing_images</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">3083</span>     <span class="n">renderer</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">artists</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">suppressComposite</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3085</span> <span class="k">for</span> <span class="n">sfig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">subfigs</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">3086</span>     <span class="n">sfig</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/image.py:131,</span> in <span class="ni">_draw_list_compositing_images</span><span class="nt">(renderer, parent, artists, suppress_composite)</span>
<span class="g g-Whitespace">    </span><span class="mi">129</span> <span class="k">if</span> <span class="n">not_composite</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">has_images</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">130</span>     <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">artists</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">131</span>         <span class="n">a</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">132</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">133</span>     <span class="c1"># Composite any adjacent images together</span>
<span class="g g-Whitespace">    </span><span class="mi">134</span>     <span class="n">image_group</span> <span class="o">=</span> <span class="p">[]</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/artist.py:51,</span> in <span class="ni">allow_rasterization.&lt;locals&gt;.draw_wrapper</span><span class="nt">(artist, renderer)</span>
<span class="g g-Whitespace">     </span><span class="mi">48</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">49</span>         <span class="n">renderer</span><span class="o">.</span><span class="n">start_filter</span><span class="p">()</span>
<span class="ne">---&gt; </span><span class="mi">51</span>     <span class="k">return</span> <span class="n">draw</span><span class="p">(</span><span class="n">artist</span><span class="p">,</span> <span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">52</span> <span class="k">finally</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">53</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/axes/_base.py:3100,</span> in <span class="ni">_AxesBase.draw</span><span class="nt">(self, renderer)</span>
<span class="g g-Whitespace">   </span><span class="mi">3097</span>         <span class="n">a</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3098</span>     <span class="n">renderer</span><span class="o">.</span><span class="n">stop_rasterizing</span><span class="p">()</span>
<span class="ne">-&gt; </span><span class="mi">3100</span> <span class="n">mimage</span><span class="o">.</span><span class="n">_draw_list_compositing_images</span><span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">3101</span>     <span class="n">renderer</span><span class="p">,</span> <span class="bp">self</span><span class="p">,</span> <span class="n">artists</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">suppressComposite</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3103</span> <span class="n">renderer</span><span class="o">.</span><span class="n">close_group</span><span class="p">(</span><span class="s1">&#39;axes&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3104</span> <span class="bp">self</span><span class="o">.</span><span class="n">stale</span> <span class="o">=</span> <span class="kc">False</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/image.py:131,</span> in <span class="ni">_draw_list_compositing_images</span><span class="nt">(renderer, parent, artists, suppress_composite)</span>
<span class="g g-Whitespace">    </span><span class="mi">129</span> <span class="k">if</span> <span class="n">not_composite</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">has_images</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">130</span>     <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">artists</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">131</span>         <span class="n">a</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">132</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">133</span>     <span class="c1"># Composite any adjacent images together</span>
<span class="g g-Whitespace">    </span><span class="mi">134</span>     <span class="n">image_group</span> <span class="o">=</span> <span class="p">[]</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/artist.py:51,</span> in <span class="ni">allow_rasterization.&lt;locals&gt;.draw_wrapper</span><span class="nt">(artist, renderer)</span>
<span class="g g-Whitespace">     </span><span class="mi">48</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">49</span>         <span class="n">renderer</span><span class="o">.</span><span class="n">start_filter</span><span class="p">()</span>
<span class="ne">---&gt; </span><span class="mi">51</span>     <span class="k">return</span> <span class="n">draw</span><span class="p">(</span><span class="n">artist</span><span class="p">,</span> <span class="n">renderer</span><span class="p">)</span>
<span class="g g-Whitespace">     </span><span class="mi">52</span> <span class="k">finally</span><span class="p">:</span>
<span class="g g-Whitespace">     </span><span class="mi">53</span>     <span class="k">if</span> <span class="n">artist</span><span class="o">.</span><span class="n">get_agg_filter</span><span class="p">()</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

<span class="nn">File /maps/projects/rasmussen/data/enhancer_logic_project/claster_env/lib/python3.11/site-packages/matplotlib/lines.py:866,</span> in <span class="ni">Line2D.draw</span><span class="nt">(self, renderer)</span>
<span class="g g-Whitespace">    </span><span class="mi">863</span> <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">864</span>     <span class="c1"># Don&#39;t scale for pixels, and don&#39;t stroke them</span>
<span class="g g-Whitespace">    </span><span class="mi">865</span>     <span class="n">marker_trans</span> <span class="o">=</span> <span class="n">marker_trans</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">866</span> <span class="n">renderer</span><span class="o">.</span><span class="n">draw_markers</span><span class="p">(</span><span class="n">gc</span><span class="p">,</span> <span class="n">marker_path</span><span class="p">,</span> <span class="n">marker_trans</span><span class="p">,</span>
<span class="g g-Whitespace">    </span><span class="mi">867</span>                       <span class="n">subsampled</span><span class="p">,</span> <span class="n">affine</span><span class="o">.</span><span class="n">frozen</span><span class="p">(),</span>
<span class="g g-Whitespace">    </span><span class="mi">868</span>                       <span class="n">fc_rgba</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">870</span> <span class="n">alt_marker_path</span> <span class="o">=</span> <span class="n">marker</span><span class="o">.</span><span class="n">get_alt_path</span><span class="p">()</span>
<span class="g g-Whitespace">    </span><span class="mi">871</span> <span class="k">if</span> <span class="n">alt_marker_path</span><span class="p">:</span>

<span class="ne">KeyboardInterrupt</span>: 
</pre></div>
</div>
<img alt="../_images/b36be6c67f3cd3917a36ad54dd750bfb4f8f4b8585e333e6284de5338c1a8141.png" src="../_images/b36be6c67f3cd3917a36ad54dd750bfb4f8f4b8585e333e6284de5338c1a8141.png" />
</div>
</div>
<p><strong>Micro-C Matrices: Original and rotated</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000059552.13_forward.npy&quot;</span> <span class="c1">#&quot;ENSMUSG00000024406.16_forward.npy&quot; # ENSMUSG00000028948.16_forward ENSMUSG00000078626.2_forward ENSMUSG00000046667.14_rev # Original paper example &quot;ENSMUST00000105369.7.npy&quot; #&quot;ENSMUST00000002350.10_flipped&quot;</span>
<span class="n">pathlist</span> <span class="o">=</span> <span class="p">[</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/microC/training/&#39;</span><span class="p">),</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/microC_rotated/training/&#39;</span><span class="p">)]</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">num_bins</span> <span class="o">=</span> <span class="mi">626</span>
<span class="n">window_of_observation</span> <span class="o">=</span> <span class="mf">707.10678118</span>  <span class="c1"># kbp (diagonal is not observing the same length in the genome.)</span>

<span class="n">plot_microc_original_and_rotated</span><span class="p">(</span><span class="n">pathlist</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Micro-C attributions</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figure_3/&quot;</span><span class="p">)</span>

<span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">201</span><span class="p">)</span>
<span class="n">num_bins</span> <span class="o">=</span> <span class="mi">626</span>
<span class="n">window_of_observation</span> <span class="o">=</span> <span class="mf">707.10678118</span>  <span class="c1"># kbp (diagonal is not observing the same length in the genome.)</span>
<span class="c1">#bins_per_kbp = 0.885297690046</span>
<span class="n">ATTR_type_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;abs&quot;</span><span class="p">,</span><span class="s2">&quot;signed&quot;</span><span class="p">]</span>
<span class="n">conditions</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
<span class="n">sigma</span> <span class="o">=</span> <span class="mi">3</span>

<span class="n">plot_microc_attributions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">,</span> <span class="n">ATTR_type_list</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">sigma</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.885297690053755
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/tmp/ipykernel_122633/2171253213.py:587: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  plt.tight_layout()
/tmp/ipykernel_122633/2171253213.py:663: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  plt.tight_layout()
</pre></div>
</div>
</div>
</div>
<p><strong>Filter visualization</strong></p>
<p>We can visualize the activations of a set of predefined filters, in this case for the structural branch handling Micro-C matrices. We can see what is the network propagating forward or what features are extracted.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Conv 0: first set of filters</span>
<span class="n">root_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_pure_conv_latents/latents/latent_outputs/3600/&quot;</span><span class="p">)</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">root_path</span> <span class="o">/</span> <span class="s2">&quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span> 
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/filters/&quot;</span><span class="p">)</span>
<span class="c1">#path = Path(&quot;/Users/wjq311/Desktop/eir/runs/gene_expression_exformer_unlimited_all/latents/latent_outputs/29800/input_modules.contact_maps.feature_extractor.conv.0.conv_1/&quot;) </span>
<span class="n">visualize_filters</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="n">savepath</span><span class="p">,</span> <span class="n">first_validation_sample_index</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">last_validation_sample_index</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">layer_name</span><span class="o">=</span> <span class="s2">&quot;layer_0.1&quot;</span><span class="p">)</span> 

<span class="c1"># # Conv 1: deeper set of filters</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">root_path</span> <span class="o">/</span> <span class="s2">&quot;input_modules.contact_maps.feature_extractor.conv.1.conv_1&quot;</span>
<span class="n">visualize_filters</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="n">savepath</span><span class="p">,</span> <span class="n">first_validation_sample_index</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">last_validation_sample_index</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">layer_name</span><span class="o">=</span> <span class="s2">&quot;layer_1.1&quot;</span><span class="p">)</span> 

<span class="c1"># # Conv 2: deeper set of filters</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">root_path</span> <span class="o">/</span> <span class="s2">&quot;input_modules.contact_maps.feature_extractor.conv.2.conv_1&quot;</span> 
<span class="n">visualize_filters</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">first_validation_sample_index</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">last_validation_sample_index</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">layer_name</span><span class="o">=</span> <span class="s2">&quot;layer_2.1&quot;</span><span class="p">)</span> 

<span class="c1"># # Conv 3: deeper set of filters</span>
<span class="n">path</span> <span class="o">=</span> <span class="n">root_path</span> <span class="o">/</span> <span class="s2">&quot;input_modules.contact_maps.feature_extractor.conv.3.conv_1&quot;</span> 
<span class="n">visualize_filters</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">first_validation_sample_index</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">last_validation_sample_index</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">layer_name</span><span class="o">=</span> <span class="s2">&quot;layer_3.1&quot;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-2.021034002304077 1.9138600826263428
-2.020750045776367 1.8846609592437744
-1.9546396732330322 1.9160393476486206
-1.8921006917953491 1.9064550399780273
-2.0078046321868896 1.928229570388794
-1.9475384950637817 1.9080767631530762
-7.166399002075195 2.429628372192383
-7.975045680999756 2.919764757156372
-7.415754318237305 2.5516605377197266
-7.684394359588623 2.1006875038146973
-7.582004547119141 2.4546499252319336
-7.244106769561768 3.086484909057617
-3.6593098640441895 1.1407158374786377
-3.393815517425537 1.2011475563049316
-3.6953325271606445 0.9598491787910461
-3.430828332901001 1.0976873636245728
-3.5928092002868652 1.1078323125839233
-3.9535763263702393 1.1592458486557007
-5.639556884765625 1.5351132154464722
-5.4008917808532715 1.2717441320419312
-5.332563877105713 1.413964033126831
-4.86221170425415 1.3372340202331543
-4.822977066040039 1.4820897579193115
-5.077983379364014 1.2805430889129639
</pre></div>
</div>
<img alt="../_images/a806cdd260751f14567fd3aff974f0c9e07a249fac9f1c06b040f4cbe07bd38b.png" src="../_images/a806cdd260751f14567fd3aff974f0c9e07a249fac9f1c06b040f4cbe07bd38b.png" />
</div>
</div>
<p><strong>Smoothing and downresolving output profiles</strong>
The main signal in the target EU-seq profiles oscillates at a hbp-kbp scale. Keeping bp resolution would:</p>
<ul class="simple">
<li><p>create a high number of outputs, complicating the computations of attributions for every input output pair.</p></li>
<li><p>Constrain us to reduce the output window per sample. We could not be able to assess the effects of the perturbations on as many genes.</p></li>
<li><p>It is not clear that the information to predict those wiggles in the signal is encoded in the inputs (core chrom. landscape marks).</p></li>
</ul>
<p>Supplementary figure showing the downresolution of the targets</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">scipy.ndimage</span> <span class="kn">import</span> <span class="n">gaussian_filter1d</span>
<span class="kn">import</span> <span class="nn">pyBigWig</span>

<span class="k">def</span> <span class="nf">smoothed_profiles</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                      <span class="n">savepath</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                      <span class="n">output_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">n_output_bins</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                      <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">sigma</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                      <span class="n">binsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Targets are centered at the TSS, and obtained at a 20 bp resolution, smoothed and downsized to 1kbp resolution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span><span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span><span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span><span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
    <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span><span class="kc">None</span><span class="p">)</span>
    
    <span class="n">original_array</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="n">gaussian_filter1d</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
    <span class="n">target_cond</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">stats</span><span class="p">)</span>

    <span class="c1"># Averaging over a number of bins</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">binsize</span><span class="p">):</span>
        <span class="n">target_cond</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">binsize</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="o">-</span><span class="n">j</span><span class="p">)</span>

    <span class="n">averaged_array</span> <span class="o">=</span> <span class="n">target_cond</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="n">final_array</span> <span class="o">=</span> <span class="n">target_cond</span><span class="p">[::</span><span class="n">binsize</span><span class="p">]</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">axs</span><span class="p">)):</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">60</span><span class="p">))</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Reads&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Bin number&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">original_array</span><span class="p">)),</span> <span class="n">original_array</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.1</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">averaged_array</span><span class="p">)),</span> <span class="n">averaged_array</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">final_array</span><span class="p">)),</span> <span class="n">final_array</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Smoothed_targets.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="n">data_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../GEO_files/EU_Seq_Ctrl.bw&quot;</span><span class="p">)</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/&quot;</span><span class="p">)</span>
<span class="n">output_shift</span> <span class="o">=</span> <span class="mi">200010</span> <span class="c1"># Output in 20 bp resolution (to be smoothed and downsampled)</span>
<span class="n">n_output_bins</span> <span class="o">=</span> <span class="mi">20001</span>
<span class="n">chrom</span> <span class="o">=</span> <span class="s1">&#39;chr4&#39;</span>
<span class="n">TSS</span> <span class="o">=</span> <span class="mi">152039321</span>
<span class="n">sigma</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">binsize</span> <span class="o">=</span> <span class="mi">50</span>

<span class="n">smoothed_profiles</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span>
                  <span class="n">savepath</span><span class="p">,</span> 
                  <span class="n">output_shift</span><span class="p">,</span>
                  <span class="n">n_output_bins</span><span class="p">,</span>
                  <span class="n">TSS</span><span class="p">,</span>
                  <span class="n">chrom</span><span class="p">,</span>
                  <span class="n">sigma</span><span class="p">,</span>
                  <span class="n">binsize</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b1e819cc24fac0b58a40cef10c6595258a21d8f5eab6981ff47832330e9a860d.png" src="../_images/b1e819cc24fac0b58a40cef10c6595258a21d8f5eab6981ff47832330e9a860d.png" />
</div>
</div>
<p><strong>Hyena and Enformer Embeddings</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span> 
<span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">57</span>
<span class="n">SHIFT</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/DNA_sequences&quot;</span><span class="p">)</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000066324.2_forward.npy&quot;</span> <span class="c1">#&quot;ENSMUSG00000003282.9_forward.npy&quot; </span>
<span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;test_embeddings_Enformer&quot;</span> <span class="o">/</span> <span class="n">sample_id</span> <span class="p">)</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/test_targets.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
<span class="n">d</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;test_embeddings_Hyena-DNA&quot;</span> <span class="o">/</span> <span class="n">sample_id</span><span class="p">)</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/supplementary_figures/&quot;</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
<span class="n">axs</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">b</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sample_id</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]][</span><span class="n">SHIFT</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">:</span><span class="n">SHIFT</span><span class="o">+</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;darkgreen&#39;</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;Control EU-seq target&quot;</span><span class="p">)</span>
<span class="c1">#axs.fill_between(np.arange(len(b.loc[sample_id[:-4]])//2),b.loc[sample_id[:-4]][len(b.loc[sample_id[:-4]])//2:], lw=1, color=&#39;goldenrod&#39;, alpha=.5, label=&quot;Treated&quot;)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;EU-seq (reads)&quot;</span><span class="p">,</span> <span class="n">rotation</span> <span class="o">=</span> <span class="mi">90</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="c1">#axs.set_yticks([0,6],[0,6])</span>
<span class="c1">#axs.set_ylim((0,6.5))</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Profile.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>


<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">15</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;inferno&#39;</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to the TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">448.5</span><span class="p">,</span><span class="mi">896</span><span class="p">],[</span><span class="o">-</span><span class="mf">57.408</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">57.408</span><span class="p">])</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">3071</span><span class="p">],[</span><span class="mi">1</span><span class="p">,</span><span class="mi">3072</span><span class="p">])</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Embedding depth&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">896</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Enformer_embedding.png&quot;</span><span class="p">,</span> <span class="n">dpi</span> <span class="o">=</span> <span class="mi">200</span><span class="p">)</span>


<span class="n">central_bins</span> <span class="o">=</span> <span class="mi">896</span>
<span class="n">bp_per_bin</span> <span class="o">=</span> <span class="mi">128</span>
<span class="n">central_width</span> <span class="o">=</span> <span class="n">central_bins</span><span class="o">*</span><span class="n">bp_per_bin</span>
<span class="n">crop_distance</span> <span class="o">=</span> <span class="p">(</span><span class="mi">160002</span><span class="o">-</span><span class="n">central_width</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span>

<span class="n">fig3</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">4.5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">d</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">T</span><span class="p">[:,</span><span class="n">crop_distance</span><span class="p">:</span><span class="o">-</span><span class="n">crop_distance</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span>
<span class="n">axs</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">,</span> <span class="n">vmin</span> <span class="o">=</span> <span class="o">-</span><span class="mi">4</span><span class="p">,</span> <span class="n">vmax</span> <span class="o">=</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]],[</span><span class="o">-</span><span class="mf">57.408</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">57.408</span><span class="p">])</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to the TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">axs</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Embedding depth&quot;</span><span class="p">)</span>
<span class="n">fig3</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Hyena_embeddings.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ea4b8b99b7137408f989aed69b12602d7421d0777ec0f6ec177543bf56f2b54f.png" src="../_images/ea4b8b99b7137408f989aed69b12602d7421d0777ec0f6ec177543bf56f2b54f.png" />
<img alt="../_images/cbc8c6c2aae1686c13c230af0dc602805e57a7a467e6a1261bf9201fb7ffd70e.png" src="../_images/cbc8c6c2aae1686c13c230af0dc602805e57a7a467e6a1261bf9201fb7ffd70e.png" />
<img alt="../_images/16c62f59f8ccd3e1fcf7a5854e983b88764feebf7742e431610e8af13c8174f0.png" src="../_images/16c62f59f8ccd3e1fcf7a5854e983b88764feebf7742e431610e8af13c8174f0.png" />
</div>
</div>
<p><strong>Benchmark model loss evolution:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="k">def</span> <span class="nf">plot_losses</span><span class="p">(</span><span class="n">hyena_log_file</span><span class="p">,</span> <span class="n">enformer_log_file</span><span class="p">,</span> <span class="n">savepath</span><span class="p">):</span>
    <span class="c1"># Function to read log file and extract losses</span>
    <span class="k">def</span> <span class="nf">extract_losses</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="n">train_pattern</span><span class="p">,</span> <span class="n">test_pattern</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">log_data</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        
        <span class="n">train_losses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">val_losses</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">epochs</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">train_matches</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">train_pattern</span><span class="p">,</span> <span class="n">log_data</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">test_pattern</span><span class="p">:</span>
            <span class="n">test_matches</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">test_pattern</span><span class="p">,</span> <span class="n">log_data</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">test_matches</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">train_matches</span><span class="p">:</span>
            <span class="n">epochs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">match</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
            <span class="n">train_losses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">match</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">match</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">:</span>
                <span class="n">val_losses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">match</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>

        <span class="k">if</span> <span class="n">test_pattern</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">test_matches</span><span class="p">:</span>
                <span class="n">val_losses</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">match</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">epochs</span><span class="p">,</span> <span class="n">train_losses</span><span class="p">,</span> <span class="n">val_losses</span>

    <span class="c1"># Extract HyenaDNA-32k data</span>
    <span class="n">hyena_train_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;Train Epoch: (\d+)\s+Average Loss: ([\d.]+)&quot;</span>
    <span class="n">hyena_test_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;Test set: Average loss: ([\d.]+)&quot;</span>
    <span class="n">hyena_epochs</span><span class="p">,</span> <span class="n">hyena_train_losses</span><span class="p">,</span> <span class="n">hyena_test_losses</span> <span class="o">=</span> <span class="n">extract_losses</span><span class="p">(</span><span class="n">hyena_log_file</span><span class="p">,</span> <span class="n">hyena_train_pattern</span><span class="p">,</span> <span class="n">hyena_test_pattern</span><span class="p">)</span>

    <span class="c1"># Extract Enformer data</span>
    <span class="n">enformer_pattern</span> <span class="o">=</span> <span class="sa">r</span><span class="s2">&quot;Epoch \[(\d+)/\d+\], Training Loss: ([\d.]+), Training R\^2: [\d.]+, Validation Loss: ([\d.]+), Validation R\^2: [\d.]+&quot;</span>
    <span class="n">enformer_epochs</span><span class="p">,</span> <span class="n">enformer_train_losses</span><span class="p">,</span> <span class="n">enformer_val_losses</span> <span class="o">=</span> <span class="n">extract_losses</span><span class="p">(</span><span class="n">enformer_log_file</span><span class="p">,</span> <span class="n">enformer_pattern</span><span class="p">)</span>

    <span class="c1"># Plotting</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

    <span class="c1"># HyenaDNA-32k plot</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">hyena_epochs</span><span class="p">,</span> <span class="n">hyena_train_losses</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Training Loss&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">hyena_epochs</span><span class="p">,</span> <span class="n">hyena_test_losses</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Test Loss&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Epoch&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Loss&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;HyenaDNA-32k&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="c1"># Enformer plot</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">enformer_epochs</span><span class="p">,</span> <span class="n">enformer_train_losses</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Training Loss&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">enformer_epochs</span><span class="p">,</span> <span class="n">enformer_val_losses</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Validation Loss&#39;</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Epoch&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Loss&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Enformer&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

<span class="c1"># File paths</span>
<span class="n">hyena_log_file</span> <span class="o">=</span> <span class="s1">&#39;../checkpoints/hyenadna_32k.log&#39;</span>
<span class="n">enformer_log_file</span> <span class="o">=</span> <span class="s1">&#39;../benchmarks/Enformer/training.log&#39;</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="s1">&#39;../figures/supplementary_figures/Train_test_loss_curves_benchmark.png&#39;</span>

<span class="c1"># Plot the losses</span>
<span class="n">plot_losses</span><span class="p">(</span><span class="n">hyena_log_file</span><span class="p">,</span> <span class="n">enformer_log_file</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/2f4003fd5b4c67823fc75ca657a56a0125ffb0b30b98249b23ad656d260d3062.png" src="../_images/2f4003fd5b4c67823fc75ca657a56a0125ffb0b30b98249b23ad656d260d3062.png" />
</div>
</div>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#"><center> DATA ANALYSIS NOTEBOOK <center></center></center></a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#packages">Packages</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functions">Functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model-performance-on-the-baseline-prediction-task"><center> Model performance on the baseline prediction task <center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#in-silico-perturbations"><center> <em>In silico</em> perturbations <center></center></center></a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#attributions"><center> Attributions <center></center></center></a></li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Marc Pielies Avelli
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022, Marc Pielies Avelli.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=3ee479438cf8b5e0d341"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=3ee479438cf8b5e0d341"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>