
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>REVISIONS: CODE ADDED &#8212; CLASTER</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=6771b56e"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'scripts/IV_Revisions';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="DATA ANALYSIS NOTEBOOK" href="III_Data_analysis.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">CLASTER</p>
  
</a></div>
        <div class="sidebar-primary-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/RasmussenLab/CLASTER" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Claster
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="0_Tutorial.html">TUTORIAL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Scripts:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="I_Data_obtention.html">1. DATA OBTENTION &amp; PREPROCESSING</a></li>
<li class="toctree-l1"><a class="reference internal" href="II_Run_CLASTER.html">2. CREATE &amp; RUN CLASTER</a></li>
<li class="toctree-l1"><a class="reference internal" href="IIb_Run_HyenaDNA_and_Enformer.html">2b. BENCHMARKING CLASTER WITH HYENA-DNA AND ENFORMER</a></li>
<li class="toctree-l1"><a class="reference internal" href="III_Data_analysis.html">DATA ANALYSIS NOTEBOOK</a></li>

<li class="toctree-l1 current active"><a class="current reference internal" href="#"><center> REVISIONS: CODE ADDED <center></center></center></a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/RasmussenLab/CLASTER/blob/master/scripts/IV_Revisions.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Colab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Colab logo" src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/RasmussenLab/CLASTER/edit/master/scripts/IV_Revisions.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/scripts/IV_Revisions.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1><center> REVISIONS: CODE ADDED <center></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#i-packages-functions">I) Packages &amp;  Functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-creation-of-eir-config-files-to-define-claster">II) Creation of EIR config files to define CLASTER</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-enhancer-centric-perturbational-analysis">III) Enhancer-centric perturbational analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#iv-new-models-and-analyses">IV) New models and analyses <center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-assessing-how-the-model-learns-to-predict-transcription"><center> A) Assessing how the model learns to predict transcription<center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#b-creating-a-shorter-context-model"><center> B) Creating a shorter context model: <center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#c-changing-the-last-activation-function-or-the-loss"><center> C) Changing the last activation function or the loss:</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#d-testing-on-another-chromosome-chr19"><center> D) Testing on another chromosome (chr19):</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#e-training-without-the-enhancer-mark-h3k27ac-as-an-input"><center> E) Training without the enhancer mark (H3K27ac) as an input.</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#f-predict-untransformed-eu-seq-only-binned-at-1kbp-resolution"><center> F) Predict untransformed EU-seq, only binned at 1kbp resolution. <center></center></center></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-predictions-and-attributions-of-the-new-models">Analyze predictions and attributions of the new models</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v-performance-evaluations">V) Performance evaluations:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-data-distributions">VI) Data distributions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-promoter-capture-hi-c-analyses">VII) Promoter-Capture Hi-C analyses</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-extending-claster-to-new-cell-types-k562-human"><center> VI) Extending CLASTER to new cell types: K562 (human) <center></center></center></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#predicting-mature-rna-seq-counts-rna-polii-chip-seq">Predicting mature RNA-seq counts &amp; RNA-polII ChIP-seq</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-centered-analyses-of-rna-seq-and-polr2a-binding-differences-after-in-silico-enhancer-ablation">Enhancer-centered analyses of RNA-seq and POLR2A binding differences after <em>in silico</em> enhancer ablation:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-gene-pairing-analysis-on-predicted-rna-seq-and-polr2a-chip-seq">Enhancer-gene pairing analysis on predicted RNA-seq and POLR2A Chip-seq.</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#editing-gene-annotations-table-to-add-whether-they-were-included-or-not-in-the-analyses">Editing gene annotations table to add whether they were included or not in the analyses</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="revisions-code-added">
<h1><center> REVISIONS: CODE ADDED <center><a class="headerlink" href="#revisions-code-added" title="Link to this heading">#</a></h1>
<p><strong>Summary:</strong></p>
<p>This notebook contains the analyses that were added during the reviewing process. It borrows elements from the rest of the notebooks, but it was made so that it can be run completely independently from the others. Some cells can be run locally, but others will create separate python files that were run on a slurm-based HPC.</p>
<p><strong>SECTIONS:</strong></p>
<ul class="simple">
<li><p>I) Packages &amp; Functions.</p></li>
<li><p>II) Creation of EIR config files to define CLASTER.</p></li>
<li><p>III) Enhancer-centric perturbational analysis.</p></li>
<li><p>IV) New models and analyses.</p></li>
<li><p>V) Performance evaluations.</p></li>
<li><p>VI) Data distributions.</p></li>
<li><p>VII) Obtain Promoter-Capture Hi-C for mESCs.</p></li>
<li><p>VIII) Perform analyses for K562.</p></li>
</ul>
<blockquote>
<div><p>Notes:</p>
<ul class="simple">
<li><p>We used eir version 0.1.42.</p></li>
<li><p>We run the models on an A100 GPU.</p></li>
<li><p>We added the genomic context of the predictions as IGV tracks describing:</p>
<ul>
<li><p>Gene annotations (Default IGV annotations for mm10)</p></li>
<li><p>Regulatory element annotations: latest ENCODE release fro mm10 candidate Cis-Regulatory Elements (cCREs). Data was obtained from <a class="reference external" href="https://hgdownload.soe.ucsc.edu/gbdb/mm10/encode3/ccre/">this download server</a></p></li>
</ul>
</li>
</ul>
</div></blockquote>
<section id="i-packages-functions">
<h2>I) Packages &amp;  Functions<a class="headerlink" href="#i-packages-functions" title="Link to this heading">#</a></h2>
<p><strong>Global packages:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">######### Packages: ########</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.colors</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">colors</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">patches</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">mpl_toolkits.axes_grid1.inset_locator</span><span class="w"> </span><span class="kn">import</span> <span class="n">inset_axes</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.integrate</span><span class="w"> </span><span class="kn">import</span> <span class="n">simpson</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">pearsonr</span><span class="p">,</span> <span class="n">spearmanr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">import_ipynb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">claster_utils</span> <span class="c1"># Import functions from data analysis notebook</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">r2_score</span><span class="p">,</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">mean_absolute_error</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">wget</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">gridspec</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.ndimage</span><span class="w"> </span><span class="kn">import</span> <span class="n">gaussian_filter</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Old functions:</strong></p>
<p>These are mostly functions used in the original analyses with slight adaptations.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">track_dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
            <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
            <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Enhancer&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;b&quot;</span><span class="p">},</span>
            <span class="mi">3</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>

<span class="c1">######### Functions: ########</span>
<span class="k">def</span><span class="w"> </span><span class="nf">visualize_input_array</span><span class="p">(</span><span class="n">a</span><span class="p">,</span>
                     <span class="n">cropped_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4800</span><span class="p">,</span> <span class="c1">#4600</span>
                     <span class="n">scaling_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.</span><span class="p">,</span>
                     <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to visualize an input numpy array with inset zoomed plots.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create single column of plots</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">track_dict</span><span class="p">),</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
    
    <span class="c1"># Handle case where there&#39;s only one track (axs would not be array)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">track_dict</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">axs</span> <span class="o">=</span> <span class="p">[</span><span class="n">axs</span><span class="p">]</span>

    <span class="c1"># Plot the scaled arrays</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="c1"># Main plot (full window)</span>
        <span class="n">x_full</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_full</span><span class="p">,</span> <span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> 
                   <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> 
                   <span class="n">label</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> 
                   <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_full</span><span class="p">,</span> <span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> 
                          <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> 
                          <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
        
        <span class="c1"># Create inset axes</span>
        <span class="n">inset_ax</span> <span class="o">=</span> <span class="n">inset_axes</span><span class="p">(</span><span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> 
                             <span class="n">width</span><span class="o">=</span><span class="s2">&quot;40%&quot;</span><span class="p">,</span> <span class="c1"># width = 30% of parent_bbox</span>
                             <span class="n">height</span><span class="o">=</span><span class="s2">&quot;50%&quot;</span><span class="p">,</span> <span class="c1"># height : 30%</span>
                             <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        
        <span class="c1"># Plot zoomed data in inset</span>
        <span class="n">x_zoom</span> <span class="o">=</span> <span class="n">x_full</span><span class="p">[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">]</span>
        <span class="n">y_zoom</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">]</span><span class="o">*</span><span class="n">scaling_factor</span>
        <span class="n">inset_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_zoom</span><span class="p">,</span> <span class="n">y_zoom</span><span class="p">,</span> 
                     <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> 
                     <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">inset_ax</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_zoom</span><span class="p">,</span> <span class="n">y_zoom</span><span class="p">,</span> 
                            <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> 
                            <span class="n">alpha</span><span class="o">=</span><span class="mf">.6</span><span class="p">)</span>
        <span class="n">inset_ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_zoom</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">x_zoom</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">inset_ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">tick_right</span><span class="p">()</span>
        

    <span class="c1"># Set the bottom x-label</span>
    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to TSS (hbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>

    <span class="c1"># Adjust spacing between subplots</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">_get_predictions</span><span class="p">(</span>
    <span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">condition_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">is_training</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">batch_num</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">CLIP</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads prediction files from EIR for each target bin condition.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: Path where the predictions are stored</span>
<span class="sd">        N_BINS: Number of bins to one side of the central bin (e.g., 200 or 57 for Enformer)</span>
<span class="sd">        condition_list: List of conditions (e.g., [&quot;_ctrl&quot;])</span>
<span class="sd">        is_training: If True, uses training file structure, else uses test structure</span>
<span class="sd">        batch_num: Batch number for training predictions (e.g., &quot;30300&quot;)</span>
<span class="sd">        CLIP: If True, applies ReLU to predictions (clips at 0)</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        Tuple containing:</span>
<span class="sd">            - ids: List of sample names</span>
<span class="sd">            - predicted: DataFrame of predicted untransformed values</span>
<span class="sd">            - actual: DataFrame of actual values</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predicted_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">batch_num</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;batch_num must be specified for training predictions&quot;</span><span class="p">)</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/samples/</span><span class="si">{</span><span class="n">batch_num</span><span class="si">}</span><span class="s2">/regression_predictions.csv&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/predictions.csv&quot;</span>
            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prediction file not found: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>

            <span class="c1"># Apply ReLU to model predictions if specified</span>
            <span class="k">if</span> <span class="n">CLIP</span><span class="p">:</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Handle different column names for training vs test</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2"> Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;True Label Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">predicted_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_column</span><span class="p">)</span>
            <span class="n">actual_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_column</span><span class="p">)</span>

    <span class="c1"># Concatenate all DataFrames horizontally</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predicted_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">actual</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">actual_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_target_predictions</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Plot the predicted profiles. </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">sharex</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_p</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;royalblue&#39;</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">.3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Predicted EU-seq&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span><span class="n">line_a</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;silver&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.6</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;EU-seq&quot;</span><span class="p">)</span>          
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>    
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>  
    <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Plot multiple predicted profiles and actual data.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    predictions: list of dicts, each containing:</span>
<span class="sd">        - &#39;profile&#39;: array-like, the prediction data</span>
<span class="sd">        - &#39;color&#39;: str, color for the plot</span>
<span class="sd">        - &#39;alpha&#39;: float, transparency value</span>
<span class="sd">        - &#39;label&#39;: str, label for the legend</span>
<span class="sd">    actual_data: dict containing:</span>
<span class="sd">        - &#39;profile&#39;: array-like, the actual data</span>
<span class="sd">        - &#39;color&#39;: str, default &#39;silver&#39;</span>
<span class="sd">        - &#39;alpha&#39;: float, default 0.6</span>
<span class="sd">        - &#39;label&#39;: str, default &#39;EU-seq&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create figure</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">4</span><span class="p">),</span> <span class="n">sharex</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Plot all predictions in top subplot</span>
    <span class="n">x_range</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">pred</span> <span class="ow">in</span> <span class="n">predictions</span><span class="p">:</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span>
                           <span class="n">pred</span><span class="p">[</span><span class="s1">&#39;profile&#39;</span><span class="p">],</span>
                           <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                           <span class="n">color</span><span class="o">=</span><span class="n">pred</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                           <span class="n">alpha</span><span class="o">=</span><span class="n">pred</span><span class="p">[</span><span class="s1">&#39;alpha&#39;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span>
                           <span class="n">pred</span><span class="p">[</span><span class="s1">&#39;profile&#39;</span><span class="p">],</span>
                           <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                           <span class="n">color</span><span class="o">=</span><span class="n">pred</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                           <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                           <span class="n">ls</span><span class="o">=</span><span class="n">pred</span><span class="p">[</span><span class="s1">&#39;ls&#39;</span><span class="p">],</span>
                           <span class="n">label</span><span class="o">=</span><span class="n">pred</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
    
    <span class="c1"># Plot actual data in bottom subplot</span>
    <span class="k">for</span> <span class="n">actual</span> <span class="ow">in</span> <span class="n">actual_data</span><span class="p">:</span>
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span>
                            <span class="n">actual</span><span class="p">[</span><span class="s1">&#39;profile&#39;</span><span class="p">],</span>
                            <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                            <span class="n">color</span><span class="o">=</span><span class="n">actual</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                            <span class="n">alpha</span><span class="o">=</span><span class="n">actual</span><span class="p">[</span><span class="s1">&#39;alpha&#39;</span><span class="p">])</span>
        
        <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_range</span><span class="p">,</span>
                    <span class="n">actual</span><span class="p">[</span><span class="s1">&#39;profile&#39;</span><span class="p">],</span>
                    <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                    <span class="n">color</span><span class="o">=</span><span class="n">actual</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                    <span class="n">ls</span><span class="o">=</span><span class="n">actual</span><span class="p">[</span><span class="s1">&#39;ls&#39;</span><span class="p">],</span>
                    <span class="n">label</span><span class="o">=</span><span class="n">actual</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">])</span>
        
    <span class="c1"># Set labels and limits</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance from TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>    
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">65</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">65</span><span class="p">)</span>
    
    <span class="c1"># Add legend</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">ncol</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
                             <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">l_in</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10001</span><span class="p">,</span>
                             <span class="n">n_in</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                             <span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                             <span class="n">SPLIT</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is aimed to read attribution arrays from the chromatin landscape branch. Attribution arrays score the </span>
<span class="sd">    contribution of each position in the input arrays (4,10.001) towards each output node/position (401).</span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: path where the attribution arrays are stored.</span>
<span class="sd">        figure_path: path where we want to store the outputs</span>
<span class="sd">        n_central_bins: number of central bins (from -N_BINS to  N_BINS+1)</span>
<span class="sd">        track_dict: dictionary containing the names of the tracks and plot details</span>
<span class="sd">        l_in: input length </span>
<span class="sd">        n_out: number of output nodes</span>
<span class="sd">        n_in: number of input channels</span>
<span class="sd">        TRAIN_ATTR: whether this is a training or test run (folder structure changes).</span>
<span class="sd">        SPLIT: if Train, we need to know which of the stored batches it is</span>

<span class="sd">    Returns:</span>
<span class="sd">        Plots with attribution scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">N_IN_BINS</span> <span class="o">=</span> <span class="n">l_in</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">x_val</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_IN_BINS</span><span class="p">,</span><span class="n">N_IN_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">500</span><span class="p">)</span>

    <span class="c1"># Getting attributions:</span>
    <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
    <span class="n">landscape_avg</span> <span class="o">=</span> <span class="p">{</span><span class="n">condition</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_in</span><span class="p">,</span><span class="n">l_in</span><span class="p">))</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">}</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
            <span class="c1"># CHROMATIN MARK ATTRIBUTIONS:</span>
            <span class="n">added_folders</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;/samples/</span><span class="si">{</span><span class="n">SPLIT</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">TRAIN_ATTR</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="c1"># Different folder structure</span>
            <span class="n">att_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}{</span><span class="n">added_folders</span><span class="si">}</span><span class="s2">/attributions/gene_expression/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">att_path</span><span class="p">):</span>
                <span class="n">attributions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">att_path</span><span class="p">)</span>
                <span class="c1">#if (i == -73) and (condition == &quot;_ctrl&quot;):</span>
                <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span> 
                <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="n">j</span><span class="p">]),</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Chromatin_landscape_attributions_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">_ctrl.png&quot;</span><span class="p">,</span><span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

                <span class="c1"># Add a small phase to remove high frequency fluctuations:</span>
                <span class="n">eps</span> <span class="o">=</span> <span class="mi">10</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                
                <span class="n">centered_attr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">attributions</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="nb">int</span><span class="p">(</span><span class="n">eps</span><span class="p">),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">centered_attr</span><span class="p">))</span>

    <span class="n">SCALING_FACTOR</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">landscape_avg</span><span class="p">[</span><span class="n">cond</span><span class="p">]</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span>   
        <span class="n">sub_axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">axes</span><span class="p">([</span><span class="mf">.615</span><span class="p">,</span> <span class="mf">.57</span><span class="p">,</span> <span class="mf">.27</span><span class="p">,</span> <span class="mf">.27</span><span class="p">])</span>  
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
            <span class="n">sub_axes</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">markersize</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span><span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
        
        <span class="c1">#sub_axes.set_xlim(xlim)</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">,</span><span class="mi">5000</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">501</span><span class="p">,</span><span class="mi">500</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5000</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">xlim</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#(0,.002))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Attribution score&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to the predicted locus (kbp)&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">400</span><span class="p">,</span><span class="mi">401</span><span class="p">,</span><span class="mi">100</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">40</span><span class="p">,</span><span class="mi">41</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Landscape_absolute_average_</span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">read_gene_positions</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function reads a csv containing genes and enhancers (entities), and creates a dictionary with the relative coordinates of</span>
<span class="sd">    all entities to the TSS of the reference gene (central gene), which gives name to the sample.</span>

<span class="sd">    Args:</span>
<span class="sd">        csv_path: path to the csv with ref genes, entities and their relative coordinates.</span>
<span class="sd">        resolution: distance unit we want our relative distances to be given with (1-&gt; bp, 1000-&gt;kbp)</span>

<span class="sd">    Returns:</span>
<span class="sd">        pos_dict: dictionary given pos_dict[ref gene][entity_name] = (rel_start, rel_end)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load the results from a CSV file</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">csv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Initialize the dictionary to store positions</span>
    <span class="n">pos_dict</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="c1"># Filter the DataFrame to get only gene entities </span>
    <span class="n">gene_entities</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ENSMUSG&#39;</span><span class="p">)]</span>  <span class="c1"># Adjust the condition according to your data</span>
    
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_entities</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">ref_gene_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span>
        <span class="n">entity_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span>
        <span class="n">rel_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_Start&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span><span class="p">))</span>
        <span class="n">rel_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Entity_rel_End&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">resolution</span><span class="p">))</span>
        
        <span class="c1"># Ensure the dictionary has the necessary structure</span>
        <span class="k">if</span> <span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pos_dict</span><span class="p">:</span>
            <span class="n">pos_dict</span><span class="p">[</span><span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        
        <span class="c1"># Store the scaled and rounded start and end positions</span>
        <span class="n">pos_dict</span><span class="p">[</span><span class="n">ref_gene_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">][</span><span class="n">entity_id</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">pos_dict</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_area_gene</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Args:</span>
<span class="sd">        line_p: vector of predicted values</span>
<span class="sd">        line_a: vector of actual values</span>
<span class="sd">        ID: reference gene, i.e. sample name</span>
<span class="sd">        pos_dict: dictionary with relative gene coordinates</span>
<span class="sd">        N_BINS: which length do we set as a threshold to integrate (in the new resolution).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_center</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">line_p</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">rel_start</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">rel_end</span> <span class="o">=</span> <span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">seq_center</span> <span class="o">+</span> <span class="n">pos_dict</span><span class="p">[</span><span class="n">ID</span><span class="p">][</span><span class="n">ID</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span> <span class="k">else</span> <span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span>

    <span class="n">WIDTH</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">rel_end</span> <span class="o">-</span> <span class="n">rel_start</span><span class="p">))</span>

    <span class="c1"># Appending area</span>
    <span class="k">if</span> <span class="n">rel_start</span> <span class="o">!=</span> <span class="n">rel_end</span><span class="p">:</span>
        <span class="n">pred_area</span> <span class="o">=</span> <span class="n">simpson</span><span class="p">(</span><span class="n">line_p</span><span class="p">[</span><span class="n">rel_start</span><span class="p">:</span><span class="n">rel_end</span><span class="p">])</span><span class="o">/</span><span class="n">WIDTH</span>
        <span class="n">actual_area</span> <span class="o">=</span> <span class="n">simpson</span><span class="p">(</span><span class="n">line_a</span><span class="p">[</span><span class="n">rel_start</span><span class="p">:</span><span class="n">rel_end</span><span class="p">])</span><span class="o">/</span><span class="n">WIDTH</span>

    <span class="k">else</span><span class="p">:</span> <span class="c1"># Annotation is shorter than bin_width:</span>
        <span class="n">pred_area</span> <span class="o">=</span> <span class="n">line_p</span><span class="p">[</span><span class="n">seq_center</span><span class="p">]</span>
        <span class="n">actual_area</span> <span class="o">=</span> <span class="n">line_a</span><span class="p">[</span><span class="n">seq_center</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">pred_area</span><span class="p">,</span> <span class="n">actual_area</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_correlations</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span> <span class="n">is_training</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">batch_num</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">CLIP</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">True</span> <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function returns a set of dictionaries containing the actual EU-seq values and the predicted values. </span>
<span class="sd">    This is done both on a bin by bin basis and integrated over the target gene length (normalized by gene length).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Storing gene position&#39;s dict:</span>
    <span class="n">pos_dict</span> <span class="o">=</span> <span class="n">read_gene_positions</span><span class="p">(</span><span class="n">gene_pos_path</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">,</span> <span class="n">condition_list</span><span class="p">,</span><span class="n">is_training</span><span class="o">=</span><span class="n">is_training</span><span class="p">,</span><span class="n">batch_num</span><span class="o">=</span><span class="n">batch_num</span><span class="p">,</span> <span class="n">CLIP</span><span class="o">=</span><span class="n">CLIP</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">)</span>

    <span class="c1"># Normalized Area per bin for the target gene (central gene in observation window)</span>
    <span class="n">pred_list_A_per_bin_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">actual_list_A_per_bin_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Predicted and actual EU-seq values in all positions:</span>
    <span class="n">pred_list_values_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">actual_list_values_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;_rev&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ID</span><span class="p">):</span> <span class="c1">#Only add the forward strand to the analyses</span>
            <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
                <span class="c1"># Extend binwise prediction and target lists</span>
                <span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">line_p</span><span class="p">))</span>
                <span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">line_a</span><span class="p">))</span>

                <span class="c1"># Extend area predictions</span>
                <span class="n">pred_area</span><span class="p">,</span> <span class="n">actual_area</span><span class="p">,</span> <span class="n">rel_start</span><span class="p">,</span> <span class="n">rel_end</span> <span class="o">=</span> <span class="n">calculate_area_gene</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">pos_dict</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">)</span>

                <span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">pred_area</span><span class="p">])</span>
                <span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">actual_area</span><span class="p">])</span>

                <span class="c1"># Adding area to extended table:</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">ID</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Ref_gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Entity_ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">),</span><span class="s1">&#39;Area&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred_area</span> <span class="c1">#Edit existing table</span>

            <span class="n">PLOT</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">ID_LIST</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000078673.10_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000067261.4_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000028234.6_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000028280.9_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000035969.15_forward&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">PLOT</span> <span class="ow">and</span> <span class="p">(</span><span class="n">ID</span> <span class="ow">in</span> <span class="n">ID_LIST</span><span class="p">):</span>
                <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_target_predictions</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">,</span> <span class="n">N_BINS</span><span class="p">)</span> 
                <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">input_table_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">pred_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">actual_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">pred_list_values_dict</span><span class="p">,</span> <span class="n">actual_list_values_dict</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">title</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">cmap</span><span class="p">:</span> <span class="n">colors</span><span class="o">.</span><span class="n">Colormap</span> <span class="o">=</span> <span class="s2">&quot;afmhot&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="p">:</span> <span class="nb">tuple</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">50</span><span class="p">),</span> <span class="n">density</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">DELTA</span><span class="p">:</span><span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">min_val</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">.0</span><span class="p">,</span> <span class="n">max_val</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    This function plots predicted vs. actual values and provides the Spearman and Pearson correlations of the regressions. </span>
<span class="sd">    The style of the plot changes depending on whether the predictions are at a bin level or a gene level.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the counts histogram to plot correlation with density colormap</span>
    <span class="n">counts</span><span class="p">,</span> <span class="n">xedges</span><span class="p">,</span> <span class="n">yedges</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">hist2d</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">binlims</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">binlims</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">200</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">binlims</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">binlims</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="mi">200</span><span class="p">)],</span> <span class="n">density</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Plot regression of single predicted values vs. real</span>
    <span class="n">pearson_r</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spearman_r</span> <span class="o">=</span> <span class="n">spearmanr</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">m</span><span class="p">,</span> <span class="n">b</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">deg</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># Actual fit:</span>
    <span class="n">a1</span><span class="p">,</span> <span class="n">a2</span><span class="p">,</span> <span class="n">a3</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">deg</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="c1"># Regression from origin:</span>
    <span class="c1">#m_0 , _, _, _ = np.linalg.lstsq(np.array(actual_list_values)[:,np.newaxis], pred_list_values) #x needs to be a column vector for this function</span>

    <span class="c1"># Create the figure</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>    
    <span class="k">if</span> <span class="n">density</span><span class="p">:</span>
        <span class="n">eps</span><span class="o">=</span> <span class="mi">1</span>
        <span class="c1"># Counts matrix is transposed: the origin convention for plt.imshow is (top, left), for plt.hist2D is (bottom,left)</span>
        <span class="n">counts</span><span class="p">[</span><span class="n">counts</span> <span class="o">&lt;</span> <span class="n">eps</span><span class="p">]</span> <span class="o">=</span><span class="mi">1</span>
        <span class="n">hist</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s1">&#39;lower&#39;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="n">xedges</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">xedges</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">yedges</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">yedges</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]],</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span>  <span class="n">norm</span><span class="o">=</span><span class="n">colors</span><span class="o">.</span><span class="n">LogNorm</span><span class="p">())</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">hist</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axline</span><span class="p">(</span><span class="n">xy1</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="n">slope</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Linear fit:</span><span class="se">\n</span><span class="s1">$ y = </span><span class="si">{</span><span class="n">m</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">x </span><span class="si">{</span><span class="n">b</span><span class="si">:</span><span class="s1">+.3f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_p:</span><span class="si">{</span><span class="n">pearson_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_s:</span><span class="si">{</span><span class="n">spearman_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;royalblue&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#_ = ax.plot(np.arange(min_val,max_val,step), a1*(np.arange(min_val,max_val,step)**2)+a2*(np.arange(min_val,max_val,step))+a3, color=&#39;royalblue&#39;, ls = &quot;--&quot;, lw=1, label= f&quot;Polynomial fit:\n$y={a1:.3f}x^2+{a2:.3f}x+{a3:.3f}$&quot;)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;scaled&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_val</span><span class="p">,</span><span class="n">max_val</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">min_val</span><span class="p">,</span><span class="n">max_val</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;log$_2$(True +1)&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;log$_2$(Predicted +1)&quot;</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.48</span><span class="p">,</span><span class="mf">.8</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">actual_list_values</span><span class="p">,</span><span class="n">pred_list_values</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span>  <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">markeredgewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;scaled&quot;</span><span class="p">)</span>
        <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axline</span><span class="p">(</span><span class="n">xy1</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">b</span><span class="p">),</span> <span class="n">slope</span><span class="o">=</span><span class="n">m</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Linear fit:</span><span class="se">\n</span><span class="s1">$ y = </span><span class="si">{</span><span class="n">m</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">x </span><span class="si">{</span><span class="n">b</span><span class="si">:</span><span class="s1">+.3f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_p:</span><span class="si">{</span><span class="n">pearson_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$</span><span class="se">\n</span><span class="s1">$r_s:</span><span class="si">{</span><span class="n">spearman_r</span><span class="si">:</span><span class="s1">.4f</span><span class="si">}</span><span class="s1">$&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;royalblue&quot;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#ax.plot(np.arange(-20,200,1), a1*(np.arange(-20,200,1)**2)+a2*(np.arange(-20,200,1))+a3, color=&#39;royalblue&#39;, ls = &quot;--&quot;, lw=1, label= f&quot;Polynomial fit:\n$y={a1:.3f}x^2+{a2:.3f}x{a3:.3f}$&quot;)</span>
        <span class="k">if</span> <span class="n">DELTA</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="o">-</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Actual $\Delta_</span><span class="si">{exp}</span><span class="s2">$  (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted $\Delta_</span><span class="si">{exp}</span><span class="s2">$ (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.53</span><span class="p">,</span><span class="mf">.87</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Actual averaged expression (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">_</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Predicted averaged expression (reads/kbp)&quot;</span><span class="p">)</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.53</span><span class="p">,</span><span class="mf">.87</span><span class="p">),</span> <span class="n">fancybox</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">shadow</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

    <span class="c1">#ax.plot(np.arange(0,500,1), np.arange(0,500,1), color=&#39;k&#39;, ls = &quot;dashed&quot;, lw=1)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Regression_histogram_</span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
    
<span class="k">def</span><span class="w"> </span><span class="nf">r2_score_computed</span><span class="p">(</span><span class="n">targets</span><span class="p">,</span> <span class="n">outputs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to compute R2 coefficient between predictions and targets.</span>
<span class="sd">    This allows for all operations to happen in the GPU )if available)</span>
<span class="sd">    without moving back and forth to the CPU.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">target_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">targets</span><span class="p">)</span>
    <span class="n">ss_tot</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">targets</span> <span class="o">-</span> <span class="n">target_mean</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">ss_res</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">((</span><span class="n">targets</span> <span class="o">-</span> <span class="n">outputs</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">r2</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">ss_res</span> <span class="o">/</span> <span class="n">ss_tot</span>
    <span class="k">return</span> <span class="n">r2</span>

<span class="k">def</span><span class="w"> </span><span class="nf">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">destination_folder</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function downloads the files in links and stores them in destination_folder.</span>
<span class="sd">    Args:</span>
<span class="sd">        links: dictionary where keys are file names and values are links.</span>
<span class="sd">        destination_folder: path where we&#39;ll store the files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span><span class="n">link</span> <span class="ow">in</span> <span class="n">links</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Ensure the destination folder exists</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">):</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">)</span>

            <span class="c1"># Specify the output path for the downloaded file</span>
            <span class="n">output_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">destination_folder</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span>

            <span class="c1"># Download the file to the specified destination folder</span>
            <span class="n">wget</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">link</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">output_path</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Successfully downloaded </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> to </span><span class="si">{</span><span class="n">destination_folder</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Failed to download </span><span class="si">{</span><span class="n">link</span><span class="si">}</span><span class="s2"> because : </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="ii-creation-of-eir-config-files-to-define-claster">
<h2>II) Creation of EIR config files to define CLASTER<a class="headerlink" href="#ii-creation-of-eir-config-files-to-define-claster" title="Link to this heading">#</a></h2>
<p>This cell writes all config files at once, easing reproducibility.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">config_paths</span> <span class="o">=</span> <span class="p">[</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_predict_prom/&quot;</span><span class="p">),</span> <span class="c1">#0</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_20kbp_context/&quot;</span><span class="p">),</span> <span class="c1">#1</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_softplus_no_bias/&quot;</span><span class="p">),</span> <span class="c1">#2</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_softplus_poisson/&quot;</span><span class="p">),</span><span class="c1">#3</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_20kbp_context_test/&quot;</span><span class="p">),</span><span class="c1">#4</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_chr19_holdout_train/&quot;</span><span class="p">),</span><span class="c1">#5</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_chr19_holdout_test/&quot;</span><span class="p">),</span><span class="c1">#6</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_no_H3K27ac_train/&quot;</span><span class="p">),</span><span class="c1">#7</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_no_H3K27ac_test/&quot;</span><span class="p">),</span><span class="c1">#8</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_20kbp_context_test/&quot;</span><span class="p">),</span><span class="c1">#9</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_1kbp_binning_train/&quot;</span><span class="p">),</span><span class="c1">#10</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_1kbp_binning_test/&quot;</span><span class="p">),</span><span class="c1">#11</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_chr19_enhancer_centric/&quot;</span><span class="p">),</span><span class="c1">#12</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_K562_train/&quot;</span><span class="p">),</span><span class="c1">#13</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_K562_test/&quot;</span><span class="p">),</span><span class="c1">#14</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_prom_CHiC_train/&quot;</span><span class="p">),</span> <span class="c1"># 15</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_prom_CHiC_test/&quot;</span><span class="p">),</span> <span class="c1"># 16</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_K562_POLR2A_train/&quot;</span><span class="p">),</span><span class="c1">#17</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/&quot;</span><span class="p">),</span><span class="c1">#18</span>
                <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../configurations/conf_pure_conv_K562_RNA_enhancer_centric/&quot;</span><span class="p">),</span><span class="c1">#19</span>
                <span class="p">]</span>
               
<span class="k">for</span> <span class="n">config_path</span> <span class="ow">in</span> <span class="n">config_paths</span><span class="p">:</span>
  <span class="n">config_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">predict_prom_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;  </span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/perturbed_landscape_arrays/inserted_promoter_arrays/  #inserted_enhancer_test #inputs/silenced_arrays/silenced_arrays_H2B_S.D </span>
<span class="s2">  #./data/parsed_data/inputs/arrays_train_100bp_no_H3K27ac/ #arrays_train_100bp_no_H3K27ac_uncoupled/</span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/prom_perturbed_targets.csv </span>
<span class="s2">  #./data/parsed_data/targets/target_arrays_perturbational_inserted_enhancer.csv #target_arrays_perturbational_S.D.csv #target_arrays_1kbp_401_bins_2_conditions_decareads_abs.csv</span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span> <span class="p">}</span>

<span class="n">training_20kbp_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_20kbp_context/  #gene_expression_prediction_no_H3K27ac_uncoupled/     #gene_expression_cnn_1kbp_401_bins_2_cond_reloaded/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs is 30300 #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training_20kbp_context/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [1,1]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 1</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">training_softplus_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_softplus_no_bias/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2"># latent_sampling: </span>
<span class="s2">#   layers_to_sample:</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.1.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.2.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.3.conv_2&quot;</span>
<span class="s2"># attribution_background_samples: 512</span>
<span class="s2"># attributions_every_sample_factor: 1</span>
<span class="s2">#pretrained_checkpoint: best_models/gene_expression_exformer_unlimited_chrom_and_micro_with_attention_model_117600_perf-average=0.8435.pt </span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">training_softplus_poisson_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias_log_input_false/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2"># latent_sampling: </span>
<span class="s2">#   layers_to_sample:</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.1.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.2.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.3.conv_2&quot;</span>
<span class="s2"># attribution_background_samples: 512</span>
<span class="s2"># attributions_every_sample_factor: 1</span>
<span class="s2">#pretrained_checkpoint: best_models/gene_expression_exformer_unlimited_chrom_and_micro_with_attention_model_117600_perf-average=0.8435.pt </span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;PoissonNLLLoss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_20kbp_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">#output_folder: ./runs/test_runs/gene_expression_only_chrom_pure_conv_20kbp_context/  #gene_expression_prediction_no_H3K27ac_uncoupled/     #gene_expression_cnn_1kbp_401_bins_2_cond_reloaded/ </span>
<span class="s2">#manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs is 30300 #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test_20kbp_context/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [1,1]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 1</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">training_chr19_holdout_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_chr19_holdout/  #gene_expression_prediction_no_H3K27ac_uncoupled/     #gene_expression_cnn_1kbp_401_bins_2_cond_reloaded/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr19.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2"># latent_sampling: </span>
<span class="s2">#   layers_to_sample:</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.1.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.2.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.3.conv_2&quot;</span>
<span class="s2"># attribution_background_samples: 512</span>
<span class="s2"># attributions_every_sample_factor: 1</span>
<span class="s2">#pretrained_checkpoint: best_models/gene_expression_exformer_unlimited_chrom_and_micro_with_attention_model_117600_perf-average=0.8435.pt </span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_chr19_holdout_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;  </span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>

<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test_chr19/  #inserted_enhancer_test #inputs/silenced_arrays/silenced_arrays_H2B_S.D </span>
<span class="s2">  #./data/parsed_data/inputs/arrays_train_100bp_no_H3K27ac/ #arrays_train_100bp_no_H3K27ac_uncoupled/</span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets_chr19.csv </span>
<span class="s2">  #./data/parsed_data/targets/target_arrays_perturbational_inserted_enhancer.csv #target_arrays_perturbational_S.D.csv #target_arrays_1kbp_401_bins_2_conditions_decareads_abs.csv</span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span> <span class="p">}</span>


<span class="n">training_no_H3K27ac_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_no_H3K27ac/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2"># latent_sampling: </span>
<span class="s2">#   layers_to_sample:</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.1.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.2.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.3.conv_2&quot;</span>
<span class="s2"># attribution_background_samples: 512</span>
<span class="s2"># attributions_every_sample_factor: 1</span>
<span class="s2">#pretrained_checkpoint: best_models/gene_expression_exformer_unlimited_chrom_and_micro_with_attention_model_117600_perf-average=0.8435.pt </span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training_no_H3K27ac/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 3 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_no_H3K27ac_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;  </span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>

<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test_no_H3K27ac/  #inserted_enhancer_test #inputs/silenced_arrays/silenced_arrays_H2B_S.D </span>
<span class="s2">  #./data/parsed_data/inputs/arrays_train_100bp_no_H3K27ac/ #arrays_train_100bp_no_H3K27ac_uncoupled/</span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 3 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets.csv </span>
<span class="s2">  #./data/parsed_data/targets/target_arrays_perturbational_inserted_enhancer.csv #target_arrays_perturbational_S.D.csv #target_arrays_1kbp_401_bins_2_conditions_decareads_abs.csv</span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span> <span class="p">}</span>


<span class="n">predict_20kbp_context_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs is 30300 #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test_20kbp_context/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [1,1]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 1</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span><span class="p">}</span>


<span class="n">training_1kbp_binning_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_pure_conv_1kbp_binning/  #gene_expression_prediction_no_H3K27ac_uncoupled/     #gene_expression_cnn_1kbp_401_bins_2_cond_reloaded/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2"># latent_sampling: </span>
<span class="s2">#   layers_to_sample:</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.0.conv_1&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.1.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.2.conv_2&quot;</span>
<span class="s2">#     - &quot;input_modules.contact_maps.feature_extractor.conv.3.conv_2&quot;</span>
<span class="s2"># attribution_background_samples: 512</span>
<span class="s2"># attributions_every_sample_factor: 1</span>
<span class="s2">#pretrained_checkpoint: best_models/gene_expression_exformer_unlimited_chrom_and_micro_with_attention_model_117600_perf-average=0.8435.pt </span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets_1kbp_binning.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_1kbp_binning_yaml_contents</span> <span class="o">=</span>  <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets_1kbp_binning.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">enhancer_centric_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;  </span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>

<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/perturbed_landscape_arrays/enhancer_arrays/  #inserted_enhancer_test #inputs/silenced_arrays/silenced_arrays_H2B_S.D </span>
<span class="s2">  #./data/parsed_data/inputs/arrays_train_100bp_no_H3K27ac/ #arrays_train_100bp_no_H3K27ac_uncoupled/</span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/Enhancer_centered_targets.csv </span>
<span class="s2">  #./data/parsed_data/targets/target_arrays_perturbational_inserted_enhancer.csv #target_arrays_perturbational_S.D.csv #target_arrays_1kbp_401_bins_2_conditions_decareads_abs.csv</span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span> <span class="p">}</span>

<span class="n">training_K562_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_K562_train/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/human/manual_validation_ids_chr17_human.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/K562/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/K562/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_K562_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/K562/test/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/K562/test_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">training_prom_CHiC_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_prom_CHiC_pure_conv/  #gene_expression_prediction_no_H3K27ac_uncoupled/     #gene_expression_cnn_1kbp_401_bins_2_cond_reloaded/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/manual_validation_ids_chr17.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 </span>
<span class="s2">sample_interval: 30300 </span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 </span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="sd">&quot;&quot;&quot;input_cnn_microc.yaml&quot;&quot;&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/Promoter-CHiC/training/ </span>
<span class="s2">  input_name: contact_maps</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 #before fc_repr_dim</span>
<span class="s2">    layers: [1,2]</span>
<span class="s2">    kernel_height: 5</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    down_stride_height: 2 #5</span>
<span class="s2">    kernel_width: 5  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 2</span>
<span class="s2">    channel_exp_base: 2  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #128 #256</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/training_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_prom_CHiC_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 </span>
<span class="s2">sample_interval: 30300 </span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 </span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/test/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="sd">&quot;&quot;&quot;input_cnn_microc.yaml&quot;&quot;&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/Promoter-CHiC/test/ </span>
<span class="s2">  input_name: contact_maps</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 #before fc_repr_dim</span>
<span class="s2">    layers: [1,2]</span>
<span class="s2">    kernel_height: 5</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    down_stride_height: 2 #5</span>
<span class="s2">    kernel_width: 5  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 2</span>
<span class="s2">    channel_exp_base: 2  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #128 #256</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/test_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">training_K562_POLR2A_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_folder: ./runs/gene_expression_only_chrom_K562_POLR2A_train/ </span>
<span class="s2">manual_valid_ids_file: ./annotations/human/manual_validation_ids_chr17_human.txt  #manual_validation_ids_chr17_uncoupled.txt</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: false</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/landscape_arrays/K562/training/ </span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/K562/polii/training_polii_targets.csv </span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>

<span class="n">test_K562_POLR2A_enhancer_centric_yaml_contents</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;globals.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">checkpoint_interval: 30300 # 100 epochs #60000</span>
<span class="s2">sample_interval: 30300 #60000</span>
<span class="s2">n_epochs: 120</span>
<span class="s2">batch_size: 64</span>
<span class="s2">optimizer: &quot;adamw&quot;</span>
<span class="s2">lr: 0.0001 #0.0001</span>
<span class="s2">device: &quot;cuda&quot;</span>
<span class="s2">compute_attributions: true</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;input_cnn.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot; </span>
<span class="s2">input_info:</span>
<span class="s2">  input_source: ./inputs/perturbed_landscape_arrays/human_enhancer_arrays/</span>
<span class="s2">  input_name: gene_expression</span>
<span class="s2">  input_type: array</span>

<span class="s2">model_config:</span>
<span class="s2">  model_type: cnn</span>
<span class="s2">  #pre_normalization: &quot;instancenorm&quot;</span>
<span class="s2">  model_init_config:</span>
<span class="s2">    num_output_features: 512 # before fc_repr_dim</span>
<span class="s2">    layers: [4,4]</span>
<span class="s2">    kernel_height: 1</span>
<span class="s2">    down_stride_width: 2</span>
<span class="s2">    first_stride_expansion_width: 1</span>
<span class="s2">    first_kernel_expansion_height: 4 #5</span>
<span class="s2">    kernel_width: 10  #10</span>
<span class="s2">    dilation_factor_width: 2</span>
<span class="s2">    dilation_factor_height: 1</span>
<span class="s2">    channel_exp_base: 5  #3 #-1</span>
<span class="s2">    first_channel_expansion: 2</span>
<span class="s2">    rb_do: .3</span>
<span class="s2">    stochastic_depth_p: .1</span>
<span class="s2">    attention_inclusion_cutoff: 1 #50 #256</span>

<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;fusion.yaml&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">model_config:</span>
<span class="s2">  fc_do: 0.4</span>
<span class="s2">  fc_task_dim: 256</span>
<span class="s2">  layers:</span>
<span class="s2">  - 2</span>
<span class="s2">  rb_do: 0.4</span>
<span class="s2">  stochastic_depth_p: 0.5</span>
<span class="s2">model_type: &quot;mlp-residual&quot;</span>
<span class="s2">&quot;&quot;&quot;</span><span class="p">,</span>
<span class="s2">&quot;outputs_2_cond.yaml&quot;</span><span class="p">:</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">output_info:</span>
<span class="s2">  output_name: expression_output</span>
<span class="s2">  output_source: ./targets/K562/Enhancer_centered_targets_human.csv</span>
<span class="s2">  output_type: tabular</span>

<span class="s2">model_config: # &lt;- new</span>
<span class="s2">  model_type: linear # &lt;- new</span>

<span class="s2">output_type_info:</span>
<span class="s2">  con_loss_name: &quot;SmoothL1Loss&quot;</span>
<span class="s2">  target_con_columns:</span>
<span class="s2">  - &quot;-200_ctrl&quot;</span>
<span class="s2">  - &quot;-199_ctrl&quot;</span>
<span class="s2">  - &quot;-198_ctrl&quot;</span>
<span class="s2">  - &quot;-197_ctrl&quot;</span>
<span class="s2">  - &quot;-196_ctrl&quot;</span>
<span class="s2">  - &quot;-195_ctrl&quot;</span>
<span class="s2">  - &quot;-194_ctrl&quot;</span>
<span class="s2">  - &quot;-193_ctrl&quot;</span>
<span class="s2">  - &quot;-192_ctrl&quot;</span>
<span class="s2">  - &quot;-191_ctrl&quot;</span>
<span class="s2">  - &quot;-190_ctrl&quot;</span>
<span class="s2">  - &quot;-189_ctrl&quot;</span>
<span class="s2">  - &quot;-188_ctrl&quot;</span>
<span class="s2">  - &quot;-187_ctrl&quot;</span>
<span class="s2">  - &quot;-186_ctrl&quot;</span>
<span class="s2">  - &quot;-185_ctrl&quot;</span>
<span class="s2">  - &quot;-184_ctrl&quot;</span>
<span class="s2">  - &quot;-183_ctrl&quot;</span>
<span class="s2">  - &quot;-182_ctrl&quot;</span>
<span class="s2">  - &quot;-181_ctrl&quot;</span>
<span class="s2">  - &quot;-180_ctrl&quot;</span>
<span class="s2">  - &quot;-179_ctrl&quot;</span>
<span class="s2">  - &quot;-178_ctrl&quot;</span>
<span class="s2">  - &quot;-177_ctrl&quot;</span>
<span class="s2">  - &quot;-176_ctrl&quot;</span>
<span class="s2">  - &quot;-175_ctrl&quot;</span>
<span class="s2">  - &quot;-174_ctrl&quot;</span>
<span class="s2">  - &quot;-173_ctrl&quot;</span>
<span class="s2">  - &quot;-172_ctrl&quot;</span>
<span class="s2">  - &quot;-171_ctrl&quot;</span>
<span class="s2">  - &quot;-170_ctrl&quot;</span>
<span class="s2">  - &quot;-169_ctrl&quot;</span>
<span class="s2">  - &quot;-168_ctrl&quot;</span>
<span class="s2">  - &quot;-167_ctrl&quot;</span>
<span class="s2">  - &quot;-166_ctrl&quot;</span>
<span class="s2">  - &quot;-165_ctrl&quot;</span>
<span class="s2">  - &quot;-164_ctrl&quot;</span>
<span class="s2">  - &quot;-163_ctrl&quot;</span>
<span class="s2">  - &quot;-162_ctrl&quot;</span>
<span class="s2">  - &quot;-161_ctrl&quot;</span>
<span class="s2">  - &quot;-160_ctrl&quot;</span>
<span class="s2">  - &quot;-159_ctrl&quot;</span>
<span class="s2">  - &quot;-158_ctrl&quot;</span>
<span class="s2">  - &quot;-157_ctrl&quot;</span>
<span class="s2">  - &quot;-156_ctrl&quot;</span>
<span class="s2">  - &quot;-155_ctrl&quot;</span>
<span class="s2">  - &quot;-154_ctrl&quot;</span>
<span class="s2">  - &quot;-153_ctrl&quot;</span>
<span class="s2">  - &quot;-152_ctrl&quot;</span>
<span class="s2">  - &quot;-151_ctrl&quot;</span>
<span class="s2">  - &quot;-150_ctrl&quot;</span>
<span class="s2">  - &quot;-149_ctrl&quot;</span>
<span class="s2">  - &quot;-148_ctrl&quot;</span>
<span class="s2">  - &quot;-147_ctrl&quot;</span>
<span class="s2">  - &quot;-146_ctrl&quot;</span>
<span class="s2">  - &quot;-145_ctrl&quot;</span>
<span class="s2">  - &quot;-144_ctrl&quot;</span>
<span class="s2">  - &quot;-143_ctrl&quot;</span>
<span class="s2">  - &quot;-142_ctrl&quot;</span>
<span class="s2">  - &quot;-141_ctrl&quot;</span>
<span class="s2">  - &quot;-140_ctrl&quot;</span>
<span class="s2">  - &quot;-139_ctrl&quot;</span>
<span class="s2">  - &quot;-138_ctrl&quot;</span>
<span class="s2">  - &quot;-137_ctrl&quot;</span>
<span class="s2">  - &quot;-136_ctrl&quot;</span>
<span class="s2">  - &quot;-135_ctrl&quot;</span>
<span class="s2">  - &quot;-134_ctrl&quot;</span>
<span class="s2">  - &quot;-133_ctrl&quot;</span>
<span class="s2">  - &quot;-132_ctrl&quot;</span>
<span class="s2">  - &quot;-131_ctrl&quot;</span>
<span class="s2">  - &quot;-130_ctrl&quot;</span>
<span class="s2">  - &quot;-129_ctrl&quot;</span>
<span class="s2">  - &quot;-128_ctrl&quot;</span>
<span class="s2">  - &quot;-127_ctrl&quot;</span>
<span class="s2">  - &quot;-126_ctrl&quot;</span>
<span class="s2">  - &quot;-125_ctrl&quot;</span>
<span class="s2">  - &quot;-124_ctrl&quot;</span>
<span class="s2">  - &quot;-123_ctrl&quot;</span>
<span class="s2">  - &quot;-122_ctrl&quot;</span>
<span class="s2">  - &quot;-121_ctrl&quot;</span>
<span class="s2">  - &quot;-120_ctrl&quot;</span>
<span class="s2">  - &quot;-119_ctrl&quot;</span>
<span class="s2">  - &quot;-118_ctrl&quot;</span>
<span class="s2">  - &quot;-117_ctrl&quot;</span>
<span class="s2">  - &quot;-116_ctrl&quot;</span>
<span class="s2">  - &quot;-115_ctrl&quot;</span>
<span class="s2">  - &quot;-114_ctrl&quot;</span>
<span class="s2">  - &quot;-113_ctrl&quot;</span>
<span class="s2">  - &quot;-112_ctrl&quot;</span>
<span class="s2">  - &quot;-111_ctrl&quot;</span>
<span class="s2">  - &quot;-110_ctrl&quot;</span>
<span class="s2">  - &quot;-109_ctrl&quot;</span>
<span class="s2">  - &quot;-108_ctrl&quot;</span>
<span class="s2">  - &quot;-107_ctrl&quot;</span>
<span class="s2">  - &quot;-106_ctrl&quot;</span>
<span class="s2">  - &quot;-105_ctrl&quot;</span>
<span class="s2">  - &quot;-104_ctrl&quot;</span>
<span class="s2">  - &quot;-103_ctrl&quot;</span>
<span class="s2">  - &quot;-102_ctrl&quot;</span>
<span class="s2">  - &quot;-101_ctrl&quot;</span>
<span class="s2">  - &quot;-100_ctrl&quot;</span>
<span class="s2">  - &quot;-99_ctrl&quot;</span>
<span class="s2">  - &quot;-98_ctrl&quot;</span>
<span class="s2">  - &quot;-97_ctrl&quot;</span>
<span class="s2">  - &quot;-96_ctrl&quot;</span>
<span class="s2">  - &quot;-95_ctrl&quot;</span>
<span class="s2">  - &quot;-94_ctrl&quot;</span>
<span class="s2">  - &quot;-93_ctrl&quot;</span>
<span class="s2">  - &quot;-92_ctrl&quot;</span>
<span class="s2">  - &quot;-91_ctrl&quot;</span>
<span class="s2">  - &quot;-90_ctrl&quot;</span>
<span class="s2">  - &quot;-89_ctrl&quot;</span>
<span class="s2">  - &quot;-88_ctrl&quot;</span>
<span class="s2">  - &quot;-87_ctrl&quot;</span>
<span class="s2">  - &quot;-86_ctrl&quot;</span>
<span class="s2">  - &quot;-85_ctrl&quot;</span>
<span class="s2">  - &quot;-84_ctrl&quot;</span>
<span class="s2">  - &quot;-83_ctrl&quot;</span>
<span class="s2">  - &quot;-82_ctrl&quot;</span>
<span class="s2">  - &quot;-81_ctrl&quot;</span>
<span class="s2">  - &quot;-80_ctrl&quot;</span>
<span class="s2">  - &quot;-79_ctrl&quot;</span>
<span class="s2">  - &quot;-78_ctrl&quot;</span>
<span class="s2">  - &quot;-77_ctrl&quot;</span>
<span class="s2">  - &quot;-76_ctrl&quot;</span>
<span class="s2">  - &quot;-75_ctrl&quot;</span>
<span class="s2">  - &quot;-74_ctrl&quot;</span>
<span class="s2">  - &quot;-73_ctrl&quot;</span>
<span class="s2">  - &quot;-72_ctrl&quot;</span>
<span class="s2">  - &quot;-71_ctrl&quot;</span>
<span class="s2">  - &quot;-70_ctrl&quot;</span>
<span class="s2">  - &quot;-69_ctrl&quot;</span>
<span class="s2">  - &quot;-68_ctrl&quot;</span>
<span class="s2">  - &quot;-67_ctrl&quot;</span>
<span class="s2">  - &quot;-66_ctrl&quot;</span>
<span class="s2">  - &quot;-65_ctrl&quot;</span>
<span class="s2">  - &quot;-64_ctrl&quot;</span>
<span class="s2">  - &quot;-63_ctrl&quot;</span>
<span class="s2">  - &quot;-62_ctrl&quot;</span>
<span class="s2">  - &quot;-61_ctrl&quot;</span>
<span class="s2">  - &quot;-60_ctrl&quot;</span>
<span class="s2">  - &quot;-59_ctrl&quot;</span>
<span class="s2">  - &quot;-58_ctrl&quot;</span>
<span class="s2">  - &quot;-57_ctrl&quot;</span>
<span class="s2">  - &quot;-56_ctrl&quot;</span>
<span class="s2">  - &quot;-55_ctrl&quot;</span>
<span class="s2">  - &quot;-54_ctrl&quot;</span>
<span class="s2">  - &quot;-53_ctrl&quot;</span>
<span class="s2">  - &quot;-52_ctrl&quot;</span>
<span class="s2">  - &quot;-51_ctrl&quot;</span>
<span class="s2">  - &quot;-50_ctrl&quot;</span>
<span class="s2">  - &quot;-49_ctrl&quot;</span>
<span class="s2">  - &quot;-48_ctrl&quot;</span>
<span class="s2">  - &quot;-47_ctrl&quot;</span>
<span class="s2">  - &quot;-46_ctrl&quot;</span>
<span class="s2">  - &quot;-45_ctrl&quot;</span>
<span class="s2">  - &quot;-44_ctrl&quot;</span>
<span class="s2">  - &quot;-43_ctrl&quot;</span>
<span class="s2">  - &quot;-42_ctrl&quot;</span>
<span class="s2">  - &quot;-41_ctrl&quot;</span>
<span class="s2">  - &quot;-40_ctrl&quot;</span>
<span class="s2">  - &quot;-39_ctrl&quot;</span>
<span class="s2">  - &quot;-38_ctrl&quot;</span>
<span class="s2">  - &quot;-37_ctrl&quot;</span>
<span class="s2">  - &quot;-36_ctrl&quot;</span>
<span class="s2">  - &quot;-35_ctrl&quot;</span>
<span class="s2">  - &quot;-34_ctrl&quot;</span>
<span class="s2">  - &quot;-33_ctrl&quot;</span>
<span class="s2">  - &quot;-32_ctrl&quot;</span>
<span class="s2">  - &quot;-31_ctrl&quot;</span>
<span class="s2">  - &quot;-30_ctrl&quot;</span>
<span class="s2">  - &quot;-29_ctrl&quot;</span>
<span class="s2">  - &quot;-28_ctrl&quot;</span>
<span class="s2">  - &quot;-27_ctrl&quot;</span>
<span class="s2">  - &quot;-26_ctrl&quot;</span>
<span class="s2">  - &quot;-25_ctrl&quot;</span>
<span class="s2">  - &quot;-24_ctrl&quot;</span>
<span class="s2">  - &quot;-23_ctrl&quot;</span>
<span class="s2">  - &quot;-22_ctrl&quot;</span>
<span class="s2">  - &quot;-21_ctrl&quot;</span>
<span class="s2">  - &quot;-20_ctrl&quot;</span>
<span class="s2">  - &quot;-19_ctrl&quot;</span>
<span class="s2">  - &quot;-18_ctrl&quot;</span>
<span class="s2">  - &quot;-17_ctrl&quot;</span>
<span class="s2">  - &quot;-16_ctrl&quot;</span>
<span class="s2">  - &quot;-15_ctrl&quot;</span>
<span class="s2">  - &quot;-14_ctrl&quot;</span>
<span class="s2">  - &quot;-13_ctrl&quot;</span>
<span class="s2">  - &quot;-12_ctrl&quot;</span>
<span class="s2">  - &quot;-11_ctrl&quot;</span>
<span class="s2">  - &quot;-10_ctrl&quot;</span>
<span class="s2">  - &quot;-9_ctrl&quot;</span>
<span class="s2">  - &quot;-8_ctrl&quot;</span>
<span class="s2">  - &quot;-7_ctrl&quot;</span>
<span class="s2">  - &quot;-6_ctrl&quot;</span>
<span class="s2">  - &quot;-5_ctrl&quot;</span>
<span class="s2">  - &quot;-4_ctrl&quot;</span>
<span class="s2">  - &quot;-3_ctrl&quot;</span>
<span class="s2">  - &quot;-2_ctrl&quot;</span>
<span class="s2">  - &quot;-1_ctrl&quot;</span>
<span class="s2">  - &quot;0_ctrl&quot;</span>
<span class="s2">  - &quot;1_ctrl&quot;</span>
<span class="s2">  - &quot;2_ctrl&quot;</span>
<span class="s2">  - &quot;3_ctrl&quot;</span>
<span class="s2">  - &quot;4_ctrl&quot;</span>
<span class="s2">  - &quot;5_ctrl&quot;</span>
<span class="s2">  - &quot;6_ctrl&quot;</span>
<span class="s2">  - &quot;7_ctrl&quot;</span>
<span class="s2">  - &quot;8_ctrl&quot;</span>
<span class="s2">  - &quot;9_ctrl&quot;</span>
<span class="s2">  - &quot;10_ctrl&quot;</span>
<span class="s2">  - &quot;11_ctrl&quot;</span>
<span class="s2">  - &quot;12_ctrl&quot;</span>
<span class="s2">  - &quot;13_ctrl&quot;</span>
<span class="s2">  - &quot;14_ctrl&quot;</span>
<span class="s2">  - &quot;15_ctrl&quot;</span>
<span class="s2">  - &quot;16_ctrl&quot;</span>
<span class="s2">  - &quot;17_ctrl&quot;</span>
<span class="s2">  - &quot;18_ctrl&quot;</span>
<span class="s2">  - &quot;19_ctrl&quot;</span>
<span class="s2">  - &quot;20_ctrl&quot;</span>
<span class="s2">  - &quot;21_ctrl&quot;</span>
<span class="s2">  - &quot;22_ctrl&quot;</span>
<span class="s2">  - &quot;23_ctrl&quot;</span>
<span class="s2">  - &quot;24_ctrl&quot;</span>
<span class="s2">  - &quot;25_ctrl&quot;</span>
<span class="s2">  - &quot;26_ctrl&quot;</span>
<span class="s2">  - &quot;27_ctrl&quot;</span>
<span class="s2">  - &quot;28_ctrl&quot;</span>
<span class="s2">  - &quot;29_ctrl&quot;</span>
<span class="s2">  - &quot;30_ctrl&quot;</span>
<span class="s2">  - &quot;31_ctrl&quot;</span>
<span class="s2">  - &quot;32_ctrl&quot;</span>
<span class="s2">  - &quot;33_ctrl&quot;</span>
<span class="s2">  - &quot;34_ctrl&quot;</span>
<span class="s2">  - &quot;35_ctrl&quot;</span>
<span class="s2">  - &quot;36_ctrl&quot;</span>
<span class="s2">  - &quot;37_ctrl&quot;</span>
<span class="s2">  - &quot;38_ctrl&quot;</span>
<span class="s2">  - &quot;39_ctrl&quot;</span>
<span class="s2">  - &quot;40_ctrl&quot;</span>
<span class="s2">  - &quot;41_ctrl&quot;</span>
<span class="s2">  - &quot;42_ctrl&quot;</span>
<span class="s2">  - &quot;43_ctrl&quot;</span>
<span class="s2">  - &quot;44_ctrl&quot;</span>
<span class="s2">  - &quot;45_ctrl&quot;</span>
<span class="s2">  - &quot;46_ctrl&quot;</span>
<span class="s2">  - &quot;47_ctrl&quot;</span>
<span class="s2">  - &quot;48_ctrl&quot;</span>
<span class="s2">  - &quot;49_ctrl&quot;</span>
<span class="s2">  - &quot;50_ctrl&quot;</span>
<span class="s2">  - &quot;51_ctrl&quot;</span>
<span class="s2">  - &quot;52_ctrl&quot;</span>
<span class="s2">  - &quot;53_ctrl&quot;</span>
<span class="s2">  - &quot;54_ctrl&quot;</span>
<span class="s2">  - &quot;55_ctrl&quot;</span>
<span class="s2">  - &quot;56_ctrl&quot;</span>
<span class="s2">  - &quot;57_ctrl&quot;</span>
<span class="s2">  - &quot;58_ctrl&quot;</span>
<span class="s2">  - &quot;59_ctrl&quot;</span>
<span class="s2">  - &quot;60_ctrl&quot;</span>
<span class="s2">  - &quot;61_ctrl&quot;</span>
<span class="s2">  - &quot;62_ctrl&quot;</span>
<span class="s2">  - &quot;63_ctrl&quot;</span>
<span class="s2">  - &quot;64_ctrl&quot;</span>
<span class="s2">  - &quot;65_ctrl&quot;</span>
<span class="s2">  - &quot;66_ctrl&quot;</span>
<span class="s2">  - &quot;67_ctrl&quot;</span>
<span class="s2">  - &quot;68_ctrl&quot;</span>
<span class="s2">  - &quot;69_ctrl&quot;</span>
<span class="s2">  - &quot;70_ctrl&quot;</span>
<span class="s2">  - &quot;71_ctrl&quot;</span>
<span class="s2">  - &quot;72_ctrl&quot;</span>
<span class="s2">  - &quot;73_ctrl&quot;</span>
<span class="s2">  - &quot;74_ctrl&quot;</span>
<span class="s2">  - &quot;75_ctrl&quot;</span>
<span class="s2">  - &quot;76_ctrl&quot;</span>
<span class="s2">  - &quot;77_ctrl&quot;</span>
<span class="s2">  - &quot;78_ctrl&quot;</span>
<span class="s2">  - &quot;79_ctrl&quot;</span>
<span class="s2">  - &quot;80_ctrl&quot;</span>
<span class="s2">  - &quot;81_ctrl&quot;</span>
<span class="s2">  - &quot;82_ctrl&quot;</span>
<span class="s2">  - &quot;83_ctrl&quot;</span>
<span class="s2">  - &quot;84_ctrl&quot;</span>
<span class="s2">  - &quot;85_ctrl&quot;</span>
<span class="s2">  - &quot;86_ctrl&quot;</span>
<span class="s2">  - &quot;87_ctrl&quot;</span>
<span class="s2">  - &quot;88_ctrl&quot;</span>
<span class="s2">  - &quot;89_ctrl&quot;</span>
<span class="s2">  - &quot;90_ctrl&quot;</span>
<span class="s2">  - &quot;91_ctrl&quot;</span>
<span class="s2">  - &quot;92_ctrl&quot;</span>
<span class="s2">  - &quot;93_ctrl&quot;</span>
<span class="s2">  - &quot;94_ctrl&quot;</span>
<span class="s2">  - &quot;95_ctrl&quot;</span>
<span class="s2">  - &quot;96_ctrl&quot;</span>
<span class="s2">  - &quot;97_ctrl&quot;</span>
<span class="s2">  - &quot;98_ctrl&quot;</span>
<span class="s2">  - &quot;99_ctrl&quot;</span>
<span class="s2">  - &quot;100_ctrl&quot;</span>
<span class="s2">  - &quot;101_ctrl&quot;</span>
<span class="s2">  - &quot;102_ctrl&quot;</span>
<span class="s2">  - &quot;103_ctrl&quot;</span>
<span class="s2">  - &quot;104_ctrl&quot;</span>
<span class="s2">  - &quot;105_ctrl&quot;</span>
<span class="s2">  - &quot;106_ctrl&quot;</span>
<span class="s2">  - &quot;107_ctrl&quot;</span>
<span class="s2">  - &quot;108_ctrl&quot;</span>
<span class="s2">  - &quot;109_ctrl&quot;</span>
<span class="s2">  - &quot;110_ctrl&quot;</span>
<span class="s2">  - &quot;111_ctrl&quot;</span>
<span class="s2">  - &quot;112_ctrl&quot;</span>
<span class="s2">  - &quot;113_ctrl&quot;</span>
<span class="s2">  - &quot;114_ctrl&quot;</span>
<span class="s2">  - &quot;115_ctrl&quot;</span>
<span class="s2">  - &quot;116_ctrl&quot;</span>
<span class="s2">  - &quot;117_ctrl&quot;</span>
<span class="s2">  - &quot;118_ctrl&quot;</span>
<span class="s2">  - &quot;119_ctrl&quot;</span>
<span class="s2">  - &quot;120_ctrl&quot;</span>
<span class="s2">  - &quot;121_ctrl&quot;</span>
<span class="s2">  - &quot;122_ctrl&quot;</span>
<span class="s2">  - &quot;123_ctrl&quot;</span>
<span class="s2">  - &quot;124_ctrl&quot;</span>
<span class="s2">  - &quot;125_ctrl&quot;</span>
<span class="s2">  - &quot;126_ctrl&quot;</span>
<span class="s2">  - &quot;127_ctrl&quot;</span>
<span class="s2">  - &quot;128_ctrl&quot;</span>
<span class="s2">  - &quot;129_ctrl&quot;</span>
<span class="s2">  - &quot;130_ctrl&quot;</span>
<span class="s2">  - &quot;131_ctrl&quot;</span>
<span class="s2">  - &quot;132_ctrl&quot;</span>
<span class="s2">  - &quot;133_ctrl&quot;</span>
<span class="s2">  - &quot;134_ctrl&quot;</span>
<span class="s2">  - &quot;135_ctrl&quot;</span>
<span class="s2">  - &quot;136_ctrl&quot;</span>
<span class="s2">  - &quot;137_ctrl&quot;</span>
<span class="s2">  - &quot;138_ctrl&quot;</span>
<span class="s2">  - &quot;139_ctrl&quot;</span>
<span class="s2">  - &quot;140_ctrl&quot;</span>
<span class="s2">  - &quot;141_ctrl&quot;</span>
<span class="s2">  - &quot;142_ctrl&quot;</span>
<span class="s2">  - &quot;143_ctrl&quot;</span>
<span class="s2">  - &quot;144_ctrl&quot;</span>
<span class="s2">  - &quot;145_ctrl&quot;</span>
<span class="s2">  - &quot;146_ctrl&quot;</span>
<span class="s2">  - &quot;147_ctrl&quot;</span>
<span class="s2">  - &quot;148_ctrl&quot;</span>
<span class="s2">  - &quot;149_ctrl&quot;</span>
<span class="s2">  - &quot;150_ctrl&quot;</span>
<span class="s2">  - &quot;151_ctrl&quot;</span>
<span class="s2">  - &quot;152_ctrl&quot;</span>
<span class="s2">  - &quot;153_ctrl&quot;</span>
<span class="s2">  - &quot;154_ctrl&quot;</span>
<span class="s2">  - &quot;155_ctrl&quot;</span>
<span class="s2">  - &quot;156_ctrl&quot;</span>
<span class="s2">  - &quot;157_ctrl&quot;</span>
<span class="s2">  - &quot;158_ctrl&quot;</span>
<span class="s2">  - &quot;159_ctrl&quot;</span>
<span class="s2">  - &quot;160_ctrl&quot;</span>
<span class="s2">  - &quot;161_ctrl&quot;</span>
<span class="s2">  - &quot;162_ctrl&quot;</span>
<span class="s2">  - &quot;163_ctrl&quot;</span>
<span class="s2">  - &quot;164_ctrl&quot;</span>
<span class="s2">  - &quot;165_ctrl&quot;</span>
<span class="s2">  - &quot;166_ctrl&quot;</span>
<span class="s2">  - &quot;167_ctrl&quot;</span>
<span class="s2">  - &quot;168_ctrl&quot;</span>
<span class="s2">  - &quot;169_ctrl&quot;</span>
<span class="s2">  - &quot;170_ctrl&quot;</span>
<span class="s2">  - &quot;171_ctrl&quot;</span>
<span class="s2">  - &quot;172_ctrl&quot;</span>
<span class="s2">  - &quot;173_ctrl&quot;</span>
<span class="s2">  - &quot;174_ctrl&quot;</span>
<span class="s2">  - &quot;175_ctrl&quot;</span>
<span class="s2">  - &quot;176_ctrl&quot;</span>
<span class="s2">  - &quot;177_ctrl&quot;</span>
<span class="s2">  - &quot;178_ctrl&quot;</span>
<span class="s2">  - &quot;179_ctrl&quot;</span>
<span class="s2">  - &quot;180_ctrl&quot;</span>
<span class="s2">  - &quot;181_ctrl&quot;</span>
<span class="s2">  - &quot;182_ctrl&quot;</span>
<span class="s2">  - &quot;183_ctrl&quot;</span>
<span class="s2">  - &quot;184_ctrl&quot;</span>
<span class="s2">  - &quot;185_ctrl&quot;</span>
<span class="s2">  - &quot;186_ctrl&quot;</span>
<span class="s2">  - &quot;187_ctrl&quot;</span>
<span class="s2">  - &quot;188_ctrl&quot;</span>
<span class="s2">  - &quot;189_ctrl&quot;</span>
<span class="s2">  - &quot;190_ctrl&quot;</span>
<span class="s2">  - &quot;191_ctrl&quot;</span>
<span class="s2">  - &quot;192_ctrl&quot;</span>
<span class="s2">  - &quot;193_ctrl&quot;</span>
<span class="s2">  - &quot;194_ctrl&quot;</span>
<span class="s2">  - &quot;195_ctrl&quot;</span>
<span class="s2">  - &quot;196_ctrl&quot;</span>
<span class="s2">  - &quot;197_ctrl&quot;</span>
<span class="s2">  - &quot;198_ctrl&quot;</span>
<span class="s2">  - &quot;199_ctrl&quot;</span>
<span class="s2">  - &quot;200_ctrl&quot;</span>
<span class="s2">  &quot;&quot;&quot;</span>
<span class="p">}</span>


<span class="c1"># Predict promoter perturbations: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">predict_prom_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>
  
<span class="c1"># Training shorter context model: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_20kbp_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Training only chrom pure conv with softplus activation SmoothL1 loss: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_softplus_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Training only chrom pure conv with softplus activation and Poisson NLL: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_softplus_poisson_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Testing shorter context model: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_20kbp_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Training with chr19 holdout: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_chr19_holdout_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Training with chr19 holdout: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_chr19_holdout_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Training with chr19 holdout: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_no_H3K27ac_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Test with chr19 holdout: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_no_H3K27ac_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="c1"># Test with chr19 holdout: configs</span>
<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">predict_20kbp_context_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_1kbp_binning_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_1kbp_binning_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">enhancer_centric_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_K562_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_K562_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">14</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_prom_CHiC_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">15</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_prom_CHiC_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">16</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">training_K562_POLR2A_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">17</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>

<span class="k">for</span> <span class="n">file</span><span class="p">,</span><span class="n">content</span> <span class="ow">in</span> <span class="n">test_K562_POLR2A_enhancer_centric_yaml_contents</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
  <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_paths</span><span class="p">[</span><span class="mi">18</span><span class="p">]</span> <span class="o">/</span> <span class="n">file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
      <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">content</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="iii-enhancer-centric-perturbational-analysis">
<h2>III) Enhancer-centric perturbational analysis<a class="headerlink" href="#iii-enhancer-centric-perturbational-analysis" title="Link to this heading">#</a></h2>
<p>Here we will follow an enhancer-centric approach to identify the genes that feel the perturbations the most.</p>
<blockquote>
<div><p><em>Pipeline:</em></p>
<ul class="simple">
<li><p>Create new enhancer-centered input samples:</p>
<ul>
<li><p>Inputs centered at active enhancers (H3K27ac &gt; 10) in chr 19 (alternative test chr) for mESCs.</p></li>
</ul>
</li>
<li><p>Replace these for a silenced chromatin state with no accessibility, H3K4me3 and H3K27ac.</p></li>
<li><p>Predict EU-seq for baseline and perturbed inputs.</p></li>
<li><p>Measure output changes:</p>
<ul>
<li><p>For each output position / bin.</p></li>
<li><p>Inside gene boundaries:</p>
<ul>
<li><p>Identify genes inside predicted boundaries and integrate predicted EU-seq within gene limits.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div></blockquote>
<p>DATA ANALYSIS PLOTS:</p>
<ul class="simple">
<li><p>Distance relative effect.</p></li>
<li><p>Integrated EU-seq change (relative) vs. distance plot.</p></li>
<li><p>Enhancer width distribution.</p></li>
<li><p>Ranking of gene index of the most affected gene per enhancer.</p></li>
</ul>
<p><strong>Get enhancer-centered baseline &amp; perturbed inputs:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_enhancer_centered_inputs.py
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pyBigWig</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                      <span class="n">input_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">n_input_tracks</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
                      <span class="n">n_input_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">bw_input_track_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
                      <span class="n">enhancer_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">center</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates input arrays centered at enhancer midpoint</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bw_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bw_input_track_list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">center</span><span class="o">-</span><span class="n">input_shift</span><span class="p">,</span> <span class="n">center</span><span class="o">+</span><span class="n">input_shift</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_input_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>  <span class="c1"># ReLU</span>
            <span class="n">input_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">stats</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2"> input landscape coordinates are out of bounds.&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">input_sample</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_perturbed_array</span><span class="p">(</span><span class="n">input_array</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates perturbed version with all marks except H3K27me3 silenced within enhancer boundaries</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">perturbed</span> <span class="o">=</span> <span class="n">input_array</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># Silence ATAC-seq, H3K4me3, and H3K27ac (indices 0, 1, 2) only within enhancer boundaries</span>
    <span class="n">perturbed</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">start_idx</span><span class="p">:</span><span class="n">end_idx</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">return</span> <span class="n">perturbed</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_enhancer</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                    <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">row_tuple</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">activity_threshold</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process a single enhancer and create its input arrays if enhancer is active</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">enhancer_id</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">row_tuple</span>  <span class="c1"># Now unpacking from tuple</span>
    <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
    <span class="n">center</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="n">end</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>

    <span class="c1"># Create regular input array</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                                    <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">enhancer_id</span><span class="p">,</span> <span class="n">center</span><span class="p">,</span> <span class="n">chrom</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">input_sample</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">INPUT_SHAPE</span><span class="p">:</span>
        <span class="c1"># Calculate enhancer boundaries in array coordinates</span>
        <span class="n">central_idx</span> <span class="o">=</span> <span class="n">n_input_bins</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="n">start_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="p">(</span><span class="n">start</span> <span class="o">-</span> <span class="n">center</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
        <span class="n">end_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">center</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>

        <span class="c1"># Check H3K27ac activity within enhancer boundaries</span>
        <span class="n">h3k27ac_idx</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># Index for H3K27ac track</span>
        <span class="n">enhancer_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">input_sample</span><span class="p">[</span><span class="n">h3k27ac_idx</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">:</span><span class="n">end_idx</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">enhancer_mean</span> <span class="o">&gt;</span> <span class="n">activity_threshold</span><span class="p">:</span>
            <span class="c1"># Save regular array</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;enhancer_arrays&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">input_sample</span><span class="p">)</span>
            
            <span class="c1"># Create and save perturbed array</span>
            <span class="n">perturbed_sample</span> <span class="o">=</span> <span class="n">create_perturbed_array</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">,</span> <span class="n">end_idx</span><span class="p">)</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_enhancer_arrays&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_perturbed.npy&quot;</span><span class="p">,</span> <span class="n">perturbed_sample</span><span class="p">)</span>
            
            <span class="k">return</span> <span class="p">(</span><span class="n">enhancer_id</span><span class="p">,</span> <span class="n">enhancer_id</span> <span class="o">+</span> <span class="s2">&quot;_perturbed&quot;</span><span class="p">)</span>  <span class="c1"># Return both IDs</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: Enhancer activity below threshold (</span><span class="si">{</span><span class="n">enhancer_mean</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> ≤ </span><span class="si">{</span><span class="n">activity_threshold</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: Input did not match the expected shape.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="c1"># Setup paths</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files&quot;</span>

    <span class="c1"># Create output directories</span>
    <span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;enhancer_arrays&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_enhancer_arrays&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Start logger</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;/Enhancer_data_creation.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="c1"># Load enhancer annotations</span>
    <span class="n">enhancer_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_Enhancer_annotation.tsv&quot;</span>
    <span class="n">enhancer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">enhancer_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Filter for chr19 (alternative test)</span>
    <span class="n">chr4_enhancers</span> <span class="o">=</span> <span class="n">enhancer_df</span><span class="p">[</span><span class="n">enhancer_df</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;chr19&#39;</span><span class="p">]</span> <span class="c1"># Rather small chromosome, 8869 enhancer-like signatures including CTCF bound, 1491 pass the activity filter.</span>

    <span class="c1"># Setup parameters</span>
    <span class="n">input_shift</span> <span class="o">=</span> <span class="mi">500050</span>
    <span class="n">n_input_bins</span> <span class="o">=</span> <span class="mi">10001</span>
    <span class="n">n_input_tracks</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">INPUT_SHAPE</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">10001</span><span class="p">)</span>
    <span class="n">bw_input_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC_Seq.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3.bw&quot;</span><span class="p">]</span>

    <span class="c1"># Process enhancers in parallel</span>
    <span class="n">process_args</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                   <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
    
    <span class="n">activity_threshold</span> <span class="o">=</span> <span class="mi">10</span>

    <span class="c1"># Convert DataFrame rows to tuples for multiprocessing</span>
    <span class="n">row_tuples</span> <span class="o">=</span> <span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">chr4_enhancers</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>

    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">process_enhancer</span><span class="p">,</span> 
                    <span class="p">[(</span><span class="n">process_args</span> <span class="o">+</span> <span class="p">(</span><span class="n">row_tuple</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="n">activity_threshold</span><span class="p">))</span> <span class="k">for</span> <span class="n">row_tuple</span> <span class="ow">in</span> <span class="n">row_tuples</span><span class="p">])</span>
        
        <span class="c1"># Count how many enhancers were processed successfully</span>
        <span class="n">processed_count</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span> <span class="k">if</span> <span class="n">result</span><span class="p">)</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="n">processed_count</span><span class="si">}</span><span class="s2"> active enhancers out of </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">chr4_enhancers</span><span class="p">)</span><span class="si">}</span><span class="s2"> total enhancers in chr4&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Create dummy targets for the enhancer-centered predictions:</strong></p>
<p>For eir to run we need to provide a dummy target file, although it will not be used. Hence we create a target file with target IDs but full of zeros.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">get_file_ids</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get all file IDs from a directory without the .npy extension&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.npy&#39;</span><span class="p">)]</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_unified_target_file</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">,</span> <span class="n">perturbed_ids</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a unified target file with zeros for both regular and perturbed IDs&quot;&quot;&quot;</span>
    <span class="c1"># Create column names</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">_ctrl&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">n_bins</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
    
    <span class="c1"># Combine all IDs</span>
    <span class="n">all_ids</span> <span class="o">=</span> <span class="n">baseline_ids</span> <span class="o">+</span> <span class="n">perturbed_ids</span>
    
    <span class="c1"># Create DataFrame with zeros</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_ids</span><span class="p">)),</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
    
    <span class="c1"># Fill ID column</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_ids</span>
    
    <span class="c1"># Save to CSV</span>
    <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">all_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> entries&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- Regular samples: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;- Perturbed samples: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">perturbed_ids</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Setup paths</span>
<span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
<span class="n">baseline_dir</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;enhancer_arrays&quot;</span>
<span class="n">perturbed_dir</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_enhancer_arrays&quot;</span>
<span class="n">output_dir</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span>

<span class="c1"># Create output directory if it doesn&#39;t exist</span>
<span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Get IDs from both directories</span>
<span class="n">baseline_ids</span> <span class="o">=</span> <span class="n">get_file_ids</span><span class="p">(</span><span class="n">baseline_dir</span><span class="p">)</span>
<span class="n">perturbed_ids</span> <span class="o">=</span> <span class="n">get_file_ids</span><span class="p">(</span><span class="n">perturbed_dir</span><span class="p">)</span>

<span class="c1"># Create unified target file</span>
<span class="n">create_unified_target_file</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">,</span> <span class="n">perturbed_ids</span><span class="p">,</span>
                            <span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;Enhancer_centered_targets.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Move perturbed inputs to the same folder to predict all at once.</span>
<span class="o">!</span><span class="w"> </span>mv<span class="w"> </span>../inputs/perturbed_landscape_arrays/perturbed_enhancer_arrays/*<span class="w"> </span>../inputs/perturbed_landscape_arrays/enhancer_arrays/
<span class="o">!</span><span class="w"> </span>rm<span class="w"> </span>-r<span class="w"> </span>../inputs/perturbed_landscape_arrays/perturbed_enhancer_arrays/
</pre></div>
</div>
</div>
</div>
<p><strong>Predict enhancer-centered samples: baseline and perturbed</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_enhancer_centric/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_enhancer_centric/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_enhancer_centric/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_enhancer_centric/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv_chr19_holdout/saved_models/gene_expression_only_chrom_pure_conv_chr19_holdout_model_30300_perf-average<span class="o">=</span><span class="m">0</span>.6487.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/perturbation_runs/gene_expression_only_chrom_pure_conv_chr19_enhancer_centric
</pre></div>
</div>
<p><strong>Quantify differences between baseline and perturbed predictions and store them in a table:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.integrate</span><span class="w"> </span><span class="kn">import</span> <span class="n">simps</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_annotations</span><span class="p">(</span><span class="n">enhancer_path</span><span class="p">,</span> <span class="n">gene_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load and return enhancer and gene annotations.&quot;&quot;&quot;</span>
    <span class="n">enhancers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">enhancer_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">enhancers_df</span><span class="p">,</span> <span class="n">genes_df</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_window_boundaries</span><span class="p">(</span><span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">200500</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate window boundaries given enhancer center.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">enhancer_center</span> <span class="o">-</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">enhancer_center</span> <span class="o">+</span> <span class="n">window_size</span>

<span class="k">def</span><span class="w"> </span><span class="nf">find_genes_in_window</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find genes fully contained within the window.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">genes_df</span><span class="p">[</span>
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> 
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">window_start</span><span class="p">)</span> <span class="o">&amp;</span> 
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">window_end</span><span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_gene_area</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate area under the curve for a gene using Simpson&#39;s rule.&quot;&quot;&quot;</span>
    <span class="n">gene_values</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="n">gene_start_bin</span><span class="p">:</span><span class="n">gene_end_bin</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_values</span><span class="p">))</span>
    <span class="n">area</span> <span class="o">=</span> <span class="n">simps</span><span class="p">(</span><span class="n">gene_values</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
    <span class="n">length</span> <span class="o">=</span> <span class="n">gene_end_bin</span> <span class="o">-</span> <span class="n">gene_start_bin</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">area</span> <span class="o">/</span> <span class="n">length</span> <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_gene_tss</span><span class="p">(</span><span class="n">gene_row</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get TSS position based on strand.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_distance_and_order</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate distances to enhancer and determine gene order.&quot;&quot;&quot;</span>
    <span class="c1"># Calculate TSS for each gene and distance to enhancer</span>
    <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">tss</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">enhancer_center</span> <span class="o">-</span> <span class="n">tss</span><span class="p">)</span>
        <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span>
            <span class="s1">&#39;distance&#39;</span><span class="p">:</span> <span class="n">distance</span><span class="p">,</span>
            <span class="s1">&#39;tss&#39;</span><span class="p">:</span> <span class="n">tss</span>
        <span class="p">})</span>
    
    <span class="c1"># Sort genes by distance</span>
    <span class="n">sorted_distances</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">distances</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">])</span>
    
    <span class="c1"># Create distance mapping</span>
    <span class="n">distance_map</span> <span class="o">=</span> <span class="p">{</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]:</span> <span class="p">{</span>
        <span class="s1">&#39;distance&#39;</span><span class="p">:</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span>
        <span class="s1">&#39;order&#39;</span><span class="p">:</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
        <span class="s1">&#39;tss&#39;</span><span class="p">:</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;tss&#39;</span><span class="p">]</span>
    <span class="p">}</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_distances</span><span class="p">)}</span>
    
    <span class="k">return</span> <span class="n">distance_map</span>

<span class="k">def</span><span class="w"> </span><span class="nf">check_gene_adjacency</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">distance_map</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a gene is adjacent to the enhancer (no genes between them).&quot;&quot;&quot;</span>
    <span class="n">gene_tss</span> <span class="o">=</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene_id</span><span class="p">][</span><span class="s1">&#39;tss&#39;</span><span class="p">]</span>
    <span class="n">gene_distance</span> <span class="o">=</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene_id</span><span class="p">][</span><span class="s1">&#39;distance&#39;</span><span class="p">]</span>
    
    <span class="c1"># Get the gene&#39;s info</span>
    <span class="n">gene_info</span> <span class="o">=</span> <span class="n">window_genes</span><span class="p">[</span><span class="n">window_genes</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene_id</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    
    <span class="c1"># Check for any genes between this gene and the enhancer</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">other_gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">other_gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene_id</span><span class="p">:</span>
            <span class="k">continue</span>
            
        <span class="n">other_tss</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">other_gene</span><span class="p">)</span>
        <span class="n">other_distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">other_tss</span> <span class="o">-</span> <span class="n">gene_tss</span><span class="p">)</span>
        
        <span class="c1"># If there&#39;s a gene with smaller distance, this gene is not adjacent</span>
        <span class="k">if</span> <span class="n">other_distance</span> <span class="o">&lt;</span> <span class="n">gene_distance</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
            
    <span class="k">return</span> <span class="kc">True</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_enhancer_predictions</span><span class="p">(</span><span class="n">enhancer_row</span><span class="p">,</span> <span class="n">genes_df</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">200500</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process predictions for one enhancer and its genes.&quot;&quot;&quot;</span>
    <span class="n">enhancer_id</span> <span class="o">=</span> <span class="n">enhancer_row</span><span class="p">[</span><span class="s1">&#39;cCRE_accession&#39;</span><span class="p">]</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">enhancer_row</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="n">enhancer_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2"> not found in predictions&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
    
    <span class="n">enhancer_center</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">enhancer_row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">enhancer_row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]))</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span> <span class="o">=</span> <span class="n">get_window_boundaries</span><span class="p">(</span><span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>
    
    <span class="n">window_genes</span> <span class="o">=</span> <span class="n">find_genes_in_window</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span><span class="p">)</span>
    
    <span class="c1"># Calculate distances and gene order</span>
    <span class="n">distance_map</span> <span class="o">=</span> <span class="n">calculate_distance_and_order</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">)</span>
    
    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="n">baseline_preds</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enhancer_id</span><span class="p">]</span>
        <span class="n">perturbed_preds</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_perturbed&quot;</span><span class="p">]</span>
        
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">gene_start_rel</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">enhancer_center</span>
            <span class="n">gene_end_rel</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">enhancer_center</span>
            
            <span class="n">gene_start_bin</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene_start_rel</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
            <span class="n">gene_end_bin</span> <span class="o">=</span> <span class="p">(</span><span class="n">gene_end_rel</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
            
            <span class="n">baseline_values</span> <span class="o">=</span> <span class="n">baseline_preds</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
            <span class="n">perturbed_values</span> <span class="o">=</span> <span class="n">perturbed_preds</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
            
            <span class="n">baseline_area</span> <span class="o">=</span> <span class="n">calculate_gene_area</span><span class="p">(</span><span class="n">baseline_values</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="p">)</span>
            <span class="n">perturbed_area</span> <span class="o">=</span> <span class="n">calculate_gene_area</span><span class="p">(</span><span class="n">perturbed_values</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="p">)</span>
            
            <span class="c1"># Check if gene is adjacent</span>
            <span class="n">is_adjacent</span> <span class="o">=</span> <span class="n">check_gene_adjacency</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">distance_map</span><span class="p">,</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">])</span>
            
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s1">&#39;baseline_area&#39;</span><span class="p">:</span> <span class="n">baseline_area</span><span class="p">,</span>
                <span class="s1">&#39;perturbed_area&#39;</span><span class="p">:</span> <span class="n">perturbed_area</span><span class="p">,</span>
                <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_length&#39;</span><span class="p">:</span> <span class="n">gene_end_bin</span> <span class="o">-</span> <span class="n">gene_start_bin</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">:</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]][</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_order&#39;</span><span class="p">:</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]][</span><span class="s1">&#39;order&#39;</span><span class="p">],</span>
                <span class="s1">&#39;is_adjacent&#39;</span><span class="p">:</span> <span class="n">is_adjacent</span>
            <span class="p">})</span>
    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing enhancer </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unexpected error processing enhancer </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span>

<span class="c1">########### Main script ############################</span>
<span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
<span class="n">enhancer_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_Enhancer_annotation.tsv&quot;</span>
<span class="n">gene_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations.tsv&quot;</span>
<span class="n">results_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbation_runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;gene_expression_only_chrom_pure_conv_chr19_enhancer_centric&quot;</span> 
<span class="n">output_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;tables&quot;</span> 
<span class="n">output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">enhancers_df</span><span class="p">,</span> <span class="n">genes_df</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">enhancer_path</span><span class="p">,</span> <span class="n">gene_path</span><span class="p">)</span>
<span class="n">_</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">results_path</span><span class="p">,</span>
                                        <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
                                        <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span>
                                        <span class="n">is_training</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of enhancers in annotation:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">enhancers_df</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of predictions:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">predictions_df</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample of prediction indices:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">predictions_df</span><span class="o">.</span><span class="n">index</span><span class="p">)[:</span><span class="mi">5</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sample of enhancer IDs:&quot;</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">enhancers_df</span><span class="p">[</span><span class="s1">&#39;cCRE_accession&#39;</span><span class="p">])[:</span><span class="mi">5</span><span class="p">])</span>

<span class="n">all_results</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">enhancer</span> <span class="ow">in</span> <span class="n">enhancers_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">process_enhancer_predictions</span><span class="p">(</span><span class="n">enhancer</span><span class="p">,</span> <span class="n">genes_df</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">results</span><span class="p">:</span>
        <span class="n">all_results</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>

<span class="k">if</span> <span class="n">all_results</span><span class="p">:</span>
    <span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">all_results</span><span class="p">)</span>
    <span class="n">results_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;Predicted_gene_areas_in_enhancer_centric_approach.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Processing complete:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancers_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancers&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">results_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Results saved to </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">else</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">No results were generated. Please check the error messages above.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Visualize enhancer results:</strong></p>
<ul class="simple">
<li><p>Distance dependent distribution of absolute differences between perturbed and baseline.</p></li>
<li><p>Gene area absolute differences as a function of distance.</p></li>
<li><p>Enhancer width distribution.</p></li>
<li><p>Histogram of most affected gene per enhancer.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>

<span class="c1"># Set global font size</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;font.size&#39;</span><span class="p">:</span> <span class="mi">12</span><span class="p">})</span>

<span class="k">def</span><span class="w"> </span><span class="nf">analyze_enhancer_perturbations</span><span class="p">():</span>
    <span class="c1"># Create save directory</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/Enhancer_perturbations/&quot;</span><span class="p">)</span>
    <span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Read enhancer annotations</span>
    <span class="n">enhancer_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_Enhancer_annotation.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">enhancer_df</span><span class="p">[</span><span class="s1">&#39;width&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enhancer_df</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">enhancer_df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span>

    <span class="c1"># Read the predictions</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span>
        <span class="n">results_path</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_only_chrom_pure_conv_chr19_enhancer_centric/&quot;</span><span class="p">),</span> 
        <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
        <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span>
        <span class="n">is_training</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>

    <span class="c1"># Calculate absolute differences and get enhancer widths</span>
    <span class="n">abs_diff_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">enhancer_widths</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">enhancer_id</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="o">~</span><span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;perturbed&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="c1"># Get regular and perturbed predictions</span>
        <span class="n">regular</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="n">enhancer_id</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">perturbed</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">index</span> <span class="o">==</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_perturbed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        
        <span class="c1"># Get enhancer width</span>
        <span class="n">base_id</span> <span class="o">=</span> <span class="n">enhancer_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_forward&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;_forward&#39;</span> <span class="ow">in</span> <span class="n">enhancer_id</span> <span class="k">else</span> <span class="n">enhancer_id</span>
        <span class="n">width</span> <span class="o">=</span> <span class="n">enhancer_df</span><span class="p">[</span><span class="n">enhancer_df</span><span class="p">[</span><span class="s1">&#39;cCRE_accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">base_id</span><span class="p">][</span><span class="s1">&#39;width&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">enhancer_widths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">width</span><span class="p">)</span>
        
        <span class="c1"># Calculate differences for each position</span>
        <span class="k">for</span> <span class="n">pos</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_ctrl&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">))</span>
            <span class="n">abs_diff</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">regular</span><span class="p">[</span><span class="n">pos</span><span class="p">]</span> <span class="o">-</span> <span class="n">perturbed</span><span class="p">[</span><span class="n">pos</span><span class="p">])</span>
            <span class="n">abs_diff_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;Position&#39;</span><span class="p">:</span> <span class="n">position</span><span class="p">,</span>
                <span class="s1">&#39;Absolute_Difference&#39;</span><span class="p">:</span> <span class="n">abs_diff</span>
            <span class="p">})</span>

    <span class="n">abs_diff_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">abs_diff_data</span><span class="p">)</span>

    <span class="c1"># Calculate statistics for each position</span>
    <span class="n">stats_df</span> <span class="o">=</span> <span class="n">abs_diff_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Position&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
        <span class="s1">&#39;Absolute_Difference&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> 
                               <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">25</span><span class="p">),</span>
                               <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">75</span><span class="p">)]</span>
    <span class="p">})</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">stats_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Position&#39;</span><span class="p">,</span> <span class="s1">&#39;mean&#39;</span><span class="p">,</span> <span class="s1">&#39;std&#39;</span><span class="p">,</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span> <span class="s1">&#39;max&#39;</span><span class="p">,</span> <span class="s1">&#39;q1&#39;</span><span class="p">,</span> <span class="s1">&#39;q3&#39;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">stats_df</span><span class="p">,</span> <span class="n">enhancer_widths</span>

<span class="k">def</span><span class="w"> </span><span class="nf">analyze_gene_areas</span><span class="p">():</span>
    <span class="c1"># Read the data</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../tables/Predicted_gene_areas_in_enhancer_centric_approach.csv&#39;</span><span class="p">))</span>

    <span class="c1"># Calculate area difference</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;area_difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;perturbed_area&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;baseline_area&#39;</span><span class="p">])</span>

    <span class="c1"># Calculate rolling statistics</span>
    <span class="n">sorted_df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">window_size</span> <span class="o">=</span> <span class="mi">5000</span>  <span class="c1"># 5kbp</span>

    <span class="c1"># Initialize lists for rolling statistics</span>
    <span class="n">x_vals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">means</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">stds</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Calculate rolling statistics manually</span>
    <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">sorted_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">lower_bound</span> <span class="o">=</span> <span class="n">center</span> <span class="o">-</span> <span class="n">window_size</span><span class="o">/</span><span class="mi">2</span>
        <span class="n">upper_bound</span> <span class="o">=</span> <span class="n">center</span> <span class="o">+</span> <span class="n">window_size</span><span class="o">/</span><span class="mi">2</span>
        
        <span class="n">window_mask</span> <span class="o">=</span> <span class="p">(</span><span class="n">sorted_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">lower_bound</span><span class="p">)</span> <span class="o">&amp;</span> \
                     <span class="p">(</span><span class="n">sorted_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">upper_bound</span><span class="p">)</span>
        <span class="n">window_values</span> <span class="o">=</span> <span class="n">sorted_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">window_mask</span><span class="p">,</span> <span class="s1">&#39;area_difference&#39;</span><span class="p">]</span>
        
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">window_values</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">x_vals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">center</span><span class="o">/</span><span class="mi">1000</span><span class="p">)</span>  <span class="c1"># Convert to kbp</span>
            <span class="n">means</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">window_values</span><span class="o">.</span><span class="n">mean</span><span class="p">())</span>
            <span class="n">stds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">window_values</span><span class="o">.</span><span class="n">std</span><span class="p">())</span>

    <span class="c1"># Get maximum effect genes</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_enhancer_id</span><span class="p">(</span><span class="n">row_id</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">row_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_ENSMUSG&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">get_enhancer_id</span><span class="p">)</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;area_difference&#39;</span><span class="p">])</span>
    <span class="n">max_diff_genes</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">)[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()]</span>

    <span class="k">return</span> <span class="n">df</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x_vals</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">means</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stds</span><span class="p">),</span> <span class="n">max_diff_genes</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_combined_figure</span><span class="p">():</span>
    <span class="c1"># Get data from both analyses</span>
    <span class="n">stats_df</span><span class="p">,</span> <span class="n">enhancer_widths</span> <span class="o">=</span> <span class="n">analyze_enhancer_perturbations</span><span class="p">()</span>
    <span class="n">df</span><span class="p">,</span> <span class="n">x_vals</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">stds</span><span class="p">,</span> <span class="n">max_diff_genes</span> <span class="o">=</span> <span class="n">analyze_gene_areas</span><span class="p">()</span>

    <span class="c1"># Create figure with custom layout</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    
    <span class="c1"># Create GridSpec with custom ratios</span>
    <span class="c1"># First row is for the absolute differences plot (full width)</span>
    <span class="c1"># Second row has three plots with ratios 1:2:3</span>
    <span class="n">gs</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_gridspec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    
    <span class="c1"># Top row: Absolute differences plot (spans all columns)</span>
    <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:])</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;Position&#39;</span><span class="p">],</span> <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">],</span> 
             <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Mean&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;Position&#39;</span><span class="p">],</span> 
                     <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">],</span>
                     <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;mean&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;std&#39;</span><span class="p">],</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Standard deviation&#39;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;Position&#39;</span><span class="p">],</span> <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;min&#39;</span><span class="p">],</span> 
             <span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Max&#39;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;Position&#39;</span><span class="p">],</span> <span class="n">stats_df</span><span class="p">[</span><span class="s1">&#39;max&#39;</span><span class="p">],</span> 
             <span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Distance to the center of the silenced enhancer (kbp)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span> <span class="mi">200</span><span class="p">))</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mi">55</span><span class="p">))</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="c1"># Bottom row, first plot (1/6 width): Enhancer width distribution</span>
    <span class="n">ax2</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="n">enhancer_widths</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">,</span> 
                <span class="n">showfliers</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;lightgray&#39;</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Enhancer Width (bp)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>

    <span class="c1"># Bottom row, third plot (1/3 width): Gene Order Histogram (limited to order 10)</span>
    <span class="n">ax3</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">:])</span>
    <span class="c1"># Filter data for orders 1-10</span>
    <span class="n">max_diff_genes_filtered</span> <span class="o">=</span> <span class="n">max_diff_genes</span><span class="p">[</span><span class="n">max_diff_genes</span><span class="p">[</span><span class="s1">&#39;gene_order&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mi">10</span><span class="p">]</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">max_diff_genes_filtered</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;gene_order&#39;</span><span class="p">,</span> <span class="n">discrete</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                 <span class="n">ax</span><span class="o">=</span><span class="n">ax3</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">ordinal</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="s2">&quot;th&quot;</span> <span class="k">if</span> <span class="mi">4</span><span class="o">&lt;=</span><span class="n">n</span><span class="o">&lt;=</span><span class="mi">20</span> <span class="k">else</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span><span class="s2">&quot;st&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">:</span><span class="s2">&quot;nd&quot;</span><span class="p">,</span><span class="mi">3</span><span class="p">:</span><span class="s2">&quot;rd&quot;</span><span class="p">}</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">n</span><span class="o">%</span><span class="k">10</span>, &quot;th&quot;))
    
    <span class="n">ax3</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">))</span>
    <span class="n">ax3</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">([</span><span class="n">ordinal</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">11</span><span class="p">)],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">90</span><span class="p">)</span>
    <span class="n">ax3</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Gene Order&#39;</span><span class="p">)</span>
    <span class="n">ax3</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Count&#39;</span><span class="p">)</span>

    <span class="c1"># Bottom row, second plot (1/2 width): Distance vs Area Difference</span>
    <span class="n">ax4</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">])</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">/</span><span class="mi">1000</span><span class="p">,</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;area_difference&#39;</span><span class="p">],</span> 
                <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> 
                <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Individual genes&#39;</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">means</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> 
             <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Rolling Mean (5kbp window)&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_vals</span><span class="p">,</span> <span class="n">means</span> <span class="o">-</span> <span class="n">stds</span><span class="p">,</span> <span class="n">means</span> <span class="o">+</span> <span class="n">stds</span><span class="p">,</span>
                     <span class="n">alpha</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Rolling STD (5kbp window)&#39;</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Distance between perturbed enhancer center and TSS (kbp)&#39;</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Normalized gene area difference&#39;</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">)</span>
    <span class="n">max_dist</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">/</span><span class="mi">1000</span><span class="o">/</span><span class="mi">10</span><span class="p">)</span><span class="o">*</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_dist</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
    <span class="n">ax4</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    
    <span class="c1"># Save the figure</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/Enhancer_perturbations/&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Combined_enhancer_analysis.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    
    <span class="c1"># Print statistics</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Enhancer Width Statistics:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">enhancer_widths</span><span class="p">)</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Median: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">enhancer_widths</span><span class="p">)</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Std: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">enhancer_widths</span><span class="p">)</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Min: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">enhancer_widths</span><span class="p">)</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Max: </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">enhancer_widths</span><span class="p">)</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2"> bp&quot;</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Statistics for genes with maximum absolute difference (up to order 10):&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total number of enhancers: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">max_diff_genes_filtered</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Gene Order distribution:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">max_diff_genes_filtered</span><span class="p">[</span><span class="s1">&#39;gene_order&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">sort_index</span><span class="p">())</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_distance_histogram</span><span class="p">(</span><span class="n">results_df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create histogram data for gene-enhancer distances.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_df: DataFrame containing the enhancer-gene analysis results</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Calculate bin edges (every 20kb up to 200kb)</span>
    <span class="n">bin_size</span> <span class="o">=</span> <span class="mi">20000</span>  <span class="c1"># 20kb bins</span>
    <span class="n">max_distance</span> <span class="o">=</span> <span class="mi">200000</span>  <span class="c1"># 200kb max</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_distance</span> <span class="o">+</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">bin_size</span><span class="p">)</span>
    
    <span class="c1"># Create the histogram</span>
    <span class="n">hist</span><span class="p">,</span> <span class="n">bin_edges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">histogram</span><span class="p">(</span><span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    
    <span class="c1"># Create bin labels for plotting</span>
    <span class="n">bin_labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="mi">1000</span><span class="p">)</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">/</span><span class="mi">1000</span><span class="p">)</span><span class="si">}</span><span class="s2">kb&quot;</span> 
                 <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bin_edges</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
    
    <span class="c1"># Plot with matplotlib</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">bin_labels</span><span class="p">,</span> <span class="n">hist</span><span class="p">,</span> <span class="n">align</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Distance from Enhancer (kb)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Gene-Enhancer Distances&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    
    <span class="c1"># Save the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s2">&quot;gene_enhancer_distance_histogram.png&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_distance_histogram</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create and save a histogram showing the distribution of gene-enhancer distances.&quot;&quot;&quot;</span>
    <span class="c1"># Read the data</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../tables/Predicted_gene_areas_in_enhancer_centric_approach.csv&#39;</span><span class="p">))</span>
    
    <span class="c1"># Create save directory</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/Enhancer_perturbations/&quot;</span><span class="p">)</span>
    <span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Create the figure</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Calculate bin edges (every 20kb up to 200kb)</span>
    <span class="n">bin_size</span> <span class="o">=</span> <span class="mi">20000</span>  <span class="c1"># 20kb bins</span>
    <span class="n">max_distance</span> <span class="o">=</span> <span class="mi">200000</span>  <span class="c1"># 200kb max</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">max_distance</span> <span class="o">+</span> <span class="n">bin_size</span><span class="p">,</span> <span class="n">bin_size</span><span class="p">)</span>
    
    <span class="c1"># Create the histogram</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>
    
    <span class="c1"># Customize the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Distance between perturbed enhancer center and TSS (kbp)&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Number of Genes&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    
    <span class="c1"># Convert x-axis ticks to kbp</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">bins</span><span class="p">[::</span><span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">/</span><span class="mi">1000</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">bins</span><span class="p">[::</span><span class="mi">2</span><span class="p">]],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    
    <span class="c1"># Save the figure</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Gene_enhancer_distance_histogram.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="c1"># Print statistics</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Gene-Enhancer Distance Statistics:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean distance: </span><span class="si">{</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">/</span><span class="mi">1000</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2"> kbp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Median distance: </span><span class="si">{</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">median</span><span class="p">()</span><span class="o">/</span><span class="mi">1000</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2"> kbp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Std distance: </span><span class="si">{</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span><span class="o">/</span><span class="mi">1000</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2"> kbp&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of genes: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">create_combined_figure</span><span class="p">()</span>
    <span class="n">create_distance_histogram</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Enhancer Width Statistics:
Mean: 257 bp
Median: 251 bp
Std: 68 bp
Min: 150 bp
Max: 350 bp

Statistics for genes with maximum absolute difference (up to order 10):
Total number of enhancers: 1396

Gene Order distribution:
gene_order
1     932
2     232
3      90
4      50
5      28
6      25
7      12
8      11
9      12
10      4
Name: count, dtype: int64

Gene-Enhancer Distance Statistics:
Mean distance: 89.3 kbp
Median distance: 85.5 kbp
Std distance: 57.2 kbp
Number of genes: 15276
</pre></div>
</div>
<img alt="../_images/90958f949c4716cbfa593bdfa3e973ebf3903b48cf6daec8fdfcbe4cb796885e.png" src="../_images/90958f949c4716cbfa593bdfa3e973ebf3903b48cf6daec8fdfcbe4cb796885e.png" />
</div>
</div>
</section>
<section id="iv-new-models-and-analyses">
<h2>IV) New models and analyses <center><a class="headerlink" href="#iv-new-models-and-analyses" title="Link to this heading">#</a></h2>
<p>The reviewers asked us to perform a number of tests regarding alternative models and data processing strategies:</p>
<blockquote>
<div><p><strong>A) Assessing how the model learns to predict transcription</strong></p>
<p><strong>B) Assessing the impact of shortening the context window</strong></p>
<p><strong>C) Changing the last activation function or the loss</strong></p>
<p><strong>D) Testing on another chromosome (chr19)</strong></p>
<p><strong>E) Training without the enhancer mark (H3K27ac) as an input.</strong></p>
<p><strong>F) Assessing the impact of smoothing</strong></p>
</div></blockquote>
</section>
<section id="a-assessing-how-the-model-learns-to-predict-transcription">
<h2><center> A) Assessing how the model learns to predict transcription<center><a class="headerlink" href="#a-assessing-how-the-model-learns-to-predict-transcription" title="Link to this heading">#</a></h2>
<p>Here we extend the perturbational framework to further explore what rules does CLASTER learn. To do that, we create a number of synthetic inputs modifying a bit the properties of the input. The reference sample is centered at the gene Akirin2 (chr4) which showed a clear directionality and had no genes to the left. We will then feed these slightly different inputs to an already pretrained model, the basic model with only the chromatin branch and pure convolutions (no attention), obtain the predictions and compare them.</p>
<p>The following perturbations will be performed:</p>
<ul class="simple">
<li><p>Modulate promoter activity (height).</p></li>
<li><p>Modulate the extension of H3K4me3 and ATAC-seq signal towards both directions to determine direction of transcription. We will do that by flipping the promoter region.</p></li>
<li><p>Plug-in active superenhancer close to the promoter.</p></li>
<li><p>Plug-in isolated promoter in intergenic region.</p></li>
</ul>
<p><strong>Modifying promoter height:</strong></p>
<p>In the region -100 to + 100 bins (-10kbp,10kbp) centered at the TSS of the gene Akirin2 (ENSMUSG00000028291.7) we will:</p>
<ul class="simple">
<li><p>Get the chromatin state in this region and paste it in the intergenic region to the left.</p></li>
<li><p>Increase the signal (multiply by 2 all marks)</p></li>
<li><p>Decrease the signal (divide by 2 all marks)</p></li>
</ul>
<p><strong>Identifying signal drivers of transcriptional directionality:</strong></p>
<ul class="simple">
<li><p>Check effects of accessibility / H3K27ac extension towards gene body. We will flip the promoter-proximal region.</p></li>
</ul>
<p><strong>Adding superenhancers:</strong></p>
<p>Superenhancers for mESCs in mm9 can be found <a class="reference external" href="https://asntech.org/dbsuper/adv_search.php?genome=mm9&amp;amp;cell_type%5B%5D=C_110&amp;amp;cell_type%5B%5D=C_111&amp;amp;cell_type%5B%5D=C_112&amp;amp;cell_type%5B%5D=C_113&amp;amp;cell_type%5B%5D=C_114&amp;amp;cell_type%5B%5D=C_115&amp;amp;cell_type%5B%5D=C_116&amp;amp;cell_type%5B%5D=C_087&amp;amp;cell_type%5B%5D=C_106&amp;amp;cell_type%5B%5D=C_117&amp;amp;cell_type%5B%5D=C_118&amp;amp;cell_type%5B%5D=C_119&amp;amp;cell_type%5B%5D=C_120&amp;amp;cell_type%5B%5D=C_121&amp;amp;cell_type%5B%5D=C_092&amp;amp;cell_type%5B%5D=C_122&amp;amp;cell_type%5B%5D=C_090&amp;amp;cell_type%5B%5D=C_123&amp;amp;cell_type%5B%5D=C_124&amp;amp;cell_type%5B%5D=C_089&amp;amp;cell_type%5B%5D=C_125&amp;amp;cell_type%5B%5D=C_126&amp;amp;cell_type%5B%5D=C_091&amp;amp;cell_type%5B%5D=C_127&amp;amp;cell_type%5B%5D=C_107&amp;amp;gene=&amp;amp;locus=&amp;amp;method=&amp;amp;submit=1">here</a></p>
<ul class="simple">
<li><p>mESCs have an active superenhancer region directly to the left of KLF4. We get that region and plug it in immediately to the left of Akirin2 (with x0.5 promoter) to observe the boost.</p></li>
</ul>
<p><strong>Adding a new gene in an intergenic region:</strong></p>
<p>We get the promoter of Akirin2 and plug it again upstream.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)</span>
<span class="n">outputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>

<span class="c1"># Interesting locus: ENSMUSG00000028291.7_forward_prediction    ENSMUSG00000035696.15_forward.npy: Rnf38</span>

<span class="n">EXAMPLE_GENE</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000003032.8_forward.npy&quot;</span> <span class="c1">#&lt;- KLF4 # Akirin2: &quot;ENSMUSG00000028291.7_forward.npy&quot;</span>
<span class="c1"># Inputs:</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">input_id_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">input_path</span><span class="p">):</span>
    <span class="c1">#if i % 100 == 0:</span>
    <span class="k">if</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">EXAMPLE_GENE</span><span class="p">:</span>
        <span class="n">name</span> <span class="o">=</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">removesuffix</span><span class="p">(</span><span class="s1">&#39;.npy&#39;</span><span class="p">)</span>
        <span class="n">input_id_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="n">a</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">_input.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">+=</span><span class="mi">1</span>

<span class="c1"># Outputs:</span>
<span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">outputs_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">])</span>

<span class="k">for</span> <span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>   
    <span class="k">if</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">input_id_list</span><span class="p">:</span>
        <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
        <span class="n">line_a</span> <span class="o">=</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_target_predictions</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_prediction.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/perturbed_landscape_arrays/inserted_promoter_arrays/&quot;</span><span class="p">)</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>

<span class="n">akirin2_baseline_input</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="s2">&quot;ENSMUSG00000028291.7_forward.npy&quot;</span><span class="p">)</span>
<span class="n">SEQLEN</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">N_BINS_SHIFT</span> <span class="o">=</span> <span class="mi">100</span> 

<span class="n">akirin2_promoter</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span>

<span class="n">sample_names</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000028291.7_forward&quot;</span><span class="p">,</span> 
                <span class="s2">&quot;ENSMUSG00000028291.7_forward_0.5prom&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ENSMUSG00000028291.7_forward_2prom&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ENSMUSG00000028291.7_forward_flipprom&quot;</span><span class="p">,</span>
                <span class="s2">&quot;ENSMUSG00000028291.7_forward_proxenh&quot;</span> <span class="p">,</span>
                <span class="s2">&quot;ENSMUSG00000028291.7_forward_proxprom&quot;</span><span class="p">]</span>

<span class="c1"># Baseline sequence:</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_baseline_input</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">akirin2_baseline_input</span><span class="p">)</span>

<span class="c1"># A) 2X less active promoter</span>
<span class="n">akirin2_less_prom</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">akirin2_less_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="n">akirin2_less_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_less_prom</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span><span class="n">akirin2_less_prom</span><span class="p">)</span>

<span class="c1"># B) 2X more active promoter</span>
<span class="n">akirin2_more_prom</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">akirin2_more_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">akirin2_more_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_more_prom</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">akirin2_more_prom</span><span class="p">)</span>

<span class="c1"># C) Flipped promoter extension:</span>
<span class="c1"># B) 2X less active promoter</span>
<span class="n">akirin2_flip_prom</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">akirin2_flip_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">akirin2_flip_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_flip_prom</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">akirin2_flip_prom</span><span class="p">)</span>


<span class="c1"># D) KLF4 SUPERENHANCER ADDED</span>
<span class="n">akirin2_prox_enh</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="c1"># - Lower promoter activity (it is super active, genes with strong promoters don&#39;t rely on enhancers that much, see Narita et al.)</span>
<span class="n">akirin2_prox_enh</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="n">akirin2_prox_enh</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span> <span class="c1"># small promoter effect</span>
<span class="c1"># - Copy paste KLF4 superenhancer state.</span>
<span class="n">ENH_WIDTH</span> <span class="o">=</span> <span class="mi">400</span>
<span class="n">ENH_START</span> <span class="o">=</span> <span class="n">ENH_WIDTH</span> <span class="o">+</span>  <span class="n">ENH_WIDTH</span> <span class="o">//</span> <span class="mi">2</span> <span class="c1">#Enhancer starts at -60 kbp kbp from TSS</span>
<span class="n">ENH_TRUE_POS</span> <span class="o">=</span> <span class="o">-</span><span class="mi">800</span>
<span class="n">klf4_baseline_input</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">input_path</span> <span class="o">/</span> <span class="s2">&quot;ENSMUSG00000003032.8_forward.npy&quot;</span><span class="p">)</span>
<span class="n">klf4_superenhancer</span> <span class="o">=</span> <span class="n">klf4_baseline_input</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">ENH_TRUE_POS</span> <span class="p">:</span> <span class="p">(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">ENH_TRUE_POS</span> <span class="o">+</span> <span class="n">ENH_WIDTH</span><span class="p">]</span>
<span class="n">akirin2_prox_enh</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="n">ENH_START</span> <span class="p">:</span> <span class="p">(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="n">ENH_START</span> <span class="o">+</span> <span class="n">ENH_WIDTH</span><span class="p">]</span> <span class="o">=</span> <span class="n">klf4_superenhancer</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_prox_enh</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">akirin2_prox_enh</span><span class="p">)</span>

<span class="c1"># E) Akirin2 Promoter added:</span>
<span class="n">PROM_POS</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1000</span> <span class="c1"># bins to the left where we plug in the artificial promoter</span>
<span class="n">akirin2_add_prom</span> <span class="o">=</span> <span class="n">akirin2_baseline_input</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">akirin2_add_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span> <span class="o">+</span> <span class="n">PROM_POS</span> <span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="o">+</span> <span class="n">PROM_POS</span><span class="p">]</span> <span class="o">=</span> <span class="n">akirin2_add_prom</span><span class="p">[:,(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span><span class="n">N_BINS_SHIFT</span><span class="p">:(</span><span class="n">SEQLEN</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span><span class="n">N_BINS_SHIFT</span> <span class="p">]</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">akirin2_add_prom</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Inputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
<span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">,</span> <span class="n">akirin2_add_prom</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Dummy sythetic targets (for eir to work we need also targets matching inputs, which will be ignored)</span>
<span class="n">targets_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets/&quot;</span><span class="p">)</span>
<span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
<span class="c1"># Create DataFrame</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> 
                 <span class="n">index</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Index</span><span class="p">(</span><span class="n">sample_names</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s1">&#39;ID&#39;</span><span class="p">),</span>  <span class="c1"># Set index with name &#39;ID&#39;</span>
                 <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">_ctrl&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span>  <span class="c1"># Set columns</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">targets_path</span> <span class="o">/</span> <span class="s2">&quot;prom_perturbed_targets.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Running new models:</strong></p>
<blockquote>
<div><p><em>Note:</em>
For GPU trained models, we need to test using the same device.</p>
</div></blockquote>
<p>To run on a SLURM based cluster:</p>
<p>srun –partition=gpuqueue –gres=gpu:a100:1 –</p>
<p><strong>Predict promoter perturbed inputs:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_predict_prom/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_predict_prom/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_predict_prom/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_predict_prom/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv/saved_models/gene_expression_only_chrom_pure_conv_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.8161.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/perturbation_runs/gene_expression_only_chrom_pure_conv_prom_perturb<span class="w"> </span>
</pre></div>
</div>
<p><strong>Plotting predicted profiles after the perturbations:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Outputs:</span>
<span class="n">original_outputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span>
<span class="n">outputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_only_chrom_pure_conv_prom_perturb/&quot;</span><span class="p">)</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">outputs_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">])</span>

<span class="n">color_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;purple&#39;</span><span class="p">,</span><span class="s1">&#39;black&#39;</span><span class="p">,</span><span class="s1">&#39;lightseagreen&#39;</span><span class="p">,</span> <span class="s1">&#39;orange&#39;</span><span class="p">,</span> <span class="s1">&#39;firebrick&#39;</span><span class="p">,</span> <span class="s1">&#39;darkgreen&#39;</span><span class="p">]</span>
<span class="n">predicted_tracks</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;baseline&#39;</span><span class="p">,</span><span class="s1">&#39;x0.5 Promoter&#39;</span><span class="p">,</span> <span class="s1">&#39;x2 Promoter&#39;</span><span class="p">,</span> <span class="s1">&#39;Flipped promoter region&#39;</span><span class="p">,</span> <span class="s1">&#39;Proximal enhancer added&#39;</span><span class="p">,</span> <span class="s1">&#39;New promoter added&#39;</span><span class="p">]</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">(</span><span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">)):</span>   
    <span class="nb">print</span><span class="p">(</span><span class="n">ID</span><span class="p">)</span>
    <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
    <span class="n">predictions</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;profile&#39;</span><span class="p">:</span><span class="n">line_p</span><span class="p">,</span>
                        <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">color_list</span><span class="p">[</span><span class="n">k</span><span class="p">],</span>
                        <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
                        <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="n">predicted_tracks</span><span class="p">[</span><span class="n">k</span><span class="p">],</span>
                        <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
    <span class="p">})</span>


<span class="c1"># Actual baseline signal:</span>
<span class="n">ID</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000028291.7_forward&quot;</span>
<span class="o">*</span><span class="n">_</span> <span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">original_outputs_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">])</span>

<span class="n">actual_data</span> <span class="o">=</span> <span class="p">[{</span>
    <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
    <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;silver&#39;</span><span class="p">,</span>
    <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">,</span>
    <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;EU-seq&#39;</span><span class="p">,</span>
    <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
<span class="p">}]</span>

<span class="c1"># Create plot</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="s1">&#39;Perturbed_promoter.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="b-creating-a-shorter-context-model">
<h2><center> B) Creating a shorter context model: <center><a class="headerlink" href="#b-creating-a-shorter-context-model" title="Link to this heading">#</a></h2>
<p>We reduced the input context to 20 kbp and the output context to the central 18 kbp.</p>
<p><strong>Creating the new inputs and outputs:</strong></p>
<p>This is a highly parallelizable process, so we will use the python multiprocessing package to exploit our slurm-based cluster. We ran:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>srun<span class="w"> </span>--cpus-per-task<span class="o">=</span><span class="m">16</span><span class="w"> </span>--<span class="w"> </span>python<span class="w"> </span>./create_cropped_inputs.py
</pre></div>
</div>
<blockquote>
<div><p><strong>Data processing:</strong>
We just cropped the exact same inputs and outputs to the new desired length.</p>
<p><strong>Model changes:</strong>
We reduced the size of the model accordingly:</p>
<ul class="simple">
<li><p>Smaller feature extractor: too many convolutions would compress too much the input sequences.</p></li>
<li><p>Smaller MLP.</p></li>
</ul>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_cropped_inputs.py
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_sample</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">original_path</span><span class="p">,</span> <span class="n">new_path</span><span class="p">,</span> <span class="n">new_dim</span><span class="o">=</span><span class="mi">201</span><span class="p">,</span> <span class="n">original_dim</span><span class="o">=</span><span class="mi">10001</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process a single sample file.&quot;&quot;&quot;</span>
    <span class="n">crop_shift</span> <span class="o">=</span> <span class="p">(</span><span class="n">original_dim</span> <span class="o">-</span> <span class="n">new_dim</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">original_path</span> <span class="o">/</span> <span class="n">filename</span><span class="p">)</span>
        <span class="n">new_array</span> <span class="o">=</span> <span class="n">array</span><span class="p">[:,</span> <span class="n">crop_shift</span><span class="p">:</span><span class="o">-</span><span class="n">crop_shift</span><span class="p">]</span>
        <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">new_path</span> <span class="o">/</span> <span class="n">filename</span><span class="p">,</span> <span class="n">new_array</span><span class="p">)</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;Successfully processed </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;Error processing </span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
    <span class="c1"># Define paths</span>
    <span class="n">original_inputs_paths</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/training/&quot;</span><span class="p">),</span>
        <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">new_inputs_paths</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/training_20kbp_context/&quot;</span><span class="p">),</span>
        <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test_20kbp_context/&quot;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># Create output directories if they don&#39;t exist</span>
    <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">new_inputs_paths</span><span class="p">:</span>
        <span class="n">path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Number of processes to use (leave one core free for system processes)</span>
    <span class="n">num_processes</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">original_inputs_paths</span><span class="p">):</span>
        <span class="c1"># Get list of filenames instead of DirEntry objects</span>
        <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
        
        <span class="c1"># Create partial function with fixed arguments</span>
        <span class="n">process_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span>
            <span class="n">process_sample</span><span class="p">,</span>
            <span class="n">original_path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span>
            <span class="n">new_path</span><span class="o">=</span><span class="n">new_inputs_paths</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="p">)</span>
        
        <span class="c1"># Process files in parallel</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing files in </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2"> using </span><span class="si">{</span><span class="n">num_processes</span><span class="si">}</span><span class="s2"> processes...&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">num_processes</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_func</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
        
        <span class="c1"># Print results</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Training eir only chrom pure conv for 20kbp context:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing eir only chrom pure conv for 20kbp context:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context_test/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_20kbp_context_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv_20kbp_context/saved_models/gene_expression_only_chrom_pure_conv_20kbp_context_model_30300_perf-average<span class="o">=</span><span class="m">0</span>.7867.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_only_chrom_pure_conv_20kbp_context_test<span class="w"> </span>
</pre></div>
</div>
</section>
<section id="c-changing-the-last-activation-function-or-the-loss">
<h2><center> C) Changing the last activation function or the loss:<a class="headerlink" href="#c-changing-the-last-activation-function-or-the-loss" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>Output activation to softplus.</p></li>
</ul>
<blockquote>
<div><p>⚠️ WARNING ⚠️ : This is a temporary hotfix that overwrites the linear head to introduce the least changes to the eir code.</p>
</div></blockquote>
<p><strong>Modifications to eir’s (version 0.1.42) code:</strong></p>
<p><em>Changing activation function of the last layer:</em></p>
<p>We will change the file defining the last layer activation functions for tabular outputs, which in on our environment can be found at <code class="docutils literal notranslate"><span class="pre">claster_env/lib/python3.11/site-packages/eir/models/output/tabular/linear.py</span></code>.
Here we will edit the function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>
<span class="c1">#Original linear head</span>
<span class="k">def</span><span class="w"> </span><span class="nf">_get_linear_multi_task_branches</span><span class="p">(</span>
    <span class="n">input_dimension</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">num_outputs_per_target</span><span class="p">:</span> <span class="s2">&quot;al_num_outputs_per_target&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nn</span><span class="o">.</span><span class="n">ModuleDict</span><span class="p">:</span>
    <span class="n">multi_task_branches</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">ModuleDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="n">target</span><span class="p">:</span> <span class="n">nn</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span>
                <span class="n">in_features</span><span class="o">=</span><span class="n">input_dimension</span><span class="p">,</span>
                <span class="n">out_features</span><span class="o">=</span><span class="n">num_outputs</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">target</span><span class="p">,</span> <span class="n">num_outputs</span> <span class="ow">in</span> <span class="n">num_outputs_per_target</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">multi_task_branches</span>
</pre></div>
</div>
<p>The new function will simply pass the output through a softplus transformation. The same principles would apply for ReLU:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># New head</span>
<span class="k">def</span><span class="w"> </span><span class="nf">_get_linear_multi_task_branches</span><span class="p">(</span>
    <span class="n">input_dimension</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">num_outputs_per_target</span><span class="p">:</span> <span class="s2">&quot;al_num_outputs_per_target&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">nn</span><span class="o">.</span><span class="n">ModuleDict</span><span class="p">:</span>
    <span class="n">multi_task_branches</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">ModuleDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="n">target</span><span class="p">:</span> <span class="n">nn</span><span class="o">.</span><span class="n">Sequential</span><span class="p">(</span>
                <span class="n">nn</span><span class="o">.</span><span class="n">Linear</span><span class="p">(</span>
                    <span class="n">in_features</span><span class="o">=</span><span class="n">input_dimension</span><span class="p">,</span>
                    <span class="n">out_features</span><span class="o">=</span><span class="n">num_outputs</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="n">nn</span><span class="o">.</span><span class="n">Softplus</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">target</span><span class="p">,</span> <span class="n">num_outputs</span> <span class="ow">in</span> <span class="n">num_outputs_per_target</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">multi_task_branches</span>
</pre></div>
</div>
<p><em>Setting the log_input to False in PoissonNLLL:</em></p>
<p>This is the edit in the file <code class="docutils literal notranslate"><span class="pre">claster_env/lib/python3.11/site-packages/eir/train_utils/criteria.py</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">_calc_con_loss</span><span class="p">(</span><span class="nb">input</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">,</span> <span class="n">target</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">,</span> <span class="n">loss_func</span><span class="p">:</span> <span class="n">al_con_losses</span><span class="p">):</span>
    <span class="k">match</span> <span class="n">loss_func</span><span class="p">:</span>
        <span class="k">case</span> <span class="n">nn</span><span class="o">.</span><span class="n">PoissonNLLLoss</span><span class="p">():</span>
            <span class="c1"># Before we had:</span>
            <span class="c1">#return loss_func(log_input=input.squeeze(), target=target.squeeze())</span>
            <span class="c1"># Create a new PoissonNLLLoss with log_input=False</span>
            <span class="n">poisson_loss</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">PoissonNLLLoss</span><span class="p">(</span><span class="n">log_input</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="c1"># Note that despite the parameter name being log_input, we pass the raw input tensor</span>
            <span class="k">return</span> <span class="n">poisson_loss</span><span class="p">(</span><span class="nb">input</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(),</span> <span class="n">target</span><span class="o">.</span><span class="n">squeeze</span><span class="p">())</span>
        <span class="k">case</span><span class="w"> </span><span class="k">_</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">loss_func</span><span class="p">(</span><span class="nb">input</span><span class="o">=</span><span class="nb">input</span><span class="o">.</span><span class="n">squeeze</span><span class="p">(),</span> <span class="n">target</span><span class="o">=</span><span class="n">target</span><span class="o">.</span><span class="n">squeeze</span><span class="p">())</span>

</pre></div>
</div>
<p><strong>Run models:</strong></p>
<p>Training eir only chrom pure conv with softplus:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus/outputs_2_cond.yaml

eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus_poisson/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus_poisson/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus_poisson/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_softplus_poisson/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Visualize predictions for Softplus and Poisson losses:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Outputs:</span>
<span class="n">original_outputs_path_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias/results/&quot;</span><span class="p">):[</span><span class="s2">&quot;Predicted EU-seq Softplus Poisson&quot;</span><span class="p">,</span> <span class="s1">&#39;royalblue&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">],</span>
                              <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_no_bias/results/&quot;</span><span class="p">):[</span><span class="s2">&quot;Predicted EU-seq Softplus SmoothL1&quot;</span><span class="p">,</span> <span class="s1">&#39;orange&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">],</span>
                              <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias_log_input_false/results/&quot;</span><span class="p">):[</span><span class="s2">&quot;Predicted EU-seq Softplus Poisson log input false&quot;</span><span class="p">,</span> <span class="s1">&#39;darkred&#39;</span><span class="p">,</span><span class="s1">&#39;-&#39;</span><span class="p">]}</span>

<span class="n">training_targets_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../targets/training_targets.csv&quot;</span><span class="p">)</span>

<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000051977.16_forward&quot;</span><span class="p">,</span><span class="s2">&quot;ENSMUSG00000109212.3_forward&quot;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">ID</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">ids</span><span class="p">):</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">original_outputs_path</span><span class="p">,</span><span class="n">attributes</span> <span class="ow">in</span> <span class="n">original_outputs_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span> <span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">original_outputs_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">is_training</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">batch_num</span><span class="o">=</span><span class="mi">60600</span><span class="p">,</span> <span class="n">CLIP</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
        <span class="n">predictions</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;profile&#39;</span><span class="p">:</span><span class="n">line_p</span><span class="p">,</span>
                            <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="n">attributes</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                            <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.05</span><span class="p">,</span>
                            <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="n">attributes</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="n">attributes</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                            <span class="p">})</span>
        

    <span class="c1"># Actual baseline signal:</span>
    <span class="n">actual_data</span> <span class="o">=</span> <span class="p">[{</span>
        <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
        <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;silver&#39;</span><span class="p">,</span>
        <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">,</span>
        <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;EU-seq&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
    <span class="p">}]</span>

    <span class="c1"># Averaged actual targets across samples:    </span>
    <span class="n">avg_training_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">training_targets_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">predictions</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;profile&#39;</span><span class="p">:</span><span class="n">avg_training_targets</span><span class="p">,</span>
                            <span class="s1">&#39;color&#39;</span><span class="p">:</span><span class="s1">&#39;darkgreen&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
                            <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;Mean EU-seq per bin&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;--&#39;</span>
                            <span class="p">})</span>


    <span class="c1"># Create plot</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;Softplus_no_bias_</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="d-testing-on-another-chromosome-chr19">
<h2><center> D) Testing on another chromosome (chr19):<a class="headerlink" href="#d-testing-on-another-chromosome-chr19" title="Link to this heading">#</a></h2>
<p>We will:</p>
<ul class="simple">
<li><p>Keep the samples on chr4 as hold out.</p></li>
<li><p>Additionally mask samples from chr19 from training (set them as validation)</p></li>
<li><p>Introduce samples in chr17 in the training (were used for validation originally).</p></li>
<li><p>Note that we will not validate now since we already have the model’s hyperparameters defined.</p></li>
</ul>
<p><strong>Training eir for chr19 holdout:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_train/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing eir for chr19 holdout:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_test/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_chr19_holdout_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv_chr19_holdout/saved_models/gene_expression_only_chrom_pure_conv_chr19_holdout_model_30300_perf-average<span class="o">=</span><span class="m">0</span>.6487.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_only_chrom_pure_conv_chr19_holdout_test
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># From the notebook II_run_CLASTER:</span>
<span class="k">def</span><span class="w"> </span><span class="nf">create_validation_ids</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">val_chrom</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; All genes encoded in chr19 will be used as a validation set.&quot;&quot;&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
    <span class="n">missing_samples_report</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;missing_samples_report.csv&quot;</span><span class="p">)[</span><span class="s2">&quot;Sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">gene_annotations_df</span><span class="p">[</span><span class="n">gene_annotations_df</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">val_chrom</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>

    <span class="c1"># Prepare the strings to write in the file</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">_forward</span><span class="se">\n</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">_rev</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">indices</span> <span class="k">if</span> <span class="n">index</span><span class="o">+</span><span class="s1">&#39;_forward&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">missing_samples_report</span><span class="p">]</span>

    <span class="c1"># Write to the text file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;manual_validation_ids_</span><span class="si">{</span><span class="n">val_chrom</span><span class="si">}</span><span class="s2">.txt&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/&quot;</span><span class="p">)</span>
<span class="n">create_validation_ids</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ids_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/manual_validation_ids_chr19.txt&quot;</span><span class="p">)</span>
<span class="n">sample_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ids_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">training_samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/training/&quot;</span><span class="p">))]</span>

<span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="n">sample_ids</span><span class="p">:</span>
    <span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.npy&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">training_samples</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">sample</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">))</span>

<span class="n">b</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/Final_gene_annotations.tsv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">b</span> <span class="o">=</span> <span class="n">b</span><span class="p">[</span><span class="n">b</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;chr19&#39;</span><span class="p">]</span>
<span class="n">b</span>
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p><em>Note:</em> samples ENSMUSG00000095993.1_forward and ENSMUSG00000095993.1_rev, corresponding to the last gene annotated in chr19 (Gm21060) were not in the original input training set, so we removed their corresponding ids in the hold out manual validation list for chr19.**</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">##################### Copying chr19</span>
<span class="c1"># Read sample IDs using pandas</span>
<span class="n">ids_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/manual_validation_ids_chr19.txt&quot;</span><span class="p">)</span>
<span class="n">sample_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ids_path</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Define paths using pathlib</span>
<span class="n">source_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/landscape_arrays/training&#39;</span><span class="p">)</span>
<span class="n">dest_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/landscape_arrays/test_chr19&#39;</span><span class="p">)</span>

<span class="c1"># Create destination directory if it doesn&#39;t exist</span>
<span class="n">dest_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># COPY matching .npy files (not move)</span>
<span class="n">moved_count</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">source_dir</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.npy&#39;</span><span class="p">):</span>
    <span class="c1"># Check if any sample ID from our list is in the filename</span>
    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">sample_id</span> <span class="ow">in</span> <span class="n">file_path</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">sample_ids</span><span class="p">):</span>
        <span class="n">dest_path</span> <span class="o">=</span> <span class="n">dest_dir</span> <span class="o">/</span> <span class="n">file_path</span><span class="o">.</span><span class="n">name</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">file_path</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">dest_path</span><span class="p">))</span>
        <span class="n">moved_count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Copied: </span><span class="si">{</span><span class="n">file_path</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Total files moved: </span><span class="si">{</span><span class="n">moved_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1">################## Retrieving the targets from the original file ################</span>
<span class="c1"># Read sample IDs using pandas</span>
<span class="n">manual_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../annotations/manual_validation_ids_chr19.txt&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># Read the training targets file</span>
<span class="n">training_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../targets/training_targets.csv&#39;</span><span class="p">)</span>

<span class="c1"># Filter rows where ID matches our manual validation IDs</span>
<span class="n">test_targets</span> <span class="o">=</span> <span class="n">training_targets</span><span class="p">[</span><span class="n">training_targets</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">manual_ids</span><span class="p">)]</span>

<span class="c1"># Save to new file in the test directory</span>
<span class="n">dest_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../targets/&#39;</span><span class="p">)</span>
<span class="n">test_targets</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">dest_dir</span> <span class="o">/</span> <span class="s1">&#39;test_targets_chr19.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Original targets shape: </span><span class="si">{</span><span class="n">training_targets</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Test targets shape: </span><span class="si">{</span><span class="n">test_targets</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="e-training-without-the-enhancer-mark-h3k27ac-as-an-input">
<h2><center> E) Training without the enhancer mark (H3K27ac) as an input.<a class="headerlink" href="#e-training-without-the-enhancer-mark-h3k27ac-as-an-input" title="Link to this heading">#</a></h2>
<p>We removed the row corresponding to the enhancer mark H3K27ac from all the input numpy arrays.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_inputs_without_H3K27ac.py
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mp</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_file</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">):</span>
    <span class="c1"># Load the array</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    
    <span class="c1"># Remove H3K27ac row (ATAC, H3K4me3, H3K27ac, H3K27me3)</span>
    <span class="n">data_filtered</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">delete</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    
    <span class="c1"># Save to new location with same filename</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">savepath</span> <span class="o">/</span> <span class="n">file_path</span><span class="o">.</span><span class="n">name</span>
    <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">data_filtered</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
    <span class="n">input_folder</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/landscape_arrays/test/&#39;</span><span class="p">)</span> <span class="c1">#Path(&#39;../inputs/landscape_arrays/training/&#39;)</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../inputs/landscape_arrays/test_no_H3K27ac/&#39;</span><span class="p">)</span> <span class="c1"># Path(&#39;../inputs/landscape_arrays/training_no_H3K27ac/&#39;)</span>
    
    <span class="c1"># Create output directory if it doesn&#39;t exist</span>
    <span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Get list of all files</span>
    <span class="n">files</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">input_folder</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.npy&#39;</span><span class="p">))</span>
    
    <span class="c1"># Set up multiprocessing</span>
    <span class="n">n_cpus</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    
    <span class="c1"># Create partial function with fixed savepath</span>
    <span class="n">process_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">process_file</span><span class="p">,</span> <span class="n">savepath</span><span class="o">=</span><span class="n">savepath</span><span class="p">)</span>
    
    <span class="c1"># Create pool and map files to processes</span>
    <span class="k">with</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">n_cpus</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_func</span><span class="p">,</span> <span class="n">files</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Processing complete!&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Training eir only chrom pure conv with no H3K27ac:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_train/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing eir only chrom pure conv with no H3K27ac:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_test/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_no_H3K27ac_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv_no_H3K27ac/saved_models/gene_expression_only_chrom_pure_conv_no_H3K27ac_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.7807.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_only_chrom_pure_conv_no_H3K27ac_test
</pre></div>
</div>
</section>
<section id="f-predict-untransformed-eu-seq-only-binned-at-1kbp-resolution">
<h2><center> F) Predict untransformed EU-seq, only binned at 1kbp resolution. <center><a class="headerlink" href="#f-predict-untransformed-eu-seq-only-binned-at-1kbp-resolution" title="Link to this heading">#</a></h2>
<p>Reviewer 2 was concerned about potential oversmoothing of the signals and the impact that this would cause in model performance. We proceeded to train CLASTER aiming to predict EU-seq profiles directly binned at a 1kbp resolution.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_1kbp_binned_targets.py
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pyBigWig</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                        <span class="n">output_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                        <span class="n">n_output_bins</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                        <span class="n">bw_target_track_list</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span>
                        <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                        <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                        <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Targets are centered at the TSS, and obtained at a 20 bp resolution, smoothed and downsized to 1kbp resolution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize stats as None</span>
    <span class="n">stats</span> <span class="o">=</span> <span class="kc">None</span>
    
    <span class="k">for</span> <span class="n">bw_path</span> <span class="ow">in</span> <span class="n">bw_target_track_list</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span><span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span><span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span><span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">stats</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># Only process if we got valid stats</span>
                <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
                <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span><span class="mi">0</span><span class="p">,</span><span class="kc">None</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> no valid stats obtained.&quot;</span><span class="p">)</span>
                <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_output_bins</span><span class="p">)</span>  <span class="c1"># Return zeros if no valid stats</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> target coordinates are out of bounds. Error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_output_bins</span><span class="p">)</span>  <span class="c1"># Return zeros in case of error</span>
    
    <span class="k">if</span> <span class="n">stats</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_output_bins</span><span class="p">)</span>  <span class="c1"># Fallback if no valid stats were obtained</span>
        
    <span class="k">return</span> <span class="n">stats</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">stats</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_target_data</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> <span class="n">target_files</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
    <span class="n">ID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">Start</span><span class="p">,</span> <span class="n">End</span><span class="p">,</span> <span class="n">Strand</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">CRE_type</span> <span class="o">=</span> <span class="n">row</span>
    <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
    <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

    <span class="n">target_sample_pos</span><span class="p">,</span> <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">)</span> <span class="o">==</span> <span class="n">OUTPUT_LENGTH</span><span class="p">):</span>
        <span class="n">target_file_path</span> <span class="o">=</span> <span class="n">target_files</span><span class="p">[</span><span class="n">split</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">target_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_pos</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_neg</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files&quot;</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span>
    <span class="n">target_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">split_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>

    <span class="c1"># Start logger</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;/Landscape_data_obtention_1kbp_binned.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>  

    <span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations.tsv&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">output_shift</span> <span class="o">=</span> <span class="mi">200500</span> <span class="c1">#200500 # Output in 1kbp resolution (500 extra per side to have extra central bin)</span>
    <span class="n">n_output_bins</span> <span class="o">=</span> <span class="mi">401</span>
    <span class="n">OUTPUT_LENGTH</span> <span class="o">=</span> <span class="mi">401</span> <span class="c1">#802</span>
    <span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
    <span class="n">bw_target_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;EU_Seq_Ctrl.bw&quot;</span><span class="p">]</span> <span class="c1"># &quot;EU_Seq_Treated.bw&quot;</span>

    <span class="n">target_files</span> <span class="o">=</span> <span class="p">{</span><span class="n">split</span><span class="p">:</span> <span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_targets_1kbp_binning.csv&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]}</span>
    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">target_files</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ID,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">process_args</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> <span class="n">target_files</span><span class="p">)</span>

    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">process_target_data</span><span class="p">,</span> <span class="p">[(</span><span class="n">process_args</span> <span class="o">+</span> <span class="p">(</span><span class="n">row</span><span class="p">,))</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()])</span>
</pre></div>
</div>
</div>
</div>
<p><strong>New targets analysis:</strong></p>
<ul class="simple">
<li><p>Remove samples that were missing in the original analyses.</p></li>
<li><p>Compute correlation between binned and smoothed profiles.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">new_training_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/training_targets_1kbp_binning.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
<span class="n">new_test_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/test_targets_1kbp_binning.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>

<span class="n">old_training_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/training_targets.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
<span class="n">old_test_targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/test_targets.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>


<span class="n">new_training_targets</span> <span class="o">=</span> <span class="n">new_training_targets</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">old_training_targets</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">new_test_targets</span> <span class="o">=</span> <span class="n">new_test_targets</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">old_test_targets</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">new_test_targets</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">old_test_targets</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="n">r</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">)</span>

<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions&quot;</span><span class="p">)</span>

<span class="n">area</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="s2">&quot;Smoothed_vs_1kbp_binned_correlation.png&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">),</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">min_val</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">max_val</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>


<span class="k">for</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">new_test_targets</span><span class="o">.</span><span class="n">index</span><span class="p">[::</span><span class="mi">100</span><span class="p">]:</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="p">[{</span>
        <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">new_test_targets</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
        <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;royalblue&#39;</span><span class="p">,</span>
        <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">,</span>
        <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;EU-seq binned 1kbp&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>

    <span class="p">}]</span>

    <span class="n">actual_data</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">old_test_targets</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
        <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;silver&#39;</span><span class="p">,</span>
        <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">,</span>
        <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="s1">&#39;EU-seq&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Training eir only chrom pure conv with no smoothing, directly binning the reads in 1kbp bins:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_train/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing eir only chrom pure conv with no smoothing, directly binning the reads in 1kbp bins:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_test/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_1kbp_binning_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_pure_conv_1kbp_binning/saved_models/gene_expression_only_chrom_pure_conv_1kbp_binning_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.7832.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_only_chrom_pure_conv_1kbp_binning_test
</pre></div>
</div>
<section id="analyze-predictions-and-attributions-of-the-new-models">
<h3>Analyze predictions and attributions of the new models<a class="headerlink" href="#analyze-predictions-and-attributions-of-the-new-models" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results_path_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv/results/&quot;</span><span class="p">),</span>
                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span> 
                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Gaussian&quot;</span><span class="p">,</span>
                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolution test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">),</span>
                                             <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                             <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                             <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                             <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                             <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;CLASTER test&quot;</span><span class="p">,</span>
                                             <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin with Attention training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_attention/results/&quot;</span><span class="p">),</span>
                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin with Attention test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_attention/&quot;</span><span class="p">),</span>
                                          <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                          <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                          <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                          <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                          <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                          <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Structure training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_pure_conv/results/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and structure&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Structure test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_pure_conv/&quot;</span><span class="p">),</span>
                                    <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                    <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                    <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                    <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                    <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure&quot;</span><span class="p">,</span>
                                    <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Structure rotated training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_rotated_pure_conv/results/&quot;</span><span class="p">),</span>
                                               <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                               <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                               <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                               <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                               <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                               <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Structure rotated test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_rotated_pure_conv/&quot;</span><span class="p">),</span>
                                           <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                           <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                           <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                           <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                           <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                           <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus SmoothL1 training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_no_bias/results/&quot;</span><span class="p">),</span>
                                                                   <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                   <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                   <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                   <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                   <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus SmoothL1&quot;</span><span class="p">,</span>
                                                                   <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;purple&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias/results/&quot;</span><span class="p">),</span>
                                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson&quot;</span><span class="p">,</span>
                                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;orange&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson log_input=False training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias_log_input_false/results/&quot;</span><span class="p">),</span>
                                                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson log input false&quot;</span><span class="p">,</span>
                                                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkred&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_20kbp_context/results/&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train 20kb context&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;yellow&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_20kbp_context_test&quot;</span><span class="p">),</span>
                                                          <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                          <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                          <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                          <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                          <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 20kb context&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;green&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_chr19_holdout/results/&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train chr19 holdout&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_chr19_holdout_test/&quot;</span><span class="p">),</span>
                                                          <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                          <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                          <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                          <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                          <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test chr19 holdout&quot;</span><span class="p">,</span>
                                                          <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_no_H3K27ac/results/&quot;</span><span class="p">),</span>
                                                           <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                           <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                           <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                           <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                           <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train no H3K27ac&quot;</span><span class="p">,</span>
                                                           <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_no_H3K27ac_test/&quot;</span><span class="p">),</span>
                                                       <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                       <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                       <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                       <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                       <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test no H3K27ac&quot;</span><span class="p">,</span>
                                                       <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_1kbp_binning/results/&quot;</span><span class="p">),</span>
                                                             <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                             <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                             <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                             <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                             <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;1kbp binning&quot;</span><span class="p">,</span>
                                                             <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_1kbp_binning_test/&quot;</span><span class="p">),</span>
                                                         <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                         <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                         <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                         <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                         <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 1kbp binning&quot;</span><span class="p">,</span>
                                                         <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions K562 RNA-seq training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_train/results/&quot;</span><span class="p">),</span>
                                                             <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                             <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                             <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                             <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                             <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 RNA-seq&quot;</span><span class="p">,</span>
                                                             <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Only Chromatin Pure Convolutions K562 POLR2A training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_POLR2A_train/results/&quot;</span><span class="p">),</span>
                                                            <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                            <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                            <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                            <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                            <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 POLR2A&quot;</span><span class="p">,</span>
                                                            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Prom-CHiC training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_prom_CHiC_pure_conv/results/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom-CHiC&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;Chromatin and Prom-CHiC test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_prom_CHiC_pure_conv_test/&quot;</span><span class="p">),</span>
                                    <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                    <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                    <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                    <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                    <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom CHiC test&quot;</span><span class="p">,</span>
                                    <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
    <span class="s2">&quot;test_K562_enhancer_centric_POLR2A&quot;</span><span class="p">:{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_only_chrom_K562_POLR2A_enhancer_centric/&quot;</span><span class="p">),</span>
        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span> 
        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 POLR2A enhancer-centric&quot;</span><span class="p">,</span>
        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>  
    <span class="s2">&quot;test_K562_enhancer_centric_RNA&quot;</span><span class="p">:{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span><span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/perturbation_runs/gene_expression_only_chrom_K562_RNA_enhancer_centric/&quot;</span><span class="p">),</span>
        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span> 
        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 RNA-seq enhancer-centric&quot;</span><span class="p">,</span>
        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>  
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Plot correlations:</strong></p>
<blockquote>
<div><p>Note: the function <code class="docutils literal notranslate"><span class="pre">calculate_correlations</span></code> was developed for mouse mm10 gene annotations. Please comment on human entries to the previous dictionary to run the following cell.</p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_pos_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_corrected.tsv&quot;</span><span class="p">)</span>
<span class="n">input_table_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/gene_enhancer_relationships_perturbed.csv&quot;</span><span class="p">)</span>

<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/prediction_correlations/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="c1">#[&quot;&quot;]</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;binary&quot;</span>

<span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">attributes</span> <span class="ow">in</span> <span class="n">results_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

    <span class="n">pred_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">actual_list_A_per_bin_dict</span><span class="p">,</span> <span class="n">pred_list_values_dict</span><span class="p">,</span> <span class="n">actual_list_values_dict</span> <span class="o">=</span> <span class="n">calculate_correlations</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;path&quot;</span><span class="p">],</span> 
                                                                                                                                  <span class="n">gene_pos_path</span><span class="p">,</span>
                                                                                                                                  <span class="n">figure_path</span><span class="p">,</span> 
                                                                                                                                  <span class="n">input_table_path</span><span class="p">,</span> 
                                                                                                                                <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span>
                                                                                                                                  <span class="n">is_training</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;is_training&quot;</span><span class="p">],</span>
                                    
                                                                                                                                  <span class="n">batch_num</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;batch_num&quot;</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="c1"># Prediction of all bins (1kbp resolution)</span>
        <span class="n">pred_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pred_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> 
        <span class="n">actual_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actual_list_values_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
        
        <span class="n">pointwise</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">,</span> <span class="n">title</span> <span class="o">+</span> <span class="s2">&quot;_pointwise_prediction&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>

        <span class="c1"># Central gene area (divided by gene length, equiv. to averaged EU-seq signal for the target gene)</span>
        <span class="n">pred_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pred_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span> 
        <span class="n">actual_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actual_list_A_per_bin_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">])</span>
        <span class="n">area</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">,</span> <span class="n">title</span> <span class="o">+</span> <span class="s2">&quot;_gene_area_lengthnorm_prediction&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Visualize predictions of the different models:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Outputs:</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="c1">#ID = ENSMUSG00000090115.7_rev ENSMUSG00000043592.15_rev ENSMUSG00000023266.11_rev # Nice examples &#39;ENSMUSG00000079707.10_forward&#39;</span>
<span class="n">val_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../annotations/manual_validation_ids_chr17.txt&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">for</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">val_ids</span><span class="p">[</span><span class="mi">172</span><span class="p">::</span><span class="mi">200</span><span class="p">]:</span><span class="c1">#  [&quot;ENSMUSG00000090115.7_forward&quot;, &quot;ENSMUSG00000043592.15_forward&quot;, &quot;ENSMUSG00000023266.11_forward&quot;]: #val_ids[::8]:</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">attributes</span><span class="p">)</span> <span class="ow">in</span>  <span class="nb">enumerate</span><span class="p">(</span><span class="n">results_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="k">if</span> <span class="s2">&quot;training&quot;</span> <span class="ow">in</span> <span class="n">title</span><span class="p">:</span>
            <span class="c1">#print(title)</span>
            <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;path&#39;</span><span class="p">],</span>
                                                    <span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;N_BINS&#39;</span><span class="p">],</span>
                                                    <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> 
                                                    <span class="n">is_training</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;is_training&#39;</span><span class="p">],</span> 
                                                    <span class="n">batch_num</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;batch_num&#39;</span><span class="p">])</span>
            
            <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
            <span class="n">predictions</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;profile&#39;</span><span class="p">:</span><span class="n">line_p</span><span class="p">,</span>
                                <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                                <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">,</span>
                                <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Predicted </span><span class="si">{</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                                <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
            <span class="p">})</span>

            <span class="n">EU_seq_colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;k&quot;</span><span class="p">,</span><span class="s2">&quot;red&quot;</span><span class="p">]</span>
            <span class="c1"># Actual baseline signal:</span>
            <span class="n">actual_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
                <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="n">EU_seq_colors</span><span class="p">[</span><span class="n">k</span><span class="p">],</span>
                <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">,</span>
                <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;EU-seq </span><span class="si">{</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
            <span class="p">})</span>

    <span class="c1"># Create plot</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;Model_predictions_</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
</pre></div>
</div>
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
<img alt="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" src="../_images/ff73ec4d59e270c145232ab3ef0ee67fd131965bf528e665494c67adb64f656e.png" />
</div>
</div>
<p><strong>Visualize attributions of the new models:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/attributions/&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s1">&#39;font.size&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">})</span>  <span class="c1"># Adjust the size as needed</span>

<span class="k">def</span><span class="w"> </span><span class="nf">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">figure_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
                             <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">l_in</span><span class="p">:</span><span class="nb">int</span> <span class="o">=</span> <span class="mi">10001</span><span class="p">,</span>
                             <span class="n">n_in</span><span class="p">:</span> <span class="nb">int</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                             <span class="n">TRAIN_ATTR</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                             <span class="n">SPLIT</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is aimed to read attribution arrays from the chromatin landscape branch. Attribution arrays score the </span>
<span class="sd">    contribution of each position in the input arrays (4,10.001) towards each output node/position (401).</span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: path where the attribution arrays are stored.</span>
<span class="sd">        figure_path: path where we want to store the outputs</span>
<span class="sd">        track_dict: dictionary containing the names of the tracks and plot details</span>
<span class="sd">        N_BINS: number of bins to one side of the central bin (200 normally)</span>
<span class="sd">        l_in: input length </span>
<span class="sd">        n_in: number of input channels</span>
<span class="sd">        TRAIN_ATTR: whether this is a training or test run (folder structure changes).</span>
<span class="sd">        SPLIT: if Train, we need to know which of the stored batches it is</span>

<span class="sd">    Returns:</span>
<span class="sd">        Plots with attribution scores.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">N_IN_BINS</span> <span class="o">=</span> <span class="n">l_in</span> <span class="o">//</span> <span class="mi">2</span>
    <span class="n">x_val</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">N_IN_BINS</span><span class="p">,</span><span class="n">N_IN_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">xlim</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">500</span><span class="p">)</span>

    <span class="c1"># Getting attributions:</span>
    <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
    <span class="n">landscape_avg</span> <span class="o">=</span> <span class="p">{</span><span class="n">condition</span> <span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_in</span><span class="p">,</span><span class="n">l_in</span><span class="p">))</span> <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">}</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
            <span class="c1"># CHROMATIN MARK ATTRIBUTIONS:</span>
            <span class="n">added_folders</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;/samples/</span><span class="si">{</span><span class="n">SPLIT</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">TRAIN_ATTR</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="c1"># Different folder structure</span>
            <span class="n">att_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}{</span><span class="n">added_folders</span><span class="si">}</span><span class="s2">/attributions/gene_expression/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">att_path</span><span class="p">):</span>
                <span class="n">attributions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">att_path</span><span class="p">)</span>
                <span class="c1">#if (i == -73) and (condition == &quot;_ctrl&quot;):</span>
                <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span> 
                <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="n">j</span><span class="p">]),</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])))</span>
                    <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
                    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">attributions</span><span class="p">[</span><span class="mi">1</span><span class="p">])],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                <span class="c1">#fig.savefig(figure_path / f&quot;Chromatin_landscape_attributions_{i}_ctrl.png&quot;,dpi=200)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

                <span class="c1"># Add a small phase to remove high frequency fluctuations:</span>
                <span class="n">eps</span> <span class="o">=</span> <span class="mi">10</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                
                <span class="n">centered_attr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">attributions</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="o">*</span><span class="n">i</span><span class="o">+</span><span class="nb">int</span><span class="p">(</span><span class="n">eps</span><span class="p">),</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">centered_attr</span><span class="p">))</span>

    <span class="n">SCALING_FACTOR</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">([</span><span class="n">landscape_avg</span><span class="p">[</span><span class="n">cond</span><span class="p">]</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">])</span>

    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mf">2.7</span><span class="p">))</span>   <span class="c1"># </span>
        <span class="n">sub_axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">axes</span><span class="p">([</span><span class="mf">.615</span><span class="p">,</span> <span class="mf">.57</span><span class="p">,</span> <span class="mf">.27</span><span class="p">,</span> <span class="mf">.27</span><span class="p">])</span>  
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">track_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
            <span class="n">sub_axes</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_val</span><span class="p">,</span> <span class="n">landscape_avg</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">i</span><span class="p">]</span><span class="o">/</span><span class="n">SCALING_FACTOR</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">markersize</span><span class="o">=</span><span class="mf">.5</span><span class="p">,</span><span class="n">markeredgecolor</span><span class="o">=</span><span class="s2">&quot;none&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">])</span>
        
        <span class="c1">#sub_axes.set_xlim(xlim)</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5001</span><span class="p">,</span><span class="mi">5000</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span><span class="mi">501</span><span class="p">,</span><span class="mi">500</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">((</span><span class="o">-</span><span class="mi">5000</span><span class="p">,</span><span class="mi">5000</span><span class="p">))</span>
        <span class="n">sub_axes</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">xlim</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span> <span class="c1">#(0,.002))</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Attribution score&quot;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to the predicted locus (kbp)&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">400</span><span class="p">,</span><span class="mi">401</span><span class="p">,</span><span class="mi">100</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="mi">40</span><span class="p">,</span><span class="mi">41</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Landscape_absolute_average_</span><span class="si">{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="c1"># Visualize for the runs for which we have attributions available.</span>
<span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">attributes</span> <span class="ow">in</span> <span class="n">results_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="k">if</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;attributions&quot;</span><span class="p">]:</span>
        <span class="k">if</span> <span class="n">title</span> <span class="o">==</span> <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac test&quot;</span><span class="p">:</span> <span class="c1"># one input track less</span>
            <span class="n">track_dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
            <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
            <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>
            
            <span class="n">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;path&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="s2">&quot;expression_output&quot;</span><span class="p">,</span>
                                <span class="n">figure_path</span><span class="o">=</span><span class="n">figure_path</span><span class="p">,</span>
                                <span class="n">track_dict</span><span class="o">=</span><span class="n">track_dict</span><span class="p">,</span>
                                <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span>
                                <span class="n">l_in</span><span class="o">=</span> <span class="mi">10001</span><span class="p">,</span>
                                <span class="n">n_in</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
                                <span class="n">TRAIN_ATTR</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                                <span class="n">SPLIT</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">track_dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
            <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
            <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Enhancer&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;b&quot;</span><span class="p">},</span>
            <span class="mi">3</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>
            <span class="n">read_target_attributions</span><span class="p">(</span><span class="n">results_path</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;path&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="s2">&quot;expression_output&quot;</span><span class="p">,</span>
                    <span class="n">figure_path</span><span class="o">=</span><span class="n">figure_path</span><span class="p">,</span>
                    <span class="n">track_dict</span><span class="o">=</span><span class="n">track_dict</span><span class="p">,</span>
                    <span class="n">N_BINS</span> <span class="o">=</span> <span class="n">attributes</span><span class="p">[</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">],</span>
                    <span class="n">l_in</span><span class="o">=</span> <span class="mi">10001</span><span class="p">,</span>
                    <span class="n">n_in</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
                    <span class="n">TRAIN_ATTR</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                    <span class="n">SPLIT</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/scratch/ipykernel_3833147/2664724846.py:62: DeprecationWarning: Conversion of an array with ndim &gt; 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  centered_attr = np.roll(attributions, -10*i+int(eps),axis=1)
/scratch/ipykernel_3833147/2664724846.py:85: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  plt.tight_layout()
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="v-performance-evaluations">
<h2>V) Performance evaluations:<a class="headerlink" href="#v-performance-evaluations" title="Link to this heading">#</a></h2>
<p>The following section is aimed to provide a more comprehensive and rigorous overview of CLASTER’s predictive performance.</p>
<ul class="simple">
<li><p>Performance is evaluated using <span class="math notranslate nohighlight">\(R^2\)</span>, <span class="math notranslate nohighlight">\(MSE\)</span>, <span class="math notranslate nohighlight">\(RMSE\)</span>, <span class="math notranslate nohighlight">\(MAE\)</span>, <span class="math notranslate nohighlight">\(Pearson\)</span> and <span class="math notranslate nohighlight">\(Spearman\)</span> for all CLASTER models.</p></li>
<li><p>Performance is reported before and after zero clipping using the ReLU activation.</p></li>
<li><p>Perfomance is reported with and without log2(1+x) transforming the clipped predictions.</p></li>
</ul>
<p><strong>Showing that unclipped profiles show values close to zero.</strong></p>
<p>Predictions below zero are not too negative. Therefore, the difference between clipped and unclipped profiles in terms of standard performance metrics is not that pronounced.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">)</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>

<span class="n">EXAMPLE_GENE</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000003032.8_forward.npy&quot;</span> <span class="c1">#&lt;- KLF4 # Akirin2: &quot;ENSMUSG00000028291.7_forward.npy&quot;</span>

<span class="n">input_id_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSMUSG00000028291.7_forward&quot;</span><span class="p">]</span>

<span class="c1"># Outputs:</span>
<span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">outputs_path</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">condition_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> <span class="n">CLIP</span><span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

<span class="k">for</span> <span class="n">ID</span><span class="p">,</span><span class="n">line_p</span><span class="p">,</span><span class="n">line_a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">,</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>   
    <span class="k">if</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">input_id_list</span><span class="p">:</span>
        <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
        <span class="n">line_a</span> <span class="o">=</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_target_predictions</span><span class="p">(</span><span class="n">line_p</span><span class="p">,</span> <span class="n">line_a</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_prediction_unclipped.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Calculating new performance metrics:</strong></p>
<ul class="simple">
<li><p>We will use parallel processing to speed up computations</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_evaluation_metrics_table.py
<span class="ch">#!/usr/bin/env python</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mp</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">r2_score</span><span class="p">,</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">mean_absolute_error</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Any</span>

<span class="k">def</span><span class="w"> </span><span class="nf">_get_predictions</span><span class="p">(</span>
    <span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">condition_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">is_training</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">batch_num</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">CLIP</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads prediction files from EIR for each target bin condition.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predicted_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">batch_num</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;batch_num must be specified for training predictions&quot;</span><span class="p">)</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/samples/</span><span class="si">{</span><span class="n">batch_num</span><span class="si">}</span><span class="s2">/regression_predictions.csv&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/predictions.csv&quot;</span>
            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prediction file not found: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>

            <span class="c1"># Apply ReLU to model predictions if specified</span>
            <span class="k">if</span> <span class="n">CLIP</span><span class="p">:</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Handle different column names for training vs test</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2"> Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;True Label Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">predicted_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_column</span><span class="p">)</span>
            <span class="n">actual_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_column</span><span class="p">)</span>

    <span class="c1"># Concatenate all DataFrames horizontally</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predicted_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">actual</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">actual_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span>

<span class="k">def</span><span class="w"> </span><span class="nf">evaluate_model</span><span class="p">(</span><span class="n">args</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="nb">bool</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Evaluate a model using R2, MSE, RMSE, and MAE metrics</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">title</span><span class="p">,</span> <span class="n">model_config</span><span class="p">,</span> <span class="n">is_clipping</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">state</span> <span class="o">=</span> <span class="s2">&quot;clipped&quot;</span> <span class="k">if</span> <span class="n">is_clipping</span> <span class="k">else</span> <span class="s2">&quot;linear&quot;</span>
    
    <span class="c1"># Extract parameters from model config</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;path&quot;</span><span class="p">]</span>
    <span class="n">is_training</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;is_training&quot;</span><span class="p">]</span>
    <span class="n">n_bins</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span>
    <span class="n">batch_num</span> <span class="o">=</span> <span class="n">model_config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;batch_num&quot;</span><span class="p">)</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Get predictions with the appropriate parameters</span>
        <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span>
            <span class="n">path</span><span class="p">,</span> 
            <span class="n">N_BINS</span><span class="o">=</span><span class="n">n_bins</span><span class="p">,</span> 
            <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> 
            <span class="n">is_training</span><span class="o">=</span><span class="n">is_training</span><span class="p">,</span> 
            <span class="n">batch_num</span><span class="o">=</span><span class="n">batch_num</span><span class="p">,</span> 
            <span class="n">CLIP</span><span class="o">=</span><span class="n">is_clipping</span>
        <span class="p">)</span>
        
        <span class="c1"># Calculate metrics</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;R2_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">r2_score</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;MSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean_squared_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;RMSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mean_squared_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">)),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;MAE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">)</span>
        <span class="p">}</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error evaluating </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;R2_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;MSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;RMSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;MAE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="p">}</span>
    
    <span class="k">return</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">parallel_model_evaluation</span><span class="p">(</span><span class="n">models_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span> 
                           <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> 
                           <span class="n">n_processes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Evaluate multiple models in parallel</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create output directory</span>
    <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="n">results_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Prepare arguments for parallel processing</span>
    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">config</span> <span class="ow">in</span> <span class="n">models_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1"># Add task for linear evaluation</span>
        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">title</span><span class="p">,</span> <span class="n">config</span><span class="p">,</span> <span class="kc">False</span><span class="p">))</span>
        <span class="c1"># Add task for clipped evaluation</span>
        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">title</span><span class="p">,</span> <span class="n">config</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span>
    
    <span class="c1"># Initialize DataFrame for results</span>
    <span class="n">r2_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R2_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;R2_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;MSE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;MSE_clipped&#39;</span><span class="p">,</span> 
                <span class="s1">&#39;RMSE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;RMSE_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;MAE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;MAE_clipped&#39;</span><span class="p">],</span> 
        <span class="n">index</span><span class="o">=</span><span class="n">models_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">)</span>
    
    <span class="c1"># Run evaluations in parallel</span>
    <span class="k">with</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_processes</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">evaluate_model</span><span class="p">,</span> <span class="n">tasks</span><span class="p">)</span>
    
    <span class="c1"># Process results</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">metric_name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">metrics</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">r2_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">title</span><span class="p">,</span> <span class="n">metric_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    
    <span class="c1"># Save results to file</span>
    <span class="n">r2_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">results_path</span> <span class="o">/</span> <span class="s2">&quot;performance_metrics_comparison.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">r2_df</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c1"># Define paths</span>
    <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Revised_performance_metrics/&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define models to evaluate</span>
    <span class="n">results_path_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv/results/&quot;</span><span class="p">),</span>
                                                      <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                      <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                      <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span> 
                                                      <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Gaussian&quot;</span><span class="p">,</span>
                                                      <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolution test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">),</span>
                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;CLASTER test&quot;</span><span class="p">,</span>
                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin with Attention training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_attention/results/&quot;</span><span class="p">),</span>
                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin with Attention test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_attention/&quot;</span><span class="p">),</span>
                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_pure_conv/results/&quot;</span><span class="p">),</span>
                                            <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                            <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                            <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                            <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                            <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and structure&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_pure_conv/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure rotated training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_rotated_pure_conv/results/&quot;</span><span class="p">),</span>
                                                   <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                   <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                   <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                   <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                   <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                                   <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure rotated test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_rotated_pure_conv/&quot;</span><span class="p">),</span>
                                               <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                               <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                               <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                               <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                               <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                               <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus SmoothL1 training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_no_bias/results/&quot;</span><span class="p">),</span>
                                                                       <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                       <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                       <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                       <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                       <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus SmoothL1&quot;</span><span class="p">,</span>
                                                                       <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;purple&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias/results/&quot;</span><span class="p">),</span>
                                                                     <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                     <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                     <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                     <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                     <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson&quot;</span><span class="p">,</span>
                                                                     <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;orange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson log_input=False training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias_log_input_false/results/&quot;</span><span class="p">),</span>
                                                                                      <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson log input false&quot;</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkred&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_20kbp_context/results/&quot;</span><span class="p">),</span>
                                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train 20kb context&quot;</span><span class="p">,</span>
                                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;yellow&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_20kbp_context_test&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 20kb context&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;green&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_chr19_holdout/results/&quot;</span><span class="p">),</span>
                                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train chr19 holdout&quot;</span><span class="p">,</span>
                                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_chr19_holdout_test/&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test chr19 holdout&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_no_H3K27ac/results/&quot;</span><span class="p">),</span>
                                                               <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                               <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                               <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                               <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                               <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train no H3K27ac&quot;</span><span class="p">,</span>
                                                               <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_no_H3K27ac_test/&quot;</span><span class="p">),</span>
                                                           <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                           <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                           <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                           <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                           <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test no H3K27ac&quot;</span><span class="p">,</span>
                                                           <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_1kbp_binning/results/&quot;</span><span class="p">),</span>
                                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;1kbp binning&quot;</span><span class="p">,</span>
                                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_1kbp_binning_test/&quot;</span><span class="p">),</span>
                                                             <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                             <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                             <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                             <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                             <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 1kbp binning&quot;</span><span class="p">,</span>
                                                             <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions K562 RNA-seq training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_train/results/&quot;</span><span class="p">),</span>
                                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 RNA-seq&quot;</span><span class="p">,</span>
                                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions K562 POLR2A training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_POLR2A_train/results/&quot;</span><span class="p">),</span>
                                                                <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 POLR2A&quot;</span><span class="p">,</span>
                                                                <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Prom-CHiC training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_prom_CHiC_pure_conv/results/&quot;</span><span class="p">),</span>
                                            <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                            <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                            <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                            <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                            <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom-CHiC&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Prom-CHiC test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_prom_CHiC_pure_conv_test/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom CHiC test&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># Run evaluation with parallel processing</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">parallel_model_evaluation</span><span class="p">(</span><span class="n">results_path_dict</span><span class="p">,</span> <span class="n">results_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Adding Pearson and Spearman correlations:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_evaluation_metrics_table.py
<span class="ch">#!/usr/bin/env python</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">mp</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">r2_score</span><span class="p">,</span> <span class="n">mean_squared_error</span><span class="p">,</span> <span class="n">mean_absolute_error</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">spearmanr</span><span class="p">,</span> <span class="n">pearsonr</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Any</span>

<span class="k">def</span><span class="w"> </span><span class="nf">_get_predictions</span><span class="p">(</span>
    <span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">condition_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">is_training</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">batch_num</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">CLIP</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">include_rev</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads prediction files from EIR for each target bin condition.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: Path to the results directory</span>
<span class="sd">        N_BINS: Number of bins to read</span>
<span class="sd">        condition_list: List of conditions to read</span>
<span class="sd">        is_training: Whether to read training or test predictions</span>
<span class="sd">        batch_num: Batch number for training predictions</span>
<span class="sd">        CLIP: Whether to clip predictions to non-negative values</span>
<span class="sd">        include_rev: Whether to include reversed samples (for data augmentation)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predicted_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">batch_num</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;batch_num must be specified for training predictions&quot;</span><span class="p">)</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/samples/</span><span class="si">{</span><span class="n">batch_num</span><span class="si">}</span><span class="s2">/regression_predictions.csv&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/predictions.csv&quot;</span>
            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prediction file not found: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
            
            <span class="c1"># Filter out samples with IDs containing &quot;_rev&quot; if requested</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">include_rev</span><span class="p">:</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="o">~</span><span class="n">predictions</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;_rev&quot;</span><span class="p">)]</span>

            <span class="c1"># Apply ReLU to model predictions if specified</span>
            <span class="k">if</span> <span class="n">CLIP</span><span class="p">:</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Handle different column names for training vs test</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2"> Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;True Label Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">predicted_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_column</span><span class="p">)</span>
            <span class="n">actual_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_column</span><span class="p">)</span>

    <span class="c1"># Concatenate all DataFrames horizontally</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predicted_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">actual</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">actual_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span>

<span class="k">def</span><span class="w"> </span><span class="nf">evaluate_model</span><span class="p">(</span><span class="n">args</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">],</span> <span class="nb">bool</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Evaluate a model using R2, MSE, RMSE, MAE, Spearman correlation, and Pearson correlation metrics</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">title</span><span class="p">,</span> <span class="n">model_config</span><span class="p">,</span> <span class="n">is_clipping</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">state</span> <span class="o">=</span> <span class="s2">&quot;clipped&quot;</span> <span class="k">if</span> <span class="n">is_clipping</span> <span class="k">else</span> <span class="s2">&quot;linear&quot;</span>
    
    <span class="c1"># Extract parameters from model config</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;path&quot;</span><span class="p">]</span>
    <span class="n">is_training</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;is_training&quot;</span><span class="p">]</span>
    <span class="n">n_bins</span> <span class="o">=</span> <span class="n">model_config</span><span class="p">[</span><span class="s2">&quot;N_BINS&quot;</span><span class="p">]</span>
    <span class="n">batch_num</span> <span class="o">=</span> <span class="n">model_config</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;batch_num&quot;</span><span class="p">)</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Get predictions with the appropriate parameters</span>
        <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span>
            <span class="n">path</span><span class="p">,</span> 
            <span class="n">N_BINS</span><span class="o">=</span><span class="n">n_bins</span><span class="p">,</span> 
            <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> 
            <span class="n">is_training</span><span class="o">=</span><span class="n">is_training</span><span class="p">,</span> 
            <span class="n">batch_num</span><span class="o">=</span><span class="n">batch_num</span><span class="p">,</span> 
            <span class="n">CLIP</span><span class="o">=</span><span class="n">is_clipping</span>
        <span class="p">)</span>
        
        <span class="c1"># Flatten matrices for correlation calculations</span>
        <span class="n">actual_flat</span> <span class="o">=</span> <span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">predicted_flat</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        
        <span class="c1"># Calculate metrics</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;R2_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">r2_score</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;MSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean_squared_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;RMSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mean_squared_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">)),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;MAE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">mean_absolute_error</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="p">),</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;Spearman_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">actual_flat</span><span class="p">,</span> <span class="n">predicted_flat</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">4</span><span class="p">),</span>
            <span class="sa">f</span><span class="s1">&#39;Pearson_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="nb">round</span><span class="p">(</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">actual_flat</span><span class="p">,</span> <span class="n">predicted_flat</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">4</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="c1"># Calculate metrics log2(x+1): </span>
        <span class="c1"># metrics = {</span>
        <span class="c1">#     f&#39;R2_{state}&#39;: round(r2_score(np.log2(actual.values+1), np.log2(predicted.values+1)), 4),</span>
        <span class="c1">#     f&#39;MSE_{state}&#39;: round(mean_squared_error(np.log2(actual.values+1), np.log2(predicted.values+1)), 4),</span>
        <span class="c1">#     f&#39;RMSE_{state}&#39;: round(np.sqrt(mean_squared_error(np.log2(actual.values+1), np.log2(predicted.values+1))), 4),</span>
        <span class="c1">#     f&#39;MAE_{state}&#39;: round(mean_absolute_error(np.log2(actual.values+1), np.log2(predicted.values+1)), 4),</span>
        <span class="c1">#     f&#39;Spearman_{state}&#39;: round(spearmanr(np.log2(actual_flat+1), np.log2(predicted_flat+1))[0], 4),</span>
        <span class="c1">#     f&#39;Pearson_{state}&#39;: round(pearsonr(np.log2(actual_flat+1), np.log2(predicted_flat+1))[0], 4)</span>
        <span class="c1"># }</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error evaluating </span><span class="si">{</span><span class="n">title</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">metrics</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;R2_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;MSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;RMSE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;MAE_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;Spearman_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span>
            <span class="sa">f</span><span class="s1">&#39;Pearson_</span><span class="si">{</span><span class="n">state</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="p">}</span>
    
    <span class="k">return</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">parallel_model_evaluation</span><span class="p">(</span><span class="n">models_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">]],</span> 
                           <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> 
                           <span class="n">n_processes</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Evaluate multiple models in parallel</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create output directory</span>
    <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="n">results_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Prepare arguments for parallel processing</span>
    <span class="n">tasks</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">config</span> <span class="ow">in</span> <span class="n">models_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1"># Add task for linear evaluation</span>
        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">title</span><span class="p">,</span> <span class="n">config</span><span class="p">,</span> <span class="kc">False</span><span class="p">))</span>
        <span class="c1"># Add task for clipped evaluation</span>
        <span class="n">tasks</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">title</span><span class="p">,</span> <span class="n">config</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span>
    
    <span class="c1"># Initialize DataFrame for results</span>
    <span class="n">r2_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;R2_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;R2_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;MSE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;MSE_clipped&#39;</span><span class="p">,</span> 
                <span class="s1">&#39;RMSE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;RMSE_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;MAE_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;MAE_clipped&#39;</span><span class="p">,</span>
                <span class="s1">&#39;Spearman_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;Spearman_clipped&#39;</span><span class="p">,</span> <span class="s1">&#39;Pearson_linear&#39;</span><span class="p">,</span> <span class="s1">&#39;Pearson_clipped&#39;</span><span class="p">],</span> 
        <span class="n">index</span><span class="o">=</span><span class="n">models_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
    <span class="p">)</span>
    
    <span class="c1"># Run evaluations in parallel</span>
    <span class="k">with</span> <span class="n">mp</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_processes</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">evaluate_model</span><span class="p">,</span> <span class="n">tasks</span><span class="p">)</span>
    
    <span class="c1"># Process results</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">metrics</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">metric_name</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">metrics</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">r2_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">title</span><span class="p">,</span> <span class="n">metric_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
    
    <span class="c1"># Save results to file</span>
    <span class="n">r2_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">results_path</span> <span class="o">/</span> <span class="s2">&quot;performance_metrics_comparison.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">r2_df</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c1"># Define paths</span>
    <span class="n">results_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../benchmarks/Revised_performance_metrics/&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define models to evaluate</span>
    <span class="n">results_path_dict</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv/results/&quot;</span><span class="p">),</span>
                                                      <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                      <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                      <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span> 
                                                      <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Gaussian&quot;</span><span class="p">,</span>
                                                      <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolution test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv/&quot;</span><span class="p">),</span>
                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;CLASTER test&quot;</span><span class="p">,</span>
                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin with Attention training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_attention/results/&quot;</span><span class="p">),</span>
                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin with Attention test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_attention/&quot;</span><span class="p">),</span>
                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Only chrom attention&quot;</span><span class="p">,</span>
                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_pure_conv/results/&quot;</span><span class="p">),</span>
                                            <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                            <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                            <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                            <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                            <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and structure&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_pure_conv/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure rotated training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_microc_rotated_pure_conv/results/&quot;</span><span class="p">),</span>
                                                   <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                   <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                   <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                   <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                   <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                                   <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Structure rotated test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_microc_rotated_pure_conv/&quot;</span><span class="p">),</span>
                                               <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                               <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                               <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                               <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                               <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Chrom and Structure rotated&quot;</span><span class="p">,</span>
                                               <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus SmoothL1 training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_no_bias/results/&quot;</span><span class="p">),</span>
                                                                       <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                       <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                       <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                       <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                       <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus SmoothL1&quot;</span><span class="p">,</span>
                                                                       <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;purple&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias/results/&quot;</span><span class="p">),</span>
                                                                     <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                     <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                     <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                     <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                     <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson&quot;</span><span class="p">,</span>
                                                                     <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;orange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions Softplus Poisson log_input=False training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_softplus_poisson_no_bias_log_input_false/results/&quot;</span><span class="p">),</span>
                                                                                      <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train Softplus Poisson log input false&quot;</span><span class="p">,</span>
                                                                                      <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkred&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_20kbp_context/results/&quot;</span><span class="p">),</span>
                                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train 20kb context&quot;</span><span class="p">,</span>
                                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;yellow&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 20kbp context test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_20kbp_context_test&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 20kb context&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;green&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_chr19_holdout/results/&quot;</span><span class="p">),</span>
                                                                  <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                  <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                  <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                  <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                  <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train chr19 holdout&quot;</span><span class="p">,</span>
                                                                  <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions chr19 holdout test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_chr19_holdout_test/&quot;</span><span class="p">),</span>
                                                              <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                              <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                              <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                              <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                              <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test chr19 holdout&quot;</span><span class="p">,</span>
                                                              <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;lightblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_no_H3K27ac/results/&quot;</span><span class="p">),</span>
                                                               <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                               <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                               <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                               <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                               <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;train no H3K27ac&quot;</span><span class="p">,</span>
                                                               <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions no H3K27ac test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_no_H3K27ac_test/&quot;</span><span class="p">),</span>
                                                           <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                           <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                           <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                           <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                           <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test no H3K27ac&quot;</span><span class="p">,</span>
                                                           <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkgreen&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_pure_conv_1kbp_binning/results/&quot;</span><span class="p">),</span>
                                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;1kbp binning&quot;</span><span class="p">,</span>
                                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions 1kbp binning test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_only_chrom_pure_conv_1kbp_binning_test/&quot;</span><span class="p">),</span>
                                                             <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                             <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                             <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                                             <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                             <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;test 1kbp binning&quot;</span><span class="p">,</span>
                                                             <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions K562 RNA-seq training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_train/results/&quot;</span><span class="p">),</span>
                                                                 <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                 <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                 <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                                                 <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                 <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 RNA-seq&quot;</span><span class="p">,</span>
                                                                 <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkorange&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Only Chromatin Pure Convolutions K562 POLR2A training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_only_chrom_K562_POLR2A_train/results/&quot;</span><span class="p">),</span>
                                                                <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                                                <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                                                <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">30300</span><span class="p">,</span>
                                                                <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                                                <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;K562 POLR2A&quot;</span><span class="p">,</span>
                                                                <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Prom-CHiC training&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/gene_expression_prom_CHiC_pure_conv/results/&quot;</span><span class="p">),</span>
                                            <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                            <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                            <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="mi">60600</span><span class="p">,</span>
                                            <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                            <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom-CHiC&quot;</span><span class="p">,</span>
                                            <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;blue&quot;</span><span class="p">},</span>
        <span class="s2">&quot;Chromatin and Prom-CHiC test&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;path&quot;</span><span class="p">:</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../runs/test_runs/gene_expression_prom_CHiC_pure_conv_test/&quot;</span><span class="p">),</span>
                                        <span class="s2">&quot;is_training&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                                        <span class="s2">&quot;N_BINS&quot;</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span>
                                        <span class="s2">&quot;batch_num&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                        <span class="s2">&quot;attributions&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                                        <span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="s2">&quot;Prom CHiC test&quot;</span><span class="p">,</span>
                                        <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;darkblue&quot;</span><span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># Run evaluation with parallel processing</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">parallel_model_evaluation</span><span class="p">(</span><span class="n">results_path_dict</span><span class="p">,</span> <span class="n">results_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overwriting create_evaluation_metrics_table.py
</pre></div>
</div>
</div>
</div>
</section>
<section id="vi-data-distributions">
<h2>VI) Data distributions<a class="headerlink" href="#vi-data-distributions" title="Link to this heading">#</a></h2>
<p><strong>Visualizing data distributions:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_data_distribution_plot.py
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/data_distributions_K562/&quot;</span><span class="p">)</span>
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_single_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="p">):</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">if</span> <span class="kc">True</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">,:]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mark</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">marks</span><span class="p">)}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_data</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">marks</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3&quot;</span><span class="p">]</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
    
    <span class="n">load_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">load_single_file</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="o">=</span><span class="n">marks</span><span class="p">)</span>
    
    <span class="n">n_cores</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_cores</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">load_func</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
    
    <span class="n">combined</span> <span class="o">=</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">mark</span> <span class="ow">in</span> <span class="n">marks</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">mark</span> <span class="ow">in</span> <span class="n">marks</span><span class="p">:</span>
            <span class="n">combined</span><span class="p">[</span><span class="n">mark</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">mark</span><span class="p">])</span>
            
    <span class="k">return</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">arrays</span><span class="p">)</span> <span class="k">for</span> <span class="n">mark</span><span class="p">,</span> <span class="n">arrays</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<span class="c1"># Load Micro-C data - using same loading function but with single mark</span>
<span class="k">def</span><span class="w"> </span><span class="nf">load_microc_data</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
    
    <span class="n">load_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">load_single_file</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">])</span>
    
    <span class="n">n_cores</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_cores</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">load_func</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
    
    <span class="n">combined</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:</span> <span class="p">[]}</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="n">combined</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">])</span>
            
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">arrays</span><span class="p">)</span> <span class="k">for</span> <span class="n">mark</span><span class="p">,</span> <span class="n">arrays</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<span class="c1"># Define paths</span>
<span class="n">training_inputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/K562/training/&quot;</span><span class="p">)</span>
<span class="n">test_inputs_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/K562/test/&quot;</span><span class="p">)</span>
<span class="c1">#training_microc_path = Path(&quot;../inputs/microC/training/&quot;)</span>
<span class="c1">#test_microc_path = Path(&quot;../inputs/microC/test/&quot;)</span>
<span class="n">training_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/training/&quot;</span><span class="p">)</span>
<span class="n">test_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/test/&quot;</span><span class="p">)</span>

<span class="c1"># Load data in parallel</span>
<span class="n">train</span> <span class="o">=</span> <span class="n">load_data</span><span class="p">(</span><span class="n">training_inputs_path</span><span class="p">)</span>
<span class="n">test</span> <span class="o">=</span> <span class="n">load_data</span><span class="p">(</span><span class="n">test_inputs_path</span><span class="p">)</span>

<span class="c1"># Load EU-seq data from CSV files</span>
<span class="n">train_euseq</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/K562/training_targets.csv&quot;</span><span class="p">,</span>  <span class="n">index_col</span><span class="o">=</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="n">test_euseq</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/K562/test_targets.csv&quot;</span><span class="p">,</span>  <span class="n">index_col</span><span class="o">=</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="c1"># Load Micro-C data</span>
<span class="n">train_microc</span> <span class="o">=</span> <span class="n">load_microc_data</span><span class="p">(</span><span class="n">training_microc_path</span><span class="p">)</span>
<span class="n">test_microc</span> <span class="o">=</span> <span class="n">load_microc_data</span><span class="p">(</span><span class="n">test_microc_path</span><span class="p">)</span>

<span class="c1"># Create and plot subplots (2x3 grid to accommodate the EU-seq plot)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

<span class="n">marks</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3&quot;</span><span class="p">]</span>

<span class="c1"># Define custom bounds for each mark</span>
<span class="n">bounds</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ATAC&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K4me3&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K27ac&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K27me3&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="c1">#&quot;EU-seq&quot;: (0, 400),</span>
    <span class="s1">&#39;RNA-seq&#39;</span><span class="p">:(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1000</span><span class="p">),</span>
    <span class="c1">#&quot;POLR2A&quot;: (0,1000),</span>
    <span class="c1">#&quot;Micro-C&quot;: (-15,2),</span>
    <span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:(</span><span class="o">-</span><span class="mi">15</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">eu_seq_ax</span> <span class="o">=</span> <span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>  <span class="c1"># Move EU-seq to position 5</span>

<span class="k">for</span> <span class="p">(</span><span class="n">mark</span><span class="p">,</span> <span class="n">ax</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">marks</span><span class="p">,</span> <span class="n">axes</span><span class="o">.</span><span class="n">flat</span><span class="p">):</span>
    <span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span> <span class="o">=</span> <span class="n">bounds</span><span class="p">[</span><span class="n">mark</span><span class="p">]</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Wider bins (step of 2 instead of 0.5)</span>
    
    <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">train</span><span class="p">[</span><span class="n">mark</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
            <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Train</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">train</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">train</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">train</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> 
            <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">test</span><span class="p">[</span><span class="n">mark</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> 
            <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Test</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">test</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">test</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">test</span><span class="p">[</span><span class="n">mark</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> 
            <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
    
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">mark</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.8</span><span class="p">,</span> <span class="mf">.999</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Plot EU-seq distributions</span>
<span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span> <span class="o">=</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;RNA-seq&quot;</span><span class="p">]</span>
<span class="n">bins_euseq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Same wider bins for consistency</span>

<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">train_euseq</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_euseq</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Train</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">train_euseq</span><span class="p">))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">train_euseq</span><span class="p">))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">train_euseq</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">test_euseq</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_euseq</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Test</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">test_euseq</span><span class="p">))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">test_euseq</span><span class="p">))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">test_euseq</span><span class="p">))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">)</span>
<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;RNA-seq&#39;</span><span class="p">)</span>
<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">eu_seq_ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.8</span><span class="p">,</span> <span class="mf">.99</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Data_distribution-RNA.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

<span class="c1"># Micro-C</span>
<span class="c1"># Before creating the EU-seq subplot, add Micro-C plotting:</span>
<span class="n">fig_microC</span><span class="p">,</span> <span class="n">microc_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span> <span class="c1"># Use position 4 for Micro-C</span>
<span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span> <span class="o">=</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]</span>
<span class="n">bins_microc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

<span class="n">microc_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_microc</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Train</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_microc</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Test</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">1e-6</span><span class="p">,</span><span class="mf">2e1</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;prom-CHiC&#39;</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.3</span><span class="p">,</span> <span class="mf">.99</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">fig_microC</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;prom_CHiC_distribution.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Overwriting create_data_distribution_plot.py
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Before creating the EU-seq subplot, add Micro-C plotting:</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>

<span class="n">training_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/training/&quot;</span><span class="p">)</span>
<span class="n">test_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/test/&quot;</span><span class="p">)</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/data_distributions_K562/&quot;</span><span class="p">)</span>
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_single_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="p">):</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">if</span> <span class="kc">True</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">,:]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">mark</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">marks</span><span class="p">)}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_data</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">marks</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3&quot;</span><span class="p">]</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
    
    <span class="n">load_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">load_single_file</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="o">=</span><span class="n">marks</span><span class="p">)</span>
    
    <span class="n">n_cores</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_cores</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">load_func</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
    
    <span class="n">combined</span> <span class="o">=</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">mark</span> <span class="ow">in</span> <span class="n">marks</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">mark</span> <span class="ow">in</span> <span class="n">marks</span><span class="p">:</span>
            <span class="n">combined</span><span class="p">[</span><span class="n">mark</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="n">mark</span><span class="p">])</span>
            
    <span class="k">return</span> <span class="p">{</span><span class="n">mark</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">arrays</span><span class="p">)</span> <span class="k">for</span> <span class="n">mark</span><span class="p">,</span> <span class="n">arrays</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<span class="c1"># Define custom bounds for each mark</span>
<span class="n">bounds</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ATAC&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K4me3&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K27ac&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="s2">&quot;H3K27me3&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1000</span><span class="p">),</span>
    <span class="c1">#&quot;EU-seq&quot;: (0, 400),</span>
    <span class="s1">&#39;RNA-seq&#39;</span><span class="p">:(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1000</span><span class="p">),</span>
    <span class="c1">#&quot;POLR2A&quot;: (0,1000),</span>
    <span class="c1">#&quot;Micro-C&quot;: (-15,2),</span>
    <span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:(</span><span class="o">-</span><span class="mi">15</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="p">}</span>

<span class="c1"># Load Micro-C data - using same loading function but with single mark</span>
<span class="k">def</span><span class="w"> </span><span class="nf">load_microc_data</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">array</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">array</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
    
    <span class="n">load_func</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">load_single_file</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">,</span> <span class="n">marks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">])</span>
    
    <span class="n">n_cores</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_cores</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">load_func</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
    
    <span class="n">combined</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:</span> <span class="p">[]}</span>
    <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="n">combined</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">])</span>
            
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">arrays</span><span class="p">)</span> <span class="k">for</span> <span class="n">mark</span><span class="p">,</span> <span class="n">arrays</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

<span class="c1"># Load Micro-C data</span>
<span class="n">train_microc</span> <span class="o">=</span> <span class="n">load_microc_data</span><span class="p">(</span><span class="n">training_microc_path</span><span class="p">)</span>
<span class="n">test_microc</span> <span class="o">=</span> <span class="n">load_microc_data</span><span class="p">(</span><span class="n">test_microc_path</span><span class="p">)</span>

<span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>

<span class="n">training_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/training/&quot;</span><span class="p">)</span>
<span class="n">test_microc_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/test/&quot;</span><span class="p">)</span>

<span class="n">fig_microC</span><span class="p">,</span> <span class="n">microc_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span> <span class="c1"># Use position 4 for Micro-C</span>
<span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span> <span class="o">=</span> <span class="n">bounds</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]</span>
<span class="n">bins_microc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">min_bound</span><span class="p">,</span> <span class="n">max_bound</span><span class="p">,</span>  <span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>

<span class="n">microc_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_microc</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Train</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">train_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins_microc</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Test</span><span class="se">\n</span><span class="s1">Min: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Max: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="se">\n</span><span class="s1">Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">test_microc</span><span class="p">[</span><span class="s2">&quot;prom-CHiC&quot;</span><span class="p">]))</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
               <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mf">1e-6</span><span class="p">,</span><span class="mf">2e1</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;prom-CHiC&#39;</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">.3</span><span class="p">,</span> <span class="mf">.99</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper center&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">microc_ax</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;auto&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">fig_microC</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;prom_CHiC_distribution.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/bfcce55d67acc58e9b442d7be0458a49ac41d5b5a298e9af1510b084a19bb2c0.png" src="../_images/bfcce55d67acc58e9b442d7be0458a49ac41d5b5a298e9af1510b084a19bb2c0.png" />
</div>
</div>
<p><strong>Complementary analysis of Micro-C data distribution:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_average_microc_matrix.py
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">multiprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Pool</span><span class="p">,</span> <span class="n">cpu_count</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">functools</span><span class="w"> </span><span class="kn">import</span> <span class="n">partial</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_single_file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">path</span><span class="p">):</span>
    <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="n">filename</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">array</span>

<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/microC/test/&quot;</span><span class="p">)</span>
    
    <span class="c1"># Get list of files</span>
    <span class="n">filenames</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">path</span><span class="p">)]</span>
    <span class="n">scaling_factor</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">filenames</span><span class="p">)</span>
    
    <span class="c1"># Setup parallel processing</span>
    <span class="n">n_cores</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="n">process_file</span> <span class="o">=</span> <span class="n">partial</span><span class="p">(</span><span class="n">process_single_file</span><span class="p">,</span> <span class="n">path</span><span class="o">=</span><span class="n">path</span><span class="p">)</span>
    
    <span class="c1"># Process files in parallel</span>
    <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">n_cores</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_file</span><span class="p">,</span> <span class="n">filenames</span><span class="p">)</span>
    
    <span class="c1"># Sum all arrays and apply scaling</span>
    <span class="n">avg_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">626</span><span class="p">,</span> <span class="mi">626</span><span class="p">))</span>
    <span class="n">median_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="mi">626</span><span class="p">,</span> <span class="mi">626</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">array</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">results</span><span class="p">):</span>
        <span class="n">avg_array</span> <span class="o">+=</span> <span class="n">array</span> <span class="o">/</span> <span class="n">scaling_factor</span>
        <span class="n">median_array</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">array</span>

    <span class="c1"># Plot mean array</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">avg_array</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s1">&#39;Average_test_MicroC.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># Plot median array</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">median_array</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;jet&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s1">&#39;Median_test_MicroC.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>


    <span class="c1"># Plot histogram of values</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">avg_array</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Value&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Density&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Distribution of Values in Average Micro-C Matrix&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s1">&#39;Average_test_MicroC_histogram.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="vii-promoter-capture-hi-c-analyses">
<h2>VII) Promoter-Capture Hi-C analyses<a class="headerlink" href="#vii-promoter-capture-hi-c-analyses" title="Link to this heading">#</a></h2>
<p><strong>Process the raw data to obtain cooler contact matrices.</strong></p>
<ul class="simple">
<li><p>Obtain the SRA file with the raw reads corresponding to mESC Promoter-Capture HiC data:
<a class="reference external" href="https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&amp;amp;acc=SRR5972844&amp;amp;display=data-access">SRA file</a></p></li>
<li><p>Dump these into two separate fastq files (to build the 2D histograms).</p></li>
</ul>
<p>The original microC matrices were obtained at a 1600bp resolution. Promoter Capture Hi-C has lower sequencing depth, and therefore we lowered the resolution to 5kbp in order to increase the signal intensity. The scripts to obtain cooler files can be found inside the <code class="docutils literal notranslate"><span class="pre">prom_CHiC_processing</span></code> folder.</p>
<p>Minimum and maximum values in the Promoter-Capture Hi-C dataset (at a 5kbp resolution) were 8e-4 and 8.47(log10 transformed were -3.1 and 0.92). In a similar fashion as for micro-C data, we imputed with 1e-14 to keep the imputation consistent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">cooler</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="k">def</span><span class="w"> </span><span class="nf">find_dataset_min_max</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">))</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_histogram_dataset</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">),</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">mcool_summary</span><span class="p">(</span><span class="n">microc_path</span><span class="p">):</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>

    <span class="c1"># Inspecting what is inside the cooler object:</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">info</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">chromnames</span>
    <span class="n">extent</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">extent</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Metadata:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Extent:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="p">)</span>

<span class="n">input_path</span> <span class="o">=</span> <span class="s2">&quot;../GEO_files/SRR5972844.mm10.mapq_30.100.mcool::/resolutions/5000&quot;</span>
<span class="n">window_start</span> <span class="o">=</span> <span class="mi">87668723</span>
<span class="n">window_end</span> <span class="o">=</span> <span class="n">window_start</span> <span class="o">+</span> <span class="mi">1000000</span>
<span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;chr1&quot;</span>
<span class="n">imputation_value</span> <span class="o">=</span> <span class="mf">1e-14</span>

<span class="n">minimum</span><span class="p">,</span> <span class="n">maximum</span> <span class="o">=</span> <span class="n">find_dataset_min_max</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">minimum</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">maximum</span><span class="p">))</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plot_histogram_dataset</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
<span class="c1"># Fetch and process the matrix</span>
<span class="n">c_matrix</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">window_start</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">window_end</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">imputation_value</span><span class="p">)</span>
<span class="n">c_matrix</span><span class="p">[</span><span class="n">c_matrix</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputation_value</span>
<span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">c_matrix</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">))</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;YlOrRd&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">show</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">c_matrix</span><span class="o">.</span><span class="n">flatten</span><span class="p">()),</span> <span class="n">bins</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>-3.1009633451024556 0.9283546208429725
(201, 201)
-14.0 2.44088680731723
</pre></div>
</div>
<img alt="../_images/7a42efaa61e023ef29d66395fc20c0fa13c416df750da8ca54606a04cf334d96.png" src="../_images/7a42efaa61e023ef29d66395fc20c0fa13c416df750da8ca54606a04cf334d96.png" />
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/scratch/ipykernel_2566803/264489752.py:62: RuntimeWarning: invalid value encountered in log10
  plt.hist(np.log10(c_matrix.flatten()), bins=1000)
</pre></div>
</div>
<img alt="../_images/faa126c473dbc12d74219b5f007f5c6cdd3248fa27b4d7207bcf9a47fb1a325e.png" src="../_images/faa126c473dbc12d74219b5f007f5c6cdd3248fa27b4d7207bcf9a47fb1a325e.png" />
<img alt="../_images/e4972feb31c7a416b592d05c734dc219a2caa50492d73edb4f1b6c51de37e1f3.png" src="../_images/e4972feb31c7a416b592d05c734dc219a2caa50492d73edb4f1b6c51de37e1f3.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_prom_CHiC_arrays.py

<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">cooler</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.ndimage</span><span class="w"> </span><span class="kn">import</span> <span class="n">rotate</span>

<span class="c1">########### Functions #####################################</span>
<span class="k">def</span><span class="w"> </span><span class="nf">find_dataset_min_max</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">))</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_histogram_dataset</span><span class="p">(</span><span class="n">input_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">bins</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
    <span class="c1"># Open contact file (mcool) as a cooler object</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
    <span class="n">bin_weights</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">bins</span><span class="p">()[</span><span class="s1">&#39;weight&#39;</span><span class="p">][:]</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">bin_weights</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">mcool_summary</span><span class="p">(</span><span class="n">microc_path</span><span class="p">):</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>

    <span class="c1"># Inspecting what is inside the cooler object:</span>
    <span class="n">metadata</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">info</span>
    <span class="n">chromosomes</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">chromnames</span>
    <span class="n">extent</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">extent</span><span class="p">(</span><span class="s1">&#39;chr2&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Metadata:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">metadata</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Chromosomes:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">chromosomes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Extent:</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_prom_CHiC_data</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">imputation_value</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span> <span class="o">=</span> <span class="n">args</span>
    <span class="n">ID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">Start</span><span class="p">,</span> <span class="n">End</span><span class="p">,</span> <span class="n">Strand</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">CRE_type</span> <span class="o">=</span> <span class="n">row</span>
    <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
    <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

    <span class="n">c</span> <span class="o">=</span> <span class="n">cooler</span><span class="o">.</span><span class="n">Cooler</span><span class="p">(</span><span class="n">microc_path</span><span class="p">)</span>
    <span class="n">window_start</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">-</span> <span class="n">shift</span>
    <span class="n">window_end</span> <span class="o">=</span> <span class="n">TSS</span> <span class="o">+</span> <span class="n">shift</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Fetch and process the matrix</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">balance</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">window_start</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">window_end</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">imputation_value</span><span class="p">)</span>
        <span class="n">c_matrix</span><span class="p">[</span><span class="n">c_matrix</span> <span class="o">==</span> <span class="mf">0.</span><span class="p">]</span> <span class="o">=</span> <span class="n">imputation_value</span>
        <span class="n">c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">c_matrix</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="p">(</span><span class="mi">201</span><span class="p">,</span> <span class="mi">201</span><span class="p">):</span>
            <span class="c1"># Save direct contact matrices</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Promoter-CHiC&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">c_matrix</span><span class="p">)</span>
            <span class="n">anti_c_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">fliplr</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flipud</span><span class="p">(</span><span class="n">c_matrix</span><span class="p">))</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Promoter-CHiC&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">anti_c_matrix</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> raised an error: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_promoter_CHiC_arrays_parallel</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="o">=</span><span class="mi">500000</span><span class="p">,</span> <span class="n">imputation_value</span><span class="o">=</span><span class="mf">1e-14</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parallel version for creating Micro C arrays.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="n">args</span> <span class="o">=</span> <span class="p">[(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">microc_path</span><span class="p">,</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">shift</span><span class="p">,</span> <span class="n">imputation_value</span><span class="p">,</span> <span class="n">row</span><span class="p">)</span> <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>

    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">process_prom_CHiC_data</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="c1"># Define paths</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">prom_CHiC_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span>  <span class="s2">&quot;GEO_files/SRR5972844.mm10.mapq_30.100.mcool::/resolutions/5000&quot;</span>
    <span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations/Final_gene_annotations.tsv&quot;</span>
    <span class="n">savepath</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span>
    <span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Setup logging</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;Prom_CHiC_data_obtention.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">savepath</span> <span class="o">/</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># Create directories for output</span>
    <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]:</span>
        <span class="p">(</span><span class="n">savepath</span> <span class="o">/</span> <span class="s2">&quot;Promoter-CHiC&quot;</span> <span class="o">/</span> <span class="n">split</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Run the processing in parallel</span>
    <span class="n">create_promoter_CHiC_arrays_parallel</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">prom_CHiC_path</span><span class="p">),</span> <span class="n">savepath</span><span class="p">,</span> <span class="n">imputation_value</span><span class="o">=</span><span class="mf">1e-14</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove the samples for which we do not have landscape arrays in training or test folders:</span>
<span class="c1"># Get list of files from both reference folders</span>
<span class="n">landscape_training</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/training/&quot;</span><span class="p">)]</span>
<span class="n">landscape_test</span> <span class="o">=</span> <span class="p">[</span><span class="n">file</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="s2">&quot;../inputs/landscape_arrays/test/&quot;</span><span class="p">)]</span>
<span class="n">old_samples</span> <span class="o">=</span> <span class="n">landscape_training</span> <span class="o">+</span> <span class="n">landscape_test</span>

<span class="c1"># Check current directory files</span>
<span class="k">for</span> <span class="n">folder</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="s1">&#39;../inputs/Promoter-CHiC/&#39;</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">scandir</span><span class="p">(</span><span class="n">folder</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">old_samples</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Removing: </span><span class="si">{</span><span class="n">file</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Training eir with promoter capture HiC added:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_train/input_cnn.yaml<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_train/input_cnn_microc.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing prom-CHiC:</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_test/input_cnn.yaml<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_test/input_cnn_microc.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_prom_CHiC_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_prom_CHiC_pure_conv/saved_models/gene_expression_prom_CHiC_pure_conv_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.8159.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_prom_CHiC_pure_conv_test
</pre></div>
</div>
<p><strong>Visualizing contact maps and attributions for prom CHiC data:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">visualize_prom_CHiC</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
    <span class="n">prom_CHiC_map</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span><span class="c1">#(11.5, 5))</span>
    <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>

    <span class="c1"># Main image</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="c1"># Create an axis for the colorbar</span>
    <span class="n">cax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>

    <span class="n">img_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">prom_CHiC_map</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;YlOrRd&quot;</span><span class="p">)</span>
    <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">img_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Bins (5kbp resolution)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Bins (5kbp resolution)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_prom_CHiC_attributions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">,</span> <span class="n">ATTR_type_list</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">sigma</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; </span>
<span class="sd">    Plot the attribution maps for micro-c matrices.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bin_per_kbp</span> <span class="o">=</span> <span class="n">num_bins</span><span class="o">/</span><span class="n">window_of_observation</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">bin_per_kbp</span><span class="p">)</span>
    <span class="n">max_shift</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="mi">200</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">)</span> <span class="c1"># Distance to crop because it was rolled from the other side</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">max_shift</span><span class="p">)</span>

    <span class="n">avg_att_mat_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ATTR_type</span> <span class="ow">in</span> <span class="n">ATTR_type_list</span><span class="p">:</span>
            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">ATTR_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">201</span><span class="p">,</span><span class="mi">201</span><span class="p">))</span>


    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">sample_ids</span><span class="p">:</span>
            <span class="c1">#path = Path(f&quot;../runs/training_runs/gene_expression_exformer_unlimited_all_rotated/results/expression_output/{i}{condition}/samples/17880/attributions/contact_maps/{i}{condition}.npy&quot;) </span>
            <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;../runs/test_runs/gene_expression_prom_CHiC_pure_conv_test/expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/attributions/contact_maps/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span> 
            <span class="n">att_mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
            <span class="n">att_mat_rolled</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">att_mat</span><span class="p">,</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">),</span> <span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">*</span><span class="n">bin_per_kbp</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="s2">&quot;abs&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">))</span><span class="o">*</span><span class="nb">abs</span><span class="p">(</span><span class="n">att_mat_rolled</span><span class="p">)</span>
            <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="s2">&quot;signed&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">))</span><span class="o">*</span><span class="n">att_mat_rolled</span>

    <span class="k">for</span> <span class="n">CONDITION</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ATTR_type</span> <span class="ow">in</span> <span class="n">ATTR_type_list</span><span class="p">:</span>
            <span class="n">avg_att_mat</span> <span class="o">=</span> <span class="n">avg_att_mat_dict</span><span class="p">[</span><span class="n">condition</span><span class="p">][</span><span class="n">ATTR_type</span><span class="p">]</span> 
            <span class="c1">#avg_att_mat = gaussian_filter(avg_att_mat, sigma=sigma)</span>

            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span><span class="c1">#(11.5, 5))</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">hspace</span><span class="o">=</span><span class="mf">.2</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">.02</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mf">0.1</span><span class="p">))</span>

            <span class="c1"># Main image</span>
            <span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
            <span class="c1"># Create an axis for the colorbar</span>
            <span class="n">cax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>

            <span class="k">if</span> <span class="n">ATTR_type</span> <span class="o">==</span> <span class="s2">&quot;abs&quot;</span><span class="p">:</span>
                <span class="n">img_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;inferno&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.5e-4</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s2">&quot;upper&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
                <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">img_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mf">5e-5</span><span class="p">])</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticklabels</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">img_display</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">avg_att_mat</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;seismic&quot;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mf">0.5e-4</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mf">0.5e-4</span><span class="p">,</span> <span class="n">origin</span><span class="o">=</span><span class="s2">&quot;upper&quot;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
                <span class="n">cbar</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">img_display</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">([</span><span class="o">-</span><span class="mf">5e-5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">5e-5</span><span class="p">])</span>
                <span class="n">cbar</span><span class="o">.</span><span class="n">set_ticklabels</span><span class="p">([</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">101</span><span class="p">),</span> <span class="mi">200</span><span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">101</span><span class="p">),</span> <span class="s1">&#39;k--&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">101</span><span class="p">),</span> <span class="mi">100</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="s1">&#39;b--&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.8</span><span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="mi">100</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="mi">200</span><span class="p">),</span> <span class="s1">&#39;b--&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.8</span><span class="p">)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patches</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">))</span>
                <span class="c1"># Right edge</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patches</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">201</span><span class="o">-</span><span class="mi">40</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">201</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">))</span>
                <span class="c1"># Bottom edge (excluding corners already covered)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patches</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">40</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="mi">201</span><span class="o">-</span><span class="mi">80</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">))</span>
                <span class="c1"># Top edge (excluding corners already covered)</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">patches</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">40</span><span class="p">,</span> <span class="mi">201</span><span class="o">-</span><span class="mi">40</span><span class="p">),</span> <span class="mi">201</span><span class="o">-</span><span class="mi">80</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">))</span>

            <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Bins (5kbp resolution)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Bins (5kbp resolution)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
            <span class="c1">#ax.set_xlim(201,0)</span>
            <span class="c1">#ax.set_ylim(201,0)</span>

            <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
            <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;Structural_attributions_</span><span class="si">{</span><span class="n">ATTR_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">CONDITION</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">sigma</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../inputs/Promoter-CHiC/test/&quot;</span><span class="p">)</span>
<span class="n">figure_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/attributions/&quot;</span><span class="p">)</span>
<span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="mi">200</span><span class="p">,</span><span class="mi">201</span><span class="p">)</span>
<span class="n">sample</span> <span class="o">=</span> <span class="s2">&quot;ENSMUSG00000028287.4_forward&quot;</span>
<span class="n">num_bins</span> <span class="o">=</span> <span class="mi">201</span>
<span class="n">window_of_observation</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">ATTR_type_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;abs&quot;</span><span class="p">,</span><span class="s2">&quot;signed&quot;</span><span class="p">]</span>
<span class="n">conditions</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span>
<span class="n">sigma</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_prom_CHiC</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.npy&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">.png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plot_prom_CHiC_attributions</span><span class="p">(</span><span class="n">figure_path</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">num_bins</span><span class="p">,</span> <span class="n">window_of_observation</span><span class="p">,</span> <span class="n">ATTR_type_list</span><span class="p">,</span> <span class="n">conditions</span><span class="p">,</span> <span class="n">sigma</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.201
40
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/scratch/ipykernel_625962/1240926082.py:85: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
  plt.tight_layout()
</pre></div>
</div>
<img alt="../_images/93cd72744efbebe0ba307cef4638a2100cbdc0b4bcc0d858a38c78f7555939ae.png" src="../_images/93cd72744efbebe0ba307cef4638a2100cbdc0b4bcc0d858a38c78f7555939ae.png" />
</div>
</div>
</section>
<section id="vi-extending-claster-to-new-cell-types-k562-human">
<h2><center> VI) Extending CLASTER to new cell types: K562 (human) <center><a class="headerlink" href="#vi-extending-claster-to-new-cell-types-k562-human" title="Link to this heading">#</a></h2>
<p>Reviewer 1 asked us to benchmark our <em>in silico</em> perturbations with experimental data. Most experimental data on genome-wide enhancer KOs is obtained in K562 cells. We did not find nascent transcription data matching our protocol, and hence decided to predict two widespread transcriptional readouts: RNA-seq and POLR2A ChIP-seq.</p>
<section id="predicting-mature-rna-seq-counts-rna-polii-chip-seq">
<h3>Predicting mature RNA-seq counts &amp; RNA-polII ChIP-seq<a class="headerlink" href="#predicting-mature-rna-seq-counts-rna-polii-chip-seq" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Obtain data (hg19).</p></li>
<li><p>Train models.</p></li>
<li><p>Evaluate performance.</p></li>
<li><p>Perform downstream perturbational analyses.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Mature RNA-seq predictions: hg19</span>
<span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="c1">#&quot;EU_Seq_Ctrl_plus.bw&quot;:&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2072364&amp;format=file&amp;file=GSM2072364%5FENCFF001SEN%5Fplus%5Fstrand%5Fsignal%5Fhg19%2EbigWig&quot;,</span>
         <span class="c1">#&quot;EU_Seq_Ctrl_minus.bw&quot;:&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2072364&amp;format=file&amp;file=GSM2072364%5FENCFF001SEO%5Fminus%5Fstrand%5Fsignal%5Fhg19%2EbigWig&quot;,</span>
         <span class="s2">&quot;POLR_IIA_Ctrl.bw&quot;</span><span class="p">:</span> <span class="s2">&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE91426&amp;format=file&amp;file=GSE91426</span><span class="si">%5F</span><span class="s2">ENCFF541PYW</span><span class="si">%5F</span><span class="s2">signal</span><span class="si">%5F</span><span class="s2">p%2Dvalue</span><span class="si">%5F</span><span class="s2">hg19</span><span class="si">%2E</span><span class="s2">bigWig&quot;</span><span class="p">,</span> <span class="c1">#Stanford ENCODE hg19</span>
         <span class="c1">#&quot;ATAC_Seq.bw&quot;: &quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4971133&amp;format=file&amp;file=GSM4971133%5F11283%5F11142%5F109459%5FHKLV3BGXC%5FATAC%5FATAC%5F0h%5FTTCTGC%2Ebw&quot;, #Danko</span>
         <span class="c1">#&quot;H3K4me3.bw&quot;:&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4971098&amp;format=file&amp;file=GSM4971098%5FmergedMasked%5FK4me3%5F0h%5Fbr1%2Eraw%2Ebw&quot;, #Danko</span>
         <span class="c1">#&quot;H3K27ac.bw&quot;:&quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4971064&amp;format=file&amp;file=GSM4971064%5FmergedMasked%5FK27ac%5F0h%5Fbr1%2Eraw%2Ebw&quot;, #Danko</span>
         <span class="c1">#&quot;H3K27me3.bw&quot;: &quot;https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM4971075&amp;format=file&amp;file=GSM4971075%5FmergedMasked%5FK27me3%5F0h%5Fbr1%2Eraw%2Ebw&quot;</span>
         <span class="p">}</span> <span class="c1">#Danko</span>

<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../GEO_files/K562/&quot;</span><span class="p">)</span>
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Gene and Enhancer annotations:</strong></p>
<ul class="simple">
<li><p>Gene annotations were obtained for hg19 from <a class="reference external" href="https://www.gencodegenes.org/human/release_19.html">gencode</a> (Input data from Danko lab was mapped to hg19 originally).</p></li>
<li><p>Enhancer annotations could be obtained from:</p>
<ul>
<li><p><a class="reference external" href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-020-2493-4/MediaObjects/41586_2020_2493_MOESM12_ESM.txt">Supplementary Table 10</a>.</p></li>
<li><p>We however used the <a class="reference external" href="https://github.com/EngreitzLab/CRISPR_comparison/blob/main/resources/crispr_data/EPCrisprBenchmark_ensemble_data_GRCh38.tsv.gz">annotations from the CRISPR enhancer KOs from the Engreitz lab</a> reported in the latest E2G paper to ease the comparison. Only enhancers whose epigenetic activity was above a certain threshold were kept.</p></li>
<li><p>The coordinates of the regulatory elements were lifted using <a class="reference external" href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">the UCSC genome browser lifting tool</a> from hg38 to hg19 to match the input profiles.</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">links</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Gene_annotations_human.gtf.gz&quot;</span><span class="p">:</span><span class="s2">&quot;https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz&quot;</span><span class="p">}</span>
<span class="n">savepath</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/human/&quot;</span><span class="p">)</span>
<span class="n">savepath</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">download_files</span><span class="p">(</span><span class="n">links</span><span class="p">,</span> <span class="n">savepath</span><span class="p">)</span>

<span class="o">!</span><span class="w"> </span>gunzip<span class="w"> </span>../annotations/human/Gene_annotations_human.gtf.gz

<span class="c1"># Create a file with annotations that correspond only to genes</span>
<span class="o">!</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;$3 == &quot;gene&quot;&#39;</span><span class="w"> </span>../annotations/human/Gene_annotations_human.gtf<span class="w"> </span>&gt;<span class="w"> </span>../annotations/human/Filtered_gene_annotations_human.gtf

<span class="c1"># Further filter the number of fields/columns required</span>

<span class="n">file</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/human/Filtered_gene_annotations_human.gtf&quot;</span><span class="p">)</span>
<span class="n">outfile</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/human/Final_gene_annotations_human.tsv&quot;</span><span class="p">)</span>
<span class="n">gene_annotations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

<span class="n">ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>

<span class="n">final_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">,</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;Start&quot;</span><span class="p">,</span><span class="s2">&quot;End&quot;</span><span class="p">,</span><span class="s2">&quot;Strand&quot;</span><span class="p">,</span><span class="s2">&quot;Name&quot;</span><span class="p">,</span><span class="s2">&quot;type&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">strand</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">features</span><span class="p">)</span> <span class="ow">in</span> <span class="n">gene_annotations</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="n">splitline</span> <span class="o">=</span>  <span class="n">features</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>

    <span class="nb">id</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">gene_type</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">name</span> <span class="o">=</span> <span class="n">splitline</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">gene_type</span> <span class="o">==</span> <span class="s1">&#39;&quot;protein_coding&quot;&#39;</span> <span class="ow">and</span> <span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;chrM&#39;</span><span class="p">,</span> <span class="s1">&#39;chrX&#39;</span><span class="p">,</span><span class="s1">&#39;chrY&#39;</span><span class="p">]:</span> <span class="c1"># or gene_type == &#39;&quot;lincRNA&quot;&#39;:</span>
        <span class="n">final_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">id</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[</span><span class="n">chrom</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">end</span><span class="p">,</span><span class="n">strand</span><span class="p">,</span><span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">gene_type</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>

<span class="n">final_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Manual validation ids (chr17) in human samples:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">create_validation_ids</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">val_chrom</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;chr17&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; All genes encoded in chr17 will be used as a validation set.&quot;&quot;&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations_human.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">gene_annotations_df</span><span class="p">[</span><span class="n">gene_annotations_df</span><span class="p">[</span><span class="s2">&quot;chr&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">val_chrom</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>

    <span class="c1"># Prepare the strings to write in the file</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">_forward</span><span class="se">\n</span><span class="si">{</span><span class="n">index</span><span class="si">}</span><span class="s2">_rev</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">]</span>

    <span class="c1"># Write to the text file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;manual_validation_ids_</span><span class="si">{</span><span class="n">val_chrom</span><span class="si">}</span><span class="s2">_human.txt&quot;</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">file</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span>

<span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../annotations/human/&quot;</span><span class="p">)</span>
<span class="n">create_validation_ids</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We removed the following IDs from manual validation id file:</p>
<ul class="simple">
<li><p>‘ENSG00000272636.1_forward’,</p></li>
<li><p>‘ENSG00000272636.1_rev’,</p></li>
<li><p>‘ENSG00000175711.4_forward’,</p></li>
<li><p>‘ENSG00000175711.4_rev’,</p></li>
<li><p>‘ENSG00000176845.8_forward’,</p></li>
<li><p>‘ENSG00000176845.8_rev’.</p></li>
</ul>
<p>Causes explained in Landscape_data_obtention_human.log: the input samples are out of bounds since these genes lie at the beginning/end of the chromosome.</p>
<p><strong>Create CLASTER inputs and outputs from human (K562) samples:</strong></p>
<p>Similar code as in I_Data_Obtention.ipynb, better handling of Nans in input data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_landscape_arrays_and_targets_human.py

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pyBigWig</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.ndimage</span><span class="w"> </span><span class="kn">import</span> <span class="n">gaussian_filter1d</span><span class="p">,</span> <span class="n">rotate</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy.ma</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ma</span>

<span class="n">track_dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;ATAC-seq&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin accessibility&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;k&quot;</span><span class="p">},</span>
            <span class="mi">1</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K4me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;r&quot;</span><span class="p">},</span>
            <span class="mi">2</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27ac&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Enhancer&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;b&quot;</span><span class="p">},</span>
            <span class="mi">3</span><span class="p">:{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;H3K27me3&quot;</span><span class="p">,</span><span class="s2">&quot;function&quot;</span><span class="p">:</span><span class="s2">&quot;Chromatin silencing&quot;</span><span class="p">,</span><span class="s2">&quot;color&quot;</span><span class="p">:</span><span class="s2">&quot;g&quot;</span><span class="p">}}</span>


<span class="k">def</span><span class="w"> </span><span class="nf">clean_array</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">max_clip</span><span class="o">=</span><span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Comprehensive cleaning function for array data.</span>
<span class="sd">    Handles NaN, None, inf, and invalid values.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        stats: Input array or list of values</span>
<span class="sd">        max_clip: Maximum value to clip to</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        Cleaned numpy array</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert to numpy array if not already</span>
    <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
    
    <span class="c1"># Create mask for invalid values (None, NaN, inf)</span>
    <span class="n">invalid_mask</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">(</span><span class="n">array</span> <span class="o">==</span> <span class="kc">None</span><span class="p">)</span> <span class="o">|</span> 
        <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">array</span><span class="p">)</span> <span class="o">|</span> 
        <span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">array</span><span class="p">)</span> <span class="o">|</span>
        <span class="p">(</span><span class="n">array</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span>  <span class="c1"># Negative values</span>
    <span class="p">)</span>
    
    <span class="c1"># Replace invalid values with 0</span>
    <span class="n">array</span><span class="p">[</span><span class="n">invalid_mask</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
    
    <span class="c1"># Clip values to valid range</span>
    <span class="n">array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">array</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">array</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                      <span class="n">input_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">n_input_tracks</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
                      <span class="n">n_input_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">bw_input_track_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
                      <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">max_clip</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Modified input array creation with improved error handling and data cleaning.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">))</span>  <span class="c1"># Initialize with zeros</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bw_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bw_input_track_list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">TSS</span><span class="o">-</span><span class="n">input_shift</span><span class="p">,</span> <span class="n">TSS</span><span class="o">+</span><span class="n">input_shift</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_input_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="c1"># Clean the data using our comprehensive function</span>
            <span class="n">input_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
            
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> track </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># Row already initialized to zeros, so we can continue</span>
            <span class="k">continue</span>
    
    <span class="c1"># Final validation of the entire array</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">input_sample</span><span class="p">))</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">input_sample</span><span class="p">)):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> still contains invalid values after cleaning. Setting to zero.&quot;</span><span class="p">)</span>
        <span class="n">input_sample</span><span class="p">[</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isfinite</span><span class="p">(</span><span class="n">input_sample</span><span class="p">)]</span> <span class="o">=</span> <span class="mf">0.0</span>
    
    <span class="k">return</span> <span class="n">input_sample</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_single_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                             <span class="n">output_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">n_output_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">bw_target_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>  <span class="c1"># Single bigWig file path</span>
                             <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                             <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                             <span class="n">sigma</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">binsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                             <span class="n">max_clip</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates target array from a single bigWig file without summing strands.</span>
<span class="sd">    Used for RNA-PolII data that&#39;s already processed.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        bw_target_path: String with the path to the bigWig file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Read the data</span>
        <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_target_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span> <span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span> 
                        <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
        <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        
        <span class="c1"># Clean the data</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
        
        <span class="c1"># Apply Gaussian smoothing</span>
        <span class="n">stats</span> <span class="o">=</span> <span class="n">gaussian_filter1d</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
        
        <span class="c1"># Initialize target condition array</span>
        <span class="n">target_cond</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">stats</span><span class="p">)</span>
        
        <span class="c1"># Averaging over bins</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">binsize</span><span class="p">):</span>
            <span class="n">target_cond</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">binsize</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="o">-</span><span class="n">i</span><span class="p">)</span>
        <span class="n">target_cond</span> <span class="o">=</span> <span class="n">target_cond</span><span class="p">[::</span><span class="n">binsize</span><span class="p">]</span>
        
        <span class="c1"># Final cleanup of target condition array</span>
        <span class="n">target_cond</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">target_cond</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
        
        <span class="c1"># Create forward and reverse arrays</span>
        <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">target_cond</span>
        <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">target_cond</span><span class="p">)</span>
        
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> RNA-PolII target processing error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="c1"># Return arrays of appropriate size filled with zeros in case of error</span>
        <span class="n">zeros</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_output_bins</span> <span class="o">//</span> <span class="n">binsize</span><span class="p">)</span>
        <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">zeros</span>
        <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">zeros</span>
    
    <span class="c1"># Final validation</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">))</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">)):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> RNA-PolII targets contain NaN values after processing. Setting to zero.&quot;</span><span class="p">)</span>
        <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
        <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">target_sample_pos</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">target_sample_neg</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                       <span class="n">output_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">n_output_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">bw_target_track_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
                       <span class="n">ID</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                       <span class="n">TSS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                       <span class="n">sigma</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">binsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                       <span class="n">max_clip</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Modified target array creation with improved error handling and data cleaning.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([])</span>
    
    <span class="k">for</span> <span class="n">plus_bw_path</span><span class="p">,</span> <span class="n">minus_bw_path</span> <span class="ow">in</span> <span class="n">bw_target_track_list</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># Read plus strand data</span>
            <span class="n">plus_bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">plus_bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">plus_stats</span> <span class="o">=</span> <span class="n">plus_bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span> <span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span> 
                                     <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
            <span class="n">plus_bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="c1"># Read minus strand data</span>
            <span class="n">minus_bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">minus_bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">minus_stats</span> <span class="o">=</span> <span class="n">minus_bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">TSS</span><span class="o">-</span><span class="n">output_shift</span><span class="p">,</span> <span class="n">TSS</span><span class="o">+</span><span class="n">output_shift</span><span class="p">,</span> 
                                       <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_output_bins</span><span class="p">)</span>
            <span class="n">minus_bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="c1"># Clean both arrays</span>
            <span class="n">plus_stats</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">plus_stats</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
            <span class="n">minus_stats</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">minus_stats</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
            
            <span class="c1"># Combine plus and minus strand signals</span>
            <span class="n">combined_stats</span> <span class="o">=</span> <span class="n">plus_stats</span> <span class="o">+</span> <span class="n">minus_stats</span>
            
            <span class="c1"># Apply Gaussian smoothing</span>
            <span class="n">combined_stats</span> <span class="o">=</span> <span class="n">gaussian_filter1d</span><span class="p">(</span><span class="n">combined_stats</span><span class="p">,</span> <span class="n">sigma</span><span class="o">=</span><span class="n">sigma</span><span class="p">)</span>
            
            <span class="c1"># Initialize target condition array</span>
            <span class="n">target_cond</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros_like</span><span class="p">(</span><span class="n">combined_stats</span><span class="p">)</span>
            
            <span class="c1"># Averaging over bins</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">binsize</span><span class="p">):</span>
                <span class="n">target_cond</span> <span class="o">+=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">binsize</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">combined_stats</span><span class="p">,</span> <span class="o">-</span><span class="n">i</span><span class="p">)</span>
            <span class="n">target_cond</span> <span class="o">=</span> <span class="n">target_cond</span><span class="p">[::</span><span class="n">binsize</span><span class="p">]</span>
            
            <span class="c1"># Final cleanup of target condition array</span>
            <span class="n">target_cond</span> <span class="o">=</span> <span class="n">clean_array</span><span class="p">(</span><span class="n">target_cond</span><span class="p">,</span> <span class="n">max_clip</span><span class="p">)</span>
            
            <span class="c1"># Add to final arrays</span>
            <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_pos</span><span class="p">,</span> <span class="n">target_cond</span><span class="p">])</span>
            <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_neg</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">target_cond</span><span class="p">)])</span>
            
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> target processing error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># Return arrays of appropriate size filled with zeros in case of error</span>
            <span class="n">zeros</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_output_bins</span> <span class="o">//</span> <span class="n">binsize</span><span class="p">)</span>
            <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_pos</span><span class="p">,</span> <span class="n">zeros</span><span class="p">])</span>
            <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">target_sample_neg</span><span class="p">,</span> <span class="n">zeros</span><span class="p">])</span>
    
    <span class="c1"># Final validation</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">))</span> <span class="ow">or</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">)):</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> targets contain NaN values after processing. Setting to zero.&quot;</span><span class="p">)</span>
        <span class="n">target_sample_pos</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
        <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">target_sample_pos</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="n">target_sample_neg</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    
<span class="k">def</span><span class="w"> </span><span class="nf">visualize_input_array</span><span class="p">(</span><span class="n">a</span><span class="p">,</span>
                     <span class="n">cropped_bins</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4400</span><span class="p">,</span>
                     <span class="n">scaling_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.</span><span class="p">,</span>
                     <span class="n">track_dict</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="n">track_dict</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to visualize an input numpy array.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">track_dict</span><span class="p">),</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>

    <span class="c1"># Plot the scaled arrays</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span><span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
        <span class="c1"># Left plots: full window</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">line</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1">#axs[j][0].set_yticks([]) #([0,int(np.max(line))], [0,int(np.max(line))], fontsize=16)</span>
        <span class="c1">#axs[j][0].set_ylim(0,350)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Distance to TSS (kbp)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span> <span class="mi">16</span><span class="p">)</span>
        <span class="c1"># Right plots, centered window</span>
        <span class="n">x_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">)[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">]</span>
        <span class="n">y_values</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="n">cropped_bins</span><span class="p">:</span><span class="o">-</span><span class="n">cropped_bins</span><span class="p">]</span><span class="o">*</span><span class="n">scaling_factor</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="n">x_values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">x_values</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x_values</span><span class="p">,</span><span class="n">y_values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;name&quot;</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
        <span class="n">axs</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x_values</span><span class="p">,</span><span class="n">y_values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">track_dict</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="s2">&quot;color&quot;</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">.6</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>
        
<span class="k">def</span><span class="w"> </span><span class="nf">visualize_target_array</span><span class="p">(</span><span class="n">a</span><span class="p">,</span>
                           <span class="n">scaling_factor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function to visualize target profiles.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ctrl</span> <span class="o">=</span> <span class="n">a</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">ctrl</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;ctrl&quot;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span><span class="n">lw</span><span class="o">=</span><span class="mf">.2</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ctrl</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">),</span><span class="n">ctrl</span><span class="o">*</span><span class="n">scaling_factor</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;silver&quot;</span> <span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_gene_data</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">rna_polii_track</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> 
                     <span class="n">n_input_bins</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">,</span> <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> 
                     <span class="n">path</span><span class="p">,</span> <span class="n">split_list</span><span class="p">,</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">PLOT_IDS</span><span class="p">,</span> <span class="n">target_files</span><span class="p">,</span> <span class="n">polii_target_files</span><span class="p">,</span> <span class="n">row</span><span class="p">):</span>
    <span class="n">ID</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">Start</span><span class="p">,</span> <span class="n">End</span><span class="p">,</span> <span class="n">Strand</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">CRE_type</span> <span class="o">=</span> <span class="n">row</span>
    <span class="n">split</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span> <span class="k">if</span> <span class="n">chrom</span> <span class="o">==</span> <span class="s2">&quot;chr4&quot;</span> <span class="k">else</span> <span class="s2">&quot;training&quot;</span>
    <span class="n">TSS</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Start</span><span class="p">)</span> <span class="k">if</span> <span class="n">Strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="nb">int</span><span class="p">(</span><span class="n">End</span><span class="p">)</span>

    <span class="c1"># Create input sample</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                                    <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">)</span>
    
    <span class="c1"># Create regular target samples</span>
    <span class="n">target_sample_pos</span><span class="p">,</span> <span class="n">target_sample_neg</span> <span class="o">=</span> <span class="n">create_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> 
                                                             <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">)</span>
    
    <span class="c1"># Create RNA-PolII target samples</span>
    <span class="n">polii_sample_pos</span><span class="p">,</span> <span class="n">polii_sample_neg</span> <span class="o">=</span> <span class="n">create_single_target_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span>
                                                                  <span class="n">rna_polii_track</span><span class="p">,</span> <span class="n">ID</span><span class="p">,</span> <span class="n">TSS</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">input_sample</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">INPUT_SHAPE</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">)</span> <span class="o">==</span> <span class="n">OUTPUT_LENGTH</span><span class="p">):</span>
        <span class="c1"># Save input arrays</span>
        <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.npy&quot;</span><span class="p">,</span> <span class="n">input_sample</span><span class="p">)</span>
        <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span> <span class="o">/</span> <span class="n">split</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.npy&quot;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
        
        <span class="c1"># Save regular targets</span>
        <span class="n">target_file_path</span> <span class="o">=</span> <span class="n">target_files</span><span class="p">[</span><span class="n">split</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">target_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_pos</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">target_sample_neg</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Save RNA-PolII targets</span>
        <span class="n">polii_file_path</span> <span class="o">=</span> <span class="n">polii_target_files</span><span class="p">[</span><span class="n">split</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">polii_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">polii_sample_pos</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">polii_sample_neg</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2"> : Input or target did not match the expected shape.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">PLOT_IDS</span><span class="p">:</span>
        <span class="c1"># Regular input/target visualizations</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">input_sample</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">target_sample_pos</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward_target.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_input_array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">flip</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">target_sample_neg</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev_target.png&quot;</span><span class="p">)</span>
        
        <span class="c1"># RNA-PolII visualizations</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">polii_sample_pos</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_forward_polii_target.png&quot;</span><span class="p">)</span>
        <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_target_array</span><span class="p">(</span><span class="n">polii_sample_neg</span><span class="p">)</span>
        <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figure_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s2">_rev_polii_target.png&quot;</span><span class="p">)</span>

<span class="c1">############################ Script ###################</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>

    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span>
    <span class="n">figure_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;figures&quot;</span> <span class="o">/</span> <span class="s2">&quot;Figures_revisions&quot;</span> <span class="o">/</span> <span class="s2">&quot;Inputs_and_Outputs&quot;</span>
    <span class="n">figure_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">target_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span>
    <span class="n">target_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">polii_target_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span> <span class="o">/</span> <span class="s2">&quot;polii&quot;</span>
    <span class="n">polii_target_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">split_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>
    <span class="n">output_dirs</span> <span class="o">=</span> <span class="p">[(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="n">split_list</span><span class="p">]</span>

    <span class="n">PLOT_IDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSG00000269308.1&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000185619.13&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000235478.1&quot;</span><span class="p">]</span>

    <span class="c1"># Start logger</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;/Landscape_data_obtention_human.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>  

    <span class="n">gene_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;human&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations_human.tsv&quot;</span>
    <span class="n">gene_annotations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">input_shift</span> <span class="o">=</span> <span class="mi">500050</span> <span class="c1"># In basepairs, from TSS to one side + central bin (100bp res)</span>
    <span class="n">output_shift</span> <span class="o">=</span> <span class="mi">200010</span> <span class="c1"># Output in 20 bp resolution (to be smoothed and downsampled)</span>
    <span class="n">n_input_bins</span> <span class="o">=</span> <span class="mi">10001</span>
    <span class="n">n_input_tracks</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">n_output_bins</span> <span class="o">=</span> <span class="mi">20001</span>
    <span class="n">sigma</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">binsize</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">INPUT_SHAPE</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">10001</span><span class="p">)</span>
    <span class="n">OUTPUT_LENGTH</span> <span class="o">=</span> <span class="mi">401</span> <span class="c1">#802</span>
    <span class="n">N_BINS</span> <span class="o">=</span> <span class="mi">200</span>
    <span class="n">bw_input_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC_Seq.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3.bw&quot;</span><span class="p">]</span>
    <span class="n">bw_target_track_list</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;EU_Seq_Ctrl_minus.bw&quot;</span><span class="p">,</span><span class="s2">&quot;EU_Seq_Ctrl_plus.bw&quot;</span><span class="p">)]</span> 
    <span class="c1"># Add RNA-PolII track</span>
    <span class="n">rna_polii_track</span> <span class="o">=</span> <span class="s2">&quot;POLR_IIA_Ctrl.bw&quot;</span> 

    <span class="n">target_files</span> <span class="o">=</span> <span class="p">{</span><span class="n">split</span><span class="p">:</span> <span class="p">(</span> <span class="n">target_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_targets.csv&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]}</span>

    <span class="c1"># Create additional target files for RNA-PolII</span>
    <span class="n">polii_target_files</span> <span class="o">=</span> <span class="p">{</span><span class="n">split</span><span class="p">:</span> <span class="p">(</span><span class="n">polii_target_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">_polii_targets.csv&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]}</span>
    <span class="c1"># Initialize RNA-PolII target files</span>
    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">polii_target_files</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ID,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">target_files</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
            <span class="n">file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;ID,&quot;</span> <span class="o">+</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">cond</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">cond</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span><span class="n">N_BINS</span><span class="o">+</span><span class="mi">1</span><span class="p">)])</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Update process arguments to include the RNA-PolII track</span>
    <span class="n">process_args</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">bw_target_track_list</span><span class="p">,</span> <span class="n">rna_polii_track</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">output_shift</span><span class="p">,</span> 
                  <span class="n">n_input_bins</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_output_bins</span><span class="p">,</span> <span class="n">sigma</span><span class="p">,</span> <span class="n">binsize</span><span class="p">,</span> <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">OUTPUT_LENGTH</span><span class="p">,</span> 
                  <span class="n">path</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;validation&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">],</span> <span class="n">figure_path</span><span class="p">,</span> <span class="n">PLOT_IDS</span><span class="p">,</span> <span class="n">target_files</span><span class="p">,</span> <span class="n">polii_target_files</span><span class="p">)</span>

    <span class="c1"># Process gene data with multiprocessing</span>
    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">process_gene_data</span><span class="p">,</span> <span class="p">[(</span><span class="n">process_args</span> <span class="o">+</span> <span class="p">(</span><span class="n">row</span><span class="p">,))</span> <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gene_annotations_df</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()])</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Training the model on K562 to predict baseline RNA-seq</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_train/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Testing the model on K562 to predict baseline RNA-seq</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_test/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_test/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_test/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_test/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_K562_train/saved_models/gene_expression_only_chrom_K562_train_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.5028.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/test_runs/gene_expression_only_chrom_K562_test
</pre></div>
</div>
<p><strong>Training the model on K562 to predict baseline POLR2A CHiP-seq</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirtrain<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_train/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_train/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_train/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_train/outputs_2_cond.yaml
</pre></div>
</div>
<p><strong>Visualize K562 predictions (RNA-seq and POLR2A)</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Outputs:</span>
<span class="n">figures_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="c1">#ID = ENSMUSG00000090115.7_rev ENSMUSG00000043592.15_rev ENSMUSG00000023266.11_rev # Nice examples &#39;ENSMUSG00000079707.10_forward&#39;</span>
<span class="n">val_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="s1">&#39;../annotations/human/manual_validation_ids_chr17_human.txt&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="k">for</span> <span class="n">ID</span> <span class="ow">in</span> <span class="n">val_ids</span><span class="p">[</span><span class="mi">5</span><span class="p">::</span><span class="mi">200</span><span class="p">]:</span> <span class="c1">#[&quot;ENSG00000166685.7_forward&quot;, &quot;ENSG00000136436.10_forward&quot;]: #val_ids[2::200]:#   [&quot;ENSMUSG00000090115.7_forward&quot;, &quot;ENSMUSG00000043592.15_forward&quot;, &quot;ENSMUSG00000023266.11_forward&quot;]: #val_ids[::8]:</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">attributes</span><span class="p">)</span> <span class="ow">in</span>  <span class="nb">enumerate</span><span class="p">(</span><span class="n">results_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="k">if</span> <span class="s2">&quot;training&quot;</span> <span class="ow">in</span> <span class="n">title</span><span class="p">:</span>
            <span class="c1">#print(title)</span>
            <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;path&#39;</span><span class="p">],</span>
                                                    <span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;N_BINS&#39;</span><span class="p">],</span>
                                                    <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> 
                                                    <span class="n">is_training</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;is_training&#39;</span><span class="p">],</span> 
                                                    <span class="n">batch_num</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;batch_num&#39;</span><span class="p">])</span>
            
            <span class="n">line_p</span> <span class="o">=</span> <span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">]</span>
            <span class="n">predictions</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s1">&#39;profile&#39;</span><span class="p">:</span><span class="n">line_p</span><span class="p">,</span>
                                <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">],</span>
                                <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">,</span>
                                <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Predicted </span><span class="si">{</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                                <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
            <span class="p">})</span>

            <span class="n">EU_seq_colors</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;k&quot;</span><span class="p">,</span><span class="s2">&quot;red&quot;</span><span class="p">]</span>
            <span class="c1"># Actual baseline signal:</span>
            <span class="n">actual_data</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;profile&#39;</span><span class="p">:</span> <span class="n">actual</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ID</span><span class="p">],</span>
                <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="n">EU_seq_colors</span><span class="p">[</span><span class="n">k</span><span class="p">],</span>
                <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.2</span><span class="p">,</span>
                <span class="s1">&#39;label&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Target </span><span class="si">{</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s1">&#39;ls&#39;</span><span class="p">:</span><span class="s1">&#39;-&#39;</span>
            <span class="p">})</span>



    <span class="c1"># Create plot</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plot_generalized_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">actual_data</span><span class="p">,</span> <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">figures_path</span> <span class="o">/</span> <span class="s1">&#39;Outputs&#39;</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;Model_predictions_human_</span><span class="si">{</span><span class="n">ID</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>


<span class="c1"># Visualize correlations:</span>
<span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="p">(</span><span class="n">title</span><span class="p">,</span> <span class="n">attributes</span><span class="p">)</span> <span class="ow">in</span>  <span class="nb">enumerate</span><span class="p">(</span><span class="n">results_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;path&#39;</span><span class="p">],</span>
                                        <span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;N_BINS&#39;</span><span class="p">],</span>
                                        <span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span> 
                                        <span class="n">is_training</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;is_training&#39;</span><span class="p">],</span> 
                                        <span class="n">batch_num</span><span class="o">=</span><span class="n">attributes</span><span class="p">[</span><span class="s1">&#39;batch_num&#39;</span><span class="p">])</span>
    
    <span class="c1"># Plot point-wise prediction correlations:</span>
    <span class="n">pred_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> 
    <span class="n">actual_list_values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actual</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">pointwise</span> <span class="o">=</span> <span class="n">plot_correlations</span><span class="p">(</span><span class="n">figures_path</span><span class="p">,</span> <span class="n">pred_list_values</span><span class="p">,</span> <span class="n">actual_list_values</span><span class="p">,</span> <span class="n">title</span> <span class="o">+</span> <span class="s2">&quot;_pointwise_correlation_prediction&quot;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;binary&#39;</span><span class="p">,</span> <span class="n">binlims</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="enhancer-centered-analyses-of-rna-seq-and-polr2a-binding-differences-after-in-silico-enhancer-ablation">
<h3>Enhancer-centered analyses of RNA-seq and POLR2A binding differences after <em>in silico</em> enhancer ablation:<a class="headerlink" href="#enhancer-centered-analyses-of-rna-seq-and-polr2a-binding-differences-after-in-silico-enhancer-ablation" title="Link to this heading">#</a></h3>
<blockquote>
<div><p><em>Pipeline:</em></p>
<ul class="simple">
<li><p>Check enhancer coordinates in EPCrisprBenchmark</p></li>
<li><p>Create samples centered at those enhancers and create silenced enhancer samples</p>
<ul>
<li><p>We filtered out: samples with enhancer activity average (H3K27ac) below 10 reads.</p></li>
<li><p>Samples where either input or output crossed chromosome boundaries (enhancers at the ends of the chromosomes).</p></li>
</ul>
</li>
<li><p>Predict using models trained on K562 (human) data.</p></li>
<li><p>Quantify POLR2A /RNA-seq changes:</p>
<ul>
<li><p>For POLR2A: Integrate between 1 kbp upstream and 2 kbp downstream of all genes in predicted window.</p></li>
<li><p>For RNA-seq: Integrate inside gene boundaries of all genes in predicted window.</p></li>
</ul>
</li>
<li><p>Downstream analyses:</p>
<ul>
<li><p>Precision-Recall and ROC curves for the following models:</p>
<ul>
<li><p>Gene-enhancer distance: <span class="math notranslate nohighlight">\(Score = - Distance\)</span></p></li>
<li><p>RNA and POLR2A models: <span class="math notranslate nohighlight">\(Score = abs(\)</span> Area difference <span class="math notranslate nohighlight">\()\)</span></p></li>
</ul>
</li>
<li><p>Confusion matrices:</p>
<ul>
<li><p>Primary target (most affected gene in a single prediction run): True / False</p></li>
<li><p>Closest gene: True / False</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div></blockquote>
<p><strong>Create enhancer centered K562 samples and targets:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%writefile</span> create_enhancer_centered_inputs_human.py
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pyBigWig</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">logging</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
                      <span class="n">input_shift</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">n_input_tracks</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
                      <span class="n">n_input_bins</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">bw_input_track_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span>
                      <span class="n">enhancer_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                      <span class="n">center</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                      <span class="n">chrom</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates input arrays centered at enhancer midpoint</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">((</span><span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bw_path</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">bw_input_track_list</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">bw</span> <span class="o">=</span> <span class="n">pyBigWig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">data_path</span> <span class="o">/</span> <span class="n">bw_path</span><span class="p">),</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">bw</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">center</span><span class="o">-</span><span class="n">input_shift</span><span class="p">,</span> <span class="n">center</span><span class="o">+</span><span class="n">input_shift</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;mean&quot;</span><span class="p">,</span> <span class="n">nBins</span><span class="o">=</span><span class="n">n_input_bins</span><span class="p">)</span>
            <span class="n">bw</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="mf">0.</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">stats</span><span class="p">])</span>
            <span class="n">stats</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">stats</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>  <span class="c1"># ReLU</span>
            <span class="n">input_sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">stats</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2"> input landscape coordinates are out of bounds: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">input_sample</span>

<span class="k">def</span><span class="w"> </span><span class="nf">create_perturbed_array</span><span class="p">(</span><span class="n">input_array</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">end_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates perturbed version with all marks except H3K27me3 silenced within enhancer boundaries</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">perturbed</span> <span class="o">=</span> <span class="n">input_array</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># Silence ATAC-seq, H3K4me3, and H3K27ac (indices 0, 1, 2) only within enhancer boundaries</span>
    <span class="n">perturbed</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">start_idx</span><span class="p">:</span><span class="n">end_idx</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">return</span> <span class="n">perturbed</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_enhancer</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                    <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">row_data</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">activity_threshold</span><span class="o">=</span><span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process a single enhancer and create its input arrays if enhancer is active</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract data from row</span>
    <span class="n">enhancer_id</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;accession&#39;</span><span class="p">]</span>
    <span class="n">chrom_hg19</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;chrom(hg19)&#39;</span><span class="p">]</span>
    <span class="n">start_hg19</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;chromStart(hg19)&#39;</span><span class="p">])</span>
    <span class="n">end_hg19</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;chromEnd(hg19)&#39;</span><span class="p">])</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">]</span>  <span class="c1"># Current assembly chromosome</span>
    
    <span class="c1"># Use hg19 coordinates for centering</span>
    <span class="n">center_hg19</span> <span class="o">=</span> <span class="p">(</span><span class="n">start_hg19</span> <span class="o">+</span> <span class="n">end_hg19</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>

    <span class="c1"># Create regular input array centered on hg19 coordinates</span>
    <span class="n">input_sample</span> <span class="o">=</span> <span class="n">create_input_array</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                                     <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">enhancer_id</span><span class="p">,</span> <span class="n">center_hg19</span><span class="p">,</span> <span class="n">chrom_hg19</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">input_sample</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">INPUT_SHAPE</span><span class="p">:</span>
        <span class="c1"># Calculate enhancer boundaries in array coordinates</span>
        <span class="n">central_idx</span> <span class="o">=</span> <span class="n">n_input_bins</span> <span class="o">//</span> <span class="mi">2</span>
        <span class="n">start_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="p">(</span><span class="n">start_hg19</span> <span class="o">-</span> <span class="n">center_hg19</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
        <span class="n">end_idx</span> <span class="o">=</span> <span class="n">central_idx</span> <span class="o">+</span> <span class="p">(</span><span class="n">end_hg19</span> <span class="o">-</span> <span class="n">center_hg19</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
        
        <span class="c1"># Ensure indices are within bounds</span>
        <span class="n">start_idx</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">)</span>
        <span class="n">end_idx</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">n_input_bins</span><span class="p">,</span> <span class="n">end_idx</span><span class="p">)</span>

        <span class="c1"># Check H3K27ac activity within enhancer boundaries</span>
        <span class="n">h3k27ac_idx</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># Index for H3K27ac track</span>
        <span class="n">enhancer_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">input_sample</span><span class="p">[</span><span class="n">h3k27ac_idx</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">:</span><span class="n">end_idx</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">enhancer_mean</span> <span class="o">&gt;</span> <span class="n">activity_threshold</span><span class="p">:</span>
            <span class="c1"># Create output filename based on enhancer coordinates to handle duplicates</span>
            <span class="n">enhancer_coord_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">chrom_hg19</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">start_hg19</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">end_hg19</span><span class="si">}</span><span class="s2">&quot;</span>
            
            <span class="c1"># Save regular array</span>
            <span class="n">enhancer_array_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;human_enhancer_arrays&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_coord_id</span><span class="si">}</span><span class="s2">.npy&quot;</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">enhancer_array_path</span><span class="p">,</span> <span class="n">input_sample</span><span class="p">)</span>
            
            <span class="c1"># Create and save perturbed array</span>
            <span class="n">perturbed_sample</span> <span class="o">=</span> <span class="n">create_perturbed_array</span><span class="p">(</span><span class="n">input_sample</span><span class="p">,</span> <span class="n">start_idx</span><span class="p">,</span> <span class="n">end_idx</span><span class="p">)</span>
            <span class="n">perturbed_array_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;human_enhancer_arrays&quot;</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_coord_id</span><span class="si">}</span><span class="s2">_perturbed.npy&quot;</span>
            <span class="n">np</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">perturbed_array_path</span><span class="p">,</span> <span class="n">perturbed_sample</span><span class="p">)</span>
            
            <span class="c1"># Record the mapping between accession ID and enhancer coordinate ID</span>
            <span class="k">return</span> <span class="p">{</span>
                <span class="s1">&#39;accession_id&#39;</span><span class="p">:</span> <span class="n">enhancer_id</span><span class="p">,</span>
                <span class="s1">&#39;enhancer_coord_id&#39;</span><span class="p">:</span> <span class="n">enhancer_coord_id</span><span class="p">,</span>
                <span class="s1">&#39;gene&#39;</span><span class="p">:</span> <span class="n">row_data</span><span class="p">[</span><span class="s1">&#39;measuredGeneSymbol&#39;</span><span class="p">],</span>
                <span class="s1">&#39;h3k27ac_mean&#39;</span><span class="p">:</span> <span class="n">enhancer_mean</span>
            <span class="p">}</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: Enhancer activity below threshold (</span><span class="si">{</span><span class="n">enhancer_mean</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> ≤ </span><span class="si">{</span><span class="n">activity_threshold</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: Input did not match the expected shape.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="c1"># Setup paths</span>
    <span class="n">path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    <span class="n">data_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;GEO_files&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span>

    <span class="c1"># Create output directories</span>
    <span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;human_enhancer_arrays&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Start logger</span>
    <span class="n">LOG_FILENAME</span> <span class="o">=</span> <span class="s2">&quot;/Enhancer_data_creation_human.log&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">)</span> <span class="o">+</span> <span class="n">LOG_FILENAME</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

    <span class="c1"># Load enhancer-promoter pair annotations</span>
    <span class="n">ep_annotations_path</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;human&quot;</span> <span class="o">/</span> <span class="s2">&quot;lifted_hg38_to_hg19&quot;</span> <span class="o">/</span> <span class="s2">&quot;EPCrisprBenchmark_ensemble_data_GRCh38_hg19.txt&quot;</span>
    <span class="n">ep_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">ep_annotations_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Make sure numeric columns are properly formatted</span>
    <span class="c1"># Handle potential comma issues in numeric columns</span>
    <span class="n">numeric_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;chromStart(hg19)&#39;</span><span class="p">,</span> <span class="s1">&#39;chromEnd(hg19)&#39;</span><span class="p">,</span> <span class="s1">&#39;chromStart&#39;</span><span class="p">,</span> <span class="s1">&#39;chromEnd&#39;</span><span class="p">,</span> <span class="s1">&#39;startTSS&#39;</span><span class="p">,</span> <span class="s1">&#39;endTSS&#39;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">numeric_cols</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ep_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="c1"># First convert any comma-formatted numbers to proper format</span>
            <span class="k">if</span> <span class="n">ep_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="s1">&#39;object&#39;</span><span class="p">:</span>
                <span class="n">ep_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">ep_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
    
    <span class="c1"># Filter for specific chromosome if needed</span>
    <span class="c1"># Uncomment and modify the line below to filter for a specific chromosome</span>
    <span class="c1"># ep_df = ep_df[ep_df[&#39;chrom(hg19)&#39;] == &#39;chr19&#39;]  </span>

    <span class="c1"># Setup parameters</span>
    <span class="n">input_shift</span> <span class="o">=</span> <span class="mi">500050</span>
    <span class="n">n_input_bins</span> <span class="o">=</span> <span class="mi">10001</span>
    <span class="n">n_input_tracks</span> <span class="o">=</span> <span class="mi">4</span>
    <span class="n">INPUT_SHAPE</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">10001</span><span class="p">)</span>
    <span class="n">bw_input_track_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ATAC_Seq.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K4me3.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27ac.bw&quot;</span><span class="p">,</span> <span class="s2">&quot;H3K27me3.bw&quot;</span><span class="p">]</span>
    <span class="n">activity_threshold</span> <span class="o">=</span> <span class="mi">10</span>

    <span class="c1"># Process enhancers in parallel</span>
    <span class="n">process_args</span> <span class="o">=</span> <span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="n">bw_input_track_list</span><span class="p">,</span> <span class="n">input_shift</span><span class="p">,</span> <span class="n">n_input_tracks</span><span class="p">,</span> <span class="n">n_input_bins</span><span class="p">,</span> 
                    <span class="n">INPUT_SHAPE</span><span class="p">,</span> <span class="n">path</span><span class="p">)</span>
    
    <span class="c1"># Use multiprocessing to process enhancers</span>
    <span class="k">with</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">process_enhancer</span><span class="p">,</span> 
                    <span class="p">[(</span><span class="n">process_args</span> <span class="o">+</span> <span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="n">activity_threshold</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ep_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()])</span>
        
    <span class="c1"># Filter out None results and create results DataFrame</span>
    <span class="n">valid_results</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span> <span class="k">if</span> <span class="n">r</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span>
    
    <span class="c1"># Create a mapping DataFrame</span>
    <span class="n">mapping_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">valid_results</span><span class="p">)</span>
    
    <span class="c1"># Save the mapping between accession IDs and enhancer coordinate IDs</span>
    <span class="n">mapping_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;human&quot;</span> <span class="o">/</span> <span class="s2">&quot;enhancer_mapping.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="c1"># Count unique enhancers processed</span>
    <span class="n">unique_enhancers</span> <span class="o">=</span> <span class="n">mapping_df</span><span class="p">[</span><span class="s1">&#39;enhancer_coord_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">valid_results</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs, representing </span><span class="si">{</span><span class="n">unique_enhancers</span><span class="si">}</span><span class="s2"> unique enhancers&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">valid_results</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs, representing </span><span class="si">{</span><span class="n">unique_enhancers</span><span class="si">}</span><span class="s2"> unique enhancers&quot;</span><span class="p">)</span>
    
    <span class="c1"># Create dummy target files</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_file_ids</span><span class="p">(</span><span class="n">directory</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get all file IDs from a directory without the .npy extension&quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">stem</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">Path</span><span class="p">(</span><span class="n">directory</span><span class="p">)</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s1">&#39;*.npy&#39;</span><span class="p">)]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_unified_target_file</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">200</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Create a unified target file with zeros for both regular and perturbed IDs&quot;&quot;&quot;</span>
        <span class="c1"># Create column names</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">_ctrl&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">n_bins</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">+</span><span class="mi">1</span><span class="p">)]</span>
        
        <span class="c1"># Create DataFrame with zeros</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">)),</span> <span class="n">columns</span><span class="o">=</span><span class="n">columns</span><span class="p">)</span>
        
        <span class="c1"># Fill ID column</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">baseline_ids</span>
        
        <span class="c1"># Save to CSV</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Created </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> entries&quot;</span><span class="p">)</span>

    <span class="c1"># Setup paths for target creation</span>
    <span class="n">baseline_dir</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;inputs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbed_landscape_arrays&quot;</span> <span class="o">/</span> <span class="s2">&quot;human_enhancer_arrays&quot;</span>
    <span class="n">output_dir</span> <span class="o">=</span> <span class="n">path</span> <span class="o">/</span> <span class="s2">&quot;targets&quot;</span> <span class="o">/</span> <span class="s2">&quot;K562&quot;</span>

    <span class="c1"># Create output directory if it doesn&#39;t exist</span>
    <span class="n">output_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Get IDs from both directories</span>
    <span class="n">baseline_ids</span> <span class="o">=</span> <span class="n">get_file_ids</span><span class="p">(</span><span class="n">baseline_dir</span><span class="p">)</span>

    <span class="c1"># Create unified target file</span>
    <span class="n">create_unified_target_file</span><span class="p">(</span><span class="n">baseline_ids</span><span class="p">,</span>
                              <span class="n">output_dir</span> <span class="o">/</span> <span class="s2">&quot;Enhancer_centered_targets_human.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Testing the model on K562 to predict POLR2A on enhancer centered, perturbed samples</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_K562_POLR2A_train/saved_models/gene_expression_only_chrom_K562_POLR2A_train_model_30300_perf-average<span class="o">=</span><span class="m">0</span>.5586.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/perturbation_runs/gene_expression_only_chrom_K562_POLR2A_enhancer_centric
</pre></div>
</div>
<p><strong>Testing the model on K562 to predict RNA-seq of enhancer-centered, perturbed samples</strong></p>
<blockquote>
<div><p>Nothing changes on globals, the only change is the model to use for the predictions, i.e. the one aimed to predict mature RNA-seq profiles, and the output folder name:</p>
</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>eirpredict<span class="w"> </span><span class="se">\</span>
--global_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/globals.yaml<span class="w"> </span><span class="se">\</span>
--input_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/input_cnn.yaml<span class="w"> </span><span class="se">\</span>
--fusion_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/fusion.yaml<span class="w"> </span><span class="se">\</span>
--output_configs<span class="w"> </span>./configurations/conf_pure_conv_K562_POLR2A_enhancer_centric/outputs_2_cond.yaml<span class="w"> </span><span class="se">\</span>
--evaluate<span class="w"> </span><span class="se">\</span>
--model_path<span class="w"> </span>./runs/gene_expression_only_chrom_K562_train/saved_models/gene_expression_only_chrom_K562_train_model_60600_perf-average<span class="o">=</span><span class="m">0</span>.5028.pt<span class="w"> </span><span class="se">\</span>
--output_folder<span class="w"> </span>./runs/perturbation_runs/gene_expression_only_chrom_K562_RNA_enhancer_centric
</pre></div>
</div>
</section>
<section id="enhancer-gene-pairing-analysis-on-predicted-rna-seq-and-polr2a-chip-seq">
<h3>Enhancer-gene pairing analysis on predicted RNA-seq and POLR2A Chip-seq.<a class="headerlink" href="#enhancer-gene-pairing-analysis-on-predicted-rna-seq-and-polr2a-chip-seq" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.integrate</span><span class="w"> </span><span class="kn">import</span> <span class="n">simps</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">confusion_matrix</span><span class="p">,</span> <span class="n">classification_report</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_annotations</span><span class="p">(</span><span class="n">gene_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load gene annotations.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">_get_predictions</span><span class="p">(</span>
    <span class="n">results_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">condition_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">is_training</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">batch_num</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">CLIP</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads prediction files from EIR for each target bin condition.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_path: Path where the predictions are stored</span>
<span class="sd">        N_BINS: Number of bins to one side of the central bin (e.g., 200 or 57 for Enformer)</span>
<span class="sd">        condition_list: List of conditions (e.g., [&quot;_ctrl&quot;])</span>
<span class="sd">        is_training: If True, uses training file structure, else uses test structure</span>
<span class="sd">        batch_num: Batch number for training predictions (e.g., &quot;30300&quot;)</span>
<span class="sd">        CLIP: If True, applies ReLU to predictions (clips at 0)</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        Tuple containing:</span>
<span class="sd">            - ids: List of sample names</span>
<span class="sd">            - predicted: DataFrame of predicted untransformed values</span>
<span class="sd">            - actual: DataFrame of actual values</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predicted_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">actual_dfs</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">condition</span> <span class="ow">in</span> <span class="n">condition_list</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">N_BINS</span><span class="p">,</span> <span class="n">N_BINS</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">batch_num</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;batch_num must be specified for training predictions&quot;</span><span class="p">)</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/samples/</span><span class="si">{</span><span class="n">batch_num</span><span class="si">}</span><span class="s2">/regression_predictions.csv&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">results_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;expression_output/</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">/predictions.csv&quot;</span>
            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">file_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Prediction file not found: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                
            <span class="n">predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;ID&#39;</span><span class="p">)</span>

            <span class="c1"># Apply ReLU to model predictions if specified</span>
            <span class="k">if</span> <span class="n">CLIP</span><span class="p">:</span>
                <span class="n">predictions</span> <span class="o">=</span> <span class="n">predictions</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">lower</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Handle different column names for training vs test</span>
            <span class="k">if</span> <span class="n">is_training</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Predicted&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;Actual&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">predicted_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2"> Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">actual_column</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="s2">&quot;True Label Untransformed&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">i</span><span class="si">}{</span><span class="n">condition</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">predicted_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">predicted_column</span><span class="p">)</span>
            <span class="n">actual_dfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">actual_column</span><span class="p">)</span>

    <span class="c1"># Concatenate all DataFrames horizontally</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">predicted_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">actual</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">actual_dfs</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ids</span><span class="p">,</span> <span class="n">predicted</span><span class="p">,</span> <span class="n">actual</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_window_boundaries</span><span class="p">(</span><span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">200500</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate window boundaries.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">enhancer_center</span> <span class="o">-</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">enhancer_center</span> <span class="o">+</span> <span class="n">window_size</span>

<span class="k">def</span><span class="w"> </span><span class="nf">find_genes_in_window</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find genes fully contained within window.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">genes_df</span><span class="p">[</span>
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;chr&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">chrom</span><span class="p">)</span> <span class="o">&amp;</span> 
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">window_start</span><span class="p">)</span> <span class="o">&amp;</span> 
        <span class="p">(</span><span class="n">genes_df</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">window_end</span><span class="p">)</span>
    <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_gene_tss</span><span class="p">(</span><span class="n">gene_row</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get TSS position based on strand.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="k">else</span> <span class="n">gene_row</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_distance_and_order</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate distances to enhancer and determine gene order.&quot;&quot;&quot;</span>
    <span class="n">distances</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">tss</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">enhancer_center</span> <span class="o">-</span> <span class="n">tss</span><span class="p">)</span>
        <span class="n">distances</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span>
            <span class="s1">&#39;distance&#39;</span><span class="p">:</span> <span class="n">distance</span><span class="p">,</span>
            <span class="s1">&#39;tss&#39;</span><span class="p">:</span> <span class="n">tss</span>
        <span class="p">})</span>
    
    <span class="n">sorted_distances</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">distances</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">])</span>
    
    <span class="k">return</span> <span class="p">{</span><span class="n">d</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]:</span> <span class="p">{</span>
        <span class="s1">&#39;distance&#39;</span><span class="p">:</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span>
        <span class="s1">&#39;order&#39;</span><span class="p">:</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
        <span class="s1">&#39;tss&#39;</span><span class="p">:</span> <span class="n">d</span><span class="p">[</span><span class="s1">&#39;tss&#39;</span><span class="p">]</span>
    <span class="p">}</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">d</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sorted_distances</span><span class="p">)}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">check_gene_adjacency</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">distance_map</span><span class="p">,</span> <span class="n">gene_id</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Check if a gene is adjacent to the enhancer.&quot;&quot;&quot;</span>
    <span class="n">gene_tss</span> <span class="o">=</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene_id</span><span class="p">][</span><span class="s1">&#39;tss&#39;</span><span class="p">]</span>
    <span class="n">gene_distance</span> <span class="o">=</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene_id</span><span class="p">][</span><span class="s1">&#39;distance&#39;</span><span class="p">]</span>
    
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">other_gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">other_gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene_id</span><span class="p">:</span>
            <span class="k">continue</span>
            
        <span class="n">other_tss</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">other_gene</span><span class="p">)</span>
        <span class="n">other_distance</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">other_tss</span> <span class="o">-</span> <span class="n">gene_tss</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">other_distance</span> <span class="o">&lt;</span> <span class="n">gene_distance</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
            
    <span class="k">return</span> <span class="kc">True</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_bin_coordinates</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert genomic position to bin coordinate.&quot;&quot;&quot;</span>
    <span class="n">rel_pos</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">-</span> <span class="n">enhancer_center</span>
    <span class="n">bin_pos</span> <span class="o">=</span> <span class="p">(</span><span class="n">rel_pos</span> <span class="o">+</span> <span class="n">window_size</span><span class="p">)</span> <span class="o">//</span> <span class="n">resolution</span>
    <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="n">bin_pos</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_gene_bin_coordinates</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get bin coordinates for TSS, gene start, and gene end.&quot;&quot;&quot;</span>
    <span class="n">tss</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="n">tss_bin</span> <span class="o">=</span> <span class="n">get_bin_coordinates</span><span class="p">(</span><span class="n">tss</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    
    <span class="n">start_bin</span> <span class="o">=</span> <span class="n">get_bin_coordinates</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">],</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    <span class="n">end_bin</span> <span class="o">=</span> <span class="n">get_bin_coordinates</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">],</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">tss_bin</span><span class="p">,</span> <span class="n">start_bin</span><span class="p">,</span> <span class="n">end_bin</span>

<span class="c1"># Abstract integration method for different prediction types</span>
<span class="k">def</span><span class="w"> </span><span class="nf">calculate_area</span><span class="p">(</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> 
    <span class="n">integration_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_strand</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">tss_bin</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">gene_start_bin</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">gene_end_bin</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">upstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">downstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate area under the curve using different integration methods.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        predictions: Array of prediction values</span>
<span class="sd">        integration_type: Type of integration (&#39;polr2a_tss&#39; or &#39;rnaseq_gene_body&#39;)</span>
<span class="sd">        gene_strand: Strand of the gene (&#39;+&#39; or &#39;-&#39;)</span>
<span class="sd">        tss_bin: Bin index of TSS (required for &#39;polr2a_tss&#39;)</span>
<span class="sd">        gene_start_bin: Start bin of gene (required for &#39;rnaseq_gene_body&#39;)</span>
<span class="sd">        gene_end_bin: End bin of gene (required for &#39;rnaseq_gene_body&#39;)</span>
<span class="sd">        upstream_bins: Number of bins to include upstream of TSS</span>
<span class="sd">        downstream_bins: Number of bins to include downstream of TSS</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        Normalized area under the curve</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_bins</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">predictions</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;polr2a_tss&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">tss_bin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;tss_bin must be provided for polr2a_tss integration&quot;</span><span class="p">)</span>
            
        <span class="c1"># For negative strand genes, swap upstream and downstream</span>
        <span class="k">if</span> <span class="n">gene_strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
            <span class="n">upstream_bins</span><span class="p">,</span> <span class="n">downstream_bins</span> <span class="o">=</span> <span class="n">downstream_bins</span><span class="p">,</span> <span class="n">upstream_bins</span>
            
        <span class="c1"># Define the region of interest around TSS</span>
        <span class="n">start_bin</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">tss_bin</span> <span class="o">-</span> <span class="n">upstream_bins</span><span class="p">)</span>
        <span class="n">end_bin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">total_bins</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">tss_bin</span> <span class="o">+</span> <span class="n">downstream_bins</span><span class="p">)</span>
        
    <span class="k">elif</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gene_start_bin</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">gene_end_bin</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;gene_start_bin and gene_end_bin must be provided for rnaseq_gene_body integration&quot;</span><span class="p">)</span>
            
        <span class="c1"># Get bins within gene body, clipped to prediction boundaries</span>
        <span class="n">start_bin</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">)</span>
        <span class="n">end_bin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">total_bins</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="p">)</span>
        
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown integration_type: </span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
    <span class="c1"># Extract values in the selected region</span>
    <span class="n">region_values</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="n">start_bin</span><span class="p">:</span><span class="n">end_bin</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_values</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">0</span>
    
    <span class="c1"># Calculate area using Simpson&#39;s rule</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">region_values</span><span class="p">))</span>
    
    <span class="c1"># Ensure x and y have the same dimensions before using Simpson&#39;s rule</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">region_values</span><span class="p">)</span>  <span class="c1"># For single point, just return the value</span>
    
    <span class="c1"># For arrays with only 2 points, use trapezoidal rule instead</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">trapz</span><span class="p">(</span><span class="n">region_values</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
    
    <span class="n">area</span> <span class="o">=</span> <span class="n">simps</span><span class="p">(</span><span class="n">region_values</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span>
    
    <span class="c1"># Normalize by length</span>
    <span class="n">length</span> <span class="o">=</span> <span class="n">end_bin</span> <span class="o">-</span> <span class="n">start_bin</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="c1">#return area / length if length &gt; 0 else 0</span>
    <span class="k">return</span> <span class="n">area</span> <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_enhancer_predictions</span><span class="p">(</span>
    <span class="n">enhancer_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> 
    <span class="n">predictions_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
    <span class="n">genes_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> 
    <span class="n">integration_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200500</span><span class="p">,</span> 
    <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">upstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> 
    <span class="n">downstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Dict</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process predictions for one enhancer and its genes.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        enhancer_id: ID of the enhancer</span>
<span class="sd">        predictions_df: DataFrame with predictions for all enhancers</span>
<span class="sd">        genes_df: DataFrame with gene annotations</span>
<span class="sd">        integration_type: Type of integration (&#39;polr2a_tss&#39; or &#39;rnaseq_gene_body&#39;)</span>
<span class="sd">        window_size: Size of the window around enhancer center</span>
<span class="sd">        resolution: Resolution of the data in base pairs per bin</span>
<span class="sd">        upstream_bins: Number of bins to include upstream of TSS (for polr2a_tss)</span>
<span class="sd">        downstream_bins: Number of bins to include downstream of TSS (for polr2a_tss)</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        List of dictionaries with results for each enhancer-gene pair</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract enhancer information</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(chr\w+)_(\d+)_(\d+)&#39;</span><span class="p">,</span> <span class="n">enhancer_id</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Could not parse enhancer ID: </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
    
    <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
    <span class="n">enhancer_center</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="n">end</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
    
    <span class="c1"># Check if we have both regular and perturbed predictions</span>
    <span class="n">perturbed_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_perturbed&quot;</span>
    <span class="k">if</span> <span class="n">perturbed_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Perturbed predictions not found for </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[]</span>
    
    <span class="c1"># Get prediction data</span>
    <span class="n">baseline_series</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enhancer_id</span><span class="p">]</span>
    <span class="n">perturbed_series</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">perturbed_id</span><span class="p">]</span>
    
    <span class="c1"># Convert series to numpy arrays for area calculation</span>
    <span class="n">baseline_array</span> <span class="o">=</span> <span class="n">baseline_series</span><span class="o">.</span><span class="n">values</span>
    <span class="n">perturbed_array</span> <span class="o">=</span> <span class="n">perturbed_series</span><span class="o">.</span><span class="n">values</span>
    
    <span class="c1"># Determine window boundaries and find genes in window</span>
    <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span> <span class="o">=</span> <span class="n">get_window_boundaries</span><span class="p">(</span><span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>
    <span class="n">window_genes</span> <span class="o">=</span> <span class="n">find_genes_in_window</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">window_genes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    
    <span class="n">distance_map</span> <span class="o">=</span> <span class="n">calculate_distance_and_order</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">)</span>
    
    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Calculate areas for all genes in window</span>
        <span class="n">area_results</span> <span class="o">=</span> <span class="p">[]</span>
        
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">gene_strand</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span>
            
            <span class="c1"># Get necessary bin coordinates</span>
            <span class="n">tss_bin</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span> <span class="o">=</span> <span class="n">get_gene_bin_coordinates</span><span class="p">(</span>
                <span class="n">gene</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span>
            <span class="p">)</span>
            
            <span class="c1"># Calculate areas based on integration type</span>
            <span class="k">if</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;polr2a_tss&#39;</span><span class="p">:</span>
                <span class="n">baseline_area</span> <span class="o">=</span> <span class="n">calculate_area</span><span class="p">(</span>
                    <span class="n">baseline_array</span><span class="p">,</span> <span class="n">integration_type</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span> 
                    <span class="n">tss_bin</span><span class="o">=</span><span class="n">tss_bin</span><span class="p">,</span> <span class="n">upstream_bins</span><span class="o">=</span><span class="n">upstream_bins</span><span class="p">,</span> <span class="n">downstream_bins</span><span class="o">=</span><span class="n">downstream_bins</span>
                <span class="p">)</span>
                <span class="n">perturbed_area</span> <span class="o">=</span> <span class="n">calculate_area</span><span class="p">(</span>
                    <span class="n">perturbed_array</span><span class="p">,</span> <span class="n">integration_type</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span> 
                    <span class="n">tss_bin</span><span class="o">=</span><span class="n">tss_bin</span><span class="p">,</span> <span class="n">upstream_bins</span><span class="o">=</span><span class="n">upstream_bins</span><span class="p">,</span> <span class="n">downstream_bins</span><span class="o">=</span><span class="n">downstream_bins</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">:</span>
                <span class="n">baseline_area</span> <span class="o">=</span> <span class="n">calculate_area</span><span class="p">(</span>
                    <span class="n">baseline_array</span><span class="p">,</span> <span class="n">integration_type</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span>
                    <span class="n">gene_start_bin</span><span class="o">=</span><span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="o">=</span><span class="n">gene_end_bin</span>
                <span class="p">)</span>
                <span class="n">perturbed_area</span> <span class="o">=</span> <span class="n">calculate_area</span><span class="p">(</span>
                    <span class="n">perturbed_array</span><span class="p">,</span> <span class="n">integration_type</span><span class="p">,</span> <span class="n">gene_strand</span><span class="p">,</span>
                    <span class="n">gene_start_bin</span><span class="o">=</span><span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span><span class="o">=</span><span class="n">gene_end_bin</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unknown integration_type: </span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            
            <span class="c1"># Calculate gene length in bp</span>
            <span class="n">gene_length_bp</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span>
            
            <span class="c1"># Check if gene is adjacent to enhancer</span>
            <span class="n">is_adjacent</span> <span class="o">=</span> <span class="n">check_gene_adjacency</span><span class="p">(</span><span class="n">window_genes</span><span class="p">,</span> <span class="n">distance_map</span><span class="p">,</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">])</span>
            
            <span class="c1"># Store area calculations</span>
            <span class="n">area_result</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_type&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;type&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_strand&#39;</span><span class="p">:</span> <span class="n">gene_strand</span><span class="p">,</span>
                <span class="s1">&#39;gene_length&#39;</span><span class="p">:</span> <span class="n">gene_length_bp</span><span class="p">,</span>
                <span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">:</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]][</span><span class="s1">&#39;distance&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_order&#39;</span><span class="p">:</span> <span class="n">distance_map</span><span class="p">[</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]][</span><span class="s1">&#39;order&#39;</span><span class="p">],</span>
                <span class="s1">&#39;is_adjacent&#39;</span><span class="p">:</span> <span class="n">is_adjacent</span><span class="p">,</span>
                <span class="s1">&#39;baseline_area&#39;</span><span class="p">:</span> <span class="n">baseline_area</span><span class="p">,</span>
                <span class="s1">&#39;perturbed_area&#39;</span><span class="p">:</span> <span class="n">perturbed_area</span><span class="p">,</span>
                <span class="s1">&#39;area_difference&#39;</span><span class="p">:</span> <span class="n">baseline_area</span> <span class="o">-</span> <span class="n">perturbed_area</span><span class="p">,</span>
                <span class="s1">&#39;abs_area_difference&#39;</span><span class="p">:</span> <span class="nb">abs</span><span class="p">(</span><span class="n">baseline_area</span> <span class="o">-</span> <span class="n">perturbed_area</span><span class="p">)</span>
            <span class="p">}</span>
            
            <span class="n">area_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">area_result</span><span class="p">)</span>
        
        <span class="c1"># Calculate z-scores for area differences within this window</span>
        <span class="n">area_diffs</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">area_diffs</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>  <span class="c1"># Need at least 2 points for z-score</span>
            <span class="n">mean_diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">area_diffs</span><span class="p">)</span>
            <span class="n">std_diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">area_diffs</span><span class="p">,</span> <span class="n">ddof</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># Use sample standard deviation</span>
            
            <span class="k">if</span> <span class="n">std_diff</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Avoid division by zero</span>
                <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
                    <span class="n">result</span><span class="p">[</span><span class="s1">&#39;z_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">mean_diff</span><span class="p">)</span> <span class="o">/</span> <span class="n">std_diff</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
                    <span class="n">result</span><span class="p">[</span><span class="s1">&#39;z_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="s1">&#39;z_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
        
        <span class="c1"># Calculate ratio to maximum score</span>
        <span class="n">max_diff</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">area_diffs</span><span class="p">)</span> <span class="k">if</span> <span class="n">area_diffs</span> <span class="k">else</span> <span class="mf">1.0</span>
        <span class="k">if</span> <span class="n">max_diff</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Avoid division by zero</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">max_diff</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
        
        <span class="c1"># Create final result entries</span>
        <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">area_results</span><span class="p">:</span>
            <span class="c1"># Create result entry</span>
            <span class="n">entry</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s1">&#39;enhancer_id&#39;</span><span class="p">:</span> <span class="n">enhancer_id</span><span class="p">,</span>
                <span class="s1">&#39;baseline_area&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;baseline_area&#39;</span><span class="p">],</span>
                <span class="s1">&#39;perturbed_area&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;perturbed_area&#39;</span><span class="p">],</span>
                <span class="s1">&#39;area_difference&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;area_difference&#39;</span><span class="p">],</span>
                <span class="s1">&#39;abs_area_difference&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span> <span class="p">,</span>
                <span class="s1">&#39;z_score&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;z_score&#39;</span><span class="p">],</span>
                <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;ratio_to_max&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_id&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_name&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_order&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_order&#39;</span><span class="p">],</span>
                <span class="s1">&#39;is_adjacent&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;is_adjacent&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_type&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_type&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_strand&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_strand&#39;</span><span class="p">],</span>
                <span class="s1">&#39;gene_length&#39;</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;gene_length&#39;</span><span class="p">],</span>
                <span class="s1">&#39;integration_type&#39;</span><span class="p">:</span> <span class="n">integration_type</span>
            <span class="p">}</span>
            
            <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">entry</span><span class="p">)</span>
            
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error processing enhancer </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">results</span>

<span class="k">def</span><span class="w"> </span><span class="nf">identify_primary_target_genes</span><span class="p">(</span><span class="n">results_df</span><span class="p">,</span> <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Identify the primary target gene for each enhancer based on the specified scoring method.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_df: DataFrame with enhancer-gene pair results</span>
<span class="sd">        score_method: Scoring method to use (&#39;abs_area_difference&#39;, &#39;z_score&#39;, or &#39;ratio_to_max&#39;)</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        DataFrame with added &#39;is_primary_target&#39; column</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Make a copy to avoid modifying the original</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">results_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="k">if</span> <span class="n">score_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Score method &#39;</span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s2">&#39; not found in results. Using &#39;abs_area_difference&#39; instead.&quot;</span><span class="p">)</span>
        <span class="n">score_method</span> <span class="o">=</span> <span class="s1">&#39;abs_area_difference&#39;</span>
    
    <span class="c1"># Group by enhancer and find gene with maximum score</span>
    <span class="n">primary_targets</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">)[</span><span class="n">score_method</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()]</span>
    
    <span class="c1"># Create a unique ID for each enhancer-gene pair</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
    <span class="n">primary_targets</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">primary_targets</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">primary_targets</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
    
    <span class="c1"># Mark primary target genes</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;is_primary_target&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">primary_targets</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">])</span>
    
    <span class="c1"># Create is_closest flag (for the closest gene by distance)</span>
    <span class="n">closest_genes</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">)[</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">idxmin</span><span class="p">()]</span>
    <span class="n">closest_genes</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">closest_genes</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">closest_genes</span><span class="p">[</span><span class="s1">&#39;gene_id&#39;</span><span class="p">]</span>
    <span class="n">df</span><span class="p">[</span><span class="s1">&#39;is_closest&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">closest_genes</span><span class="p">[</span><span class="s1">&#39;pair_id&#39;</span><span class="p">])</span>
    
    <span class="c1"># Remove temporary column</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;pair_id&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">df</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_confusion_matrix</span><span class="p">(</span><span class="n">results_df</span><span class="p">,</span> <span class="n">score_method</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create confusion matrix comparing primary target identification methods.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_df: DataFrame with results including is_primary_target and is_closest flags</span>
<span class="sd">        score_method: Scoring method used to identify primary targets</span>
<span class="sd">        output_path: Path to save the visualization</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create figure</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">4.5</span><span class="p">,</span> <span class="mf">4.5</span><span class="p">))</span>
    
    <span class="c1"># Create confusion matrix</span>
    <span class="n">cm</span> <span class="o">=</span> <span class="n">confusion_matrix</span><span class="p">(</span><span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target&#39;</span><span class="p">],</span> <span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;is_closest&#39;</span><span class="p">])</span>
    
    <span class="c1"># Plot using seaborn</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">cm</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span>
               <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Closest&#39;</span><span class="p">,</span> <span class="s1">&#39;Closest&#39;</span><span class="p">],</span>
               <span class="n">yticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Primary Target&#39;</span><span class="p">,</span> <span class="s1">&#39;Primary Target&#39;</span><span class="p">])</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Proximity-based Prediction&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">score_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">()</span><span class="si">}</span><span class="s1">-based Prediction&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Primary Target Gene Identification: </span><span class="si">{</span><span class="n">score_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">()</span><span class="si">}</span><span class="s1"> vs. Proximity&#39;</span><span class="p">)</span>
    
    <span class="c1"># Add accuracy and other metrics</span>
    <span class="n">accuracy</span> <span class="o">=</span> <span class="p">(</span><span class="n">cm</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">cm</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="n">cm</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figtext</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.01</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Accuracy: </span><span class="si">{</span><span class="n">accuracy</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
    
    <span class="c1"># Save figure</span>
    <span class="n">out_file</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;target_identification_confusion_</span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s1">.png&#39;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="c1"># Generate classification report</span>
    <span class="n">report</span> <span class="o">=</span> <span class="n">classification_report</span><span class="p">(</span>
        <span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target&#39;</span><span class="p">],</span> 
        <span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;is_closest&#39;</span><span class="p">],</span>
        <span class="n">target_names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Primary Target&#39;</span><span class="p">,</span> <span class="s1">&#39;Primary Target&#39;</span><span class="p">],</span>
        <span class="n">output_dict</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    
    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;confusion_matrix&#39;</span><span class="p">:</span> <span class="n">cm</span><span class="p">,</span>
        <span class="s1">&#39;accuracy&#39;</span><span class="p">:</span> <span class="n">accuracy</span><span class="p">,</span>
        <span class="s1">&#39;classification_report&#39;</span><span class="p">:</span> <span class="n">report</span>
    <span class="p">}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">visualize_enhancer_predictions</span><span class="p">(</span>
    <span class="n">enhancer_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">predictions_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">genes_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">integration_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200500</span><span class="p">,</span>
    <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">upstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
    <span class="n">downstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">save_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">show_plot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">plt</span><span class="o">.</span><span class="n">Figure</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize predictions and integration domains for an enhancer.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Extract enhancer information</span>
    <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(chr\w+)_(\d+)_(\d+)&#39;</span><span class="p">,</span> <span class="n">enhancer_id</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">match</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Could not parse enhancer ID: </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span>
    <span class="n">enhancer_center</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="n">end</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>
    
    <span class="c1"># Check if we have both regular and perturbed predictions</span>
    <span class="n">perturbed_id</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_perturbed&quot;</span>
    <span class="k">if</span> <span class="n">perturbed_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Perturbed predictions not found for </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="c1"># Get prediction data</span>
    <span class="n">baseline_series</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enhancer_id</span><span class="p">]</span>
    <span class="n">perturbed_series</span> <span class="o">=</span> <span class="n">predictions_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">perturbed_id</span><span class="p">]</span>
    
    <span class="c1"># Get positions for x-axis (relative to enhancer center)</span>
    <span class="n">bin_range</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="o">-</span><span class="n">window_size</span><span class="o">//</span><span class="n">resolution</span><span class="p">,</span> <span class="n">window_size</span><span class="o">//</span><span class="n">resolution</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># inclusive</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">bin_range</span><span class="p">)</span> <span class="o">*</span> <span class="n">resolution</span>
    
    <span class="c1"># Ensure positions and values arrays have the same length</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">baseline_series</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
        <span class="c1"># Adjust the smaller array to match the larger one</span>
        <span class="n">min_length</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">baseline_series</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="n">positions</span><span class="p">[:</span><span class="n">min_length</span><span class="p">]</span>
        <span class="n">baseline_values</span> <span class="o">=</span> <span class="n">baseline_series</span><span class="o">.</span><span class="n">values</span><span class="p">[:</span><span class="n">min_length</span><span class="p">]</span>
        <span class="n">perturbed_values</span> <span class="o">=</span> <span class="n">perturbed_series</span><span class="o">.</span><span class="n">values</span><span class="p">[:</span><span class="n">min_length</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">baseline_values</span> <span class="o">=</span> <span class="n">baseline_series</span><span class="o">.</span><span class="n">values</span>
        <span class="n">perturbed_values</span> <span class="o">=</span> <span class="n">perturbed_series</span><span class="o">.</span><span class="n">values</span>
    
    <span class="c1"># Create figure</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    
    <span class="c1"># Plot baseline and perturbed predictions</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">baseline_values</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Baseline&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">perturbed_values</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Perturbed&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    
    <span class="c1"># Determine window boundaries and find genes in window</span>
    <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span> <span class="o">=</span> <span class="n">get_window_boundaries</span><span class="p">(</span><span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>
    <span class="n">window_genes</span> <span class="o">=</span> <span class="n">find_genes_in_window</span><span class="p">(</span><span class="n">genes_df</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">window_start</span><span class="p">,</span> <span class="n">window_end</span><span class="p">)</span>
    
    <span class="c1"># Mark enhancer position</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Enhancer Center&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">start</span> <span class="o">-</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">end</span> <span class="o">-</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Enhancer Region&#39;</span><span class="p">)</span>
    
    <span class="c1"># Add gene visualization</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">tab10</span><span class="o">.</span><span class="n">colors</span>
    
    <span class="c1"># Calculate y-axis limits for proper positioning of gene visualizations</span>
    <span class="n">y_min</span><span class="p">,</span> <span class="n">y_max</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_ylim</span><span class="p">()</span>
    <span class="n">highlight_height</span> <span class="o">=</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">y_max</span> <span class="o">-</span> <span class="n">y_min</span><span class="p">)</span>  <span class="c1"># Height for visualization elements</span>
    
    <span class="c1"># Add legend entries for genes</span>
    <span class="n">gene_legend_handles</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">gene</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">window_genes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()):</span>
        <span class="n">gene_strand</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Strand&#39;</span><span class="p">]</span>
        <span class="n">gene_color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="nb">len</span><span class="p">(</span><span class="n">colors</span><span class="p">)]</span>
        
        <span class="c1"># Get necessary bin coordinates</span>
        <span class="n">tss_bin</span><span class="p">,</span> <span class="n">gene_start_bin</span><span class="p">,</span> <span class="n">gene_end_bin</span> <span class="o">=</span> <span class="n">get_gene_bin_coordinates</span><span class="p">(</span>
            <span class="n">gene</span><span class="p">,</span> <span class="n">enhancer_center</span><span class="p">,</span> <span class="n">window_size</span><span class="p">,</span> <span class="n">resolution</span>
        <span class="p">)</span>
        
        <span class="c1"># Convert bin positions to genomic coordinates relative to enhancer center</span>
        <span class="n">tss_rel</span> <span class="o">=</span> <span class="n">get_gene_tss</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">-</span> <span class="n">enhancer_center</span>
        <span class="n">start_rel</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Start&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">enhancer_center</span>
        <span class="n">end_rel</span> <span class="o">=</span> <span class="n">gene</span><span class="p">[</span><span class="s1">&#39;End&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">enhancer_center</span>
        
        <span class="c1"># Plot TSS vertical line</span>
        <span class="n">tss_line</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">tss_rel</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
        
        <span class="c1"># Highlight integration regions based on type</span>
        <span class="k">if</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;polr2a_tss&#39;</span><span class="p">:</span>
            <span class="c1"># For POLR2A, highlight region around TSS</span>
            <span class="k">if</span> <span class="n">gene_strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
                <span class="n">domain_start</span> <span class="o">=</span> <span class="n">tss_rel</span> <span class="o">-</span> <span class="p">(</span><span class="n">upstream_bins</span> <span class="o">*</span> <span class="n">resolution</span><span class="p">)</span>
                <span class="n">domain_end</span> <span class="o">=</span> <span class="n">tss_rel</span> <span class="o">+</span> <span class="p">(</span><span class="n">downstream_bins</span> <span class="o">*</span> <span class="n">resolution</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># &#39;-&#39; strand</span>
                <span class="n">domain_start</span> <span class="o">=</span> <span class="n">tss_rel</span> <span class="o">-</span> <span class="p">(</span><span class="n">downstream_bins</span> <span class="o">*</span> <span class="n">resolution</span><span class="p">)</span>
                <span class="n">domain_end</span> <span class="o">=</span> <span class="n">tss_rel</span> <span class="o">+</span> <span class="p">(</span><span class="n">upstream_bins</span> <span class="o">*</span> <span class="n">resolution</span><span class="p">)</span>
            
            <span class="c1"># Shade integration domain</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="n">domain_start</span><span class="p">,</span> <span class="n">domain_end</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene_color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
            
        <span class="k">elif</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">:</span>
            <span class="c1"># For RNA-seq, highlight entire gene body</span>
            <span class="n">rect_y_bottom</span> <span class="o">=</span> <span class="n">y_min</span>
            <span class="n">rect_y_top</span> <span class="o">=</span> <span class="n">y_min</span> <span class="o">+</span> <span class="n">highlight_height</span>
            
            <span class="c1"># Create rectangle for gene body</span>
            <span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="kn">import</span> <span class="n">Rectangle</span>
            <span class="n">gene_rect</span> <span class="o">=</span> <span class="n">Rectangle</span><span class="p">((</span><span class="n">start_rel</span><span class="p">,</span> <span class="n">rect_y_bottom</span><span class="p">),</span> <span class="n">end_rel</span> <span class="o">-</span> <span class="n">start_rel</span><span class="p">,</span> <span class="n">highlight_height</span><span class="p">,</span>
                                 <span class="n">facecolor</span><span class="o">=</span><span class="n">gene_color</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;none&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">add_patch</span><span class="p">(</span><span class="n">gene_rect</span><span class="p">)</span>
            
            <span class="c1"># Add arrow to indicate strand</span>
            <span class="n">arrow_y</span> <span class="o">=</span> <span class="n">rect_y_bottom</span> <span class="o">+</span> <span class="n">highlight_height</span> <span class="o">/</span> <span class="mi">2</span>
            <span class="n">arrow_length</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mf">0.1</span> <span class="o">*</span> <span class="p">(</span><span class="n">end_rel</span> <span class="o">-</span> <span class="n">start_rel</span><span class="p">),</span> <span class="mi">5000</span><span class="p">)</span>  <span class="c1"># 10% of gene length or max 5kb</span>
            
            <span class="k">if</span> <span class="n">gene_strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
                <span class="n">arrow_x</span> <span class="o">=</span> <span class="n">end_rel</span> <span class="o">-</span> <span class="n">arrow_length</span>
                <span class="n">dx</span> <span class="o">=</span> <span class="n">arrow_length</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># &#39;-&#39; strand</span>
                <span class="n">arrow_x</span> <span class="o">=</span> <span class="n">start_rel</span>
                <span class="n">dx</span> <span class="o">=</span> <span class="o">-</span><span class="n">arrow_length</span>
                
            <span class="n">ax</span><span class="o">.</span><span class="n">arrow</span><span class="p">(</span><span class="n">arrow_x</span><span class="p">,</span> <span class="n">arrow_y</span><span class="p">,</span> <span class="n">dx</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">head_width</span><span class="o">=</span><span class="n">highlight_height</span><span class="o">*</span><span class="mf">0.8</span><span class="p">,</span> 
                    <span class="n">head_length</span><span class="o">=</span><span class="nb">min</span><span class="p">(</span><span class="mi">2000</span><span class="p">,</span> <span class="n">arrow_length</span><span class="o">*</span><span class="mf">0.3</span><span class="p">),</span> <span class="n">fc</span><span class="o">=</span><span class="n">gene_color</span><span class="p">,</span> <span class="n">ec</span><span class="o">=</span><span class="n">gene_color</span><span class="p">)</span>
        
        <span class="c1"># Create custom legend entry for this gene</span>
        <span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="kn">import</span> <span class="n">Patch</span>
        <span class="n">gene_patch</span> <span class="o">=</span> <span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">gene_color</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> 
                         <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;Name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">gene_strand</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
        <span class="n">gene_legend_handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_patch</span><span class="p">)</span>
    
    <span class="c1"># Set x-axis limits to match the data range</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="n">window_size</span><span class="p">,</span> <span class="n">window_size</span><span class="p">)</span>
    
    <span class="c1"># Set x-axis ticks in kbp</span>
    <span class="n">x_ticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">window_size</span><span class="p">,</span> <span class="n">window_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">50000</span><span class="p">)</span>  <span class="c1"># Every 50 kbp</span>
    <span class="n">x_tick_labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="o">/</span><span class="mi">1000</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">x_ticks</span><span class="p">]</span>  <span class="c1"># Format as integer</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">x_ticks</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">x_tick_labels</span><span class="p">)</span>
    
    <span class="c1"># Set labels and title</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Position relative to enhancer center (kbp)&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Prediction value&#39;</span><span class="p">)</span>
    <span class="n">method_name</span> <span class="o">=</span> <span class="s2">&quot;POLR2A around TSS&quot;</span> <span class="k">if</span> <span class="n">integration_type</span> <span class="o">==</span> <span class="s1">&#39;polr2a_tss&#39;</span> <span class="k">else</span> <span class="s2">&quot;RNA-seq across gene body&quot;</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Enhancer </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s1"> predictions (</span><span class="si">{</span><span class="n">method_name</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
    
    <span class="c1"># Add vertical line legend entries</span>
    <span class="n">line_legend_handles</span><span class="p">,</span> <span class="n">line_legend_labels</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
    
    <span class="c1"># Combine both legend entries</span>
    <span class="n">all_handles</span> <span class="o">=</span> <span class="n">line_legend_handles</span> <span class="o">+</span> <span class="n">gene_legend_handles</span>
    
    <span class="c1"># Add legend (positioned outside the plot)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">all_handles</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    
    <span class="c1"># Adjust layout</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    
    <span class="c1"># Save figure if path is provided</span>
    <span class="k">if</span> <span class="n">save_path</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    
    <span class="c1"># Show plot if requested</span>
    <span class="k">if</span> <span class="n">show_plot</span><span class="p">:</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    
    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">visualize_primary_targets</span><span class="p">(</span><span class="n">results_df</span><span class="p">,</span> <span class="n">score_method</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize statistics about primary targets vs. proximity-based targets.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        results_df: DataFrame with results including primary target flags</span>
<span class="sd">        score_method: Scoring method used to identify primary targets</span>
<span class="sd">        output_path: Path to save visualizations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create figure for distance distribution</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Create distance histograms for primary targets vs non-primary</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">histplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">results_df</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s1">&#39;distance_to_enhancer&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;is_primary_target&#39;</span><span class="p">,</span>
                <span class="n">multiple</span><span class="o">=</span><span class="s1">&#39;dodge&#39;</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">log_scale</span><span class="o">=</span><span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Distance to Enhancer (bp)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Count (log scale)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Distance Distribution: </span><span class="si">{</span><span class="n">score_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">()</span><span class="si">}</span><span class="s1"> Primary Target Genes vs. Others&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;primary_target_distance_distribution_</span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="c1"># Create figure for gene order comparison</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Count occurrences of each gene order for primary targets</span>
    <span class="n">order_counts</span> <span class="o">=</span> <span class="n">results_df</span><span class="p">[</span><span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target&#39;</span><span class="p">]][</span><span class="s1">&#39;gene_order&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
    
    <span class="c1"># Plot gene order distribution</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">barplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">order_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">order_counts</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Gene Order (by Distance)&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Count of Primary Targets&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Distribution of </span><span class="si">{</span><span class="n">score_method</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot; &quot;</span><span class="p">)</span><span class="o">.</span><span class="n">title</span><span class="p">()</span><span class="si">}</span><span class="s1"> Primary Target Genes by Distance Order&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s1">&#39;primary_target_order_distribution_</span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s1">.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="k">def</span><span class="w"> </span><span class="nf">batch_process_enhancers</span><span class="p">(</span>
    <span class="n">enhancer_ids</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">predictions_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">genes_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">integration_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">visualize_top_n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">score_methods</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">,</span> <span class="s1">&#39;z_score&#39;</span><span class="p">,</span> <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">],</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process multiple enhancers and generate visualizations.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        enhancer_ids: List of enhancer IDs to process</span>
<span class="sd">        predictions_df: DataFrame with predictions</span>
<span class="sd">        genes_df: DataFrame with gene annotations</span>
<span class="sd">        integration_type: Type of integration (&#39;polr2a_tss&#39; or &#39;rnaseq_gene_body&#39;)</span>
<span class="sd">        output_path: Path to save results</span>
<span class="sd">        visualize_top_n: Number of top enhancers to visualize</span>
<span class="sd">        score_methods: List of scoring methods to use</span>
<span class="sd">        **kwargs: Additional arguments for process_enhancer_predictions</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">        DataFrame with combined results</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Create visualization output directory</span>
    <span class="n">vis_output_path</span> <span class="o">=</span> <span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">_visualizations&quot;</span>
    <span class="n">vis_output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Process each enhancer</span>
    <span class="n">all_results</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">enhancer_id</span> <span class="ow">in</span> <span class="n">enhancer_ids</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">process_enhancer_predictions</span><span class="p">(</span>
            <span class="n">enhancer_id</span><span class="p">,</span> 
            <span class="n">predictions_df</span><span class="p">,</span> 
            <span class="n">genes_df</span><span class="p">,</span> 
            <span class="n">integration_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">all_results</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
    
    <span class="c1"># Save results</span>
    <span class="k">if</span> <span class="n">all_results</span><span class="p">:</span>
        <span class="c1"># Create results DataFrame</span>
        <span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">all_results</span><span class="p">)</span>
        
        <span class="c1"># Process for each scoring method</span>
        <span class="n">all_results_with_targets</span> <span class="o">=</span> <span class="p">{}</span>
        
        <span class="k">for</span> <span class="n">score_method</span> <span class="ow">in</span> <span class="n">score_methods</span><span class="p">:</span>
            <span class="c1"># Add primary target flags based on current scoring method</span>
            <span class="n">results_with_targets</span> <span class="o">=</span> <span class="n">identify_primary_target_genes</span><span class="p">(</span><span class="n">results_df</span><span class="p">,</span> <span class="n">score_method</span><span class="p">)</span>
            <span class="n">all_results_with_targets</span><span class="p">[</span><span class="n">score_method</span><span class="p">]</span> <span class="o">=</span> <span class="n">results_with_targets</span>
            
            <span class="c1"># Save method-specific results</span>
            <span class="n">results_with_targets</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s2">_results_with_targets.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            
            <span class="c1"># Create visualization for primary target stats</span>
            <span class="n">visualize_primary_targets</span><span class="p">(</span><span class="n">results_with_targets</span><span class="p">,</span> <span class="n">score_method</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
            
            <span class="c1"># Create confusion matrix for primary vs. closest identification</span>
            <span class="n">cm_results</span> <span class="o">=</span> <span class="n">plot_confusion_matrix</span><span class="p">(</span><span class="n">results_with_targets</span><span class="p">,</span> <span class="n">score_method</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
            
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Processing complete for </span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2"> using </span><span class="si">{</span><span class="n">score_method</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processed </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancer_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancers&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">results_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Primary target identification accuracy: </span><span class="si">{</span><span class="n">cm_results</span><span class="p">[</span><span class="s1">&#39;accuracy&#39;</span><span class="p">]</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Save base results</span>
        <span class="n">results_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">_base_results.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
        <span class="c1"># Use the standard abs_area_difference for further operations</span>
        <span class="n">results_with_targets</span> <span class="o">=</span> <span class="n">all_results_with_targets</span><span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span>
        
        <span class="c1"># Generate visualizations for top enhancers by effect size</span>
        <span class="k">if</span> <span class="n">visualize_top_n</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">top_enhancers</span> <span class="o">=</span> <span class="n">results_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">)[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="n">visualize_top_n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Visualizing top </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">top_enhancers</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancers by effect size...&quot;</span><span class="p">)</span>
            
            <span class="k">for</span> <span class="n">enhancer_id</span> <span class="ow">in</span> <span class="n">top_enhancers</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Visualizing </span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
                <span class="n">save_path</span> <span class="o">=</span> <span class="n">vis_output_path</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">enhancer_id</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">_visualization.png&quot;</span>
                
                <span class="n">fig</span> <span class="o">=</span> <span class="n">visualize_enhancer_predictions</span><span class="p">(</span>
                    <span class="n">enhancer_id</span><span class="o">=</span><span class="n">enhancer_id</span><span class="p">,</span>
                    <span class="n">predictions_df</span><span class="o">=</span><span class="n">predictions_df</span><span class="p">,</span>
                    <span class="n">genes_df</span><span class="o">=</span><span class="n">genes_df</span><span class="p">,</span>
                    <span class="n">integration_type</span><span class="o">=</span><span class="n">integration_type</span><span class="p">,</span>
                    <span class="n">save_path</span><span class="o">=</span><span class="n">save_path</span><span class="p">,</span>
                    <span class="n">show_plot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="o">**</span><span class="n">kwargs</span>
                <span class="p">)</span>
                
                <span class="k">if</span> <span class="n">fig</span><span class="p">:</span>
                    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
            
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Visualizations saved to </span><span class="si">{</span><span class="n">vis_output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">all_results_with_targets</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">No results were generated for </span><span class="si">{</span><span class="n">integration_type</span><span class="si">}</span><span class="s2">. Please check the error messages above.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">run_analysis</span><span class="p">(</span>
    <span class="n">base_path</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
    <span class="n">gene_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">polr2a_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">rna_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="n">Path</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">window_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200500</span><span class="p">,</span>
    <span class="n">resolution</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">upstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
    <span class="n">downstream_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
    <span class="n">N_BINS</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span>
    <span class="n">visualize_top_n</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="n">score_methods</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">,</span> <span class="s1">&#39;z_score&#39;</span><span class="p">,</span> <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run the full analysis pipeline.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        base_path: Base directory path</span>
<span class="sd">        gene_path: Path to gene annotations (default: base_path/annotations/human/Final_gene_annotations_human.tsv)</span>
<span class="sd">        polr2a_path: Path to POLR2A results (default: base_path/runs/perturbation_runs/gene_expression_only_chrom_K562_POLR2A_enhancer_centric)</span>
<span class="sd">        rna_path: Path to RNA-seq results (default: base_path/runs/perturbation_runs/gene_expression_only_chrom_K562_RNA_enhancer_centric)</span>
<span class="sd">        output_path: Path to save output (default: base_path/tables)</span>
<span class="sd">        window_size: Size of the window around enhancer center</span>
<span class="sd">        resolution: Resolution of the data in base pairs per bin</span>
<span class="sd">        upstream_bins: Number of bins to include upstream of TSS (for polr2a_tss)</span>
<span class="sd">        downstream_bins: Number of bins to include downstream of TSS (for polr2a_tss)</span>
<span class="sd">        N_BINS: Number of bins to one side of the central bin</span>
<span class="sd">        visualize_top_n: Number of top enhancers to visualize (0 to skip visualization)</span>
<span class="sd">        score_methods: List of scoring methods to use</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        Dictionary containing DataFrames with results for each analysis type</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Set default paths if not provided</span>
    <span class="k">if</span> <span class="n">gene_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">gene_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;human&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations_human.tsv&quot;</span>
    
    <span class="k">if</span> <span class="n">polr2a_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">polr2a_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbation_runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;gene_expression_only_chrom_K562_POLR2A_enhancer_centric&quot;</span>
    
    <span class="k">if</span> <span class="n">rna_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">rna_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbation_runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;gene_expression_only_chrom_K562_RNA_enhancer_centric&quot;</span>
    
    <span class="k">if</span> <span class="n">output_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;tables&quot;</span>
    
    <span class="n">output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Load gene annotations</span>
    <span class="n">genes_df</span> <span class="o">=</span> <span class="n">load_annotations</span><span class="p">(</span><span class="n">gene_path</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">genes_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> gene annotations&quot;</span><span class="p">)</span>
    
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    
    <span class="c1"># Process POLR2A data if path exists</span>
    <span class="k">if</span> <span class="n">polr2a_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Processing POLR2A data...&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading predictions from </span><span class="si">{</span><span class="n">polr2a_path</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
        
        <span class="n">ids</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span>
            <span class="n">results_path</span><span class="o">=</span><span class="n">polr2a_path</span><span class="p">,</span>
            <span class="n">N_BINS</span><span class="o">=</span><span class="n">N_BINS</span><span class="p">,</span>
            <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span>
            <span class="n">is_training</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded predictions for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> samples&quot;</span><span class="p">)</span>
        
        <span class="c1"># Get unique enhancer IDs (excluding perturbed ones)</span>
        <span class="n">enhancer_ids</span> <span class="o">=</span> <span class="p">[</span><span class="nb">id</span> <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">ids</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">id</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;_perturbed&#39;</span><span class="p">)]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancer_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> unique enhancers&quot;</span><span class="p">)</span>
        
        <span class="c1"># Process enhancers</span>
        <span class="n">polr2a_results</span> <span class="o">=</span> <span class="n">batch_process_enhancers</span><span class="p">(</span>
            <span class="n">enhancer_ids</span><span class="o">=</span><span class="n">enhancer_ids</span><span class="p">,</span>
            <span class="n">predictions_df</span><span class="o">=</span><span class="n">predictions_df</span><span class="p">,</span>
            <span class="n">genes_df</span><span class="o">=</span><span class="n">genes_df</span><span class="p">,</span>
            <span class="n">integration_type</span><span class="o">=</span><span class="s1">&#39;polr2a_tss&#39;</span><span class="p">,</span>
            <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span>
            <span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span>
            <span class="n">resolution</span><span class="o">=</span><span class="n">resolution</span><span class="p">,</span>
            <span class="n">upstream_bins</span><span class="o">=</span><span class="n">upstream_bins</span><span class="p">,</span>
            <span class="n">downstream_bins</span><span class="o">=</span><span class="n">downstream_bins</span><span class="p">,</span>
            <span class="n">visualize_top_n</span><span class="o">=</span><span class="n">visualize_top_n</span><span class="p">,</span>
            <span class="n">score_methods</span><span class="o">=</span><span class="n">score_methods</span>
        <span class="p">)</span>
        
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;polr2a_tss&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">polr2a_results</span>
    
    <span class="c1"># Process RNA-seq data if path exists</span>
    <span class="k">if</span> <span class="n">rna_path</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Processing RNA-seq data...&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading predictions from </span><span class="si">{</span><span class="n">rna_path</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
        
        <span class="n">ids</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">_get_predictions</span><span class="p">(</span>
            <span class="n">results_path</span><span class="o">=</span><span class="n">rna_path</span><span class="p">,</span>
            <span class="n">N_BINS</span><span class="o">=</span><span class="n">N_BINS</span><span class="p">,</span>
            <span class="n">condition_list</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;_ctrl&quot;</span><span class="p">],</span>
            <span class="n">is_training</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded predictions for </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> samples&quot;</span><span class="p">)</span>
        
        <span class="c1"># Get unique enhancer IDs (excluding perturbed ones)</span>
        <span class="n">enhancer_ids</span> <span class="o">=</span> <span class="p">[</span><span class="nb">id</span> <span class="k">for</span> <span class="nb">id</span> <span class="ow">in</span> <span class="n">ids</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">id</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;_perturbed&#39;</span><span class="p">)]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancer_ids</span><span class="p">)</span><span class="si">}</span><span class="s2"> unique enhancers&quot;</span><span class="p">)</span>
        
        <span class="c1"># Process enhancers</span>
        <span class="n">rna_results</span> <span class="o">=</span> <span class="n">batch_process_enhancers</span><span class="p">(</span>
            <span class="n">enhancer_ids</span><span class="o">=</span><span class="n">enhancer_ids</span><span class="p">,</span>
            <span class="n">predictions_df</span><span class="o">=</span><span class="n">predictions_df</span><span class="p">,</span>
            <span class="n">genes_df</span><span class="o">=</span><span class="n">genes_df</span><span class="p">,</span>
            <span class="n">integration_type</span><span class="o">=</span><span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">,</span>
            <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span>
            <span class="n">window_size</span><span class="o">=</span><span class="n">window_size</span><span class="p">,</span>
            <span class="n">resolution</span><span class="o">=</span><span class="n">resolution</span><span class="p">,</span>
            <span class="n">visualize_top_n</span><span class="o">=</span><span class="n">visualize_top_n</span><span class="p">,</span>
            <span class="n">score_methods</span><span class="o">=</span><span class="n">score_methods</span>
        <span class="p">)</span>
        
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">rna_results</span>
    
    <span class="k">return</span> <span class="n">results</span>

<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main function to run the analysis pipeline.&quot;&quot;&quot;</span>
    <span class="c1"># Define base path</span>
    <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    
    <span class="c1"># Run analysis</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Starting enhancer effect analysis pipeline...&quot;</span><span class="p">)</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">run_analysis</span><span class="p">(</span><span class="n">base_path</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Analysis complete!&quot;</span><span class="p">)</span>
    
    <span class="c1"># If both POLR2A and RNA-seq results are available, run comparison</span>
    <span class="k">if</span> <span class="s1">&#39;polr2a_tss&#39;</span> <span class="ow">in</span> <span class="n">results</span> <span class="ow">and</span> <span class="s1">&#39;rnaseq_gene_body&#39;</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Both POLR2A and RNA-seq results available. Running comparison...&quot;</span><span class="p">)</span>
        
        <span class="c1"># Create comparison output directory</span>
        <span class="n">comparison_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;results_CRISPR&quot;</span> <span class="o">/</span> <span class="s2">&quot;Integration_comparison&quot;</span>
        <span class="n">comparison_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
        <span class="c1"># Save individual result files for reference</span>
        <span class="n">polr2a_file</span> <span class="o">=</span> <span class="n">comparison_path</span> <span class="o">/</span> <span class="s2">&quot;polr2a_results_with_targets.csv&quot;</span>
        <span class="n">rna_file</span> <span class="o">=</span> <span class="n">comparison_path</span> <span class="o">/</span> <span class="s2">&quot;rna_results_with_targets.csv&quot;</span>
        
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;polr2a_tss&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">polr2a_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;rnaseq_gene_body&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">rna_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved individual result files to </span><span class="si">{</span><span class="n">comparison_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Call comparison function (to be implemented in comparison module)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;To compare results, use the comparison module with the generated CSV files.&quot;</span><span class="p">)</span>
        
        <span class="k">return</span> <span class="n">polr2a_file</span><span class="p">,</span> <span class="n">rna_file</span>
    
    <span class="k">return</span> <span class="kc">None</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c1"># Define base path</span>
    <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define paths for analysis</span>
    <span class="n">gene_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations&quot;</span> <span class="o">/</span> <span class="s2">&quot;human&quot;</span> <span class="o">/</span> <span class="s2">&quot;Final_gene_annotations_human.tsv&quot;</span>
    <span class="n">polr2a_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbation_runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;gene_expression_only_chrom_K562_POLR2A_enhancer_centric&quot;</span>
    <span class="n">rna_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;perturbation_runs&quot;</span> <span class="o">/</span> <span class="s2">&quot;gene_expression_only_chrom_K562_RNA_enhancer_centric&quot;</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;tables&quot;</span>
    
    <span class="c1"># Define scoring methods</span>
    <span class="n">score_methods</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">,</span> <span class="s1">&#39;z_score&#39;</span><span class="p">,</span> <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span>
    
    <span class="c1"># Run both analyses with custom parameters</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">run_analysis</span><span class="p">(</span>
        <span class="n">base_path</span><span class="o">=</span><span class="n">base_path</span><span class="p">,</span>
        <span class="n">gene_path</span><span class="o">=</span><span class="n">gene_path</span><span class="p">,</span>
        <span class="n">polr2a_path</span><span class="o">=</span><span class="n">polr2a_path</span><span class="p">,</span>
        <span class="n">rna_path</span><span class="o">=</span><span class="n">rna_path</span><span class="p">,</span>
        <span class="n">output_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span>
        <span class="n">window_size</span><span class="o">=</span><span class="mi">200500</span><span class="p">,</span>
        <span class="n">resolution</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
        <span class="n">upstream_bins</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">downstream_bins</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
        <span class="n">N_BINS</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
        <span class="n">visualize_top_n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
        <span class="n">score_methods</span><span class="o">=</span><span class="n">score_methods</span>
    <span class="p">)</span>
    
    <span class="c1"># Check if comparison can be run</span>
    <span class="k">if</span> <span class="s1">&#39;polr2a_tss&#39;</span> <span class="ow">in</span> <span class="n">results</span> <span class="ow">and</span> <span class="s1">&#39;rnaseq_gene_body&#39;</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Both analyses completed successfully. To run comparison:&quot;</span><span class="p">)</span>
        
        <span class="c1"># Define comparison path and CRISPR benchmark file</span>
        <span class="n">crispr_file</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations/human/lifted_hg38_to_hg19/EPCrisprBenchmark_ensemble_data_GRCh38_hg19.txt&quot;</span>
        <span class="n">comparison_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;results_CRISPR&quot;</span> <span class="o">/</span> <span class="s2">&quot;Integration_comparison&quot;</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;CRISPR benchmark file: </span><span class="si">{</span><span class="n">crispr_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Comparison output path: </span><span class="si">{</span><span class="n">comparison_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Use the comparison module with the generated result files.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Analysis completed. Not all result types are available for comparison.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Loaded 19430 gene annotations

Processing POLR2A data...
Loading predictions from ../runs/perturbation_runs/gene_expression_only_chrom_K562_POLR2A_enhancer_centric...
Loaded predictions for 4240 samples
Found 2120 unique enhancers

Processing complete for polr2a_tss using abs_area_difference:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.857

Processing complete for polr2a_tss using z_score:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.857

Processing complete for polr2a_tss using ratio_to_max:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.857
Visualizing top 10 enhancers by effect size...
  Visualizing chr19_13262245_13262985...
  Visualizing chr6_27863221_27863721...
  Visualizing chr6_27156062_27157040...
  Visualizing chr6_27125069_27126302...
  Visualizing chr6_27446342_27447683...
  Visualizing chr3_130466803_130467303...
  Visualizing chr19_49378405_49379925...
  Visualizing chr12_54676805_54677305...
  Visualizing chr6_27462924_27463990...
  Visualizing chr19_13266245_13266765...
Visualizations saved to ../tables/polr2a_tss_visualizations

Processing RNA-seq data...
Loading predictions from ../runs/perturbation_runs/gene_expression_only_chrom_K562_RNA_enhancer_centric...
Loaded predictions for 4240 samples
Found 2120 unique enhancers

Processing complete for rnaseq_gene_body using abs_area_difference:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.826

Processing complete for rnaseq_gene_body using z_score:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.826

Processing complete for rnaseq_gene_body using ratio_to_max:
Processed 2120 enhancers
Found 13390 enhancer-gene pairs
Primary target identification accuracy: 0.826
Visualizing top 10 enhancers by effect size...
  Visualizing chr2_145339863_145340776...
  Visualizing chr8_101912570_101913033...
  Visualizing chr17_75066221_75067454...
  Visualizing chr8_61242560_61243567...
  Visualizing chr11_17374296_17375700...
  Visualizing chr1_51442000_51445530...
  Visualizing chr2_152616827_152617327...
  Visualizing chr6_53223073_53224863...
  Visualizing chr8_124178947_124180157...
  Visualizing chr20_45987402_45987902...
Visualizations saved to ../tables/rnaseq_gene_body_visualizations

Both analyses completed successfully. To run comparison:
CRISPR benchmark file: ../annotations/human/lifted_hg38_to_hg19/EPCrisprBenchmark_ensemble_data_GRCh38_hg19.txt
Comparison output path: ../results_CRISPR/Integration_comparison
Use the comparison module with the generated result files.
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/scratch/ipykernel_3900088/2164204773.py:217: DeprecationWarning: &#39;scipy.integrate.simps&#39; is deprecated in favour of &#39;scipy.integrate.simpson&#39; and will be removed in SciPy 1.14.0
  area = simps(region_values, x)
/scratch/ipykernel_3900088/2164204773.py:217: DeprecationWarning: &#39;scipy.integrate.simps&#39; is deprecated in favour of &#39;scipy.integrate.simpson&#39; and will be removed in SciPy 1.14.0
  area = simps(region_values, x)
</pre></div>
</div>
</div>
</div>
<p><strong>Visualize Enhancer-Gene pairing results:</strong></p>
<ul class="simple">
<li><p>Precision- Recall curves.</p></li>
<li><p>ROC curves.</p></li>
<li><p>Confusion matrices.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">confusion_matrix</span><span class="p">,</span> <span class="n">classification_report</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">precision_recall_curve</span><span class="p">,</span> <span class="n">roc_curve</span><span class="p">,</span> <span class="n">auc</span><span class="p">,</span> <span class="n">average_precision_score</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">pearsonr</span><span class="p">,</span> <span class="n">spearmanr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.gridspec</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">gridspec</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib_venn</span><span class="w"> </span><span class="kn">import</span> <span class="n">venn2</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: matplotlib_venn not installed. Venn diagrams will be skipped.&quot;</span><span class="p">)</span>

<span class="c1"># Set plotting style</span>
<span class="n">plt</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Nimbus Roman&#39;</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">42</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_binary_metrics</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate and save comprehensive classification metrics for binary predictions.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: &#39;Regulated&#39; column not found in merged data.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">classification_report</span><span class="p">,</span> <span class="n">precision_score</span><span class="p">,</span> <span class="n">recall_score</span><span class="p">,</span> <span class="n">f1_score</span><span class="p">,</span> <span class="n">accuracy_score</span>
    
    <span class="c1"># Define prediction methods to evaluate</span>
    <span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;polr2a_effect&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Effect-based Primary Target&#39;</span><span class="p">,</span>
            <span class="s1">&#39;predictions&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target_polr2a&#39;</span><span class="p">],</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#1f77b4&#39;</span>  <span class="c1"># blue</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_effect&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Effect-based Primary Target&#39;</span><span class="p">,</span>
            <span class="s1">&#39;predictions&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target_rna&#39;</span><span class="p">],</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#2ca02c&#39;</span>  <span class="c1"># green</span>
        <span class="p">},</span>
        <span class="s1">&#39;proximity_based&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Proximity-based Target&#39;</span><span class="p">,</span>
            <span class="s1">&#39;predictions&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;is_closest_polr2a&#39;</span><span class="p">],</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#d62728&#39;</span>  <span class="c1"># red</span>
        <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># True labels</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span>
    
    <span class="c1"># Calculate metrics for each method</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">method_id</span><span class="p">,</span> <span class="n">method_info</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">y_pred</span> <span class="o">=</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;predictions&#39;</span><span class="p">]</span>
        
        <span class="c1"># Calculate metrics</span>
        <span class="n">accuracy</span> <span class="o">=</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
        <span class="n">precision</span> <span class="o">=</span> <span class="n">precision_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
        <span class="n">recall</span> <span class="o">=</span> <span class="n">recall_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
        <span class="n">f1</span> <span class="o">=</span> <span class="n">f1_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
        
        <span class="c1"># Store results</span>
        <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
            <span class="s1">&#39;accuracy&#39;</span><span class="p">:</span> <span class="n">accuracy</span><span class="p">,</span>
            <span class="s1">&#39;precision&#39;</span><span class="p">:</span> <span class="n">precision</span><span class="p">,</span>
            <span class="s1">&#39;recall&#39;</span><span class="p">:</span> <span class="n">recall</span><span class="p">,</span>
            <span class="s1">&#39;f1&#39;</span><span class="p">:</span> <span class="n">f1</span><span class="p">,</span>
            <span class="s1">&#39;report&#39;</span><span class="p">:</span> <span class="n">classification_report</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">,</span> <span class="n">output_dict</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="p">}</span>
    
    <span class="c1"># Create metrics table</span>
    <span class="n">metrics_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s1">&#39;Method&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">],</span>
        <span class="s1">&#39;Accuracy&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;accuracy&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">],</span>
        <span class="s1">&#39;Precision&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;precision&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">],</span>
        <span class="s1">&#39;Recall&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;recall&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">],</span>
        <span class="s1">&#39;F1 Score&#39;</span><span class="p">:</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">m</span><span class="p">][</span><span class="s1">&#39;f1&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">]</span>
    <span class="p">})</span>
    
    <span class="c1"># Save metrics to CSV</span>
    <span class="n">metrics_table</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;binary_classification_metrics.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="c1"># Create visualization</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Plot metrics as bar chart</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Accuracy&#39;</span><span class="p">,</span> <span class="s1">&#39;Precision&#39;</span><span class="p">,</span> <span class="s1">&#39;Recall&#39;</span><span class="p">,</span> <span class="s1">&#39;F1 Score&#39;</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">metrics</span><span class="p">))</span>
    <span class="n">width</span> <span class="o">=</span> <span class="mf">0.25</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">method_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">methods</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
        <span class="n">method_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;accuracy&#39;</span><span class="p">],</span> 
                         <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;precision&#39;</span><span class="p">],</span> 
                         <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;recall&#39;</span><span class="p">],</span> 
                         <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;f1&#39;</span><span class="p">]]</span>
        
        <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">width</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">method_metrics</span><span class="p">,</span> <span class="n">width</span><span class="p">,</span> 
               <span class="n">label</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> 
               <span class="n">color</span><span class="o">=</span><span class="n">methods</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;color&#39;</span><span class="p">])</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Metric&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Classification Metrics Comparison&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;binary_classification_metrics.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Binary classification metrics:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">metrics_table</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span>

<span class="k">def</span><span class="w"> </span><span class="nf">calculate_threshold_metrics</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate optimal thresholds and metrics for continuous score methods.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: &#39;Regulated&#39; column not found in merged data.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">precision_recall_curve</span><span class="p">,</span> <span class="n">f1_score</span><span class="p">,</span> <span class="n">accuracy_score</span>
    
    <span class="c1"># Define methods to evaluate</span>
    <span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;polr2a_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#1f77b4&#39;</span>  <span class="c1"># blue</span>
        <span class="p">},</span>
        <span class="s1">&#39;polr2a_zscore&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Z-Score&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;z_score_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff7f0e&#39;</span>  <span class="c1"># orange</span>
        <span class="p">},</span>
        <span class="s1">&#39;polr2a_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#9467bd&#39;</span>  <span class="c1"># purple</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#2ca02c&#39;</span>  <span class="c1"># green</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_zscore&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Z-Score&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;z_score_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#d62728&#39;</span>  <span class="c1"># red</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#8c564b&#39;</span>  <span class="c1"># brown</span>
        <span class="p">},</span>
        <span class="s1">&#39;proximity&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Proximity-based&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="o">-</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#e377c2&#39;</span>  <span class="c1"># pink</span>
        <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># True labels</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    
    <span class="c1"># Calculate optimal thresholds and metrics</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">metrics_rows</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">method_id</span><span class="p">,</span> <span class="n">method_info</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">scores</span> <span class="o">=</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">]</span>
        
        <span class="c1"># Calculate precision-recall curve and find optimal threshold for F1</span>
        <span class="n">precision</span><span class="p">,</span> <span class="n">recall</span><span class="p">,</span> <span class="n">thresholds</span> <span class="o">=</span> <span class="n">precision_recall_curve</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">scores</span><span class="p">)</span>
        
        <span class="c1"># Calculate F1 score for each threshold</span>
        <span class="n">f1_scores</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
            <span class="n">y_pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
            <span class="n">f1</span> <span class="o">=</span> <span class="n">f1_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
            <span class="n">f1_scores</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">f1</span><span class="p">)</span>
        
        <span class="c1"># Find optimal threshold for F1 score</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">f1_scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">best_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">f1_scores</span><span class="p">)</span>
            <span class="n">best_threshold</span> <span class="o">=</span> <span class="n">thresholds</span><span class="p">[</span><span class="n">best_idx</span><span class="p">]</span>
            <span class="n">best_f1</span> <span class="o">=</span> <span class="n">f1_scores</span><span class="p">[</span><span class="n">best_idx</span><span class="p">]</span>
            
            <span class="c1"># Apply optimal threshold</span>
            <span class="n">y_pred</span> <span class="o">=</span> <span class="p">(</span><span class="n">scores</span> <span class="o">&gt;=</span> <span class="n">best_threshold</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
            
            <span class="c1"># Calculate metrics</span>
            <span class="n">accuracy</span> <span class="o">=</span> <span class="n">accuracy_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">y_pred</span><span class="p">)</span>
            <span class="n">precision_at_threshold</span> <span class="o">=</span> <span class="n">precision</span><span class="p">[</span><span class="n">best_idx</span><span class="p">]</span>
            <span class="n">recall_at_threshold</span> <span class="o">=</span> <span class="n">recall</span><span class="p">[</span><span class="n">best_idx</span><span class="p">]</span>
            
            <span class="c1"># Store results</span>
            <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;optimal_threshold&#39;</span><span class="p">:</span> <span class="n">best_threshold</span><span class="p">,</span>
                <span class="s1">&#39;accuracy&#39;</span><span class="p">:</span> <span class="n">accuracy</span><span class="p">,</span>
                <span class="s1">&#39;precision&#39;</span><span class="p">:</span> <span class="n">precision_at_threshold</span><span class="p">,</span>
                <span class="s1">&#39;recall&#39;</span><span class="p">:</span> <span class="n">recall_at_threshold</span><span class="p">,</span>
                <span class="s1">&#39;f1&#39;</span><span class="p">:</span> <span class="n">best_f1</span>
            <span class="p">}</span>
            
            <span class="c1"># Add to metrics table</span>
            <span class="n">metrics_rows</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s1">&#39;Method&#39;</span><span class="p">:</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
                <span class="s1">&#39;Optimal Threshold&#39;</span><span class="p">:</span> <span class="n">best_threshold</span><span class="p">,</span>
                <span class="s1">&#39;Accuracy&#39;</span><span class="p">:</span> <span class="n">accuracy</span><span class="p">,</span>
                <span class="s1">&#39;Precision&#39;</span><span class="p">:</span> <span class="n">precision_at_threshold</span><span class="p">,</span>
                <span class="s1">&#39;Recall&#39;</span><span class="p">:</span> <span class="n">recall_at_threshold</span><span class="p">,</span>
                <span class="s1">&#39;F1 Score&#39;</span><span class="p">:</span> <span class="n">best_f1</span>
            <span class="p">})</span>
    
    <span class="c1"># Create metrics table</span>
    <span class="n">metrics_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">metrics_rows</span><span class="p">)</span>
    
    <span class="c1"># Save metrics to CSV</span>
    <span class="n">metrics_table</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;continuous_metrics_at_optimal_threshold.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="c1"># Plot metrics</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">14</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Plot metrics as bar chart</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Accuracy&#39;</span><span class="p">,</span> <span class="s1">&#39;Precision&#39;</span><span class="p">,</span> <span class="s1">&#39;Recall&#39;</span><span class="p">,</span> <span class="s1">&#39;F1 Score&#39;</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">metrics</span><span class="p">))</span>
    <span class="n">width</span> <span class="o">=</span> <span class="mf">0.1</span>
    <span class="n">method_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">methods</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">method_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">method_ids</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">method_id</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">method_result</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="n">method_id</span><span class="p">]</span>
            <span class="n">method_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="n">method_result</span><span class="p">[</span><span class="s1">&#39;accuracy&#39;</span><span class="p">],</span> 
                             <span class="n">method_result</span><span class="p">[</span><span class="s1">&#39;precision&#39;</span><span class="p">],</span> 
                             <span class="n">method_result</span><span class="p">[</span><span class="s1">&#39;recall&#39;</span><span class="p">],</span> 
                             <span class="n">method_result</span><span class="p">[</span><span class="s1">&#39;f1&#39;</span><span class="p">]]</span>
            
            <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">width</span><span class="o">*</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">method_ids</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">),</span> <span class="n">method_metrics</span><span class="p">,</span> <span class="n">width</span><span class="p">,</span> 
                   <span class="n">label</span><span class="o">=</span><span class="n">method_result</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span> 
                   <span class="n">color</span><span class="o">=</span><span class="n">methods</span><span class="p">[</span><span class="n">method_id</span><span class="p">][</span><span class="s1">&#39;color&#39;</span><span class="p">])</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Metric&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Classification Metrics at Optimal Thresholds&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s1">&#39;best&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="s1">&#39;small&#39;</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;continuous_classification_metrics.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Continuous score metrics at optimal thresholds:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">metrics_table</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span>

<span class="k">def</span><span class="w"> </span><span class="nf">load_crispr_data</span><span class="p">(</span><span class="n">crispr_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load and preprocess CRISPR benchmark data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">crispr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">crispr_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">decimal</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
        
        <span class="n">float_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;EffectSize&#39;</span><span class="p">,</span> <span class="s1">&#39;pValueAdjusted&#39;</span><span class="p">,</span> <span class="s1">&#39;PowerAtEffectSize10&#39;</span><span class="p">,</span> <span class="s1">&#39;PowerAtEffectSize15&#39;</span><span class="p">,</span> 
                     <span class="s1">&#39;PowerAtEffectSize20&#39;</span><span class="p">,</span> <span class="s1">&#39;PowerAtEffectSize25&#39;</span><span class="p">,</span> <span class="s1">&#39;PowerAtEffectSize50&#39;</span><span class="p">]</span>
        
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">float_cols</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">crispr_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">crispr_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> 
                                                   <span class="nb">float</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;.&#39;</span><span class="p">))</span> 
                                                   <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">in</span> <span class="n">crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">==</span> <span class="kc">True</span> <span class="ow">or</span> <span class="n">x</span> <span class="o">==</span> <span class="s1">&#39;TRUE&#39;</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">))</span> <span class="k">else</span> <span class="nb">bool</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="p">)</span>
        
        <span class="n">crispr_df</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">crispr_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom(hg19)&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chromStart(hg19)&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chromEnd(hg19)&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
        <span class="p">)</span>
        
        <span class="k">if</span> <span class="s1">&#39;measuredGeneSymbol&#39;</span> <span class="ow">in</span> <span class="n">crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">crispr_df</span><span class="p">[</span><span class="s1">&#39;gene_name&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">crispr_df</span><span class="p">[</span><span class="s1">&#39;measuredGeneSymbol&#39;</span><span class="p">]</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded CRISPR data: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">crispr_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">crispr_df</span>
    
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error loading CRISPR data: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_unified_curves</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create unified ROC and PR curves with shared legend.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: &#39;Regulated&#39; column not found in merged data.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="c1"># Convert boolean Regulated to numeric (0/1)</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    
    <span class="c1"># For proximity, use negative distance as the score</span>
    <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;proximity_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer_polr2a&#39;</span><span class="p">]</span>
    
    <span class="c1"># Define methods for comparison</span>
    <span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;polr2a_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#1f77b4&#39;</span>  <span class="c1"># blue</span>
        <span class="p">},</span>
        <span class="c1">#&#39;polr2a_zscore&#39;: {</span>
        <span class="c1">#    &#39;name&#39;: &#39;POLR2A Z-Score&#39;,</span>
        <span class="c1">#    &#39;scores&#39;: merged_crispr_df[&#39;z_score_polr2a&#39;].values,</span>
        <span class="c1">#    &#39;color&#39;: &#39;#ff7f0e&#39;  # orange</span>
        <span class="c1">#},</span>
        <span class="s1">&#39;polr2a_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#9467bd&#39;</span>  <span class="c1"># purple</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#2ca02c&#39;</span>  <span class="c1"># green</span>
        <span class="p">},</span>
        <span class="c1">#&#39;rna_zscore&#39;: {</span>
        <span class="c1">#    &#39;name&#39;: &#39;RNA-seq Z-Score&#39;,</span>
        <span class="c1">#    &#39;scores&#39;: merged_crispr_df[&#39;z_score_rna&#39;].values,</span>
        <span class="c1">#    &#39;color&#39;: &#39;#d62728&#39;  # red</span>
        <span class="c1">#},</span>
        <span class="s1">&#39;rna_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff7f0e&#39;</span>  <span class="c1"># orange &#39;#8c564b&#39;  # brown</span>
        <span class="p">},</span>
        <span class="s1">&#39;proximity&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Proximity-based&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;proximity_score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#d62728&#39;</span>  <span class="c1"># red #&#39;#e377c2&#39;  # pink</span>
        <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># Create figure with GridSpec for custom layout</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">width_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    
    <span class="c1"># Store lines for legend</span>
    <span class="n">legend_lines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">legend_labels</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Create subplots</span>
    <span class="n">ax_roc</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">ax_pr</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sharey</span><span class="o">=</span><span class="n">ax_roc</span><span class="p">)</span>  <span class="c1"># Share y-axis</span>
    <span class="n">ax_legend</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

    <span class="c1"># Plot ROC curves</span>
    <span class="k">for</span> <span class="n">method_id</span><span class="p">,</span> <span class="n">method_info</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">roc_curve</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">])</span>
        <span class="n">roc_auc_value</span> <span class="o">=</span> <span class="n">auc</span><span class="p">(</span><span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">)</span>
        
        <span class="n">line</span><span class="p">,</span> <span class="o">=</span> <span class="n">ax_roc</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">])</span>
        <span class="n">legend_lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
        <span class="n">legend_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2"> (AUROC = </span><span class="si">{</span><span class="n">roc_auc_value</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># ROC plot settings</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;k--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;False Positive Rate&#39;</span><span class="p">)</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;True Positive Rate&#39;</span><span class="p">)</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;ROC Curves&#39;</span><span class="p">)</span>
    <span class="n">ax_roc</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span>

    <span class="c1"># Plot PR curves</span>
    <span class="k">for</span> <span class="n">method_id</span><span class="p">,</span> <span class="n">method_info</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">precision</span><span class="p">,</span> <span class="n">recall</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">precision_recall_curve</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">])</span>
        <span class="n">average_precision</span> <span class="o">=</span> <span class="n">average_precision_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">])</span>
        
        <span class="n">ax_pr</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">recall</span><span class="p">,</span> <span class="n">precision</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">])</span>
        <span class="c1"># Update labels with AUPRC</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">methods</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">method_id</span><span class="p">)</span>
        <span class="n">legend_labels</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;, AUPRC = </span><span class="si">{</span><span class="n">average_precision</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">)&quot;</span>

    <span class="c1"># PR plot settings</span>
    <span class="n">no_skill</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">y_true</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">y_true</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">no_skill</span><span class="p">)</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="n">no_skill</span><span class="p">,</span> <span class="n">no_skill</span><span class="p">],</span> <span class="s1">&#39;k--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Recall&#39;</span><span class="p">)</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Precision&#39;</span><span class="p">)</span>  <span class="c1"># No y-label since shared</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Precision-Recall Curves&#39;</span><span class="p">)</span>
    <span class="n">ax_pr</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span>

    <span class="c1"># Make legend subplot empty but add the legend</span>
    <span class="n">ax_legend</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">ax_legend</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">legend_lines</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">legend_labels</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;unified_curves.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">fig</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_performance_summary</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a summary plot comparing all methods&#39; performance.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: &#39;Regulated&#39; column not found in merged data.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Convert boolean Regulated to numeric (0/1)</span>
    <span class="n">y_true</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
    
    <span class="c1"># For proximity, use negative distance as the score (closer = higher score)</span>
    <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;proximity_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;distance_to_enhancer_polr2a&#39;</span><span class="p">]</span>
    
    <span class="c1"># Define all methods to compare - same as in plot_roc_curves</span>
    <span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;polr2a_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#1f77b4&#39;</span>  <span class="c1"># blue</span>
        <span class="p">},</span>
        <span class="s1">&#39;polr2a_zscore&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Z-Score&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;z_score_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff7f0e&#39;</span>  <span class="c1"># orange</span>
        <span class="p">},</span>
        <span class="s1">&#39;polr2a_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;POLR2A Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#9467bd&#39;</span>  <span class="c1"># purple</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_abs&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Abs Area Difference&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#2ca02c&#39;</span>  <span class="c1"># green</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_zscore&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Z-Score&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;z_score_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#d62728&#39;</span>  <span class="c1"># red</span>
        <span class="p">},</span>
        <span class="s1">&#39;rna_ratio&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;RNA-seq Ratio to Max&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;ratio_to_max_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#8c564b&#39;</span>  <span class="c1"># brown</span>
        <span class="p">},</span>
        <span class="s1">&#39;proximity&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;name&#39;</span><span class="p">:</span> <span class="s1">&#39;Proximity-based&#39;</span><span class="p">,</span>
            <span class="s1">&#39;scores&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;proximity_score&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;#e377c2&#39;</span>  <span class="c1"># pink</span>
        <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="c1"># Calculate performance metrics for each method</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">for</span> <span class="n">method_id</span><span class="p">,</span> <span class="n">method_info</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="c1"># Calculate AUC-ROC</span>
        <span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">roc_curve</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">])</span>
        <span class="n">roc_auc</span> <span class="o">=</span> <span class="n">auc</span><span class="p">(</span><span class="n">fpr</span><span class="p">,</span> <span class="n">tpr</span><span class="p">)</span>
        
        <span class="c1"># Calculate Average Precision</span>
        <span class="n">ap</span> <span class="o">=</span> <span class="n">average_precision_score</span><span class="p">(</span><span class="n">y_true</span><span class="p">,</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;scores&#39;</span><span class="p">])</span>
        
        <span class="c1"># Store results</span>
        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s1">&#39;Method&#39;</span><span class="p">:</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;name&#39;</span><span class="p">],</span>
            <span class="s1">&#39;AUC-ROC&#39;</span><span class="p">:</span> <span class="n">roc_auc</span><span class="p">,</span>
            <span class="s1">&#39;AP&#39;</span><span class="p">:</span> <span class="n">ap</span><span class="p">,</span>
            <span class="s1">&#39;Color&#39;</span><span class="p">:</span> <span class="n">method_info</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]</span>
        <span class="p">})</span>
    
    <span class="c1"># Create summary DataFrame</span>
    <span class="n">summary_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
    
    <span class="c1"># Create summary plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
    
    <span class="c1"># Plot as grouped bar chart</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AUC-ROC&#39;</span><span class="p">,</span> <span class="s1">&#39;AP&#39;</span><span class="p">]</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">summary_df</span><span class="p">))</span>
    <span class="n">width</span> <span class="o">=</span> <span class="mf">0.35</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">width</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;AUC-ROC&#39;</span><span class="p">],</span> <span class="n">width</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;AUC-ROC&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;navy&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="n">width</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span> <span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;AP&#39;</span><span class="p">],</span> <span class="n">width</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Average Precision&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;darkred&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Method&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Performance Comparison of Prediction Methods&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;Method&#39;</span><span class="p">],</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;performance_summary.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    
    <span class="c1"># Save summary to CSV</span>
    <span class="n">summary_df</span><span class="p">[[</span><span class="s1">&#39;Method&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC-ROC&#39;</span><span class="p">,</span> <span class="s1">&#39;AP&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;performance_summary.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">summary_df</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_primary_target_comparison</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create confusion matrix comparing methods and CRISPR data with precision/recall metrics.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: &#39;Regulated&#39; column not found in merged data.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="c1"># Ensure output directory exists</span>
    <span class="n">output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Create figure</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mf">4.3</span><span class="p">))</span>
    
    <span class="c1"># Calculate metrics for effect-based prediction</span>
    <span class="n">cm1</span> <span class="o">=</span> <span class="n">confusion_matrix</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">],</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target_polr2a&#39;</span><span class="p">])</span>
    <span class="n">tn1</span><span class="p">,</span> <span class="n">fp1</span><span class="p">,</span> <span class="n">fn1</span><span class="p">,</span> <span class="n">tp1</span> <span class="o">=</span> <span class="n">cm1</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="n">acc1</span> <span class="o">=</span> <span class="p">(</span><span class="n">tp1</span> <span class="o">+</span> <span class="n">tn1</span><span class="p">)</span> <span class="o">/</span> <span class="n">cm1</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">precision1</span> <span class="o">=</span> <span class="n">tp1</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp1</span> <span class="o">+</span> <span class="n">fp1</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">tp1</span> <span class="o">+</span> <span class="n">fp1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">recall1</span> <span class="o">=</span> <span class="n">tp1</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp1</span> <span class="o">+</span> <span class="n">fn1</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">tp1</span> <span class="o">+</span> <span class="n">fn1</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    
    <span class="c1"># Plot first confusion matrix</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">cm1</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span>
                <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Primary&#39;</span><span class="p">,</span> <span class="s1">&#39;Primary&#39;</span><span class="p">],</span>
                <span class="n">yticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Regulated&#39;</span><span class="p">,</span> <span class="s1">&#39;Regulated&#39;</span><span class="p">],</span>
                <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Effect-based Prediction&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;CRISPR Regulated&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Effect-based Primary Target vs CRISPR&#39;</span><span class="p">)</span>
    
    <span class="c1"># Calculate metrics for proximity-based prediction</span>
    <span class="n">cm2</span> <span class="o">=</span> <span class="n">confusion_matrix</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">],</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;is_closest_polr2a&#39;</span><span class="p">])</span>
    <span class="n">tn2</span><span class="p">,</span> <span class="n">fp2</span><span class="p">,</span> <span class="n">fn2</span><span class="p">,</span> <span class="n">tp2</span> <span class="o">=</span> <span class="n">cm2</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
    <span class="n">acc2</span> <span class="o">=</span> <span class="p">(</span><span class="n">tp2</span> <span class="o">+</span> <span class="n">tn2</span><span class="p">)</span> <span class="o">/</span> <span class="n">cm2</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">precision2</span> <span class="o">=</span> <span class="n">tp2</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp2</span> <span class="o">+</span> <span class="n">fp2</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">tp2</span> <span class="o">+</span> <span class="n">fp2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">recall2</span> <span class="o">=</span> <span class="n">tp2</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp2</span> <span class="o">+</span> <span class="n">fn2</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">tp2</span> <span class="o">+</span> <span class="n">fn2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">f1_score2</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">precision2</span> <span class="o">*</span> <span class="n">recall2</span> <span class="o">/</span> <span class="p">(</span><span class="n">precision2</span> <span class="o">+</span> <span class="n">recall2</span><span class="p">)</span> <span class="k">if</span> <span class="p">(</span><span class="n">precision2</span> <span class="o">+</span> <span class="n">recall2</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">0</span>
    
    <span class="c1"># Plot second confusion matrix</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">cm2</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fmt</span><span class="o">=</span><span class="s1">&#39;d&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;Blues&#39;</span><span class="p">,</span>
                <span class="n">xticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Closest&#39;</span><span class="p">,</span> <span class="s1">&#39;Closest&#39;</span><span class="p">],</span>
                <span class="n">yticklabels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Not Regulated&#39;</span><span class="p">,</span> <span class="s1">&#39;Regulated&#39;</span><span class="p">],</span>
                <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Proximity-based Prediction&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;CRISPR Regulated&#39;</span><span class="p">)</span>
    <span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Proximity-based Target vs CRISPR&#39;</span><span class="p">)</span>
    
    <span class="c1"># Add metrics below each matrix</span>
    <span class="n">y_pos</span> <span class="o">=</span> <span class="mf">0.02</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figtext</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span> <span class="n">y_pos</span><span class="o">+</span><span class="mf">0.06</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Accuracy: </span><span class="si">{</span><span class="n">acc1</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">, Precision: </span><span class="si">{</span><span class="n">precision1</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">, Recall: </span><span class="si">{</span><span class="n">recall1</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figtext</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">y_pos</span><span class="o">+</span><span class="mf">0.06</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Accuracy: </span><span class="si">{</span><span class="n">acc2</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">, Precision: </span><span class="si">{</span><span class="n">precision2</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">, Recall: </span><span class="si">{</span><span class="n">recall2</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>  <span class="c1"># Make space for metrics</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;crispr_prediction_comparison.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;effect_cm&#39;</span><span class="p">:</span> <span class="n">cm1</span><span class="p">,</span>
        <span class="s1">&#39;proximity_cm&#39;</span><span class="p">:</span> <span class="n">cm2</span><span class="p">,</span>
        <span class="s1">&#39;effect_metrics&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;accuracy&#39;</span><span class="p">:</span> <span class="n">acc1</span><span class="p">,</span>
            <span class="s1">&#39;precision&#39;</span><span class="p">:</span> <span class="n">precision1</span><span class="p">,</span>
            <span class="s1">&#39;recall&#39;</span><span class="p">:</span> <span class="n">recall1</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s1">&#39;proximity_metrics&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;accuracy&#39;</span><span class="p">:</span> <span class="n">acc2</span><span class="p">,</span>
            <span class="s1">&#39;precision&#39;</span><span class="p">:</span> <span class="n">precision2</span><span class="p">,</span>
            <span class="s1">&#39;recall&#39;</span><span class="p">:</span> <span class="n">recall2</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">plot_correlation_scatter</span><span class="p">(</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create scatter plot comparing POLR2A and RNA-seq scores with primary target highlighting.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
    
    <span class="c1"># Define colors based on primary target status</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;is_primary_target_polr2a&#39;</span><span class="p">],</span> <span class="s1">&#39;#ff7f0e&#39;</span><span class="p">,</span> <span class="s1">&#39;#1f77b4&#39;</span><span class="p">)</span>
    
    <span class="c1"># Create legend elements</span>
    <span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="kn">import</span> <span class="n">Patch</span>
    <span class="n">legend_elements</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;#ff7f0e&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Primary Target (POLR2A)&#39;</span><span class="p">),</span>
        <span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;#1f77b4&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Secondary Target&#39;</span><span class="p">)</span>
    <span class="p">]</span>
    
    <span class="c1"># Create scatter plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">],</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">],</span>
        <span class="n">c</span><span class="o">=</span><span class="n">colors</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
        <span class="n">s</span><span class="o">=</span><span class="mi">50</span>
    <span class="p">)</span>
    
    <span class="c1"># Add regression line</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">regplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">],</span>
        <span class="n">y</span><span class="o">=</span><span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">],</span>
        <span class="n">scatter</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span>
        <span class="n">line_kws</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;linestyle&#39;</span><span class="p">:</span> <span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="s1">&#39;alpha&#39;</span><span class="p">:</span> <span class="mf">0.7</span><span class="p">}</span>
    <span class="p">)</span>
    
    <span class="c1"># Calculate correlation statistics</span>
    <span class="n">pearson_r</span><span class="p">,</span> <span class="n">pearson_p</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">],</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">]</span>
    <span class="p">)</span>
    
    <span class="n">spearman_r</span><span class="p">,</span> <span class="n">spearman_p</span> <span class="o">=</span> <span class="n">spearmanr</span><span class="p">(</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_polr2a&#39;</span><span class="p">],</span>
        <span class="n">merged_df</span><span class="p">[</span><span class="s1">&#39;abs_area_difference_rna&#39;</span><span class="p">]</span>
    <span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;POLR2A Integration Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;RNA-seq Integration Score&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Score Correlation</span><span class="se">\n</span><span class="s1">Pearson r = </span><span class="si">{</span><span class="n">pearson_r</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">, Spearman ρ = </span><span class="si">{</span><span class="n">spearman_r</span><span class="si">:</span><span class="s1">.3f</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">legend_elements</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;score_correlation_with_targets.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="p">{</span>
        <span class="s1">&#39;pearson_r&#39;</span><span class="p">:</span> <span class="n">pearson_r</span><span class="p">,</span>
        <span class="s1">&#39;pearson_p&#39;</span><span class="p">:</span> <span class="n">pearson_p</span><span class="p">,</span>
        <span class="s1">&#39;spearman_r&#39;</span><span class="p">:</span> <span class="n">spearman_r</span><span class="p">,</span>
        <span class="s1">&#39;spearman_p&#39;</span><span class="p">:</span> <span class="n">spearman_p</span>
    <span class="p">}</span>


<span class="k">def</span><span class="w"> </span><span class="nf">compare_results</span><span class="p">(</span><span class="n">polr2a_path</span><span class="p">,</span> <span class="n">rna_path</span><span class="p">,</span> <span class="n">crispr_path</span><span class="p">,</span> <span class="n">output_path</span><span class="p">,</span> <span class="n">score_methods</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run comparison analysis between POLR2A and RNA-seq predictions with multiple scoring methods.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">        polr2a_path: Path to POLR2A results CSV with primary target flags</span>
<span class="sd">        rna_path: Path to RNA-seq results CSV with primary target flags</span>
<span class="sd">        crispr_path: Path to CRISPR benchmark data</span>
<span class="sd">        output_path: Path to save outputs</span>
<span class="sd">        score_methods: List of scoring methods to compare</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">        Dictionary with analysis results</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">score_methods</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">score_methods</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">,</span> <span class="s1">&#39;z_score&#39;</span><span class="p">,</span> <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span>
        
    <span class="c1"># Create output directory</span>
    <span class="n">output_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_path</span><span class="p">)</span>
    <span class="n">output_path</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Running Enhanced Enhancer-Gene Interaction Analysis ===&quot;</span><span class="p">)</span>
    
    <span class="c1"># Load POLR2A and RNA-seq results</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Loading prediction results...&quot;</span><span class="p">)</span>
    <span class="n">polr2a_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">polr2a_path</span><span class="p">)</span>
    <span class="n">rna_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">rna_path</span><span class="p">)</span>
    
    <span class="c1"># Merge datasets on enhancer-gene pairs</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Merging POLR2A and RNA-seq results...&quot;</span><span class="p">)</span>
    <span class="n">merged_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
        <span class="n">polr2a_df</span><span class="p">,</span> 
        <span class="n">rna_df</span><span class="p">,</span> 
        <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">],</span>
        <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_polr2a&#39;</span><span class="p">,</span> <span class="s1">&#39;_rna&#39;</span><span class="p">)</span>
    <span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">merged_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> common enhancer-gene pairs&quot;</span><span class="p">)</span>
    <span class="n">merged_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;merged_results_with_targets.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="c1"># Load CRISPR benchmark data if available</span>
    <span class="n">crispr_df</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">crispr_path</span> <span class="ow">and</span> <span class="n">Path</span><span class="p">(</span><span class="n">crispr_path</span><span class="p">)</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Loading CRISPR benchmark data...&quot;</span><span class="p">)</span>
        <span class="n">crispr_df</span> <span class="o">=</span> <span class="n">load_crispr_data</span><span class="p">(</span><span class="n">crispr_path</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">crispr_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Merge with CRISPR ground truth</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Merging with CRISPR benchmark data...&quot;</span><span class="p">)</span>
            <span class="n">merge_keys</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;enhancer_id&#39;</span><span class="p">,</span> <span class="s1">&#39;gene_name&#39;</span><span class="p">]</span>
            
            <span class="c1"># Select relevant columns from CRISPR data</span>
            <span class="n">crispr_cols</span> <span class="o">=</span> <span class="n">merge_keys</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">,</span> <span class="s1">&#39;EffectSize&#39;</span><span class="p">,</span> <span class="s1">&#39;Significant&#39;</span><span class="p">]</span>
            <span class="n">available_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">crispr_cols</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
            
            <span class="c1"># Perform merge</span>
            <span class="n">merged_crispr_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                <span class="n">merged_df</span><span class="p">,</span>
                <span class="n">crispr_df</span><span class="p">[</span><span class="n">available_cols</span><span class="p">],</span>
                <span class="n">on</span><span class="o">=</span><span class="n">merge_keys</span><span class="p">,</span>
                <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span>
            <span class="p">)</span>
            
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Merged results: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">)</span><span class="si">}</span><span class="s2"> enhancer-gene pairs with CRISPR data&quot;</span><span class="p">)</span>
            
            <span class="c1"># Report on regulated status</span>
            <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">regulated_count</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Regulated genes: </span><span class="si">{</span><span class="n">regulated_count</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">regulated_count</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.1f</span><span class="si">}</span><span class="s2">%)&quot;</span><span class="p">)</span>
                
            <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_path</span> <span class="o">/</span> <span class="s1">&#39;merged_with_crispr_targets.csv&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">merged_crispr_df</span> <span class="o">=</span> <span class="n">merged_df</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: Could not load CRISPR data. Proceeding with basic comparison.&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">merged_crispr_df</span> <span class="o">=</span> <span class="n">merged_df</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No CRISPR benchmark data provided. Proceeding with basic comparison.&quot;</span><span class="p">)</span>
    
    <span class="c1"># Generate plots</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generating plots...&quot;</span><span class="p">)</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">plt</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
    <span class="n">ax0</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_polr2a&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_polr2a&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span><span class="mi">1</span><span class="p">))</span>
    <span class="c1">#ax0.set_yscale(&#39;log&#39;)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">ax1</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_rna&#39;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_rna&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="n">threshold_polr2a</span> <span class="o">=</span> <span class="mi">10</span>
    <span class="n">merged_crispr_df</span> <span class="o">=</span> <span class="n">merged_crispr_df</span><span class="p">[(</span><span class="n">merged_crispr_df</span><span class="p">[</span><span class="s1">&#39;baseline_area_polr2a&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold_polr2a</span><span class="p">)]</span>
    
    <span class="c1"># 1. Plot correlation between methods</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Generating correlation scatter plot...&quot;</span><span class="p">)</span>
    <span class="n">corr_results</span> <span class="o">=</span> <span class="n">plot_correlation_scatter</span><span class="p">(</span><span class="n">merged_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
    
    <span class="c1"># 3. Generate CRISPR comparison if available</span>
    <span class="n">crispr_results</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">roc_results</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">pr_results</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">performance_summary</span> <span class="o">=</span> <span class="kc">None</span>
    
    <span class="k">if</span> <span class="s1">&#39;Regulated&#39;</span> <span class="ow">in</span> <span class="n">merged_crispr_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Generating CRISPR comparison plots...&quot;</span><span class="p">)</span>
        <span class="n">crispr_results</span> <span class="o">=</span> <span class="n">plot_primary_target_comparison</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
        
        <span class="c1"># ... other plotting functions</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Generating unified ROC and PR curves...&quot;</span><span class="p">)</span>
        <span class="n">unified_fig</span> <span class="o">=</span> <span class="n">plot_unified_curves</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Generating performance summary...&quot;</span><span class="p">)</span>
        <span class="n">performance_summary</span> <span class="o">=</span> <span class="n">plot_performance_summary</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Calculating binary classification metrics...&quot;</span><span class="p">)</span>
        <span class="n">binary_metrics</span> <span class="o">=</span> <span class="n">calculate_binary_metrics</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- Calculating continuous score metrics at optimal thresholds...&quot;</span><span class="p">)</span>
        <span class="n">threshold_metrics</span> <span class="o">=</span> <span class="n">calculate_threshold_metrics</span><span class="p">(</span><span class="n">merged_crispr_df</span><span class="p">,</span> <span class="n">output_path</span><span class="p">)</span>
    
    <span class="c1"># Compile results</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;merged&#39;</span><span class="p">:</span> <span class="n">merged_df</span><span class="p">,</span>
            <span class="s1">&#39;with_crispr&#39;</span><span class="p">:</span> <span class="n">merged_crispr_df</span> <span class="k">if</span> <span class="n">crispr_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="p">},</span>
        <span class="s1">&#39;metrics&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s1">&#39;correlation&#39;</span><span class="p">:</span> <span class="n">corr_results</span><span class="p">,</span>
            <span class="s1">&#39;crispr_comparison&#39;</span><span class="p">:</span> <span class="n">crispr_results</span><span class="p">,</span>
            <span class="s1">&#39;roc&#39;</span><span class="p">:</span> <span class="n">roc_results</span><span class="p">,</span>
            <span class="s1">&#39;pr&#39;</span><span class="p">:</span> <span class="n">pr_results</span><span class="p">,</span>
            <span class="s1">&#39;performance_summary&#39;</span><span class="p">:</span> <span class="n">performance_summary</span><span class="p">,</span>
            <span class="s1">&#39;binary_metrics&#39;</span><span class="p">:</span> <span class="n">binary_metrics</span><span class="p">,</span>
            <span class="s1">&#39;threshold_metrics&#39;</span><span class="p">:</span> <span class="n">threshold_metrics</span>
        <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== Analysis Complete ===&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;All results and visualizations saved to: </span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span>


<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Main function to run the analysis pipeline.&quot;&quot;&quot;</span>
    <span class="c1"># Define base path</span>
    <span class="n">base_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define paths for analysis and score methods</span>
    <span class="n">polr2a_file</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;tables&quot;</span> <span class="o">/</span> <span class="s2">&quot;polr2a_tss_abs_area_difference_results_with_targets.csv&quot;</span>
    <span class="n">rna_file</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;tables&quot;</span> <span class="o">/</span> <span class="s2">&quot;rnaseq_gene_body_abs_area_difference_results_with_targets.csv&quot;</span>
    <span class="n">crispr_file</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;annotations/human/lifted_hg38_to_hg19/EPCrisprBenchmark_ensemble_data_GRCh38_hg19.txt&quot;</span>
    <span class="n">comparison_path</span> <span class="o">=</span> <span class="n">base_path</span> <span class="o">/</span> <span class="s2">&quot;results_CRISPR&quot;</span> <span class="o">/</span> <span class="s2">&quot;Integration_comparison_targets&quot;</span>
    
    <span class="n">score_methods</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;abs_area_difference&#39;</span><span class="p">,</span> <span class="s1">&#39;z_score&#39;</span><span class="p">,</span> <span class="s1">&#39;ratio_to_max&#39;</span><span class="p">]</span>
    
    <span class="c1"># Run comparison</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">compare_results</span><span class="p">(</span>
        <span class="n">polr2a_path</span><span class="o">=</span><span class="n">polr2a_file</span><span class="p">,</span>
        <span class="n">rna_path</span><span class="o">=</span><span class="n">rna_file</span><span class="p">,</span>
        <span class="n">crispr_path</span><span class="o">=</span><span class="n">crispr_file</span><span class="p">,</span>
        <span class="n">output_path</span><span class="o">=</span><span class="n">comparison_path</span><span class="p">,</span>
        <span class="n">score_methods</span><span class="o">=</span><span class="n">score_methods</span>
    <span class="p">)</span>
    
    <span class="k">return</span> <span class="n">results</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning: matplotlib_venn not installed. Venn diagrams will be skipped.

=== Running Enhanced Enhancer-Gene Interaction Analysis ===
Loading prediction results...
Merging POLR2A and RNA-seq results...
Found 13390 common enhancer-gene pairs
Loading CRISPR benchmark data...
Loaded CRISPR data: 9445 enhancer-gene pairs
Merging with CRISPR benchmark data...
Merged results: 1757 enhancer-gene pairs with CRISPR data
Regulated genes: 288 (16.4%)
Generating plots...
- Generating correlation scatter plot...
- Generating CRISPR comparison plots...
- Generating unified ROC and PR curves...
0.14177693761814744
- Generating performance summary...
- Calculating binary classification metrics...
Binary classification metrics:
                                Method  Accuracy  Precision    Recall  \
0   POLR2A Effect-based Primary Target  0.718967   0.243619  0.466667   
1  RNA-seq Effect-based Primary Target  0.689981   0.233533  0.520000   
2               Proximity-based Target  0.846881   0.471154  0.653333   

   F1 Score  
0  0.320122  
1  0.322314  
2  0.547486  
- Calculating continuous score metrics at optimal thresholds...
Continuous score metrics at optimal thresholds:
                        Method  Optimal Threshold  Accuracy  Precision  \
0   POLR2A Abs Area Difference       5.224048e-08  0.144928   0.142225   
1               POLR2A Z-Score       7.071068e-01  0.664146   0.221014   
2          POLR2A Ratio to Max       8.530436e-01  0.706994   0.264706   
3  RNA-seq Abs Area Difference       6.924736e-05  0.330813   0.160032   
4              RNA-seq Z-Score       6.715896e-01  0.657845   0.216071   
5         RNA-seq Ratio to Max       7.821679e-01  0.707624   0.252588   
6              Proximity-based      -3.629500e+04  0.798362   0.382134   

     Recall  F1 Score  
0  1.000000  0.249032  
1  0.542222  0.314028  
2  0.600000  0.367347  
3  0.875556  0.270604  
4  0.537778  0.308280  
5  0.542222  0.344633  
6  0.684444  0.490446  

=== Analysis Complete ===
All results and visualizations saved to: ../results_CRISPR/Integration_comparison_targets
</pre></div>
</div>
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Figure size 640x480 with 0 Axes&gt;
</pre></div>
</div>
<img alt="../_images/ff5c9aa81c7c09e09408ca3d6fabc5e0c1b67383d27cfbc16cfd0d61d4fc4f11.png" src="../_images/ff5c9aa81c7c09e09408ca3d6fabc5e0c1b67383d27cfbc16cfd0d61d4fc4f11.png" />
<img alt="../_images/26ec944ad556e8fe5565dd63eddda04e57baa5a16e1b2e8b497c0d48328c3c17.png" src="../_images/26ec944ad556e8fe5565dd63eddda04e57baa5a16e1b2e8b497c0d48328c3c17.png" />
<img alt="../_images/7884921781abd428cd7afec5dd0c2b7a398042897f0ea155c9356bb527b74cc0.png" src="../_images/7884921781abd428cd7afec5dd0c2b7a398042897f0ea155c9356bb527b74cc0.png" />
<img alt="../_images/9f22f018064228185b1349b67a59bfe70765b02ad08cd921e0f7dfd0cf78e179.png" src="../_images/9f22f018064228185b1349b67a59bfe70765b02ad08cd921e0f7dfd0cf78e179.png" />
</div>
</div>
<p><strong>Plotting ground truth Enhancer-Gene pairs</strong></p>
<p>Here we will plot ground truth Enhancer-Gene pairs in K562 cells, obtained by CRISPR KO of enhancers and measuring the induced gene expression changes. This data was downloaded from the <a class="reference external" href="https://github.com/EngreitzLab/CRISPR_comparison/tree/main/resources/crispr_data">Engreitz lab’s github</a>, referenced as a benchmarking dataset in <a class="reference external" href="https://doi.org/10.1101/2023.11.09.563812">A. Gschwind et al.</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DOWNLOAD</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">DATA_DIR</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span> 
<span class="n">FIGURE_DIR</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;../figures/Figures_revisions/&quot;</span><span class="p">)</span>
<span class="n">FILENAME</span> <span class="o">=</span> <span class="s2">&quot;EPCrisprBenchmark_ensemble_data_GRCh38.tsv&quot;</span>
<span class="n">URL</span> <span class="o">=</span> <span class="s2">&quot;https://github.com/EngreitzLab/CRISPR_comparison/raw/refs/heads/main/resources/crispr_data/EPCrisprBenchmark_ensemble_data_GRCh38.tsv.gz&quot;</span>

<span class="c1"># Create data directory if it doesn&#39;t exist</span>
<span class="n">DATA_DIR</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">parents</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="k">if</span> <span class="n">DOWNLOAD</span><span class="p">:</span>
    <span class="c1"># Download and process file</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">urllib.request</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">gzip</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
    
    <span class="c1"># Download the gzipped file</span>
    <span class="n">gz_path</span> <span class="o">=</span> <span class="n">DATA_DIR</span> <span class="o">/</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">FILENAME</span><span class="si">}</span><span class="s2">.gz&quot;</span>
    <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">URL</span><span class="p">,</span> <span class="n">gz_path</span><span class="p">)</span>
    
    <span class="c1"># Unzip the file</span>
    <span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">gz_path</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_in</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">FILENAME</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copyfileobj</span><span class="p">(</span><span class="n">f_in</span><span class="p">,</span> <span class="n">f_out</span><span class="p">)</span>
    
    <span class="c1"># Remove the gzipped file</span>
    <span class="n">gz_path</span><span class="o">.</span><span class="n">unlink</span><span class="p">()</span>

<span class="c1"># Read the data into DataFrame</span>
<span class="n">df_path</span> <span class="o">=</span> <span class="n">DATA_DIR</span> <span class="o">/</span> <span class="n">FILENAME</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">df_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Regulated&#39;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span> <span class="c1"># Keep only true EP pairs, defined by Regulated=True</span>

<span class="c1"># Calculate distances</span>
<span class="n">enh_pos</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;chromStart&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;chromStart&#39;</span><span class="p">])</span> <span class="o">//</span> <span class="mi">2</span> <span class="c1">#Center of the regulatory element</span>
<span class="n">prom_pos</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;startTSS&#39;</span><span class="p">]</span> <span class="c1"># Only one basepair difference between start and end TSS</span>
<span class="n">x</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">enh_pos</span><span class="o">.</span><span class="n">values</span> <span class="o">-</span> <span class="n">prom_pos</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">scale</span> <span class="o">=</span> <span class="mi">1000</span> <span class="c1">#kbp</span>
<span class="n">median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="n">n_EG_pairs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>

<span class="c1">#Plot histogram:</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="mi">3</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">x</span><span class="o">/</span><span class="n">scale</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span><span class="mi">5</span><span class="p">),</span> <span class="n">density</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;# True EG pairs: </span><span class="si">{</span><span class="n">n_EG_pairs</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">median</span><span class="o">/</span><span class="n">scale</span><span class="o">*</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.04</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">ls</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Median: </span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">median</span><span class="p">)</span><span class="si">}</span><span class="s1"> bp&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Distance to TSS (kbp)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.04</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">200</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">FIGURE_DIR</span> <span class="o">/</span> <span class="s1">&#39;Ground_truth_EP_histogram.png&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/24730ce962f9f3dbebe4176fcc4ba59f63e908290ab3e1ae381fb2044ce432e2.png" src="../_images/24730ce962f9f3dbebe4176fcc4ba59f63e908290ab3e1ae381fb2044ce432e2.png" />
</div>
</div>
<p><strong>Get genomic region boundaries for fast UCSC query:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">chrom</span> <span class="o">=</span> <span class="mi">17</span>
<span class="n">start</span> <span class="o">=</span> <span class="mi">75082798</span>
<span class="n">SHIFT</span> <span class="o">=</span> <span class="mi">200050</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;chr</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">start</span><span class="o">-</span><span class="n">SHIFT</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">start</span><span class="o">+</span><span class="n">SHIFT</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>chr17:74882748-75282848
</pre></div>
</div>
</div>
</div>
<p><strong>Check which perturbations were included:</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="c1"># Load data</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/human/lifted_hg38_to_hg19/EPCrisprBenchmark_ensemble_data_GRCh38_hg19.txt&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">included</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/K562/Enhancer_centered_targets_human.csv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>

<span class="c1"># Function to parse IDs in the format chr_start_end or chr_start_end_perturbed</span>
<span class="k">def</span><span class="w"> </span><span class="nf">parse_id</span><span class="p">(</span><span class="n">id_str</span><span class="p">):</span>
    <span class="n">parts</span> <span class="o">=</span> <span class="n">id_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
    
    <span class="c1"># Check if it has the perturbed suffix</span>
    <span class="n">is_perturbed</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;perturbed&#39;</span>
    
    <span class="c1"># Extract chr, start, end</span>
    <span class="k">if</span> <span class="n">is_perturbed</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">4</span><span class="p">:</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="k">return</span> <span class="p">{</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">end</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">is_perturbed</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">parts</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">3</span><span class="p">:</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">parts</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
            <span class="k">return</span> <span class="p">{</span><span class="s1">&#39;chrom&#39;</span><span class="p">:</span> <span class="n">chrom</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="n">end</span><span class="p">}</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

<span class="c1"># Create a set of (chrom, start, end) tuples from included dataframe</span>
<span class="n">included_regions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">included</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">parsed</span> <span class="o">=</span> <span class="n">parse_id</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">parsed</span><span class="p">:</span>
        <span class="n">included_regions</span><span class="o">.</span><span class="n">add</span><span class="p">((</span><span class="n">parsed</span><span class="p">[</span><span class="s1">&#39;chrom&#39;</span><span class="p">],</span> <span class="n">parsed</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">],</span> <span class="n">parsed</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]))</span>

<span class="c1"># Add &quot;included&quot; column to df based on whether the hg19 coordinates appear in included_regions</span>
<span class="n">df</span><span class="p">[</span><span class="s1">&#39;included&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="s2">&quot;yes&quot;</span> <span class="k">if</span> <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;chrom(hg19)&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;chromStart(hg19)&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;chromEnd(hg19)&#39;</span><span class="p">])</span> <span class="ow">in</span> <span class="n">included_regions</span> <span class="k">else</span> <span class="s2">&quot;no&quot;</span><span class="p">,</span> 
    <span class="n">axis</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>

<span class="c1"># Save the updated dataframe</span>
<span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;df_with_included_flag.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Print summary</span>
<span class="n">included_count</span> <span class="o">=</span> <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;included&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;yes&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total regions in df: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Regions also in included dataframe: </span><span class="si">{</span><span class="n">included_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Percentage of df regions in included dataframe: </span><span class="si">{</span><span class="n">included_count</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total regions in df: 9445
Regions also in included dataframe: 5590
Percentage of df regions in included dataframe: 59.18%
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Load the enhancers data (update the path to your enhancers file)</span>
<span class="n">enhancers</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../annotations/CLASTER Supplementary Tables Revisited - ST3. Enhancer_Like_Signatures.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Load the targets data from paste.txt</span>
<span class="n">targets</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../targets/Enhancer_centered_targets.csv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>

<span class="c1"># Extract the target IDs from the first column</span>
<span class="n">target_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">targets</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">])</span>

<span class="c1"># Add &#39;Included&#39; column to enhancers dataframe</span>
<span class="n">enhancers</span><span class="p">[</span><span class="s1">&#39;Included&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enhancers</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;yes&quot;</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">target_ids</span> <span class="k">else</span> <span class="s2">&quot;no&quot;</span><span class="p">)</span>

<span class="c1"># Save the result</span>
<span class="n">enhancers</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;enhancers_with_included.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># Print summary</span>
<span class="n">included_count</span> <span class="o">=</span> <span class="p">(</span><span class="n">enhancers</span><span class="p">[</span><span class="s1">&#39;Included&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;yes&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total enhancers: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancers</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Enhancers included in targets file: </span><span class="si">{</span><span class="n">included_count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Percentage of enhancers in targets: </span><span class="si">{</span><span class="n">included_count</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">enhancers</span><span class="p">)</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total enhancers: 6498
Enhancers included in targets file: 1251
Percentage of enhancers in targets: 19.25%
</pre></div>
</div>
</div>
</div>
</section>
<section id="editing-gene-annotations-table-to-add-whether-they-were-included-or-not-in-the-analyses">
<h3>Editing gene annotations table to add whether they were included or not in the analyses<a class="headerlink" href="#editing-gene-annotations-table-to-add-whether-they-were-included-or-not-in-the-analyses" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">csv</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">re</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">argparse</span>

<span class="k">def</span><span class="w"> </span><span class="nf">process_annotation_file</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="s2">&quot;mouse&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Updates the gene annotations TSV file by adding columns that indicate:</span>
<span class="sd">    1. If the gene is in training, test, or missing for targets</span>
<span class="sd">    2. For each input folder type, whether the gene is in training, test, both, or missing</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    species (str): Either &quot;mouse&quot; or &quot;human&quot; to determine which files to process</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;mouse&quot;</span><span class="p">:</span>
        <span class="c1"># Paths for mouse files</span>
        <span class="n">annotation_file</span> <span class="o">=</span> <span class="s2">&quot;../annotations/Final_gene_annotations.tsv&quot;</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;../annotations/Final_gene_annotations_updated.tsv&quot;</span>
        
        <span class="c1"># Target files for mouse</span>
        <span class="n">target_paths</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;default&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/training_targets.csv&quot;</span><span class="p">,</span>
                <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/test_targets.csv&quot;</span>
            <span class="p">}</span>
        <span class="p">}</span>
        
        <span class="c1"># Input folders for mouse</span>
        <span class="n">input_types</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;landscape_arrays&quot;</span><span class="p">,</span>
            <span class="s2">&quot;microC&quot;</span><span class="p">,</span>
            <span class="s2">&quot;microC_rotated&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Promoter-CHiC&quot;</span>
        <span class="p">]</span>
        
    <span class="k">elif</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;human&quot;</span><span class="p">:</span>
        <span class="c1"># Paths for human files</span>
        <span class="n">annotation_file</span> <span class="o">=</span> <span class="s2">&quot;../annotations/human/Final_gene_annotations_human.tsv&quot;</span>
        <span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;../annotations/human/Final_gene_annotations_human_updated.tsv&quot;</span>
        
        <span class="c1"># Target files for human</span>
        <span class="n">target_paths</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;K562&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/K562/training_targets.csv&quot;</span><span class="p">,</span>
                <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/K562/test_targets.csv&quot;</span>
            <span class="p">},</span>
            <span class="s2">&quot;K562_polii&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/K562/polii/training_polii_targets.csv&quot;</span><span class="p">,</span>
                <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="s2">&quot;../targets/K562/polii/test_polii_targets.csv&quot;</span>
            <span class="p">}</span>
        <span class="p">}</span>
        
        <span class="c1"># Check for polii target files - FIX: Try multiple potential locations</span>
        <span class="n">polii_paths</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;../targets/K562/polii/training_polii_targets.csv&quot;</span><span class="p">,</span>
                <span class="s2">&quot;../targets/K562/polii/training_targets.csv&quot;</span>
            <span class="p">],</span>
            <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;../targets/K562/polii/test_polii_targets.csv&quot;</span><span class="p">,</span>
                <span class="s2">&quot;../targets/K562/polii/test_targets.csv&quot;</span>
            <span class="p">]</span>
        <span class="p">}</span>
        
        <span class="c1"># Check all potential paths for polii files</span>
        <span class="k">for</span> <span class="n">split</span><span class="p">,</span> <span class="n">paths</span> <span class="ow">in</span> <span class="n">polii_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">found</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">path</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
                    <span class="n">target_paths</span><span class="p">[</span><span class="s2">&quot;K562_polii&quot;</span><span class="p">][</span><span class="n">split</span><span class="p">]</span> <span class="o">=</span> <span class="n">path</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found polii </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> targets at: </span><span class="si">{</span><span class="n">path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="n">found</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">found</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;WARNING: No polii </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> targets found. Tried paths: </span><span class="si">{</span><span class="n">paths</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1"># Input folders for human</span>
        <span class="n">input_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;K562&quot;</span><span class="p">]</span>
        
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported species: </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">. Choose &#39;mouse&#39; or &#39;human&#39;.&quot;</span><span class="p">)</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">--- Processing </span><span class="si">{</span><span class="n">species</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2"> gene annotations ---</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Read the gene annotations file</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">annotation_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loaded </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="si">}</span><span class="s2"> gene annotations for </span><span class="si">{</span><span class="n">species</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error: Annotation file </span><span class="si">{</span><span class="n">annotation_file</span><span class="si">}</span><span class="s2"> not found.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="n">species</span><span class="p">,</span> <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;Annotation file not found: </span><span class="si">{</span><span class="n">annotation_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">}</span>
    
    <span class="c1"># Initialize dictionaries to store all target types</span>
    <span class="n">all_targets</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">target_type</span><span class="p">,</span> <span class="n">paths</span> <span class="ow">in</span> <span class="n">target_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
            <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">()</span>
        <span class="p">}</span>
    
    <span class="c1"># Initialize dictionaries to store array file information by input type</span>
    <span class="n">array_data</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">input_type</span> <span class="ow">in</span> <span class="n">input_types</span><span class="p">:</span>
        <span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;training&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
            <span class="s2">&quot;test&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">()</span>
        <span class="p">}</span>
    
    <span class="c1"># Extract IDs from all target files</span>
    <span class="n">missing_target_files</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">target_type</span><span class="p">,</span> <span class="n">paths</span> <span class="ow">in</span> <span class="n">target_paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">split</span><span class="p">,</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">paths</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing target file: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                    <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                    <span class="n">header</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>  <span class="c1"># Skip header</span>
                    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">row</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># Make sure the row is not empty</span>
                            <span class="c1"># For polii files, the ID might be in the second column if there&#39;s a dummy index</span>
                            <span class="k">if</span> <span class="n">target_type</span> <span class="o">==</span> <span class="s2">&quot;K562_polii&quot;</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                                <span class="n">full_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">full_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                            <span class="c1"># Remove only the _forward or _rev suffix</span>
                            <span class="k">if</span> <span class="s2">&quot;_forward&quot;</span> <span class="ow">in</span> <span class="n">full_id</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">full_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_forward&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="k">elif</span> <span class="s2">&quot;_rev&quot;</span> <span class="ow">in</span> <span class="n">full_id</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">full_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_rev&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">full_id</span>
                            <span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="n">split</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">base_id</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="n">split</span><span class="p">])</span><span class="si">}</span><span class="s2"> unique IDs in </span><span class="si">{</span><span class="n">target_type</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2"> targets&quot;</span><span class="p">)</span>
                <span class="c1"># Print sample IDs for verification</span>
                <span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="n">split</span><span class="p">])[:</span><span class="mi">3</span><span class="p">]</span> <span class="k">if</span> <span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="n">split</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                <span class="k">if</span> <span class="n">sample_ids</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Sample IDs: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">missing_target_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Target file </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2"> not found&quot;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">missing_target_files</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">WARNING: The following target files could not be found:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">missing_target_files</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  - </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Scan each input folder for array files - with better path checking and debugging</span>
    <span class="k">for</span> <span class="n">input_type</span> <span class="ow">in</span> <span class="n">input_types</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]:</span>
            <span class="c1"># Construct folder path based on species</span>
            <span class="k">if</span> <span class="n">species</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;mouse&quot;</span><span class="p">:</span>
                <span class="c1"># Try multiple possible folder structures for mouse</span>
                <span class="n">possible_paths</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="sa">f</span><span class="s2">&quot;../inputs/</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>  <span class="c1"># New structure</span>
                    <span class="sa">f</span><span class="s2">&quot;../</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span>          <span class="c1"># Original structure </span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>  <span class="c1"># human</span>
                <span class="c1"># Try multiple possible folder structures for human K562</span>
                <span class="k">if</span> <span class="n">input_type</span> <span class="o">==</span> <span class="s2">&quot;K562&quot;</span><span class="p">:</span>
                    <span class="n">possible_paths</span> <span class="o">=</span> <span class="p">[</span>
                        <span class="sa">f</span><span class="s2">&quot;../inputs/landscape_arrays/K562/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>      <span class="c1"># Direct K562 folder</span>
                        <span class="sa">f</span><span class="s2">&quot;../inputs/landscape_arrays/</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>  <span class="c1"># Standard path</span>
                        <span class="sa">f</span><span class="s2">&quot;../inputs/</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>  <span class="c1"># Alternative path</span>
                        <span class="sa">f</span><span class="s2">&quot;../inputs/</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">_landscape/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span>  <span class="c1"># Another alternative</span>
                    <span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">possible_paths</span> <span class="o">=</span> <span class="p">[</span>
                        <span class="sa">f</span><span class="s2">&quot;../inputs/K562/</span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span>  <span class="c1"># Nested under K562</span>
                    <span class="p">]</span>
            
            <span class="c1"># Try each possible path</span>
            <span class="n">found_path</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">folder_path</span> <span class="ow">in</span> <span class="n">possible_paths</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">folder_path</span><span class="p">):</span>
                    <span class="n">found_path</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found directory: </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    
                    <span class="c1"># Look for .npy files</span>
                    <span class="n">array_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">/*.npy&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">array_files</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">file_path</span> <span class="ow">in</span> <span class="n">array_files</span><span class="p">:</span>
                            <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
                            <span class="c1"># Extract gene ID from file name (remove .npy extension)</span>
                            <span class="n">file_id</span> <span class="o">=</span> <span class="n">file_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.npy&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="c1"># Remove orientation suffix</span>
                            <span class="k">if</span> <span class="s2">&quot;_forward&quot;</span> <span class="ow">in</span> <span class="n">file_id</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">file_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_forward&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="k">elif</span> <span class="s2">&quot;_rev&quot;</span> <span class="ow">in</span> <span class="n">file_id</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">file_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_rev&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">base_id</span> <span class="o">=</span> <span class="n">file_id</span>
                            <span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">][</span><span class="n">split</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">base_id</span><span class="p">)</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">array_files</span><span class="p">)</span><span class="si">}</span><span class="s2"> array files, extracted </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">][</span><span class="n">split</span><span class="p">])</span><span class="si">}</span><span class="s2"> unique IDs in </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="c1"># Print a few sample file names for debugging</span>
                        <span class="k">if</span> <span class="n">array_files</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Sample files: </span><span class="si">{</span><span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">f</span><span class="p">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">array_files</span><span class="p">[:</span><span class="mi">3</span><span class="p">]]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: No .npy files found in </span><span class="si">{</span><span class="n">folder_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                    
                    <span class="c1"># Break after finding a valid path</span>
                    <span class="k">break</span>
            
            <span class="k">if</span> <span class="ow">not</span> <span class="n">found_path</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Warning: Could not find any valid path for </span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">split</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Tried the following paths: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">possible_paths</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Add columns for each target type</span>
    <span class="k">for</span> <span class="n">target_type</span> <span class="ow">in</span> <span class="n">target_paths</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Included_</span><span class="si">{</span><span class="n">target_type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">target_type</span> <span class="o">!=</span> <span class="s2">&quot;default&quot;</span> <span class="k">else</span> <span class="s2">&quot;Included&quot;</span>
        <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Missing&quot;</span>
        
        <span class="c1"># Set the target status for each gene</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span>
            <span class="n">training_present</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="s2">&quot;training&quot;</span><span class="p">]</span>
            <span class="n">test_present</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">all_targets</span><span class="p">[</span><span class="n">target_type</span><span class="p">][</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>
            
            <span class="k">if</span> <span class="n">training_present</span> <span class="ow">and</span> <span class="n">test_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Both&quot;</span>
            <span class="k">elif</span> <span class="n">training_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Training&quot;</span>
            <span class="k">elif</span> <span class="n">test_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Test&quot;</span>
    
    <span class="c1"># Add columns for each input type</span>
    <span class="k">for</span> <span class="n">input_type</span> <span class="ow">in</span> <span class="n">input_types</span><span class="p">:</span>
        <span class="c1"># Create a clean column name from the input type</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="n">input_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="c1"># Initialize the column with &quot;Missing&quot;</span>
        <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Missing&quot;</span>
        
        <span class="c1"># Check if we found any arrays for this input type</span>
        <span class="n">any_arrays_found</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">split</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;training&quot;</span><span class="p">,</span> <span class="s2">&quot;test&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">][</span><span class="n">split</span><span class="p">]:</span>
                <span class="n">any_arrays_found</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
                
        <span class="k">if</span> <span class="ow">not</span> <span class="n">any_arrays_found</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;WARNING: No arrays found for input type </span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">. All genes will be marked as &#39;Missing&#39;.&quot;</span><span class="p">)</span>
        
        <span class="c1"># Update values for genes that have arrays in this input type</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">annotations</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span>
            <span class="n">training_present</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">][</span><span class="s2">&quot;training&quot;</span><span class="p">]</span>
            <span class="n">test_present</span> <span class="o">=</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">array_data</span><span class="p">[</span><span class="n">input_type</span><span class="p">][</span><span class="s2">&quot;test&quot;</span><span class="p">]</span>
            
            <span class="k">if</span> <span class="n">training_present</span> <span class="ow">and</span> <span class="n">test_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Both&quot;</span>
            <span class="k">elif</span> <span class="n">training_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Training&quot;</span>
            <span class="k">elif</span> <span class="n">test_present</span><span class="p">:</span>
                <span class="n">annotations</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">idx</span><span class="p">,</span> <span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Test&quot;</span>
    
    <span class="c1"># Count the distribution for each column</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Distribution of genes in target categories:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">target_type</span> <span class="ow">in</span> <span class="n">target_paths</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Included_</span><span class="si">{</span><span class="n">target_type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">target_type</span> <span class="o">!=</span> <span class="s2">&quot;default&quot;</span> <span class="k">else</span> <span class="s2">&quot;Included&quot;</span>
        <span class="n">dist</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">target_type</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">target_type</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;default&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;Default targets&#39;</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">dist</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;    </span><span class="si">{</span><span class="n">category</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Count the distribution for each input type</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Distribution of genes in input arrays:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">input_type</span> <span class="ow">in</span> <span class="n">input_types</span><span class="p">:</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="n">input_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">dist</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  </span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">dist</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;    </span><span class="si">{</span><span class="n">category</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Create output directory if it doesn&#39;t exist</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_file</span><span class="p">),</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    
    <span class="c1"># Save the updated annotations file</span>
    <span class="n">annotations</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Updated annotations saved to </span><span class="si">{</span><span class="n">output_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># Prepare summary statistics</span>
    <span class="n">summary</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="n">species</span><span class="p">,</span>
        <span class="s2">&quot;total_genes&quot;</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotations</span><span class="p">),</span>
        <span class="s2">&quot;target_distributions&quot;</span><span class="p">:</span> <span class="p">{},</span>
        <span class="s2">&quot;array_distributions&quot;</span><span class="p">:</span> <span class="p">{}</span>
    <span class="p">}</span>
    
    <span class="c1"># Add target distributions to summary</span>
    <span class="k">for</span> <span class="n">target_type</span> <span class="ow">in</span> <span class="n">target_paths</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Included_</span><span class="si">{</span><span class="n">target_type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">target_type</span> <span class="o">!=</span> <span class="s2">&quot;default&quot;</span> <span class="k">else</span> <span class="s2">&quot;Included&quot;</span>
        <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;target_distributions&quot;</span><span class="p">][</span><span class="n">target_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    
    <span class="c1"># Add array distributions to summary</span>
    <span class="k">for</span> <span class="n">input_type</span> <span class="ow">in</span> <span class="n">input_types</span><span class="p">:</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="n">input_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">summary</span><span class="p">[</span><span class="s2">&quot;array_distributions&quot;</span><span class="p">][</span><span class="n">input_type</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotations</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    
    <span class="k">return</span> <span class="n">summary</span>

<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
    <span class="c1"># Check if running in Jupyter</span>
    <span class="n">is_jupyter</span> <span class="o">=</span> <span class="s1">&#39;ipykernel_launcher&#39;</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">or</span> <span class="s1">&#39;ipykernel&#39;</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="n">is_jupyter</span><span class="p">:</span>
        <span class="c1"># Default to processing both species when run in Jupyter</span>
        <span class="n">species_to_process</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># When run as standalone script, use argparse</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Update gene annotations with target and input information&#39;</span><span class="p">)</span>
        <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--species&#39;</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span> <span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;both&#39;</span><span class="p">],</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;both&#39;</span><span class="p">,</span>
                            <span class="n">help</span><span class="o">=</span><span class="s1">&#39;Species to process (mouse, human, or both)&#39;</span><span class="p">)</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
        
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">species</span> <span class="o">==</span> <span class="s1">&#39;both&#39;</span><span class="p">:</span>
            <span class="n">species_to_process</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">species_to_process</span> <span class="o">=</span> <span class="p">[</span><span class="n">args</span><span class="o">.</span><span class="n">species</span><span class="p">]</span>
    
    <span class="c1"># Process requested species</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">species</span> <span class="ow">in</span> <span class="n">species_to_process</span><span class="p">:</span>
        <span class="n">results</span><span class="p">[</span><span class="n">species</span><span class="p">]</span> <span class="o">=</span> <span class="n">process_annotation_file</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="n">species</span><span class="p">)</span>
    
    <span class="c1"># Print overall summary</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">=== OVERALL SUMMARY ===&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">species</span><span class="p">,</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="si">{</span><span class="n">species</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="si">}</span><span class="s2"> ANNOTATIONS:&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;error&quot;</span> <span class="ow">in</span> <span class="n">result</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Error: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;error&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
            
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;  Total genes: </span><span class="si">{</span><span class="n">result</span><span class="p">[</span><span class="s1">&#39;total_genes&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  Target distributions:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">target_type</span><span class="p">,</span> <span class="n">dist</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;target_distributions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">target_name</span> <span class="o">=</span> <span class="n">target_type</span> <span class="k">if</span> <span class="n">target_type</span> <span class="o">!=</span> <span class="s2">&quot;default&quot;</span> <span class="k">else</span> <span class="s2">&quot;Default&quot;</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;    </span><span class="si">{</span><span class="n">target_name</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">dist</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;      </span><span class="si">{</span><span class="n">category</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;  Array distributions:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">input_type</span><span class="p">,</span> <span class="n">dist</span> <span class="ow">in</span> <span class="n">result</span><span class="p">[</span><span class="s1">&#39;array_distributions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;    </span><span class="si">{</span><span class="n">input_type</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">category</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">dist</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;      </span><span class="si">{</span><span class="n">category</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">count</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># This function can be called directly from jupyter</span>
<span class="k">def</span><span class="w"> </span><span class="nf">run_for_species</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="s2">&quot;both&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Wrapper function to run the annotation processing for specific species.</span>
<span class="sd">    Can be called directly from Jupyter notebook.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    species (str): &#39;mouse&#39;, &#39;human&#39;, or &#39;both&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">species</span> <span class="o">==</span> <span class="s1">&#39;both&#39;</span><span class="p">:</span>
        <span class="n">species_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;mouse&#39;</span><span class="p">,</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">species_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">species</span><span class="p">]</span>
        
    <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">sp</span> <span class="ow">in</span> <span class="n">species_list</span><span class="p">:</span>
        <span class="n">results</span><span class="p">[</span><span class="n">sp</span><span class="p">]</span> <span class="o">=</span> <span class="n">process_annotation_file</span><span class="p">(</span><span class="n">species</span><span class="o">=</span><span class="n">sp</span><span class="p">)</span>
        
    <span class="k">return</span> <span class="n">results</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>--- Processing MOUSE gene annotations ---

Loaded 21846 gene annotations for mouse
Processing target file: ../targets/training_targets.csv
Found 20451 unique IDs in default training targets
  Sample IDs: ENSMUSG00000056296.17, ENSMUSG00000075370.12, ENSMUSG00000050240.16
Processing target file: ../targets/test_targets.csv
Found 1295 unique IDs in default test targets
  Sample IDs: ENSMUSG00000028546.17, ENSMUSG00000028971.4, ENSMUSG00000028397.13
Found directory: ../inputs/landscape_arrays/training
Found 40902 array files, extracted 20451 unique IDs in ../inputs/landscape_arrays/training
  Sample files: [&#39;ENSMUSG00000024349.10_rev.npy&#39;, &#39;ENSMUSG00000024539.17_rev.npy&#39;, &#39;ENSMUSG00000025464.15_forward.npy&#39;]
Found directory: ../inputs/landscape_arrays/test
Found 2590 array files, extracted 1295 unique IDs in ../inputs/landscape_arrays/test
  Sample files: [&#39;ENSMUSG00000028715.2_rev.npy&#39;, &#39;ENSMUSG00000073842.10_rev.npy&#39;, &#39;ENSMUSG00000028287.4_forward.npy&#39;]
Found directory: ../inputs/microC/training
Found 40902 array files, extracted 20451 unique IDs in ../inputs/microC/training
  Sample files: [&#39;ENSMUSG00000024349.10_rev.npy&#39;, &#39;ENSMUSG00000024539.17_rev.npy&#39;, &#39;ENSMUSG00000025464.15_forward.npy&#39;]
Found directory: ../inputs/microC/test
Found 2590 array files, extracted 1295 unique IDs in ../inputs/microC/test
  Sample files: [&#39;ENSMUSG00000028715.2_rev.npy&#39;, &#39;ENSMUSG00000073842.10_rev.npy&#39;, &#39;ENSMUSG00000028287.4_forward.npy&#39;]
Found directory: ../inputs/microC_rotated/training
Found 40902 array files, extracted 20451 unique IDs in ../inputs/microC_rotated/training
  Sample files: [&#39;ENSMUSG00000024349.10_rev.npy&#39;, &#39;ENSMUSG00000024539.17_rev.npy&#39;, &#39;ENSMUSG00000025464.15_forward.npy&#39;]
Found directory: ../inputs/microC_rotated/test
Found 2590 array files, extracted 1295 unique IDs in ../inputs/microC_rotated/test
  Sample files: [&#39;ENSMUSG00000028715.2_rev.npy&#39;, &#39;ENSMUSG00000073842.10_rev.npy&#39;, &#39;ENSMUSG00000028287.4_forward.npy&#39;]
Found directory: ../inputs/Promoter-CHiC/training
Found 40896 array files, extracted 20448 unique IDs in ../inputs/Promoter-CHiC/training
  Sample files: [&#39;ENSMUSG00000024349.10_rev.npy&#39;, &#39;ENSMUSG00000024539.17_rev.npy&#39;, &#39;ENSMUSG00000025464.15_forward.npy&#39;]
Found directory: ../inputs/Promoter-CHiC/test
Found 2590 array files, extracted 1295 unique IDs in ../inputs/Promoter-CHiC/test
  Sample files: [&#39;ENSMUSG00000028715.2_rev.npy&#39;, &#39;ENSMUSG00000073842.10_rev.npy&#39;, &#39;ENSMUSG00000028287.4_forward.npy&#39;]

Distribution of genes in target categories:
  Default targets:
    Missing: 100
    Test: 1295
    Training: 20451

Distribution of genes in input arrays:
  landscape_arrays:
    Missing: 100
    Test: 1295
    Training: 20451
  microC:
    Missing: 100
    Test: 1295
    Training: 20451
  microC_rotated:
    Missing: 100
    Test: 1295
    Training: 20451
  Promoter-CHiC:
    Missing: 103
    Test: 1295
    Training: 20448

Updated annotations saved to ../annotations/Final_gene_annotations_updated.tsv
Found polii training targets at: ../targets/K562/polii/training_polii_targets.csv
Found polii test targets at: ../targets/K562/polii/test_polii_targets.csv

--- Processing HUMAN gene annotations ---

Loaded 19430 gene annotations for human
Processing target file: ../targets/K562/training_targets.csv
Found 18585 unique IDs in K562 training targets
  Sample IDs: ENSG00000197774.8, ENSG00000168118.7, ENSG00000185664.10
Processing target file: ../targets/K562/test_targets.csv
Found 759 unique IDs in K562 test targets
  Sample IDs: ENSG00000163682.11, ENSG00000171195.6, ENSG00000052802.8
Processing target file: ../targets/K562/polii/training_polii_targets.csv
Found 18585 unique IDs in K562_polii training targets
  Sample IDs: ENSG00000197774.8, ENSG00000168118.7, ENSG00000185664.10
Processing target file: ../targets/K562/polii/test_polii_targets.csv
Found 759 unique IDs in K562_polii test targets
  Sample IDs: ENSG00000163682.11, ENSG00000171195.6, ENSG00000052802.8
Found directory: ../inputs/landscape_arrays/K562/training
Found 37170 array files, extracted 18585 unique IDs in ../inputs/landscape_arrays/K562/training
  Sample files: [&#39;ENSG00000171936.1_forward.npy&#39;, &#39;ENSG00000051596.5_forward.npy&#39;, &#39;ENSG00000176390.10_forward.npy&#39;]
Found directory: ../inputs/landscape_arrays/K562/test
Found 1518 array files, extracted 759 unique IDs in ../inputs/landscape_arrays/K562/test
  Sample files: [&#39;ENSG00000151790.4_forward.npy&#39;, &#39;ENSG00000145246.9_forward.npy&#39;, &#39;ENSG00000168564.5_forward.npy&#39;]

Distribution of genes in target categories:
  K562:
    Missing: 86
    Test: 759
    Training: 18585
  K562_polii:
    Missing: 86
    Test: 759
    Training: 18585

Distribution of genes in input arrays:
  K562:
    Missing: 86
    Test: 759
    Training: 18585

Updated annotations saved to ../annotations/human/Final_gene_annotations_human_updated.tsv

=== OVERALL SUMMARY ===

MOUSE ANNOTATIONS:
  Total genes: 21846
  Target distributions:
    Default:
      Missing: 100
      Test: 1295
      Training: 20451
  Array distributions:
    landscape_arrays:
      Missing: 100
      Test: 1295
      Training: 20451
    microC:
      Missing: 100
      Test: 1295
      Training: 20451
    microC_rotated:
      Missing: 100
      Test: 1295
      Training: 20451
    Promoter-CHiC:
      Missing: 103
      Test: 1295
      Training: 20448

HUMAN ANNOTATIONS:
  Total genes: 19430
  Target distributions:
    K562:
      Missing: 86
      Test: 759
      Training: 18585
    K562_polii:
      Missing: 86
      Test: 759
      Training: 18585
  Array distributions:
    K562:
      Missing: 86
      Test: 759
      Training: 18585
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="III_Data_analysis.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">DATA ANALYSIS NOTEBOOK</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#i-packages-functions">I) Packages &amp;  Functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-creation-of-eir-config-files-to-define-claster">II) Creation of EIR config files to define CLASTER</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-enhancer-centric-perturbational-analysis">III) Enhancer-centric perturbational analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#iv-new-models-and-analyses">IV) New models and analyses <center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-assessing-how-the-model-learns-to-predict-transcription"><center> A) Assessing how the model learns to predict transcription<center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#b-creating-a-shorter-context-model"><center> B) Creating a shorter context model: <center></center></center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#c-changing-the-last-activation-function-or-the-loss"><center> C) Changing the last activation function or the loss:</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#d-testing-on-another-chromosome-chr19"><center> D) Testing on another chromosome (chr19):</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#e-training-without-the-enhancer-mark-h3k27ac-as-an-input"><center> E) Training without the enhancer mark (H3K27ac) as an input.</center></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#f-predict-untransformed-eu-seq-only-binned-at-1kbp-resolution"><center> F) Predict untransformed EU-seq, only binned at 1kbp resolution. <center></center></center></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-predictions-and-attributions-of-the-new-models">Analyze predictions and attributions of the new models</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#v-performance-evaluations">V) Performance evaluations:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-data-distributions">VI) Data distributions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vii-promoter-capture-hi-c-analyses">VII) Promoter-Capture Hi-C analyses</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#vi-extending-claster-to-new-cell-types-k562-human"><center> VI) Extending CLASTER to new cell types: K562 (human) <center></center></center></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#predicting-mature-rna-seq-counts-rna-polii-chip-seq">Predicting mature RNA-seq counts &amp; RNA-polII ChIP-seq</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-centered-analyses-of-rna-seq-and-polr2a-binding-differences-after-in-silico-enhancer-ablation">Enhancer-centered analyses of RNA-seq and POLR2A binding differences after <em>in silico</em> enhancer ablation:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#enhancer-gene-pairing-analysis-on-predicted-rna-seq-and-polr2a-chip-seq">Enhancer-gene pairing analysis on predicted RNA-seq and POLR2A Chip-seq.</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#editing-gene-annotations-table-to-add-whether-they-were-included-or-not-in-the-analyses">Editing gene annotations table to add whether they were included or not in the analyses</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Marc Pielies Avelli
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022, Marc Pielies Avelli.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>